<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Spatial dissimilarity analysis in single cells">

<title>commands</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="_commands_files/libs/clipboard/clipboard.min.js"></script>
<script src="_commands_files/libs/quarto-html/quarto.js"></script>
<script src="_commands_files/libs/quarto-html/popper.min.js"></script>
<script src="_commands_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="_commands_files/libs/quarto-html/anchor.min.js"></script>
<link href="_commands_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="_commands_files/libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="_commands_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="_commands_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="_commands_files/libs/bootstrap/bootstrap-c0367b04c37547644fece4185067e4a7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="index.css">
</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Spatial dissimilarity analysis for single-cell trajectories and supervised embeddings</h1>
</div>

<div>
  <div class="description">
    Spatial dissimilarity analysis in single cells
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="common-options-for-all-tools" class="level3">
<h3 class="anchored" data-anchor-id="common-options-for-all-tools">Common options for all tools</h3>
<dl>
<dt>
<strong>-h</strong>
</dt>
<dd>
Help information
</dd>
<dt>
<strong>-o</strong> <em>FILE</em>
</dt>
<dd>
Output file.
</dd>
<dt>
<strong>-tags</strong> <em>tag(s)</em>
</dt>
<dd>
Barcode tags to group reads, usually be cell barcode tag.
</dd>
<dt>
<strong>-umi</strong> <em>tag</em>
</dt>
<dd>
UMI tag.
</dd>
<dt>
<strong>-t/-@</strong> <em>number</em>
</dt>
<dd>
Multithreads to process data.
</dd>
<dt>
<strong>-report</strong> <em>FILE.csv</em>
</dt>
<dd>
Summary report in csv format.
</dd>
</dl>
</section>
<section id="parse" class="level3">
<h3 class="anchored" data-anchor-id="parse">parse</h3>
<p>The parse tool is specifically designed to convert FASTQ files into the extended FASTQ+ format. It utilizes the -rule option to define the library structure, accommodating various sequencing setups. For ease of use, this tool has included predefined common library structures in the release; these can be applied using the -x option. This tool is optimized for speed and supports the correction of barcodes that have up to one mismatch.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Parse cell barcode and UMI string from raw FASTQ.</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>$ PISA parse -rule CB,R1:1-10,whitelist.txt,CB,1;R1,R1:11-60;R2,R2 -report fastq.csv lane1_1.fq.gz,lane02_1.fq.gz lane1_2.fq.gz,lane2_2.fq.gz</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a> -1       <span class="co">[</span><span class="ot">fastq</span><span class="co">]</span>   Read 1 output.</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a> -2       <span class="co">[</span><span class="ot">fastq</span><span class="co">]</span>   Read 2 output.</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a> -rule    <span class="co">[</span><span class="ot">STRING</span><span class="co">]</span>  Read structure in line. See Notice.</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a> -p                 Read 1 and read 2 interleaved in the input file.</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a> -q       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Drop reads if average sequencing quality below this value.</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a> -dropN             Drop reads if N base in sequence or barcode.</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a> -report  <span class="co">[</span><span class="ot">csv</span><span class="co">]</span>     Summary report.</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a> -t       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Threads. <span class="co">[</span><span class="ot">4</span><span class="co">]</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a> -order             Keep input order.</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a> -x                 Predefined code for specific library.</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="ss">          * </span>C4      Library structure for DNBelab C4 RNA kit v1.</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>-rule accept tag rule STRING to parse input fastq following format "TAG,location,whitelist,corrected TAG,allow mismatch".</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>   For each tag rule, location part should be format like R<span class="co">[</span><span class="ot">12</span><span class="co">]</span>:start-end. Both start and end location start from 1.</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>   TAG and locaion parts are mandatory, and whitelist, corrected TAG and mismatch are optional.</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>   Futhermore, multiply tags seperated by ';'. In location part, R1 stands for raw read 1, R2 stands for raw read 2.</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>   In tag part, R1 stands for output read 1 while R2 stands for output read 2. Here are some examples.</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>$ PISA parse -rule 'CR,R1:1-18,barcodes.txt,CB,1;UR,R1:19-30;R1,R2:1-100' -1 read_1.fq raw_read_1.fq raw_read_2.fq</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a><span class="fu"># CR,R1:1-18,barcodes.txt,CB,1  - CR tag start from 1 to 18 in read 1, and barcodes.txt are barcode whitelist,</span></span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a><span class="fu">#   each barcode per line. Cell barcode will be corrected while hamming distance &lt;= 1.</span></span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a><span class="fu">#   Corrected cell barcode tag is CB. </span></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a><span class="fu"># UR,R1:19-30 - UR tag start from 19-30 in read 1.</span></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a><span class="fu"># R1,R2:1-100 - Sequence from 1 to 100 in read 2 output to read 1 file. </span></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>$ PISA parse -rule 'CR,R1:1-10,bc1.txt,CB,1;CR,R1:11-20,bc2.txt,CB,1;R1,R2:1-100' -1 read_1.fq raw_read_1.fq raw_read_2.fq</span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a><span class="fu"># CR,R1:1-10,bc1.txt,CB,1;CR,R1:11-20,bc2.txt,CB,1 - This cell barcode consist of two segments, first segment start</span></span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a><span class="fu">#   from 1 to 10 in read 1, and whitelist is bc1.txt, and second segment start from 11 to 20, and whitelist is bc2.txt.</span></span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a><span class="fu">#   These two segments will be combined after correction, because the corrected tag are the same.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Option -report can specify a quality control report in CSV format. Here is the explanation of each term in this file.</p>
<div class="center">
<table class="caption-top table">
<colgroup>
<col style="width: 36%">
<col style="width: 63%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"><strong>Terms</strong></th>
<th style="text-align: left;"><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Number of Fragments</td>
<td style="text-align: left;">The number of records in the FASTQ(s). For paired reads, each pair only count once.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Fragments pass QC</td>
<td style="text-align: left;">Reads or paired reads pass QC.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Fragments with Exactly Matched Barcodes</td>
<td style="text-align: left;">Barcodes exactly matched with any barcode in the candidate list.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Fragments with Failed Barcodes</td>
<td style="text-align: left;">No barcode found in the candidate list with similar search.</td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="fsort" class="level3">
<h3 class="anchored" data-anchor-id="fsort">fsort</h3>
<p>The fsort tool is engineered to order FASTQ+ records based on specified tags, which can be defined using the -tags option. This tool efficiently handles file sorting: for files smaller than 1 gigabyte, it performs the sort directly. However, for larger FASTQ files, where caching the entire file in memory is impractical, fsort employs a different strategy. It splits the large file into smaller segments, sorts each segment individually, and then merges the sorted segments. This method ensures efficient handling of large datasets while maintaining the integrity and order of the FASTQ+ records.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Sort reads by tags.</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>$ PISA fsort -tags CB,UR -list cell_barcodes_top10K.txt -@ 5 -o sorted.fq.gz in.fq</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a> -tags    <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>     Tags, such as CB,UR. Order of these tags is sensitive.</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a> -@       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Threads to compress file.</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a> -o       <span class="co">[</span><span class="ot">fq.gz</span><span class="co">]</span>    bgzipped output fastq file.</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a> -m       <span class="co">[</span><span class="ot">mem</span><span class="co">]</span>      Memory per thread. <span class="co">[</span><span class="ot">1G</span><span class="co">]</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a> -p                  Input fastq is smart pairing.</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a> -T       <span class="co">[</span><span class="ot">prefix</span><span class="co">]</span>   Write temporary files to PREFIX.nnnn.tmp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="stream" class="level3">
<h3 class="anchored" data-anchor-id="stream">stream</h3>
<p>The stream tool is designed as a framework to process FASTQ+ files, where each FASTQ+ block—defined by having identical tags and grouped together in the file—is handled individually. The -script option allows users to specify a custom bash script that processes each block. This user-defined script reads a ‘small’ FASTQ+ file, generating FASTQ or FASTA output that is then sent to stdout. The stream tool captures this output via a pipe and updates the tags to ensure that each block of reads retains its original tags. Finally, it consolidates all outputs into a single file. In essence, this tool efficiently divides and processes each FASTQ+ block through a user-defined method, then seamlessly merges the results.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Perform user-defined script for each FASTQ+ block.</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>$ PISA stream -script run.sh reads.fq.gz</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a> -tags    <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>     Tags to define read blocks.</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a> -script  <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     User defined bash script, process $FQ and generate results to stdout.</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a> -min     <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Mininal reads per block to process.  <span class="co">[</span><span class="ot">2</span><span class="co">]</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a> -keep               Output unprocessed FASTQ+ records.</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a> -fa                 Stream FASTQ output instead of FASTQ.</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a> -tmpdir</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a> -t                  Threads.</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a> -o       <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Path to output file.</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a> -nw                 Disable warning messages.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="write-a-script-for-pisa-stream" class="level4">
<h4 class="anchored" data-anchor-id="write-a-script-for-pisa-stream">Write a script for PISA stream</h4>
<p>The PISA stream tool generates a temporary file named _block.fq for each block of reads, storing this file in a designated temporary directory. The path to this file is set in the environment variable ${FQ} for accessibility by subprocesses. Additionally, to ensure the uniqueness of each block, an alias named the ‘unique block index’ is exported to the environment as ${UBI}.</p>
<p>The following example script demonstrates how to convert FASTQ+ to FASTA and rename the sequence ID. It is crucial for users to ensure that the script’s final output (either FASTQ+ or FASTA) is directed to stdout. All other script steps should avoid producing output to stdout or stderr, except for the last step. This precaution is necessary because PISA captures the script’s output through a pipe, and any unintended characters could disrupt the data format. Scripts can be written in a bash file or specified inline within the command.</p>
<div class="framed">
<div class="sourceCode" id="cb4"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>$ cat run.sh</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>seqtk rename ${FQ} &gt; test.fa; seqtk rename test.fa ${UBI}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="addtags" class="level3">
<h3 class="anchored" data-anchor-id="addtags">addtags</h3>
<p>The addtags tool is designed to put the new tags or update tags for FASTQ+ or BAM files.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Just add tags to reads.</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>$ PISA addtags -str CB:Z:CELL1,LB:Z:PoII2 -o out.bam in.bam</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>$ PISA addtags -str CB:Z:CELL1,LB:Z:PoII2 -o out.fastq in.fastq</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>-o    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Output file, bam or fastq, depends on input format</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>-str  <span class="co">[</span><span class="ot">string</span><span class="co">]</span>  TAGs.</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>-@    <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Threads to pack file.</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>-mapq <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Mapping quality score to filter mapped reads.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="sam2bam" class="level3">
<h3 class="anchored" data-anchor-id="sam2bam">sam2bam</h3>
<p>After alignment, the sequence ID from the FASTQ+ records is retained in the RNAME field of the SAM file. Given that the RNAME field is limited to 254 characters, we also restrict the sequence ID and any optional tag fields in FASTQ+ to this length to ensure compatibility. The sam2bam tool processes these details by parsing the tags from the RNAME and appending them to the end of the SAM optional fields.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Parse FASTQ+ read name and convert SAM to BAM.</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>$ PISA sam2bam -report alignment.csv -@ 5 -adjust-mapq -gtf genes.gtf -o aln.bam in.sam<span class="co">[</span><span class="ot">.gz</span><span class="co">]</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a> -o       <span class="co">[</span><span class="ot">BAM</span><span class="co">]</span>       Output file <span class="co">[</span><span class="ot">stdout</span><span class="co">]</span>.</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a> -t       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>       Work threads.</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a> -mito    <span class="co">[</span><span class="ot">string</span><span class="co">]</span>    Mitochondria name. Used to stat ratio of mitochondria reads.</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a> -maln    <span class="co">[</span><span class="ot">BAM</span><span class="co">]</span>       Export mitochondria reads into this file instead of standard output file.</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a> -@       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>       Threads to compress bam file.</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a> -report  <span class="co">[</span><span class="ot">csv</span><span class="co">]</span>       Alignment report.</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>Note :</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Reads map to multiple loci usually be marked as low quality and filtered at downstream analysis.</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>  But for RNAseq library, if reads map to an exonic locus but also align to 1 or more non-exonic loci,</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>  the exonic locus can be prioritized as primary alignments, and mapping quality adjusts to 255. Tag</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>  MM:i:1 will also be added for this record. Following options used to adjust mapping quality.</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Input SAM need be sorted by read name, and aligner should output all hits of a read in this SAM.</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a> -adjust-mapq         Enable adjusts mapping quality score.</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a> -gtf     <span class="co">[</span><span class="ot">GTF</span><span class="co">]</span>       GTF annotation file. This file is required to check the exonic regions.</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a> -qual    <span class="co">[</span><span class="ot">255</span><span class="co">]</span>       Updated quality score.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Option -t is to set the threads to parse the SAM records. The -@ option is to set the threads to compress alignments in BGZF format. The default compress level of BGZF is 6 in the htslib, but here PISA has reset this value to 2 to decrease the CPU times. It’s not a good design to have both -t and -@, but will require a lot work to redesign the multithreads strategy, I have put this work in my todo list.</p>
<p>Option -report can specify a quality control report in CSV format. Here is the explanation of each term in this file.</p>
<div class="center">
<table class="caption-top table">
<colgroup>
<col style="width: 26%">
<col style="width: 73%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"><strong>Terms</strong></th>
<th style="text-align: left;"><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Raw reads</td>
<td style="text-align: left;">Raw reads in the BAM files, secondary alignment will be skipped.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Mapped reads</td>
<td style="text-align: left;">Reads mapped to reference, and the ratio of raw reads.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Plus strand</td>
<td style="text-align: left;">Reads mapped to forward strand of reference.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Minus strand</td>
<td style="text-align: left;">Reads mapped to backward strand of reference.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Mitochondria ratio</td>
<td style="text-align: left;">Ratio of reads mapped to chromosome mitochondria. The default mito name is “chrM”, user should change it by -mito option if reference is different. Otherwise this value will always be 0.</td>
</tr>
</tbody>
</table>
</div>
<section id="mapq-adjust-method" class="level4">
<h4 class="anchored" data-anchor-id="mapq-adjust-method">MapQ adjust method</h4>
<p>For RNA libraries, if a read from cDNA maps to an exonic locus but also maps to one or more non-exonic regions, the exonic locus can be prioritized as primary alignments, and mapping quality adjusts to 255. In the below records below, read DP8400008965TLL1C001R0102043364 mapped to three loci, and the aligner random pick one as the primary alignment and others as secondary. Each of these alignments has low mapping quality (MAPQ == 2, is usually filtered at downstream analysis). Our adjustment method will check if only one of these alignments overlaps with exonic regions. In our example, the last alignment overlapped with gene EEF1A1, and the other two hit intergenic regions. After adjustment, the last record has been flagged as a primary hit, and the mapping quality adjusted to 255, and other alignments of the same read are updated as secondary, MAPQ adjust to 0. MM:i:1 tag also is added to the primary record after adjustment. Option -adjust-mapq is reimplemented to mirror the 10X CellRanger’s MAPQ adjustment method (<a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview#alignment" class="uri">https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview#alignment</a>).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">#  Output by aligner:</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>DP8400008965TLL1C001R0102043364 0   9   133020979   2   100M    *   0   0   GTTAATGATAACAATGCATCGTAAAACCTTCAGAAGGAAAGGAGAATGTTTTGTGGACCACTTTGGTTTTCTTGTTTGCGTGTGGCAGTTTTAAGTTTTT    ...</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>DP8400008965TLL1C001R0102043364 272 7   22510408    2   100M    *   0   0   AAAAACTTAAAACTGCCACACGCAAACAAGAAAACCAAAGTGGTCCACAAAACATTCTCCTTTCCTTCTGAAGGTTTTACGATGCATTGTTATCATTAAC    ...</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>DP8400008965TLL1C001R0102043364 272 6   73517606    2   100M    *   0   0   AAAAACTTAAAACTGCCACACGCAAACAAGAAAACCAAAGTGGTCCACAAAACATTCTCCTTTCCTTCTGAAGGTTTTACGATGCATTGTTATCATTAAC    ...</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="fu"># After adjustment (Seq and Qual in secondary alignments masked as *):</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>DP8400008965TLL1C001R0102043364 256 9   133020979   0   100M    *   0   0   *   ...</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>DP8400008965TLL1C001R0102043364 272 7   22510408    0   100M    *   0   0   *   ...</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>DP8400008965TLL1C001R0102043364 16  6   73517606    255 100M    *   0   0   AAAAACTTAAAACTGCCACACGCAAACAAGAAAACCAAAGTGGTCCACAAAACATTCTCCTTTCCTTCTGAAGGTTTTACGATGCATTGTTATCATTAAC    ... MM:i:1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>An example list here to show how to enable mapq adjuestment.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>PISA sam2bam -report alignment.csv -o out.bam -adjust-mapq -gtf hg38.gtf in.sam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="rmdup" class="level3">
<h3 class="anchored" data-anchor-id="rmdup">rmdup</h3>
<p>To effectively manage PCR duplication in single-cell experiments, it is essential to consider both cell and molecular barcodes. During the feature counting stage facilitated by PISA count, deduplication is efficiently handled by relying solely on unique UMIs. This reliance makes traditional PCR deduplication unnecessary for libraries that use UMIs. Nonetheless, producing a deduplicated BAM file remains beneficial for other analytical processes, such as variant calling, or simply to reduce file size.</p>
<p>The rmdup tool is specifically designed to remove duplicate reads that share identical barcodes, such as UMIs and cell barcodes. This selective deduplication approach ensures that only truly redundant data is removed, thus preserving the integrity and completeness of the dataset for comprehensive downstream analyses.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Deduplicate PCR reads with same barcodes.</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>$ PISA rmdup -tags CB,UR -o rmdup.bam in.bam</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>   -tags  <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>       Barcode tags to group reads.</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>   -@     <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>        Threads to unpack BAM.</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>   -o     <span class="co">[</span><span class="ot">BAM</span><span class="co">]</span>        Output bam.</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>   -q     <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>        Map Quality Score cutoff.</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>   -k                  Keep duplicates, make flag instead of remove them.</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>   -nw                 Disable warnings.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>In this version, PISA rmdup only supports single-end reads. For paired-end reads, such as scATAC data, PCR deduplication can be performed by the PISA bam2frag tool.</p>
</section>
<section id="pick" class="level3">
<h3 class="anchored" data-anchor-id="pick">pick</h3>
<p>The PISA pick tool is designed to select alignments with predefined tags and candidate values.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>$ PISA pick -tags CB,GN -list cell_barcodes.txt in.bam</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a> -tags    <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>       Barcode tags.</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a> -list    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>       Barcode white list, tag values in related column will be apply.</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a> -o       <span class="co">[</span><span class="ot">BAM</span><span class="co">]</span>        Output file.</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a> -q       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>        Map Quality Score cutoff.</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a> -@       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>        Threads to unpack BAM.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Depending on the number of tags specified by the user, the candidate list for tags can consist of either a single column or multiple columns. If multiple tags are specified but only one column is present in the list, the program will primarily compare the value of the first tag in the alignments with the list.</p>
</section>
<section id="anno" class="level3">
<h3 class="anchored" data-anchor-id="anno">anno</h3>
<p>Connecting alignment results with genomic features is essential in single-cell data analysis. PISA categorizes features into three main types: gene annotation, functional regions, and genetic or sequence variations. For gene annotation, the PISA anno tool efficiently organizes all exons, transcripts, and genes from a GTF database into a sorted hierarchical tree structure.</p>
<p>Based on their alignment status, reads are then classified into one of nine distinct types, see illustrate below <a href="#fig:fig01" data-reference-type="ref" data-reference="fig:fig01">Figure 1</a>. This detailed categorization helps in accurately assessing the transcriptional landscape and understanding the complex genomic architecture within single-cell datasets.</p>
<div id="fig:fig01" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="anno_type.png" class="img-fluid figure-img"></p>
<figcaption>Figure 1: Read types.</figcaption>
</figure>
</div>
<div class="sourceCode" id="cb11"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu"># annotate strand-specific reads</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>$ PISA anno -gtf genes.gtf -o anno.bam in.bam</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu"># annotate non-strand-specific reads, for Smartseq or bulk RNAseq</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>$ PISA anno -is -gtf genes.gtf -o anno.bam in.bam</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="fu"># also label gene name for intronic reads, i.e. RNA velocity analysis</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>$ PISA anno -velo -gtf genes.gtf -o anno.bam in.bam</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="fu"># annotate exon, junction, and exon skipped reads</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>$ PISA anno -exon -psi -gtf genes.gtf -o anno.bam in.bam</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a><span class="fu"># annotate expressed peaks</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>$ PISA anno -bed peak.bed -o anno.bam in.bam</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a><span class="fu"># annotate genetic variants (both reference allele and alternative allele) </span></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>$ PISA anno  -vcf in.vcf.gz -ref-alt -o anno.bam in.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Besides these annotation methods, PISA anno also supports a -chr-species method. This method requires a binding list for chromosome and related label relationships. The software will check the chromosome and add the related tag for each chromosome. This method, combined with PISA attrcnt can help to summarize the mixed two cell lines from different species.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># A binding list is tab-separated two columns txt file.</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>$ cat binding_list.txt</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>GRCh38_chr1    Human</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>GRCh38_chr21   Human</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>mm10_chr21     Mouse</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>$ PISA anno -chr-species binding.txt -btag SP -o anno_species.bam sorted.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The full options and descriptions list below.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Annotate SAM/BAM records with overlapped function regions. Such as gene, transcript etc.</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>$ PISA anno -bed peak.bed -tag PK -vcf in.vcf.gz -vtag VF -vcf-ss -ref-alt -o anno.bam in.bam</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>$ PISA anno -gtf genes.gtf -o anno.bam in.bam</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>$ PISA anno -gtf genes.gtf -o anno.bam -sam in.sam</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">BAM</span><span class="co">]</span>       Output bam file.</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a> -report   <span class="co">[</span><span class="ot">csv</span><span class="co">]</span>       Summary report.</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a> -@        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>       Threads to compress bam file.</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a> -q        <span class="co">[</span><span class="ot">0</span><span class="co">]</span>         Map Quality Score cutoff. MapQ smaller than this value will not be annotated.</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a> -t        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>       Threads to annotate.</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a> -chunk    <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>       Chunk size per thread.</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a> -anno-only            Export annotated reads only.</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a> -sam                  Input is SAM file, parse tags from read name.</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a> -rev                  Annotation in reverse strand; Some probe ligation library for FFPE samples create reverse fragments.</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a> -is                   Disable strand sensitive annotation of gene, genomic region and genetic variants.</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>Options for BED file :</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a> -bed      <span class="co">[</span><span class="ot">BED</span><span class="co">]</span>       Function regions. Three or four columns bed file. Col 4 could be empty or names of this region.</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a> -tag      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>       Attribute tag name. Set with -bed. Default is PK.</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>Options for mixed samples.</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a> -chr-species  <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>  Chromosome name and related species binding list.</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a> -btag     <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>       Species tag name. Set with -chr-species. Default is SP.</span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>Options for GTF file :</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a> -gtf      <span class="co">[</span><span class="ot">GTF</span><span class="co">]</span>       GTF annotation file. gene_id,transcript_id is required for each record.</span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a> -tags     <span class="co">[</span><span class="ot">TAGs</span><span class="co">]</span>      Attribute names, more details see <span class="in">`Notice`</span> below. <span class="co">[</span><span class="ot">TX,GN,GX,RE,EX,JC</span><span class="co">]</span></span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a> -splice               Reads covered exon-intron edge (ExonIntron type) will also be annotated with all tags.</span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a> -intron/-velo         Reads covered intron regions will also be annotated with all tags.</span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a> -exon                 Generate exon level and junction annotation. Put exon name (chr:start-end/[<span class="co">[</span><span class="ot">+-</span><span class="co">]</span>/Gene) in EX tag.\ </span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a>                       Also generate junction name (chr:exon1_end-exon2_start/<span class="co">[</span><span class="ot">+-</span><span class="co">]</span>/Gene) in JC tag.</span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a> -flatten              Split overlapped exons into nonoverlapped bins.</span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a> -psi                  Annotate exclude reads tag (ER) for each exon.</span>
<span id="cb13-36"><a href="#cb13-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-37"><a href="#cb13-37" aria-hidden="true" tabindex="-1"></a> -tss                  Annotate reads start from TSS, designed for capped library. **experiment**</span>
<span id="cb13-38"><a href="#cb13-38" aria-hidden="true" tabindex="-1"></a> -ctag     <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>       Tag name for TSS annotation. Need set with -tss.</span>
<span id="cb13-39"><a href="#cb13-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-40"><a href="#cb13-40" aria-hidden="true" tabindex="-1"></a>Options for VCF file :</span>
<span id="cb13-41"><a href="#cb13-41" aria-hidden="true" tabindex="-1"></a> -vcf      <span class="co">[</span><span class="ot">VCF/BCF</span><span class="co">]</span>   Varaints file in vcf or bcf format. In default, only annotate alternative alleles.</span>
<span id="cb13-42"><a href="#cb13-42" aria-hidden="true" tabindex="-1"></a> -vtag     <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>       Tag name for variants. Set with -vcf. Default is VR.</span>
<span id="cb13-43"><a href="#cb13-43" aria-hidden="true" tabindex="-1"></a> -ref-alt              Annotate ref allele.</span>
<span id="cb13-44"><a href="#cb13-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-45"><a href="#cb13-45" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb13-46"><a href="#cb13-46" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>If input is SAM format, will try to parse the tags in the read name.</span>
<span id="cb13-47"><a href="#cb13-47" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>For GTF mode, this program will set tags in default, you could also reset them by -tags.</span>
<span id="cb13-48"><a href="#cb13-48" aria-hidden="true" tabindex="-1"></a>   TX : Transcript id.</span>
<span id="cb13-49"><a href="#cb13-49" aria-hidden="true" tabindex="-1"></a>   GN : Gene name.</span>
<span id="cb13-50"><a href="#cb13-50" aria-hidden="true" tabindex="-1"></a>   GX : Gene ID.</span>
<span id="cb13-51"><a href="#cb13-51" aria-hidden="true" tabindex="-1"></a>   RE : Region type, E (exon), N (intron), C (exon and intron), S (junction reads cover isoforms properly), V (ambiguous reads),</span>
<span id="cb13-52"><a href="#cb13-52" aria-hidden="true" tabindex="-1"></a>        I (intergenic), A (antisense or antisense exon), B (antisense intron), X (one or more exons are excluded in transcrpit)</span>
<span id="cb13-53"><a href="#cb13-53" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>The following tags set with -exon.</span>
<span id="cb13-54"><a href="#cb13-54" aria-hidden="true" tabindex="-1"></a>   EX : Exon name tag.</span>
<span id="cb13-55"><a href="#cb13-55" aria-hidden="true" tabindex="-1"></a>   JC : Isoform junction name.</span>
<span id="cb13-56"><a href="#cb13-56" aria-hidden="true" tabindex="-1"></a>   FL : Flatten exon name. Only generate it with -flatten.</span>
<span id="cb13-57"><a href="#cb13-57" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>The following tags set with -psi.</span>
<span id="cb13-58"><a href="#cb13-58" aria-hidden="true" tabindex="-1"></a>   ER : Excluded exons.</span>
<span id="cb13-59"><a href="#cb13-59" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>PSI = EX/(EX+ER); EX is the exon tag, which indicate include reads in exon.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="corr" class="level3">
<h3 class="anchored" data-anchor-id="corr">corr</h3>
<p>The diversity of UMIs of each gene in one cell is used to evaluate the gene expression level, but the error of UMI comes from sequencing or PCR may introduce bias. PISA corr is designed to correct the UMI or barcode sequence based on Hamming distance. In default, two groups of UMI from the same gene of one cell with Hamming distance equal to 1, will be considered to originate from the same transcript. The group with high frequency will be selected as a real one, and another one will be corrected to the high one.</p>
<p>Because PISA corr does not require a sorted BAM, this tool will first build a correction list by caching all the raw UMIs and barcodes and then correct them in memory. After these steps, this tool will reread the file and update these records by order. This design can avoid potential bias for the same gene from different chromosomes (i.e., the HLA genes in alternative locus). However, this design also required a lot of memory for a big BAM.</p>
<p>CellRanger also introduces an algorithm to correct reads with the same UMI of one cell but mapping to more than one gene (<a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview" class="uri">https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview</a>). PISA implements this method but does not enable it in default. The -cr option can be used by users to enable this function. The reason for not enable it by default is PISA corr not only is used to correct UMIs, but also can be used to correct any other types of barcodes. The following example shows how to use PISA corr to correct cell barcodes for reads in the same gene.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>// Reads with same gene tag (GN) and UMI (UB) will be grouped and calculate the Hamming distance between each other. </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>// Only if Hamming distance == 1 will be corrected.</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>PISA corr -tag CR -new-tag CB -tags-block GN,UB -o cell_barcode_corrected.bam in.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Full options of PISA corr list below.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Correct similar barcodes (hamming distance == 1).</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>$ PISA corr -tag UR -new-tag UB -tags-block CB,GN -@ 5 -o corr.bam in.bam</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">BAM</span><span class="co">]</span>       Output bam.</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a> -tag      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>       Tag to correct.</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a> -new-tag  <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>       Create a new tag for corrected barcodes.</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a> -tags-block  <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>   Tags to define read group. For example, if set to GN (gene), reads in the same gene will be grouped together.</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a> -cr                   Enable CellRanger like UMI correction method. See <span class="in">`Examples`</span> for details.</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a> -e                    Maximal hamming distance to define similar barcode, default is 1.</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a> -@        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>       Thread to compress BAM file.</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>Examples :</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a> // Two groups of reads have same cell barcode (CB) and gene (GN) but their raw UMIs (UR) differ by only one base. The UMI of less </span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a> // supported is corrected to the UMI with higher support. UB save the checked or corrected UMI.</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>$ PISA corr -tag UR -new-tag UB -tags-block CB,GN in.bam -o corr.bam </span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a> // Same with above. Besides, if two or more groups of reads have same CB and UB but different GN, the GN with the most supporting reads</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a> // is kept for UMI counting, and the other read groups are discarded. In case of a tie for maximal read support, all read groups are</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a> // discarded, as the gene cannot be confidently assigned (Cell Ranger method).</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>$ PISA corr -cr -tag UR -new-tag UB -tags-block CB,GN in.bam -o corr.bam </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="attrcnt" class="level3">
<h3 class="anchored" data-anchor-id="attrcnt">attrcnt</h3>
<p>PISA attrcnt is used to summarize the meta information of the library. We start introduce this tool with few examples.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count reads per cell, -cb option is required to specify cell barcode tag</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>$ PISA attrcnt -cb CB in.bam</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>// the summary information output in tsv format</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>BARCODE Raw                           // the title</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>AGCTATGCTTCTAGTGTAAC-1  38            // cell barcode and raw reads per cell, separated by a tab</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>GCCGCTGATCGGCCTGCACA-1  12</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>GTCGCGATTCTGCTCAGAAG-1  13</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>GTCAGTACCATGCTCAGAAG-1  27</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>CAACTCGTCGGTTGTCTGAC-1  23</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count raw reads and reads in the peaks per cell</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>$ PISA attrcnt -cb CB -tags PK             // PK tag is annotated for reads in peak</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>             -o summary.tsv              // Summary file</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>             example/anno/demo_1.bam </span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>$ head summary.tsv </span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>BARCODE Raw PK</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>GGCATTATCGGCTCGGTATG    2   2</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>TGAGTTGTGTATACTCCTAC    2   2</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>CCGCGGCACTACACACCAGA    1   1</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>ATTAGTGGTCTCCTGGTCGG    1   1</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>TTATTGGACCACGTTGAATA    1   1</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>GGAATGCCACAAGCGCCGTA    1   1</span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>AGTCTATCGTGGCCTGCACA    5   5</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>AGGCACACCTCGCTGAATTC    3   3</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>GTCAGGATCGATAACATACG    2   2</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count UMIs per cell</span></span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>$ PISA attrcnt -cb CB -tags UB      // UB is tag of corrected UMIs</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>             -dedup               // -dedup option used to remove duplication of tag values, if not set, this </span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>                                  // command will export reads with UB tag but not the unique UMIs per cell</span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>             -o summary.tsv anno.bam</span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count UMIs and Genes per cell</span></span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a>$ PISA attrcnt -cb CB -tags UB,GN  // GN is tag for annotated gene; -tags can accept multiple tag names, and seperated by ","</span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a>             -dedup -o summary.tsv anno.bam</span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a>$ head summary.tsv</span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a>BARCODE Raw UB  GN</span>
<span id="cb16-41"><a href="#cb16-41" aria-hidden="true" tabindex="-1"></a>AGCTATGCTTCTAGTGTAAC-1  38  35  0</span>
<span id="cb16-42"><a href="#cb16-42" aria-hidden="true" tabindex="-1"></a>GCCGCTGATCGGCCTGCACA-1  12  12  0</span>
<span id="cb16-43"><a href="#cb16-43" aria-hidden="true" tabindex="-1"></a>GTCGCGATTCTGCTCAGAAG-1  13  11  0</span>
<span id="cb16-44"><a href="#cb16-44" aria-hidden="true" tabindex="-1"></a>GTCAGTACCATGCTCAGAAG-1  27  26  0</span>
<span id="cb16-45"><a href="#cb16-45" aria-hidden="true" tabindex="-1"></a>CAACTCGTCGGTTGTCTGAC-1  23  21  0</span>
<span id="cb16-46"><a href="#cb16-46" aria-hidden="true" tabindex="-1"></a>GGTACACCACAGTAGTTACG-1  12  10  0</span>
<span id="cb16-47"><a href="#cb16-47" aria-hidden="true" tabindex="-1"></a>GCGCGCCGAGGGACACTCTT-1  1   1   0</span>
<span id="cb16-48"><a href="#cb16-48" aria-hidden="true" tabindex="-1"></a>CTCTAAGCATCGAGGTTAAC-1  162 138 50</span>
<span id="cb16-49"><a href="#cb16-49" aria-hidden="true" tabindex="-1"></a>TTCGTAGCACCGATACTAGC-1  51  48  46</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Full list of options list below.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count the frequency of tag values.</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>$ PISA attrcnt -cb CB -tags UR,GN -dedup -all-tags in.bam</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a> -cb       <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Cell Barcode, or other tag used for grouping reads.</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a> -list     <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Cell barcode white list.</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a> -tags     <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>     Tags to count.</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a> -dedup               Deduplicate the atrributes in each tag.</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a> -all-tags            Only records with all tags be count.</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a> -group    <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Group tag, count all tags for each group seperately.</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Output count table.</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a> -q        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Map Quality to filter bam.</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a> -no-header           Ignore header in the output.</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a> -@        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Thread to unpack bam.</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a> -ttag     <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Region type tag. <span class="co">[</span><span class="ot">RE</span><span class="co">]</span></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a> -ttype               Region type used to count. Set <span class="in">`E,S`</span> to count exon enclosed reads. Set <span class="in">`N,C`</span> to count intron overlapped reads.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="extract" class="level3">
<h3 class="anchored" data-anchor-id="extract">extract</h3>
<p>PISA extract is designed to extract the values of tags from BAM records and generate a tab-separated file.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Extract tag values from alignments.</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>$ PISA extract -tags CB,UR,GN -o tags.tsv in.bam</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a> -tags     <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>     Tags to be extracted.</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Output file. tsv format</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a> -n                   Print read name.</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a> -q                   Map Quality Score threshold.</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a> -all                 Only export if all tags have value.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="count" class="level3">
<h3 class="anchored" data-anchor-id="count">count</h3>
<p>The PISA count tool is designed to generate a counts matrix for various features or tags, traditionally outputting a gene-by-cell digit matrix. Starting with version 0.4, this tool now supports the MEX format output, which is highly recommended for use in downstream analyses due to its superior performance. It’s important to note that MEX format consists of three files, so you should use the -outdir option to specify the directory where the output files will be saved.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count reads or fragments matrix for single-cell datasets.</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>$ PISA count -cb CB -anno-tag GN -umi UB -outdir exp aln.bam</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>$ PISA count <span class="co">[</span><span class="ot">options</span><span class="co">]</span> aln1.bam,aln2.bam</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>$ PISA count -cb RG -sample-list bam_files.txt -outdir exp</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>$ PISA count -tags Cx,Cy -anno-tag GN -umi UB -outdir exp -velo aln.bam</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a> -tags/-cb <span class="co">[</span><span class="ot">TAGs</span><span class="co">]</span>     A cell barcode tag or two tags of spatial coordinates for spatial data.</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a> -anno-tag <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Annotation tag, gene or peak.</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a> -genome-bin <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>    If genome bin size set, genome bin count matrix will be generated, conflict with -anno-tag and -chr.</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a> -is                  Ignore strand for bin counting.</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a> -chr                 Count chromosome expression level, conflict with -anno-tag and -genome-bin.</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a> -list     <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Barcode white list, used as column names at matrix. If not set, all barcodes will be count.</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a> -outdir   <span class="co">[</span><span class="ot">DIR</span><span class="co">]</span>      Output matrix in MEX format into this folder.</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a> -umi      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      UMI tag. Count once if more than one record has same UMI in one gene or peak.</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a> -one-hit             Skip if a read hits more than 1 gene or peak.</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a> -q        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Minimal map quality to filter. Default is 20.</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a> -t        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Threads.</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a> -ttag     <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Region type tag. <span class="co">[</span><span class="ot">RE</span><span class="co">]</span></span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a> -velo                Generate spliced and unspliced matrix files for RNA velocity analysis.</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a> -ttype    <span class="co">[</span><span class="ot">TYPE</span><span class="co">]</span>     Region type used to count. Set <span class="in">`E,S`</span> to count exon enclosed reads. Set <span class="in">`N,C`</span> to count intron overlapped reads.</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a> -sample-list <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>  A list of bam files. Each path per line.</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>Options for Stereoseq:</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a> -stereoseq           Stereoseq pipeline pack UMI to hex string. Need set this option to decode UMIs.</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a> -spatial-bin <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>   Bin size for spatial coordiate. Can be set if -tags specify spatial coordinates.<span class="co">[</span><span class="ot">1</span><span class="co">]</span></span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a> -dup                 Do NOT skip duplicate reads.</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Region type (RE), which label functional region reads mapped, is annotated by <span class="in">`PISA anno`</span>. Optional -ttype can be set</span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a>   to one of region types(E/S/C/N) or combination to count reads mapped to these functional regions only.</span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>If you want count from more than one bam file, there are two ways to set the parameter. By seperating bam files with ',' or by</span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a>   setting -sample-list option.</span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>If -velo set, spliced and unspliced folders will be created at outdir.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="for-smartseq-user" class="level4">
<h4 class="anchored" data-anchor-id="for-smartseq-user">For Smartseq user</h4>
<p>PISA count also support counting gene expression from multiple bam files.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>$ PISA count -file-barcode             // use alias name for each bam file as cell barcode. If this flag is not set -cb must be specified.</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>-tags CB                   // Cell barcode</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>-sample-list bam_list.txt // bam file path and alias name</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>-outdir exp/ -anno-tag GN </span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="fu"># The `-sample-list' specify multiply files, each BAM path per line.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="description-of-mex-file" class="level4">
<h4 class="anchored" data-anchor-id="description-of-mex-file">Description of MEX file</h4>
<p>The Market Exchange (MEX) format (<a href="https://math.nist.gov/MatrixMarket/formats.html" class="uri">https://math.nist.gov/MatrixMarket/formats.html</a>) is designed for representing the sparse matrix. The -outdir option specifies the output directory for one MEX fold. The MEX format consists of three files, one is cell barcodes, one is feature names (genes or peak names), and the third one defines the expression or signal values.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>$ ls</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>barcodes.tsv.gz  features.tsv.gz  matrix.mtx.gz</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>$ zcat barcodes.tsv.gz|head</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>AACCTGGTGAAGTTGTCGAA</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>AAGGAACTAAGCGCAGCACC</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>CGATAGAATACTTCTTCGTA</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>TACTATCCTCTAGCTGCTAC</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>TGACCATCCTACAGTCCACC</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>CAGATTCAACTACGAAGTGC</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>TTCGTAGCACTCTTCATCTC</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>GGCACCTTGCTTAACGTAGG</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>ACTTCGGATACGTATCGCCT</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>GACTCGCTAGTAGTCGGAAT</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>$ zcat features.tsv.gz|head</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>RP11-34P13.7</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>RP11-34P13.8</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>RP11-34P13.9</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>FO538757.3</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>FO538757.2</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>AP006222.2</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a>RP4-669L17.10</span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a>RP5-857K21.4</span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>RP11-206L10.4</span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>RP11-206L10.9</span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>$ zcat matrix.mtx.gz|head</span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a>%%MatrixMarket matrix coordinate integer general</span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a>% Generated by PISA v0.4-alpha-72-g09c4ded</span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a>23900   782761  11533380</span>
<span id="cb21-32"><a href="#cb21-32" aria-hidden="true" tabindex="-1"></a>1   1   2</span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a>1   2   2</span>
<span id="cb21-34"><a href="#cb21-34" aria-hidden="true" tabindex="-1"></a>1   3   2</span>
<span id="cb21-35"><a href="#cb21-35" aria-hidden="true" tabindex="-1"></a>1   4   1</span>
<span id="cb21-36"><a href="#cb21-36" aria-hidden="true" tabindex="-1"></a>1   5   2</span>
<span id="cb21-37"><a href="#cb21-37" aria-hidden="true" tabindex="-1"></a>1   6   1</span>
<span id="cb21-38"><a href="#cb21-38" aria-hidden="true" tabindex="-1"></a>1   7   2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The MEX file can be read by R package Yano::ReadPISA.</p>
</section>
</section>
<section id="bam2fq" class="level3">
<h3 class="anchored" data-anchor-id="bam2fq">bam2fq</h3>
<p>PISA bam2fq is designed to convert alignment records to FATSQ+ records. Option -tags specify which tags will be kept in the FASTQ+. Full options list below.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Convert BAM into fastq.</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>$ PISA bam2fq -tags CB,UB,GN -o out.fq aln.bam</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a> -tags     <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>     Export tags in read name.</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a> -f                   Filter records if specified tags not all exist.</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a> -fa                  Output fasta instead of fastq.</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">fastq</span><span class="co">]</span>    Output file.</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a> -@        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Threads to unpack BAM.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="bam2frag" class="level3">
<h3 class="anchored" data-anchor-id="bam2frag">bam2frag</h3>
<p>The fragment file is a five columns tab-separated flat file which is designed for scATAC-seq. The first column is the chromosome name, the second column is the start location of this fragment (0 based), and the third column is the end position in 1 based of this fragment. The fourth column is the cell barcode of this fragment and the last column is how many duplicates of this fragment.</p>
<p>PISA bam2frag requires the input BAM file to be sorted by coordinate. This tool will check the cell barcode and the fragment position for each paired reads, duplicates in one cell will only keep one record, and the numeber of copies for this fragment will be updated in column four.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Convert sam record to fragment file.</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>$ PISA bam2frag -cb CB -list cell_barcodes.txt -o out.tsv.gz in.bam</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a> -o       <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Output file. This file will be bgzipped and indexed.</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a> -cb      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>     Cell barcode tag.</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a> -list    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Cell barcode white list.</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a> -q       <span class="co">[</span><span class="ot">20</span><span class="co">]</span>      Mapping quality score to filter reads.</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a> -isize   <span class="co">[</span><span class="ot">2000</span><span class="co">]</span>    Skip if insert size greater than this. <span class="co">[</span><span class="ot">2KB</span><span class="co">]</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a> -bed     <span class="co">[</span><span class="ot">BED</span><span class="co">]</span>     Only convert fragments overlapped with target regions.</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a> -black-region <span class="co">[</span><span class="ot">BED</span><span class="co">]</span> Skip convert fragments overlapped with black regions.</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a> -stat    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Transposition events per cell.</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a> -@       <span class="co">[</span><span class="ot">4</span><span class="co">]</span>       Thread to unpack and pack files.<span class="co">[</span><span class="ot">4</span><span class="co">]</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a> -disable-offset    Disable Tn5 offset for each fragment.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="depth" class="level3">
<h3 class="anchored" data-anchor-id="depth">depth</h3>
<p>PISA depth generates coverage information for each position of the predefined region(s). The significant difference between PISA depth and samtools depth (<a href="http://www.htslib.org/doc/samtools-depth.html" class="uri">http://www.htslib.org/doc/samtools-depth.html</a>) is that PISA depth use UMI rather than reads. Besides, PISA depth is strand sensitive, and can produce results for target cells.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count coverage depth or unique UMIs for genome locations.</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>Usage : PISA depth <span class="co">[</span><span class="ot">options</span><span class="co">]</span> sorted.bam <span class="co">[</span><span class="ot">region</span><span class="co">]</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>$ PISA depth -cb CB -umi UB -tags GN -region in.bed -o depth.tsv sorted.bam</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>$ PISA depth -cb CB -umi UB sorted.bam chr1:1-2:+</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>Options : </span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a> -tag      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Tag used for grouping reads.</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a> -list     <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Candidate list for -tag.</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a> -umi      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      UMI tag. If set, only count unique UMIs for each location.</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a> -bed      <span class="co">[</span><span class="ot">BED</span><span class="co">]</span>      Target BED region file. If the strand in column six set, only count reads with the same strand.</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Output depth file. <span class="co">[</span><span class="ot">stdout</span><span class="co">]</span>.</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a> -q        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Minimal map quality to filter. <span class="co">[</span><span class="ot">20</span><span class="co">]</span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a> -@        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Threads to unpack bam. <span class="co">[</span><span class="ot">4</span><span class="co">]</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Require sorted and indexed BAM as input.</span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Compare with `samtools depth', PISA depth considers UMIs and strand of reads.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="callept" class="level3">
<h3 class="anchored" data-anchor-id="callept">callept</h3>
<p>The term ‘EPT’ stands for Expressed Peak Tag. EPTs can be identified by PISA within aligned reads from RNA libraries. Once identified, the relationship between each EPT and its corresponding gene can be annotated and analyzed using the Yano package. This analysis helps in understanding the functional implications of EPTs in gene expression regulation.</p>
<p>The callept tool is designed to call expressed peak tags (EPTs) for indexed BAM files. The peaks are defined with UMI depth &gt;= cutoff.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>$ PISA cellept -o epts.bed sorted.bam</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>$ PISA cellept -tag CB -list cells.txt -umi UB -o epts.bed sorted.bam</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a> -tag      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Tag used for grouping reads.</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a> -list     <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Candidate list for -tag.</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a> -umi      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      UMI tag. If set, only count unique UMIs for each location.</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a> -is                  Ignore strand.</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a> -gap      <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Maximum gap to merge nearby peaks. <span class="co">[</span><span class="ot">50</span><span class="co">]</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a> -min-length  <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>   Minimum peak length. <span class="co">[</span><span class="ot">50</span><span class="co">]</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a> -cutoff   <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Cutoff of depth. <span class="co">[</span><span class="ot">10</span><span class="co">]</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Output EPTs in bed format. <span class="co">[</span><span class="ot">stdout</span><span class="co">]</span>.</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a> -q        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Minimal map quality to filter. <span class="co">[</span><span class="ot">20</span><span class="co">]</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a> -t        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Threads. <span class="co">[</span><span class="ot">4</span><span class="co">]</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Requires sorted and indexed BAM as input.</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Compares with <span class="in">`MACS2`</span> and other peak callers, PISA callept considers UMIs and strand of reads.</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>For paired reads, strand of read 2 will be reversed to revert fragment strand.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="count2" class="level3">
<h3 class="anchored" data-anchor-id="count2">count2</h3>
<p>PISA count generates the MEX format matrix for fragments per peak.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count fragments per peak per cell matrix.</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>$ PISA count2 -bed peaks.bed -t 10 -list barcodes.txt -outdir exp fragments.tsv.gz</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a> -list     <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Barcode white list, used as column names at matrix. If not set, all barcodes will be count.</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a> -outdir   <span class="co">[</span><span class="ot">DIR</span><span class="co">]</span>      Output matrix in MEX format into this fold.</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a> -prefix   <span class="co">[</span><span class="ot">STR</span><span class="co">]</span>      Prefix of output files.</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a> -t        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Threads.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="mergebed" class="level3">
<h3 class="anchored" data-anchor-id="mergebed">mergebed</h3>
<p>Merge overlapped regions by strand and name.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>$ PISA mergebed -o merged.bed sample1.bed sample2.bed</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>$ PISA mergebed -up 500 -down 500 -o flank.bed peaks.bed</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>-o    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Output bed file.</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>-s              Ignore strand.</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>-up   <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Enlarge regions upstream.</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>-down <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Enlarge regions downstream.</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>-name           Merge regions by bed name.</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>This tool accepts 3 columns or 6 columns bed file(s), strand (+/-) is set in column 6.</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>By default, merging is done with respect to strandness unless -s is set.</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>-up/-down is set respect to strandness, so upstream of plus strand is downstream of minus strand.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="annobed" class="level3">
<h3 class="anchored" data-anchor-id="annobed">annobed</h3>
<p>Annotate the type of region and output in a BED like file. The region type can be defined into.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>$ PISA annobed -gtf genes.gtf -o anno.bed in.bed</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>-gtf    <span class="co">[</span><span class="ot">GTF</span><span class="co">]</span>     GTF database.</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>-o      <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Output bed file.</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>-report <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Summary report. Export in STDERR by default.</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>-is               Ignore strand.</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>-gene-name        Set annatated gene as bed name (column 4).</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>-skip-chrs        Skip chromosomes if not exist in GTF. Defa</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>-up  <span class="co">[</span><span class="ot">1000</span><span class="co">]</span>         Annotate intergenic regions at upstream of gene.</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>-down  <span class="co">[</span><span class="ot">1000</span><span class="co">]</span>       Annotate intergenic regions at downstream of gene.</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>Output format :</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>chromosome,start(0based),end(1based),name,score,strand,number of covered genes, cover gene name(s),type,nearest gene name,distance to nearby gene</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>This tool accepts 3 columns or 6 columns bed file(s), strand (+/-) is set in column 6.</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>By default, annotation is done with respect to strandness unless -s is set.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="flatten" class="level3">
<h3 class="anchored" data-anchor-id="flatten">flatten</h3>
<p>Convert overlapped into flatten records. If strand exist, the flatten records will be strand sensitive.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>$ PISA flatten -o flatten.bed overlapped.bed</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>-o    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Output bed file.</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>For example:</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>reg1 ===========</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>reg2       ===========</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>flattening of regions:</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>reg1 ======</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>reg2       =====</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>reg3            ======</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="gtffmt" class="level3">
<h3 class="anchored" data-anchor-id="gtffmt">gtffmt</h3>
<p>Format and reorder GTF records.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>$ PISA gtffmt in.gtf</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a> -o      <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Output GTF file</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a> -f                Only export gene, transcript, exon and CDS records.</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a> -key    <span class="co">[</span><span class="ot">all</span><span class="co">]</span>     Export selected keys in optional fields.</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a> -report <span class="co">[</span><span class="ot">stderr</span><span class="co">]</span>  Summary report file.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="gtf2bed" class="level3">
<h3 class="anchored" data-anchor-id="gtf2bed">gtf2bed</h3>
<p>Convert GTF file to BED format.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>$ PISA gtf2bed -o merged.bed in.gtf.gz</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  -o    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>                          Output bed file.</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  -type <span class="co">[</span><span class="ot">gene|transcript|exon</span><span class="co">]</span>          Covert to bed.</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>  -name <span class="co">[</span><span class="ot">none|gene|transcript|exon</span><span class="co">]</span>     Set name for bed.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="deprecated-commands" class="level2">
<h2 class="anchored" data-anchor-id="deprecated-commands">Deprecated commands</h2>
<section id="parse0" class="level3">
<h3 class="anchored" data-anchor-id="parse0">parse0</h3>
<p>Since v1.0, the old parse tool has been replaced by parse2, and renamed to parse0 for backup purposes. Due to its lower performance compared to parse, parse0 will be deleted from v2 onwards. The parse0 tool requires a configuration file to describe the library structure, including cell barcode locations, UMI locations, and tag names. Additionally, it generates various quality control reports and outputs cell barcode distributions to stdout. Below is the complete options list and a general workflow for filtering and counting reads.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>$ PISA parse -config read_struct.json -report fastq.csv -cbdis cell_dist.tsv \</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>          -1 out.fq lane1_1.fq.gz,lane02_1.fq.gz  lane1_2.fq.gz,lane2_2.fq.gz</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a> -1       <span class="co">[</span><span class="ot">fastq</span><span class="co">]</span>   Read 1 output. Default is stdout.</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a> -2       <span class="co">[</span><span class="ot">fastq</span><span class="co">]</span>   Read 2 output.</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a> -config  <span class="co">[</span><span class="ot">json</span><span class="co">]</span>    Read structure configure file in JSON format. Required.</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a> -run     <span class="co">[</span><span class="ot">string</span><span class="co">]</span>  Run code, used for different library.</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a> -cbdis   <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Read count per cell barcode.</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a> -p                 Read 1 and read 2 interleaved in the input file.</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a> -q       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Drop reads if average sequencing quality below this value.</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a> -dropN             Drop reads if N base in sequence or barcode.</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a> -report  <span class="co">[</span><span class="ot">csv</span><span class="co">]</span>     Summary report.</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a> -t       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Threads. <span class="co">[</span><span class="ot">4</span><span class="co">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>-config requires a JSON file to tell software the locations of cell barcodes and/or UMIs. An example configured file can be found at demo/demo.json.</p>
<div class="framed">
<div class="sourceCode" id="cb33"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>    "cell barcode tag":"CB",  </span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>    "cell barcode":[</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>        {</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>            "location":"R1:1-16" # the location is 1 based</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>    "UMI tag":"UR",</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>    "UMI":{</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>        "location":"R1:17-28",</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>    "read 1":{</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>        "location":"R2:1-91",</span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Option -report can specify a quality control report in CSV format. Here is the explanation of each term in this file.</p>
<div class="center">
<table class="caption-top table">
<colgroup>
<col style="width: 36%">
<col style="width: 63%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"><strong>Terms</strong></th>
<th style="text-align: left;"><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Number of Fragments</td>
<td style="text-align: left;">The number of records in the FASTQ(s). For paired reads, each pair only count once.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Fragments pass QC</td>
<td style="text-align: left;">Reads or paired reads pass QC.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Fragments with Exactly Matched Barcodes</td>
<td style="text-align: left;">Cell barcode exactly matched with any barcode in the candidate list.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Fragments with Failed Barcodes</td>
<td style="text-align: left;">No cell barcode found in the candidate list with similar search.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Fragments Filtered on Low Quality</td>
<td style="text-align: left;">Mean sequence quality of the records smaller than threshold.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Fragments Filtered on Unknown Sample Barcodes</td>
<td style="text-align: left;">Sample barcodes not matched with any barcode in the candidate list.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Q30 bases in Cell Barcode</td>
<td style="text-align: left;">Ratio of bases in cell barcode sequence with quality <span class="math inline">\(&gt;=\)</span> 30.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Q30 bases in Sample Barcode</td>
<td style="text-align: left;">Ratio of bases in sample barcode sequence with quality <span class="math inline">\(&gt;=\)</span> 30.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Q30 bases in UMI</td>
<td style="text-align: left;">Ratio of bases in UMI sequence with quality <span class="math inline">\(&gt;=\)</span> 30.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Q30 bases in Reads</td>
<td style="text-align: left;">Ratio of bases in read sequence with quality <span class="math inline">\(&gt;=\)</span> 30.</td>
</tr>
</tbody>
</table>
</div>
<p>Here is an example to parse barcodes and reads with a predefined configure JSON file, the test files can be found at demo directory.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>$ PISA parse0 -config demo/demo.json -report demo/parse.csv demo/demo_1.fq.gz demo/demo_2.fq.gz &gt; demo/demo.fq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The FASTQ+ format is a variant of the FASTQ. It is designed to store more features or annotations for FASTQ records but not change the own property of the FASTQ structure. Each FASTQ+ record consists of four lines.</p>
<ul>
<li><p>The first line starts with an ‘<code>@</code>’ character and the sequence identifier. The optional tag fields are between the sequence identifier and the first space in or at end of this line. Optional read 1 and read 2 label “<code>/[12]</code>” add after tags. Optional description words place at the end of this line, separated by a space.</p></li>
<li><p>The sequence, bases consist of A, C, G, T and N.</p></li>
<li><p>A plus character. Optional sequence identifier can also appended here.</p></li>
<li><p>The base call quality scores.</p></li>
</ul>
<p>The optional fields of FASTQ+ is consist of numbers of tags in TAG:TYPE:VALUE format. The tag format is inherited from the SAM tag format (<a href="https://samtools.github.io/hts-specs/SAMv1.pdf" class="uri">https://samtools.github.io/hts-specs/SAMv1.pdf</a>). The TAG is a two-character string that matches /[A-Za-z][A-Za-z0-9]/. TYPE is a single case-sensitive letter, which defines the format of VALUE. Tags in optional fields are separated by three ’’ characters. No space allowed in the sequence identifier and optional fields. The total length of the first line should not exceed 254 characters.</p>
</section>
</section>
<section id="an-example" class="level2">
<h2 class="anchored" data-anchor-id="an-example">An example</h2>
<p>In the example below, SEQ_ID is the sequence identifier that users can set or generate by software. CB is the recommended tag name for the corrected cell barcode, GN is the recommended tag for gene name, and UB is the recommended tag for UMI.</p>
<div class="framed">
<pre><code>@SEQ_ID|||CB:Z:ACGT|||GN:Z:BRCA1|||UB:Z:AACG
GATTTGGGGTTCAAAGCAGTATCGA
+
1***-+*''))**5&gt;&gt;CCCCCCC65</code></pre>
</div>
</section>
<section id="optional-tag-fields" class="level2">
<h2 class="anchored" data-anchor-id="optional-tag-fields">Optional tag fields</h2>
<p>All optional tags for FASTQ+ follow the SAM tag format but add two more restrictions.</p>
<ol type="1">
<li><p>SAM tags allow space in type Z, but not allowed in FASTQ+ tags.</p></li>
<li><p>The length of SAM tags is not limited, but FASTQ+ read identifier and tags are specified to be no longer than 254 characters.</p></li>
</ol>
<div class="center">
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 51%">
<col style="width: 39%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;"><strong>Type</strong></th>
<th style="text-align: left;"><strong>Regexp matching <code>VALUE</code></strong></th>
<th style="text-align: left;"><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">A</td>
<td style="text-align: left;"><code>[!-]</code></td>
<td style="text-align: left;">Printable character</td>
</tr>
<tr class="even">
<td style="text-align: center;">i</td>
<td style="text-align: left;"><code>[-+]?[0-9]+</code></td>
<td style="text-align: left;">Signed integer</td>
</tr>
<tr class="odd">
<td style="text-align: center;">f</td>
<td style="text-align: left;"><code>[-+]?[0-9]*.?[0-9]+([eE][-+]?[0-9]+)?</code></td>
<td style="text-align: left;">Signed single-precision floating number</td>
</tr>
<tr class="even">
<td style="text-align: center;">Z</td>
<td style="text-align: left;"><code>[!-]*</code></td>
<td style="text-align: left;">Printable string.</td>
</tr>
<tr class="odd">
<td style="text-align: center;">H</td>
<td style="text-align: left;"><code>([0-9A-F][0-9A-F])*</code></td>
<td style="text-align: left;">Byte array in the Hex format</td>
</tr>
<tr class="even">
<td style="text-align: center;">B</td>
<td style="text-align: left;"><code>[cCsSiIf](,[-+]?[0-9]*.?[0-9]+([eE][-+]?[0-9]+)?)*</code></td>
<td style="text-align: left;">Integer or numberic array</td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="recommended-tag-names-and-types-for-single-cell-experiments" class="level2">
<h2 class="anchored" data-anchor-id="recommended-tag-names-and-types-for-single-cell-experiments">Recommended tag names and types for single cell experiments</h2>
<p>The following tag names and types have been widely used in single-cell experiments. Although it is not required but highly recommended following the exact definition.</p>
<div class="center">
<table class="caption-top table">
<colgroup>
<col style="width: 13%">
<col style="width: 9%">
<col style="width: 76%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;"><strong>Tag name</strong></th>
<th style="text-align: center;"><strong>Type</strong></th>
<th style="text-align: left;"><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">CR</td>
<td style="text-align: center;"><code>Z</code></td>
<td style="text-align: left;">Raw cell barcode.</td>
</tr>
<tr class="even">
<td style="text-align: center;">CB</td>
<td style="text-align: center;"><code>Z</code></td>
<td style="text-align: left;">Cell barcode that is confirmed against a list of known-good barcode sequences.</td>
</tr>
<tr class="odd">
<td style="text-align: center;">UR</td>
<td style="text-align: center;"><code>Z</code></td>
<td style="text-align: left;">Raw molecular barcode. Usually be unique molecule indentifer (UMI).</td>
</tr>
<tr class="even">
<td style="text-align: center;">UB</td>
<td style="text-align: center;"><code>Z</code></td>
<td style="text-align: left;">Molecular barcode that is corrected among other molecule barcodes.</td>
</tr>
<tr class="odd">
<td style="text-align: center;">GN</td>
<td style="text-align: center;"><code>Z</code></td>
<td style="text-align: left;">Gene name.</td>
</tr>
<tr class="even">
<td style="text-align: center;">TX</td>
<td style="text-align: center;"><code>Z</code></td>
<td style="text-align: left;">Transcript id.</td>
</tr>
<tr class="odd">
<td style="text-align: center;">GX</td>
<td style="text-align: center;"><code>Z</code></td>
<td style="text-align: left;">Gene id.</td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="read-block" class="level2">
<h2 class="anchored" data-anchor-id="read-block">Read block</h2>
<p>A combination of tags can define the FASTQ+ block. Reads with the same tags can be selected and manipulated in a group. For example, SEQ1, SEQ2, and SEQ3 share the same cell barcode, and SEQ2 and SEQ3 share the same gene name but not for SEQ1. Using CB to define the block, all these three reads are from the same block. But if using CB and GN to determine the block, SEQ2 and SEQ3 are from the same block, different from SEQ1. Sorting the FASTQ+ file by tags can facilitate downstream analysis, i.e., assembly, for each block to be processed sequentially.</p>
<div class="framed">
<pre><code>@SEQ1|||CB:Z:ACGT|||GN:Z:BRCA1
GATTTGGGGTTCAAAGCAGTATCGA
+
1***-+*''))**5&gt;&gt;CCCCCCC65
@SEQ2|||CB:Z:ACGT|||GN:Z:SAA1
ACACTCGAAGATACAGAAATGAGTA
+
EEEEEE6/E/EEEAEEEE/E/EEE&lt;
@SEQ3|||CB:Z:ACGT|||GN:Z:SAA1
TATCGACAGGAAGAAGGAGGGAGGG
+
AE/EE&lt;/AEEEAEEEEE//AA/EEE</code></pre>
</div>
</section>
<section id="fastq-version-history" class="level1">
<h1>FASTQ+ Version History</h1>
<section id="v1.0-apr-2022" class="level2">
<h2 class="anchored" data-anchor-id="v1.0-apr-2022">v1.0 : Apr 2022</h2>
<p>Initial edition.</p>
<p>This website outlines the key tools and software packages developed from my research:</p>
<!-- -->
<ul>
<li><strong>FASTQ+</strong>: This format enhances the standard FASTQ file by integrating barcode information directly into the read name. This modification facilitates the harmonization of single-cell sequencing data across various platforms. FASTQ+ can be generated with PISA.</li>
<li><strong>PISA</strong>: A suite of programs specifically designed for preprocessing data from single-cell sequencing and spatial transcriptomics. It efficiently processes FASTQ+, BAM, and BED files, and generates various feature counts and annotations.</li>
<li><strong>Yano</strong>: An R package that employs the spatial dissimilarity test method as its core algorithm. Yano is tailored for exploring various biological events in single-cell and spatial transcriptomics data, including alternative splicing, allele-specific gene expression, somatic variants and more, providing comprehensive insights into cellular dynamics. <!--
**dST**: An R package designed for zonation analysis and rankable cell-cell interaction analysis. This package is currently under development.
--></li>
</ul>
<p><strong>Workflows and Short Cases for scRNA-seq and Spatial Transcriptome</strong></p>
<ul>
<li><a href="workflow1.Rmd">From raw reads to gene counts</a> (PISA)</li>
<li><a href="anno.Rmd">Annotate various features for alignment</a> (PISA)</li>
<li><a href="Yano_AS.Rmd">Alternative splicing analysis for scRNA-seq</a> (Yano)</li>
<li><a href="Yano_ASE.Rmd">Allele-specific gene expression analysis for scRNA-seq</a> (Yano)</li>
<li><a href="Yano_anno.Rmd">Annotating and prioritizing genetic variants for scRNA-seq</a> (Yano)</li>
<li><a href="Yano_trajectory.Rmd">Perform alternative splicing analysis for cell trajectory and user-defined embeddings</a> (Yano)</li>
<li><a href="Visium.Rmd">Perform alternative splicing analysis for multiple Visium samples</a> (Yano)</li>
<li><a href="dST_selector.Rmd">Select cells from reduction maps and spatial locations</a> (Yano)</li>
</ul>
</section>
<section id="synopsis" class="level2">
<h2 class="anchored" data-anchor-id="synopsis">Synopsis</h2>
<div class="sourceCode" id="cb37"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> <span class="pp">[</span><span class="ss">tool</span><span class="pp">]</span> <span class="pp">[</span><span class="ss">options</span><span class="pp">]</span> <span class="pp">[</span><span class="ss">input</span><span class="pp">-</span><span class="ss">file</span><span class="pp">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="install" class="level2">
<h2 class="anchored" data-anchor-id="install">Install</h2>
<p>PISA source code can be downloaded at <a href="https://github.com/shiquan/PISA/releases" class="uri">https://github.com/shiquan/PISA/releases</a>, or the development version from <a href="https://github.com/shiquan/PISA/" class="uri">https://github.com/shiquan/PISA/</a>. To compile PISA from sources run <strong>make</strong> in the source directory.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> git clone https://github.com/shiquan/PISA</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd PISA</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> make</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="get-started" class="level2">
<h2 class="anchored" data-anchor-id="get-started">Get Started</h2>
<p>The code snippet below demonstrates how to use the PISA tools to process test data and generate various feature counts. You can find the test data in the PISA/demo directory. This example provides a practical approach to familiarize yourself with the functionality of PISA and to validate its operations with provided sample data.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>$ cd demo</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>$ ls</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>aln.sam.gz  barcodes.txt  demo_1.fq.gz  demo_2.fq.gz  demo.gtf.gz  peaks.bed  README.md  var.vcf.gz</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>$ zcat demo_1.fq.gz|head -n 4</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>@A00984:220:HNJ7KDRXX:1:1118:2510:4586</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>AAGCATCCACACAGAGCACCCCGTTCTT</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>+</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>FFFFFFFFFFFFFFFFFFFFFFFFFFFF</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>$ zcat demo_2.fq.gz|head -n 4</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>@A00984:220:HNJ7KDRXX:1:1118:2510:4586</span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC</span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>+</span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a>FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF</span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a><span class="fu"># Convert raw FASTQ to FASTQ+ format</span></span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a>$ PISA parse -rule 'CR,R1:1-16,barcodes.txt,CB,1;UR,R1:17-28;R1,R2' demo_1.fq.gz demo_2.fq.gz -1 demo.fq</span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a>Number of Fragments,825</span>
<span id="cb39-21"><a href="#cb39-21" aria-hidden="true" tabindex="-1"></a>Fragments pass QC,825</span>
<span id="cb39-22"><a href="#cb39-22" aria-hidden="true" tabindex="-1"></a>Fragments with Exactly Matched Barcodes,805</span>
<span id="cb39-23"><a href="#cb39-23" aria-hidden="true" tabindex="-1"></a>Fragments with Failed Barcodes,0</span>
<span id="cb39-24"><a href="#cb39-24" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:21:35</span><span class="co">]</span> Real time: 0.003 sec; CPU: 0.009 sec; Peak RSS: 0.010 GB.</span>
<span id="cb39-25"><a href="#cb39-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-26"><a href="#cb39-26" aria-hidden="true" tabindex="-1"></a>$ head -n 4 demo.fq</span>
<span id="cb39-27"><a href="#cb39-27" aria-hidden="true" tabindex="-1"></a>@A00984:220:HNJ7KDRXX:1:1118:2510:4586|||CR:Z:AAGCATCCACACAGAG|||CB:Z:AAGCATCCACACAGAG|||UR:Z:CACCCCGTTCTT</span>
<span id="cb39-28"><a href="#cb39-28" aria-hidden="true" tabindex="-1"></a>GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC</span>
<span id="cb39-29"><a href="#cb39-29" aria-hidden="true" tabindex="-1"></a>+</span>
<span id="cb39-30"><a href="#cb39-30" aria-hidden="true" tabindex="-1"></a>FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF</span>
<span id="cb39-31"><a href="#cb39-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-32"><a href="#cb39-32" aria-hidden="true" tabindex="-1"></a><span class="fu"># Alignment results of FASTQ+ </span></span>
<span id="cb39-33"><a href="#cb39-33" aria-hidden="true" tabindex="-1"></a>$ samtools view aln.sam.gz|head -n 1</span>
<span id="cb39-34"><a href="#cb39-34" aria-hidden="true" tabindex="-1"></a>A00984:220:HNJ7KDRXX:1:1118:2510:4586|||CR:Z:AAGCATCCACACAGAG|||CB:Z:AAGCATCCACACAGAG|||UR:Z:CACCCCGTTCTT   0   chr11   35139165    255 26S65M  *   0   0   GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF NH:i:1  HI:i:1  AS:i:61 nM:i:1</span>
<span id="cb39-35"><a href="#cb39-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-36"><a href="#cb39-36" aria-hidden="true" tabindex="-1"></a><span class="fu"># Convert format alignment records from SAM to BAM</span></span>
<span id="cb39-37"><a href="#cb39-37" aria-hidden="true" tabindex="-1"></a>$ PISA sam2bam aln.sam.gz -o aln.bam</span>
<span id="cb39-38"><a href="#cb39-38" aria-hidden="true" tabindex="-1"></a>Raw reads,825</span>
<span id="cb39-39"><a href="#cb39-39" aria-hidden="true" tabindex="-1"></a>Mapped reads,820 (99.39%)</span>
<span id="cb39-40"><a href="#cb39-40" aria-hidden="true" tabindex="-1"></a>Plus strand,820</span>
<span id="cb39-41"><a href="#cb39-41" aria-hidden="true" tabindex="-1"></a>Minus strand,0</span>
<span id="cb39-42"><a href="#cb39-42" aria-hidden="true" tabindex="-1"></a>Mitochondria ratio,0.00%</span>
<span id="cb39-43"><a href="#cb39-43" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:26:31</span><span class="co">]</span> Real time: 0.005 sec; CPU: 0.009 sec; Peak RSS: 0.010 GB.</span>
<span id="cb39-44"><a href="#cb39-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-45"><a href="#cb39-45" aria-hidden="true" tabindex="-1"></a>$ samtools view aln.bam|head -n 1</span>
<span id="cb39-46"><a href="#cb39-46" aria-hidden="true" tabindex="-1"></a>A00984:220:HNJ7KDRXX:1:1118:2510:4586   0   chr11   35139165    255 26S65M  *   0   0   GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF NH:i:1  HI:i:1  AS:i:61 nM:i:1  CR:Z:AAGCATCCACACAGAG   CB:Z:AAGCATCCACACAGAG   UR:Z:CACCCCGTTCTT</span>
<span id="cb39-47"><a href="#cb39-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-48"><a href="#cb39-48" aria-hidden="true" tabindex="-1"></a><span class="fu"># Annotate gene names for BAM</span></span>
<span id="cb39-49"><a href="#cb39-49" aria-hidden="true" tabindex="-1"></a>$ PISA anno -gtf ./demo.gtf.gz aln.bam -o anno_gtf.bam</span>
<span id="cb39-50"><a href="#cb39-50" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:28:38</span><span class="co">]</span> GTF loading..</span>
<span id="cb39-51"><a href="#cb39-51" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:28:38</span><span class="co">]</span> Load 2 genes.</span>
<span id="cb39-52"><a href="#cb39-52" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:28:38</span><span class="co">]</span> Load time : 0.003 sec</span>
<span id="cb39-53"><a href="#cb39-53" aria-hidden="true" tabindex="-1"></a>Reads Mapped to Genome (Map Quality &gt;= 0),99.4%</span>
<span id="cb39-54"><a href="#cb39-54" aria-hidden="true" tabindex="-1"></a>Reads Mapped to Exonic Regions,99.3%</span>
<span id="cb39-55"><a href="#cb39-55" aria-hidden="true" tabindex="-1"></a>Reads Mapped to Intronic Regions,0.0%</span>
<span id="cb39-56"><a href="#cb39-56" aria-hidden="true" tabindex="-1"></a>Reads Mapped to both Exonic and Intronic Regions,0.7%</span>
<span id="cb39-57"><a href="#cb39-57" aria-hidden="true" tabindex="-1"></a>Reads Mapped Antisense to Gene,0.0%</span>
<span id="cb39-58"><a href="#cb39-58" aria-hidden="true" tabindex="-1"></a>Reads Mapped to Intergenic Regions,0.0%</span>
<span id="cb39-59"><a href="#cb39-59" aria-hidden="true" tabindex="-1"></a>Reads Mapped to Gene but Failed to Interpret Type,0.0%</span>
<span id="cb39-60"><a href="#cb39-60" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:28:38</span><span class="co">]</span> Real time: 0.026 sec; CPU: 0.086 sec; Speed : 9528 records/sec; Peak RSS: 0.034 GB.</span>
<span id="cb39-61"><a href="#cb39-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-62"><a href="#cb39-62" aria-hidden="true" tabindex="-1"></a><span class="fu"># Correct UMIs amongst other UMIs from the same cell and mapped to the same gene, and create new tag UB for corrected UMIs</span></span>
<span id="cb39-63"><a href="#cb39-63" aria-hidden="true" tabindex="-1"></a>$ PISA corr -tag UR -new-tag UB -tags-block CB,GN anno_gtf.bam -o corr.bam</span>
<span id="cb39-64"><a href="#cb39-64" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:36:21</span><span class="co">]</span> Building index ..</span>
<span id="cb39-65"><a href="#cb39-65" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:36:21</span><span class="co">]</span> Build time : 0.002 sec</span>
<span id="cb39-66"><a href="#cb39-66" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:36:21</span><span class="co">]</span> Real time: 0.077 sec; CPU: 0.085 sec</span>
<span id="cb39-67"><a href="#cb39-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-68"><a href="#cb39-68" aria-hidden="true" tabindex="-1"></a>$ samtools view corr.bam|head -n 1</span>
<span id="cb39-69"><a href="#cb39-69" aria-hidden="true" tabindex="-1"></a>A00984:220:HNJ7KDRXX:1:1118:2510:4586   0   chr11   35139165    255 26S65M  *   0   0   GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF NH:i:1  HI:i:1  AS:i:61 nM:i:1  CR:Z:AAGCATCCACACAGAG   CB:Z:AAGCATCCACACAGAG   UR:Z:CACCCCGTTCTT   RE:A:E  GX:Z:ENSG00000026508.18 GN:Z:CD44   TX:Z:ENST00000263398.10,ENST00000428726.7,ENST00000526025.2 UB:Z:CACCCCGTTCTT</span>
<span id="cb39-70"><a href="#cb39-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-71"><a href="#cb39-71" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count gene X cell features</span></span>
<span id="cb39-72"><a href="#cb39-72" aria-hidden="true" tabindex="-1"></a>$ mkdir exp</span>
<span id="cb39-73"><a href="#cb39-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-74"><a href="#cb39-74" aria-hidden="true" tabindex="-1"></a>$ PISA count  -cb CB -anno-tag GN -outdir exp -umi UB corr.bam </span>
<span id="cb39-75"><a href="#cb39-75" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:38:44</span><span class="co">]</span> Real time: 0.033 sec; CPU: 0.013 sec; Peak RSS: 0.010 GB.</span>
<span id="cb39-76"><a href="#cb39-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-77"><a href="#cb39-77" aria-hidden="true" tabindex="-1"></a><span class="fu"># Gene expression matrix generated in the Market Exchange format </span></span>
<span id="cb39-78"><a href="#cb39-78" aria-hidden="true" tabindex="-1"></a>$ ls exp/</span>
<span id="cb39-79"><a href="#cb39-79" aria-hidden="true" tabindex="-1"></a>barcodes.tsv.gz  features.tsv.gz  matrix.mtx.gz</span>
<span id="cb39-80"><a href="#cb39-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-81"><a href="#cb39-81" aria-hidden="true" tabindex="-1"></a><span class="fu"># Not just gene features, we can also annotate variants and functional regions to reads</span></span>
<span id="cb39-82"><a href="#cb39-82" aria-hidden="true" tabindex="-1"></a>$ PISA anno -bed peaks.bed -tag PK -vcf var.vcf.gz -vtag VR  corr.bam -o anno_vcf_bed.bam    </span>
<span id="cb39-83"><a href="#cb39-83" aria-hidden="true" tabindex="-1"></a>Reads Mapped to Genome (Map Quality &gt;= 0),99.4%</span>
<span id="cb39-84"><a href="#cb39-84" aria-hidden="true" tabindex="-1"></a>Reads Mapped to BED regions / Peaks,0.0%</span>
<span id="cb39-85"><a href="#cb39-85" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:43:01</span><span class="co">]</span> Real time: 0.027 sec; CPU: 0.090 sec; Speed : 9085 records/sec; Peak RSS: 0.034 GB.</span>
<span id="cb39-86"><a href="#cb39-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-87"><a href="#cb39-87" aria-hidden="true" tabindex="-1"></a>$ samtools view anno_vcf_bed.bam|grep "VR:Z"|grep "PK:Z"|head -n 1</span>
<span id="cb39-88"><a href="#cb39-88" aria-hidden="true" tabindex="-1"></a>A00984:220:HNJ7KDRXX:1:2266:27597:30843 0   chr11   35229688    255 91M *   0   0   CCCAGGGTTAATAGGGCCTGGTCCCTGGGAGGAAATTTGAATGGGTCCATTTTGCCCTTCCATAGCCTAATCCCGGGGCATTGTTTTCCAC</span>
<span id="cb39-89"><a href="#cb39-89" aria-hidden="true" tabindex="-1"></a>FFFF,FFFFFFFFFF,FFFFFFF:,:FFFFFF,FF:FFFFFFFFFFF:,,:F:F:F,F:F:F,FFFF,F:FFFF,FF,FF:FFFFFFF:F: NH:i:1  HI:i:1  AS:i:85 nM:i:2  CR:Z:ATTGTTCCAAGTCCCG   CB:Z:ATTGTTCCAAGTCCCG   UR:Z:TCTTTAAGTCAG   RE:A:E  GX:Z:ENSG00000026508.18 GN:Z:CD44   TX:Z:ENST00000263398.10,ENST00000428726.7,ENST00000425428.6,ENST00000433892.6,ENST00000525469.1 UB:Z:TCTTTAAGTCAG   PK:Z:demo_peak_14a;demo_peak_14b    VR:Z:chr11:35229771C&gt;T</span>
<span id="cb39-90"><a href="#cb39-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-91"><a href="#cb39-91" aria-hidden="true" tabindex="-1"></a><span class="fu"># Summarize the reads, UMIs, genes, peaks, and variants per cell</span></span>
<span id="cb39-92"><a href="#cb39-92" aria-hidden="true" tabindex="-1"></a>$ PISA attrcnt -cb CB -tags UB,GN,PK,VR anno_vcf_bed.bam -dedup |head -n 5</span>
<span id="cb39-93"><a href="#cb39-93" aria-hidden="true" tabindex="-1"></a>BARCODE Raw UB  GN  PK  VR</span>
<span id="cb39-94"><a href="#cb39-94" aria-hidden="true" tabindex="-1"></a>AAGCATCCACACAGAG    503 132 1   15  1</span>
<span id="cb39-95"><a href="#cb39-95" aria-hidden="true" tabindex="-1"></a>ATTGTTCCAAGTCCCG    533 124 1   13  1</span>
<span id="cb39-96"><a href="#cb39-96" aria-hidden="true" tabindex="-1"></a>AAGCATCCACACNGAG    3   1   1   1   0</span>
<span id="cb39-97"><a href="#cb39-97" aria-hidden="true" tabindex="-1"></a>AAGCNTCCACACAGAG    3   1   1   1   0</span>
<span id="cb39-98"><a href="#cb39-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-99"><a href="#cb39-99" aria-hidden="true" tabindex="-1"></a><span class="fu"># Deduplicate BAM file for each cell</span></span>
<span id="cb39-100"><a href="#cb39-100" aria-hidden="true" tabindex="-1"></a>$ PISA rmdup -tags CB corr.bam -o rmdup.bam -nw </span>
<span id="cb39-101"><a href="#cb39-101" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:59:39</span><span class="co">]</span> Deduplicating chr11</span>
<span id="cb39-102"><a href="#cb39-102" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:59:39</span><span class="co">]</span> All reads,820</span>
<span id="cb39-103"><a href="#cb39-103" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:59:39</span><span class="co">]</span> Duplicate reads,125</span>
<span id="cb39-104"><a href="#cb39-104" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:59:39</span><span class="co">]</span> Duplicate ratio,0.1524</span>
<span id="cb39-105"><a href="#cb39-105" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:59:39</span><span class="co">]</span> Real time: 0.008 sec; CPU: 0.015 sec; Peak RSS: 0.010 GB.</span>
<span id="cb39-106"><a href="#cb39-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-107"><a href="#cb39-107" aria-hidden="true" tabindex="-1"></a><span class="fu"># Deduplicate BAM file for each molecular</span></span>
<span id="cb39-108"><a href="#cb39-108" aria-hidden="true" tabindex="-1"></a>$ PISA rmdup -tags CB,UR corr.bam -o rmdup1.bam -nw</span>
<span id="cb39-109"><a href="#cb39-109" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:00:35</span><span class="co">]</span> Deduplicating chr11</span>
<span id="cb39-110"><a href="#cb39-110" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:00:35</span><span class="co">]</span> All reads,820</span>
<span id="cb39-111"><a href="#cb39-111" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:00:35</span><span class="co">]</span> Duplicate reads,0</span>
<span id="cb39-112"><a href="#cb39-112" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:00:35</span><span class="co">]</span> Duplicate ratio,0.0000</span>
<span id="cb39-113"><a href="#cb39-113" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:00:35</span><span class="co">]</span> Real time: 0.009 sec; CPU: 0.015 sec; Peak RSS: 0.011 GB.</span>
<span id="cb39-114"><a href="#cb39-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-115"><a href="#cb39-115" aria-hidden="true" tabindex="-1"></a><span class="fu"># Select all reads annotated to gene CD44</span></span>
<span id="cb39-116"><a href="#cb39-116" aria-hidden="true" tabindex="-1"></a><span class="fu"># Generate a gene candidate list</span></span>
<span id="cb39-117"><a href="#cb39-117" aria-hidden="true" tabindex="-1"></a>$  echo "CD44" &gt; gene.txt</span>
<span id="cb39-118"><a href="#cb39-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-119"><a href="#cb39-119" aria-hidden="true" tabindex="-1"></a>$ PISA pick -tags GN -list gene.txt anno_vcf_bed.bam  -o picked.bam</span>
<span id="cb39-120"><a href="#cb39-120" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:03:01</span><span class="co">]</span> Real time: 0.009 sec; CPU: 0.016 sec</span>
<span id="cb39-121"><a href="#cb39-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-122"><a href="#cb39-122" aria-hidden="true" tabindex="-1"></a><span class="fu"># Select reads with more features</span></span>
<span id="cb39-123"><a href="#cb39-123" aria-hidden="true" tabindex="-1"></a>$ awk '{printf("%s\tCD44\n", $1)}' barcodes.txt &gt; candidates.txt</span>
<span id="cb39-124"><a href="#cb39-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-125"><a href="#cb39-125" aria-hidden="true" tabindex="-1"></a>$ cat candidates.txt </span>
<span id="cb39-126"><a href="#cb39-126" aria-hidden="true" tabindex="-1"></a>AAGCATCCACACAGAG    CD44</span>
<span id="cb39-127"><a href="#cb39-127" aria-hidden="true" tabindex="-1"></a>ATTGTTCCAAGTCCCG    CD44</span>
<span id="cb39-128"><a href="#cb39-128" aria-hidden="true" tabindex="-1"></a>GCACATAGTCAGTTTG    CD44</span>
<span id="cb39-129"><a href="#cb39-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-130"><a href="#cb39-130" aria-hidden="true" tabindex="-1"></a>$ PISA pick -tags CB,GN -list candidates.txt anno_vcf_bed.bam -o picked.bam </span>
<span id="cb39-131"><a href="#cb39-131" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:09:28</span><span class="co">]</span> Real time: 0.007 sec; CPU: 0.013 sec</span>
<span id="cb39-132"><a href="#cb39-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-133"><a href="#cb39-133" aria-hidden="true" tabindex="-1"></a><span class="fu"># Convert BAM to FASTQ+</span></span>
<span id="cb39-134"><a href="#cb39-134" aria-hidden="true" tabindex="-1"></a>$ PISA bam2fq -tags CB,GN picked.bam -o pick.fq</span>
<span id="cb39-135"><a href="#cb39-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-136"><a href="#cb39-136" aria-hidden="true" tabindex="-1"></a>$ head -n 4 pick.fq </span>
<span id="cb39-137"><a href="#cb39-137" aria-hidden="true" tabindex="-1"></a>@A00984:220:HNJ7KDRXX:1:1118:2510:4586|||CB:Z:AAGCATCCACACAGAG|||GN:Z:CD44</span>
<span id="cb39-138"><a href="#cb39-138" aria-hidden="true" tabindex="-1"></a>GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC</span>
<span id="cb39-139"><a href="#cb39-139" aria-hidden="true" tabindex="-1"></a>+</span>
<span id="cb39-140"><a href="#cb39-140" aria-hidden="true" tabindex="-1"></a>FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF</span>
<span id="cb39-141"><a href="#cb39-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-142"><a href="#cb39-142" aria-hidden="true" tabindex="-1"></a><span class="fu"># Sort FASTQ+ reads by CB and GN</span></span>
<span id="cb39-143"><a href="#cb39-143" aria-hidden="true" tabindex="-1"></a>$ PISA fsort -tags CB,GN pick.fq -o fsort.fq.gz</span>
<span id="cb39-144"><a href="#cb39-144" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:22:50</span><span class="co">]</span> Write 795 records to fsort.fq.gz.0000.bgz.</span>
<span id="cb39-145"><a href="#cb39-145" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:22:50</span><span class="co">]</span> Unlink fsort.fq.gz.0000.bgz</span>
<span id="cb39-146"><a href="#cb39-146" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:22:50</span><span class="co">]</span> Create fsort.fq.gz from 1 files.</span>
<span id="cb39-147"><a href="#cb39-147" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:22:50</span><span class="co">]</span> Real time: 0.021 sec; CPU: 0.020 sec</span>
<span id="cb39-148"><a href="#cb39-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-149"><a href="#cb39-149" aria-hidden="true" tabindex="-1"></a>$ zcat fsort.fq.gz|head -n 8</span>
<span id="cb39-150"><a href="#cb39-150" aria-hidden="true" tabindex="-1"></a>@A00984:220:HNJ7KDRXX:1:1118:2510:4586|||CB:Z:AAGCATCCACACAGAG|||GN:Z:CD44</span>
<span id="cb39-151"><a href="#cb39-151" aria-hidden="true" tabindex="-1"></a>GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC</span>
<span id="cb39-152"><a href="#cb39-152" aria-hidden="true" tabindex="-1"></a>+</span>
<span id="cb39-153"><a href="#cb39-153" aria-hidden="true" tabindex="-1"></a>FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF</span>
<span id="cb39-154"><a href="#cb39-154" aria-hidden="true" tabindex="-1"></a>@A00984:220:HNJ7KDRXX:1:2143:21640:21496|||CB:Z:AAGCATCCACACAGAG|||GN:Z:CD44</span>
<span id="cb39-155"><a href="#cb39-155" aria-hidden="true" tabindex="-1"></a>CCTGCCCCGCGCCCAGAGATCCTCCAGCTCCTTTCGCCCGCGCCCTACGTTCGCTCCGGACACCATGGACAAGTATTGGTGGAACACAGCC</span>
<span id="cb39-156"><a href="#cb39-156" aria-hidden="true" tabindex="-1"></a>+</span>
<span id="cb39-157"><a href="#cb39-157" aria-hidden="true" tabindex="-1"></a>,,FFFFFFFF,F,,:F,F,F:FFF,FFFFFF,F::F,FF,F:FF:,,FF:FFF,:FFFF:FFF,::FFFF,F:F,FFFF,,F,FFF,F,::</span>
<span id="cb39-158"><a href="#cb39-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-159"><a href="#cb39-159" aria-hidden="true" tabindex="-1"></a><span class="fu"># Assembly reads mapped to CD44 of the same cell into contigs</span></span>
<span id="cb39-160"><a href="#cb39-160" aria-hidden="true" tabindex="-1"></a><span class="fu"># This step requires Trinity software and seqtk already installed in your environment</span></span>
<span id="cb39-161"><a href="#cb39-161" aria-hidden="true" tabindex="-1"></a>$ PISA stream -tags CB,GN -script 'Trinity --seqType fq --SS_lib_type F --single ${FQ} --max_memory 1G 2&gt;/dev/null 1&gt;/dev/null; seqtk rename trinity_out_dir.Trinity.fasta ${UBI}_ 2&gt;/dev/null' -t 10 -fa  -nw ./fsort.fq.gz -o assem.fa</span>
<span id="cb39-162"><a href="#cb39-162" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:24:21</span><span class="co">]</span> Real time: 5.607 sec; CPU: 0.010 sec; Peak RSS: 0.010 GB.</span>
<span id="cb39-163"><a href="#cb39-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-164"><a href="#cb39-164" aria-hidden="true" tabindex="-1"></a>$ seqtk seq assem.fa -l 100 |head </span>
<span id="cb39-165"><a href="#cb39-165" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;Z_AAGCATCCACACAGAG_Z_CD44_1|||CB:Z:AAGCATCCACACAGAG|||GN:Z:CD44 len=439 path=</span><span class="co">[</span><span class="ot">0:0-438</span><span class="co">]</span></span>
<span id="cb39-166"><a href="#cb39-166" aria-hidden="true" tabindex="-1"></a><span class="at">GAAATTAGGGCCCAATTAATAATCAGCAAGAATTTGATCGTTCCAGTTCCCACTTGGAGGCCTTTCATCCCTCGGGTGTGCTATGGATGGCTTCTAACAA</span></span>
<span id="cb39-167"><a href="#cb39-167" aria-hidden="true" tabindex="-1"></a><span class="at">AAACTACACATATGTATTCCTGATCGCCAACCTTTCCCCCACCAGCTAAGGACATTTCCCAGGGTTAATAGGGCCTGGTCCCTGGGAGGAAATTTGAATG</span></span>
<span id="cb39-168"><a href="#cb39-168" aria-hidden="true" tabindex="-1"></a><span class="at">GGTCCATTTTGCCCTTCCATAGCCTAATCCCTGGGCATTGTTTTCCACTGAGGTTGGGGGTTGGGGTGTACTAGTTACACATCTTCAACAGACCCCCTCT</span></span>
<span id="cb39-169"><a href="#cb39-169" aria-hidden="true" tabindex="-1"></a><span class="at">AGAAATTTTTCAGATGCTTCTGGGAGACACCCAAAGGGTGAAGCTATTTATCTGTAGTAAACTATTTATCTGTGTTTTTGAAATATTAAACCCTGGATCA</span></span>
<span id="cb39-170"><a href="#cb39-170" aria-hidden="true" tabindex="-1"></a><span class="at">GTCCTTTGATCAGTATAATTTTTTAAAGTTACTTTGTCA</span></span>
<span id="cb39-171"><a href="#cb39-171" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;Z_AAGCATCCACACAGAG_Z_CD44_2|||CB:Z:AAGCATCCACACAGAG|||GN:Z:CD44 len=384 path=</span><span class="co">[</span><span class="ot">0:0-383</span><span class="co">]</span></span>
<span id="cb39-172"><a href="#cb39-172" aria-hidden="true" tabindex="-1"></a><span class="at">CCTGGTAGAATTGGCTTTTCTAGCAGAACCTTTCCAAAAGTTTTATATTGAGATTCATAACAACACCAAGAATTGATTTTGTAGCCAACATTCATTCAAT</span></span>
<span id="cb39-173"><a href="#cb39-173" aria-hidden="true" tabindex="-1"></a><span class="at">ACTGTTATATCAGAGGAGTAGGAGAGAGGAAACATTTGACTTATCTGGAAAAGCAAAATGTACTTAAGAATAAGAATAACATGGTCCATTCACCTTTATG</span></span>
<span id="cb39-174"><a href="#cb39-174" aria-hidden="true" tabindex="-1"></a><span class="at">TTATAGATATGTCTTTGTGTAAATCATTTGTTTTGAGTTTTCAAAGAATAGCCCATTGTTCATTCTTGTGCTGTACAATGACCACTGTTATTGTTACTTT</span></span>
<span id="cb39-175"><a href="#cb39-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-176"><a href="#cb39-176" aria-hidden="true" tabindex="-1"></a><span class="fu"># Align assembled reads to reference and convert to BAM file</span></span>
<span id="cb39-177"><a href="#cb39-177" aria-hidden="true" tabindex="-1"></a><span class="fu"># Here I use minimap2 for simplicity</span></span>
<span id="cb39-178"><a href="#cb39-178" aria-hidden="true" tabindex="-1"></a>$ minimap2 -x splice -a ~/Documents/datasets/GRCh38/fasta/genome.fasta assem.fa 1&gt; asm_aln.sam</span>
<span id="cb39-179"><a href="#cb39-179" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::mm_idx_gen::50.495*1.81</span><span class="co">]</span> collected minimizers</span>
<span id="cb39-180"><a href="#cb39-180" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::mm_idx_gen::71.980*2.16</span><span class="co">]</span> sorted minimizers</span>
<span id="cb39-181"><a href="#cb39-181" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::main::71.980*2.16</span><span class="co">]</span> loaded/built the index for 194 target sequence(s)</span>
<span id="cb39-182"><a href="#cb39-182" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::mm_mapopt_update::75.057*2.11</span><span class="co">]</span> mid_occ = 767</span>
<span id="cb39-183"><a href="#cb39-183" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::mm_idx_stat</span><span class="co">]</span> kmer size: 15; skip: 5; is_hpc: 0; #seq: 194</span>
<span id="cb39-184"><a href="#cb39-184" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::mm_idx_stat::77.004*2.09</span><span class="co">]</span> distinct minimizers: 167225302 (35.42% are singletons); average occurrences: 6.036; average spacing: 3.071; total length: 3099750718</span>
<span id="cb39-185"><a href="#cb39-185" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::worker_pipeline::77.010*2.09</span><span class="co">]</span> mapped 13 sequences</span>
<span id="cb39-186"><a href="#cb39-186" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::main</span><span class="co">]</span> Version: 2.21-r1071</span>
<span id="cb39-187"><a href="#cb39-187" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::main</span><span class="co">]</span> CMD: minimap2 -x splice -a /home/shiquan/Documents/datasets/GRCh38/fasta/genome.fasta assem.fa</span>
<span id="cb39-188"><a href="#cb39-188" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::main</span><span class="co">]</span> Real time: 77.358 sec; CPU: 161.000 sec; Peak RSS: 18.519 GB</span>
<span id="cb39-189"><a href="#cb39-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-190"><a href="#cb39-190" aria-hidden="true" tabindex="-1"></a>$ PISA sam2bam asm_aln.sam -o asm_aln.bam</span>
<span id="cb39-191"><a href="#cb39-191" aria-hidden="true" tabindex="-1"></a>Raw reads,13</span>
<span id="cb39-192"><a href="#cb39-192" aria-hidden="true" tabindex="-1"></a>Mapped reads,13 (100.00%)</span>
<span id="cb39-193"><a href="#cb39-193" aria-hidden="true" tabindex="-1"></a>Plus strand,13</span>
<span id="cb39-194"><a href="#cb39-194" aria-hidden="true" tabindex="-1"></a>Minus strand,0</span>
<span id="cb39-195"><a href="#cb39-195" aria-hidden="true" tabindex="-1"></a>Mitochondria ratio,0.00%</span>
<span id="cb39-196"><a href="#cb39-196" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:30:20</span><span class="co">]</span> Real time: 0.001 sec; CPU: 0.006 sec; Peak RSS: 0.010 GB.</span>
<span id="cb39-197"><a href="#cb39-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-198"><a href="#cb39-198" aria-hidden="true" tabindex="-1"></a>$ samtools view asm_aln.bam|head -n 1</span>
<span id="cb39-199"><a href="#cb39-199" aria-hidden="true" tabindex="-1"></a>Z_AAGCATCCACACAGAG_Z_CD44_1 0   chr11   35229531    60  439M    *   0   0   GAAATTAGGGCCCAATTAATAATCAGCAAGAATTTGATCGTTCCAGTTCCCACTTGGAGGCCTTTCATCCCTCGGGTGTGCTATGGATGGCTTCTAACAA</span>
<span id="cb39-200"><a href="#cb39-200" aria-hidden="true" tabindex="-1"></a>AAACTACACATATGTATTCCTGATCGCCAACCTTTCCCCCACCAGCTAAGGACATTTCCCAGGGTTAATAGGGCCTGGTCCCTGGGAGGAAATTTGAATG</span>
<span id="cb39-201"><a href="#cb39-201" aria-hidden="true" tabindex="-1"></a>GGTCCATTTTGCCCTTCCATAGCCTAATCCCTGGGCATTGTTTTCCACTGAGGTTGGGGGTTGGGGTGTACTAGTTACACATCTTCAACAGACCCCCTCT</span>
<span id="cb39-202"><a href="#cb39-202" aria-hidden="true" tabindex="-1"></a>AGAAATTTTTCAGATGCTTCTGGGAGACACCCAAAGGGTGAAGCTATTTATCTGTAGTAAACTATTTATCTGTGTTTTTGAAATATTAAACCCTGGATCA</span>
<span id="cb39-203"><a href="#cb39-203" aria-hidden="true" tabindex="-1"></a>GTCCTTTGATCAGTATAATTTTTTAAAGTTACTTTGTCA *   NM:i:1  ms:i:436    AS:i:436    nn:i:0  tp:A:P  cm:i:137    s1:i:430    s2:i:0  de:f:0.0023 rl:i:0  CB:Z:AAGCATCCACACAGAG   GN:Z:CD44</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="list-of-commands" class="level2">
<h2 class="anchored" data-anchor-id="list-of-commands">List of commands</h2>
<table class="table-striped table-hover caption-top table">
<colgroup>
<col style="width: 23%">
<col style="width: 76%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Tool</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"></td>
<td style="text-align: left;"><strong><em>The following tools are used to process FASTQ/FASTQ+ files.</em></strong></td>
</tr>
<tr class="even">
<td style="text-align: left;">parse</td>
<td style="text-align: left;">Parse barcodes from FASTQ reads to FASTQ+.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">fsort</td>
<td style="text-align: left;">Sort FASTQ+ records by barcodes.</td>
</tr>
<tr class="even">
<td style="text-align: left;">stream</td>
<td style="text-align: left;">Perform user-defined process for each read block.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">addtags</td>
<td style="text-align: left;">Add tag string to FASTQ reads.</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;"><strong><em>The following tools are used to process BAM files.</em></strong></td>
</tr>
<tr class="odd">
<td style="text-align: left;">sam2bam</td>
<td style="text-align: left;">Parse FASTQ+ read name and convert SAM to BAM.</td>
</tr>
<tr class="even">
<td style="text-align: left;">rmdup</td>
<td style="text-align: left;">Remove PCR duplicates per molecular.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">pick</td>
<td style="text-align: left;">Pick alignments with tags.</td>
</tr>
<tr class="even">
<td style="text-align: left;">anno</td>
<td style="text-align: left;">Annotate functional regions or gene names.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">corr</td>
<td style="text-align: left;">Correct error prone UMIs. 1 mismatch considered.</td>
</tr>
<tr class="even">
<td style="text-align: left;">attrcnt</td>
<td style="text-align: left;">Count raw reads and tag values per cell.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">extract</td>
<td style="text-align: left;">Extract tag value from BAM.</td>
</tr>
<tr class="even">
<td style="text-align: left;">count</td>
<td style="text-align: left;">Count feature X cell matrix from BAMs.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">bam2fq</td>
<td style="text-align: left;">Convert BAM to FASTQ+ file with selected tags.</td>
</tr>
<tr class="even">
<td style="text-align: left;">bam2frag</td>
<td style="text-align: left;">Generate fragment file.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">depth</td>
<td style="text-align: left;">Coverage depth/UMI for target regions.</td>
</tr>
<tr class="even">
<td style="text-align: left;">addtags</td>
<td style="text-align: left;">Add tag string to BAM alignments.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">callept</td>
<td style="text-align: left;">Call expressed peak tags (EPTs) for RNA library.</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;"><strong><em>The following tool used to process fragment file.</em></strong></td>
</tr>
<tr class="odd">
<td style="text-align: left;">count2</td>
<td style="text-align: left;">Count peak X cell matrix from fragment file.</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;"><strong><em>The following tools used to process BED files.</em></strong></td>
</tr>
<tr class="odd">
<td style="text-align: left;">mergebed</td>
<td style="text-align: left;">Merge BED files.</td>
</tr>
<tr class="even">
<td style="text-align: left;">annobed</td>
<td style="text-align: left;">Annotate BED files with genes and functional elements.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">flatten</td>
<td style="text-align: left;">Flatten overlapped regions to nonoverlaps.</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;"><strong><em>The following tools used to process GTF files.</em></strong></td>
</tr>
<tr class="odd">
<td style="text-align: left;">gtffmt</td>
<td style="text-align: left;">Format and reorder GTF file.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gtf2bed</td>
<td style="text-align: left;">Convert GTF to BED.</td>
</tr>
</tbody>
</table>
</section>
<section id="commands-and-options" class="level2">
<h2 class="anchored" data-anchor-id="commands-and-options">Commands and options</h2>
{{&lt; include _commands.qmd &gt;}}
</section>
<section id="difference-in-alignment-annotation-between-pisa-and-cellranger" class="level2">
<h2 class="anchored" data-anchor-id="difference-in-alignment-annotation-between-pisa-and-cellranger">Difference in alignment annotation between PISA and CellRanger</h2>
<p>In most cases, PISA annotation produces results similar to CellRanger, but there are two key differences. In CellRanger, exonic reads are defined as those that overlap with an exon by at least 50% of the read length. In PISA, however, reads overlapping with an exon are classified into three distinct categories. Exonic reads are fully enclosed within an exon (E). If a read partially overlaps both an exon and an intron, it is classified as exonintron (C). Junction reads spanning more than one exon are classified as spliced reads (S). By default, only exonic (E) and spliced reads (S) are counted towards gene expression, but exonintronic (C) are skipped unless option <code>-splice</code> or <code>-intron</code> is set. Therefore, PISA anno is more stringent than CellRanger in gene annotation. However, for third-generation sequencing reads, small indels may be introduced during sequencing, causing imperfect alignment at splice sites or gene ends. In these cases, the <code>-vague-edge</code> option can be used to account for such mapping variances.</p>
<p>The second difference relates to UMI handling. In CellRanger, if a UMI maps to more than one gene, it is discarded. In contrast, PISA annotates all genes if the read is fully enclosed within the exons of these genes, and the UMI is counted for all of them. However, this type of reads can be skipped in PISA if the <code>-one-hit</code> option is set during the PISA count process.</p>
<p>For more information on CellRanger’s counting strategy, you can refer to the official documentation: <a href="https://www.10xgenomics.com/support/software/cell-ranger/7.2/algorithms-overview/cr-gex-algorithm">Cell Ranger Algorithm Overview</a>.</p>
</section>
<section id="changelog" class="level2">
<h2 class="anchored" data-anchor-id="changelog">Changelog</h2>
<details>
<summary>
v1.6 2025/05/29
</summary>
<ul>
<li>Fix a bug if gene_name or gene_id is empty in GTF, annotation may case segmental fault.</li>
<li><code>PISA bin</code> method is introduced.</li>
</ul>
</details>
<details>
<summary>
v1.5 2025/03/17
</summary>
<p>Check and label EXACT_MATCH for all cell barcode segments. Previous only check the first segment. Thanks to <span class="citation" data-cites="lishuangshuang0616">@lishuangshuang0616</span> report this bug.</p>
</details>
<details>
<summary>
v1.4 2025/03/15
</summary>
<ul>
<li>Update <code>-anno-tags</code> instead of <code>-anno-tag</code> to allow specify more than one tag. Only count tag 1 if other tags exist in the alignment.</li>
</ul>
</details>
<details>
<summary>
v1.3 2024/10/10
</summary>
<ul>
<li>Remove <code>-vcf-ss</code> option for <code>PISA anno</code>. Enable strand sensitive mode in default for genetic variant annotation.</li>
</ul>
</details>
<details>
<summary>
v1.2 2024/9/7
</summary>
<ul>
<li>Fix a bug of <code>PISA corr</code>. This bug was caused by the recent update to <code>dict.c</code>. The<code>PISA cor</code>r function was not updated accordingly. This issue will not influence the result.</li>
</ul>
</details>
<details>
<summary>
v1.1 2024/6/11
</summary>
<ul>
<li>-psi and -flatten added for <code>PISA anno</code>;</li>
<li>Add new tools <code>gtf2bed</code> and <code>flatten</code>.</li>
</ul>
</details>
<details>
<summary>
v1.0 2023/12/2
</summary>
<ul>
<li>EPT calling method added.</li>
<li><code>mergebed</code> and <code>annobed</code> for BED file added.</li>
<li>New parameters added to <code>anno</code>.</li>
</ul>
</details>
<details>
<summary>
v1.0a 2023/10/20
</summary>
<ul>
<li>Major update.</li>
<li>New <code>callept</code> method introduced. Check EPT paper for details.</li>
<li>New <code>mergebed</code> and <code>annobed</code> methods added.</li>
<li>New <code>-vcf-ss</code> and <code>-alt-ref</code> now added to <code>PISA anno</code></li>
<li><code>PISA count</code> can count bin and chromosome expression.</li>
</ul>
</details>
<details>
<summary>
v0.12 2022/06/18
</summary>
<ul>
<li>Fix the number of records in MEX file.</li>
</ul>
</details>
<details>
<summary>
v0.12b 2022/04/26
</summary>
<ul>
<li>Bugs fixed.</li>
<li>Change compress level of BGZF from 6 to 2, speed up few tools.</li>
<li>Now accept unorder GTFs</li>
<li>New tool <code>gtfmt</code> introduced to format a GTF file.</li>
<li>Add manual and FASTQ+ specification.</li>
</ul>
</details>
<details>
<summary>
v0.12a 2022/03/31
</summary>
<ul>
<li>Add <code>parse2</code>.</li>
</ul>
</details>
<details>
<summary>
v0.11a 2022/03/13
</summary>
<ul>
<li>Add <code>counts2</code>, count peaks X cells matrix from the fragment file.</li>
</ul>
</details>
<details>
<summary>
v0.10 2022/01/06
</summary>
<ul>
<li>Update <code>bam2frag</code>, export a fragment file compatible with 10X cellranger-ATAC.</li>
</ul>
</details>
<details>
<summary>
v0.10b 2021/12/09
</summary>
<ul>
<li><code>PISA count</code> now has <code>-velo</code> option to export unspliced and spliced matrix together. For velocity analysis, remember to use <code>-intron</code> to annotate reads.</li>
<li><code>PISA parse</code> support multi-threads.</li>
</ul>
</details>
<details>
<summary>
v0.10a 2021/11/06
</summary>
<ul>
<li><code>PISA count</code> support count spliced and unspliced reads.</li>
<li><code>PISA count</code> support count from multiple bam files.</li>
</ul>
</details>
<details>
<summary>
v0.9 2021/10/14
</summary>
<ul>
<li>Rewrite <code>rmdup</code>. Not support paired reads for now.</li>
</ul>
</details>
<details>
<summary>
v0.8 2021/07/20
</summary>
<ul>
<li>Reduce memory usage of <code>count</code></li>
<li>Fix region query bug of <code>anno -bed</code></li>
<li>Add <code>anno -vcf</code> method</li>
</ul>
</details>
<details>
<summary>
v0.7 2020/11/20
</summary>
<ul>
<li>Introduce the PCR deduplicate method <code>rmdup</code>.</li>
<li>Mask read and qual field as * instead of sequence for secondary alignments in the BAM file.</li>
</ul>
</details>
<details>
<summary>
v0.6 2020/10/29
</summary>
<ul>
<li><code>PISA attrcnt</code>, Skip secondary alignments before counting reads</li>
<li><code>PISA anno</code> fix segments fault bugs when loading malformed GTF</li>
</ul>
</details>
<details>
<summary>
v0.5 2020/08/27
</summary>
<ul>
<li>Add <code>PISA bam2frag</code> function (experimental).</li>
<li><code>PISA anno</code> Skip secondary alignments when counting total reads.</li>
</ul>
</details>
<details>
<summary>
v0.4 2020/07/14
</summary>
<ul>
<li><code>PISA sam2bam</code> add mapping quality adjustment method;</li>
<li>Rewrite UMI correction index structure to reduce memory use;</li>
<li>Fix bugs.</li>
</ul>
</details>
<details>
<summary>
v0.4alpha 2020/05/2
</summary>
<ul>
<li><code>PISA anno</code> use UCSC bin scheme instead of linear search for reads query gene regions. Fix the bug of misannotated antisense reads.</li>
<li><code>PISA count</code> use MEX output instead of plain cell vs gene table.</li>
</ul>
</details>
<details>
<summary>
v0.3 2020/03/26
</summary>
<p>Fix bugs and improve preformance.</p>
</details>
<details>
<summary>
0.0.0.9999 2019/05/19
</summary>
<p>Init version.</p>
</details>
</section>
<section id="todo-list" class="level2">
<h2 class="anchored" data-anchor-id="todo-list">TODO list</h2>
<ul>
<li>Improve multi-threads performance.</li>
<li><s>Support Stereo-seq and Visium </s></li>
<li><s>Speed up <code>PISA count</code> and <code>PISA corr</code>.</s></li>
<li><em>Implement parse strategy for cell hash and CITEseq (frozen)</em>.</li>
<li><s>Assemble reads original from one molecule;</s></li>
<li><s>Implement new designed and more user-friendly <code>parse</code>;</s></li>
<li><em>Support loom output (frozen)</em>;</li>
<li><s>Export unspliced matrix for velocity;</s></li>
<li><s>Upgrade <code>PISA parse</code> for faster process fastqs.</s></li>
</ul>
</section>
<section id="reporting-issues" class="level2">
<h2 class="anchored" data-anchor-id="reporting-issues">Reporting Issues</h2>
<p>If you have any suggestion or report issues, please using Github issues page <a href="https://github.com/shiquan/PISA/issues" class="uri">https://github.com/shiquan/PISA/issues</a>.</p>
</section>
<section id="citation" class="level2">
<h2 class="anchored" data-anchor-id="citation">Citation</h2>
<p>Shi Q, Liu S, Kristiansen K, Liu L. The FASTQ+ format and PISA. Bioinformatics. 2022 Sep 30;38(19):4639-4642. doi: 10.1093/bioinformatics/btac562. PMID: 35993907.</p>
</section>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Yano represents an R/C toolkit designed for conducting spatial dissimilarity analysis on single-cell RNA sequencing data. This method revolves around the core concept of examining the distinct expression patterns of a given feature (e.g.&nbsp;exon, snp allele) in relation to its associated binding feature (typically a gene or another allele at the same genomic locus) within the context of cell lineage (1D), spatial position (2D), or the multi-dimensional PCA space. The discernible differences in feature expression patterns and their binding features provide insights into a range of biological phenomena, including alternative splicing, cis-antisense RNA regulation, allele-specific gene expression, and more.</p>
<p>Yano is seamlessly integrated with Seurat, building upon the Seurat object’s framework. Users can perform conventional cell clustering analyses using the state-of-the-art Seurat pipeline and then incorporate exon, SNP counts as new “assays” within the Seurat objects. Subsequently, Yano facilitates the assessment of spatial dissimilarity between these two assays. For more details about the method, please refer to our manuscript.</p>
</section>
<section id="install-1" class="level2">
<h2 class="anchored" data-anchor-id="install-1">INSTALL</h2>
<div class="sourceCode" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>)) <span class="fu">install.packages</span>(<span class="st">'BiocManager'</span>) </span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"shiquan/Yano"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong><em>Notice:</em></strong> Multithread mode is disabled on macOS by default due to the lack of OpenMP support. However, <code>data.table</code> provides a useful tutorial on how to enable OpenMP on macOS. For more information, please refer to this guide: <a href="https://github.com/Rdatatable/data.table/wiki/Installation#enable-openmp-for-macos">https://github.com/Rdatatable/data.table/wiki/Installation#enable-openmp-for-macos</a>.</p>
</section>
<section id="get-started-1" class="level2">
<h2 class="anchored" data-anchor-id="get-started-1">Get started</h2>
<p>A typical workflow with Yano starts by using a built-in dataset.</p>
<p>```{r message=FALSE, warning=FALSE, fig.align=‘center’, fig.width=12, fig.height=6} require(Yano) data(“glbt_small”) DefaultAssay(glbt_small) &lt;- “RNA” glbt_small &lt;- NormalizeData(glbt_small) %&gt;% RunUMAP(dim = 1:20)</p>
<p>DimPlot(glbt_small, label = TRUE, label.size = 5)</p>
<p>DefaultAssay(glbt_small) &lt;- “exon” glbt_small &lt;- NormalizeData(glbt_small) glbt_small &lt;- ParseExonName(glbt_small) grep(“_wm<span class="math inline">\(",names(glbt_small), value=TRUE)
glbt_small &lt;- RunAutoCorr(glbt_small)
grep("_wm\)</span>“,names(glbt_small), value=TRUE) glbt_small &lt;- SetAutoCorrFeatures(glbt_small) glbt_small &lt;- RunBlockCorr(glbt_small, bind.name =”gene_name”, bind.assay = “RNA”)</p>
</section>
</section>
<section id="manhattan-plot-for-spatial-dissimilarity-test-result" class="level1">
<h1>Manhattan plot for spatial dissimilarity test result</h1>
<p>FbtPlot(glbt_small, val = “gene_name.padj”)</p>
<p>FeaturePlot(glbt_small, features = c(“chr19:16095264-16095454/+/TPM4”, “TPM4”), order=TRUE)</p>
</section>
<section id="track-plot-for-gene-coverage-at-different-cell-types" class="level1">
<h1>Track plot for gene coverage at different cell types</h1>
<p>db &lt;- gtf2db(“./gencode.v44.annotation.gtf.gz”) TrackPlot(bamfile=“./Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam”, gtf =db, gene = “TPM4”, junc = TRUE, cell.group = Idents(glbt_small), highlights = c(16095264,16095454))</p>
<pre><code>
See short cases for more details.

## Short cases

- [Alternative splicing analysis for scRNA-seq](https://shiquan.github.io/Yano_AS.html)
- [Allele-specific gene expression analysis for scRNA-seq](https://shiquan.github.io/Yano_ASE.html)
- [Annotating and prioritizing genetic variants for scRNA-seq](https://shiquan.github.io/Yano_anno.html)
- [Perform alternative splicing analysis for cell trajectory and user-defined embeddings](Yano_trajectory.Rmd)
- [Perform alternative splicing analysis for multiple Visium samples](https://shiquan.github.io/Visium.html)

## Changelog
&lt;details&gt;
&lt;summary&gt;1.0 2025/02/19&lt;/summary&gt;

First stable release.

&lt;/details&gt;

&lt;details&gt;
&lt;summary&gt;0.0.0.9999 2023/03/22&lt;/summary&gt;

Init version.

&lt;/details&gt;

## Issues or questions

- [https://github.com/shiquan/Yano/issues](https://github.com/shiquan/Yano/issues)
- [https://github.com/shiquan/Yano/discussions](https://github.com/shiquan/Yano/discussions)

## Functions

Function(s) | Description
|---|---|
|[annoBED](Yano/annoBED.md)|Annotate BED with preload GTF.|
|[annoVAR](Yano/annoVAR.md)|Annotate genetic variants with preload GTF and/or VCF databases.|
|[AutoCorrFeatures](Yano/AutoCorrFeatures.md)|Select spatial autocorrelated features.|
|[FbtPlot](Yano/FbtPlot.md)|Generate a mahattan plot for features.|
|[FindAltExp](Yano/FindAltExp.md)|Find group specific alternative expressed features.|
|[FindAllAltExp](Yano/FindAllAltExp.md)|Find alternative expressed features for all groups.|
|[GetWeights](Yano/GetWeights.md)|Get weight matrix.|
|[gtf2db](Yano/gtf2db.md)|Load GTF database into memory.|
|[mergeMatrix](Yano/mergeMatrix-SMatrix-method.md)|Merge two or more count matrix into one.|
|[ParseBED](Yano/ParseBED.md)|Parse chromsome name, locations and strand from BED names.|
|[ParseExonName](Yano/ParseExonName.md)|Parse chromsome name, locations, strand and gene name from Exon/Junction names.|
|[ParseVarName](Yano/ParseVarName.md)|Parse chromsome name, locations and strand from genetic variant names.|
|[PSIPlot](Yano/PSIPlot.md)|Plot Percent Spliced-IN value on a reduction map.|
|[RatioPlot](Yano/RatioPlot.md)|Plot ratio of feature and binding feature expression on a reduction map.|
|[ReadPISA](Yano/ReadPISA.md)|Load raw matrix which generated by PISA.|
|[RunAutoCorr](Yano/RunAutoCorr.md)|Perform spatial autocorrelation test for features.|
|[RunBlockCorr](Yano/RunBlockCorr.md)|Perform spatial dissimilarity test for features.|
|[RunDEXSeq](Yano/RunDEXSeq.md)|Run alternative expression analysis with DEXSeq.|
|[QuickRecipe](Yano/QuickRecipe-Seurat-method.md)|Run Seurat pipeline.|
|[SetAutoCorrFeatures](Yano/SetAutoCorrFeatures.md)|Setup spatial autocorrelated features.|
|[TrackPlot](Yano/TrackPlot.md)|Generate tracks plot from BAM(s).|
|[varanno](Yano/varanno.md)|Low level version of annoVAR, need specify chr, position manually.|


---
title: Count gene, exon, junction and SNV expression from a BAM file
format:
  html:
    code-overflow: wrap
    include-in-header:
      - text: |
          &lt;style&gt;
           .cell-output-stdout code {
            word-break: break-wor !important;
            white-space: pre-wrap !important;
            }
          &lt;/style&gt;
---

The following vignette demonstrates how to generate various feature counts from an alignment BAM file. Here we use a 10X demo data of human glioblastoma cells. The description of data can be found at &lt;https://www.10xgenomics.com/datasets/human-glioblastoma-multiforme-3-v-3-whole-transcriptome-analysis-3-standard-4-0-0&gt;.

## Prepare the raw data and reference

-   Download [alignment data (BAM)](https://cf.10xgenomics.com/samples/cell-exp/4.0.0/Parent_SC3v3_Human_Glioblastoma/Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam). (~14G).

#### The following file is used for variant calling and allele specific gene expression. It can be skipped for alternative splicing analysis.
- BAM Index File (for variant calling):
[Download BAM index](https://cf.10xgenomics.com/samples/cell-exp/4.0.0/Parent_SC3v3_Human_Glioblastoma/Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam.bai)
- Human Reference Databases and Gene Annotation Files:
    - [GRCh38.p13 genome](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh38.p13.genome.fa.gz)
    - [Gencode v44 annotation](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/gencode.v44.annotation.gtf.gz)


## Programs we will use

-   PISA (&gt;=v1.3) (&lt;https://github.com/shiquan/PISA&gt;)
-   bcftools (&lt;https://github.com/samtools/bcftools&gt;)
-   samtools (&lt;https://github.com/samtools/samtools&gt;)

## Step by step tutorial

### 1. Genetic variants calling

```bash
# Unpack the FASTA file
gzip -d GRCh38.p13.genome.fa.gz

# Build the FASTA index
samtools faidx GRCh38.p13.genome.fa

# Perform variant calling
bcftools mpileup -Ou -f GRCh38.p13.genome.fa ./Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam  | bcftools call -vmO z -o var.vcf.gz</code></pre>
<p><strong>Notes:</strong></p>
<ul>
<li>While several new methods for genetic variant calling have been developed, the basic principle remains the same: identifying differences between sequence reads and the reference genome. For simplicity, we are using the ‘traditional’ bcftools approach.</li>
<li>There is no need to filter raw genetic variants by sequencing depth or strand bias tests. Since this is strand-specific RNA sequencing data, strand bias is expected in many cases. Instead, the feature expression matrix can be filtered by cells during downstream analysis.</li>
</ul>
<section id="annotate-various-features" class="level3">
<h3 class="anchored" data-anchor-id="annotate-various-features">2. Annotate various features</h3>
<p>The following command generates gene, transcript, exon, exon skipped reads, junction and genetic variant for each alignment:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> anno <span class="at">-gtf</span> gencode.v44.annotation.gtf.gz <span class="at">-exon</span> <span class="at">-psi</span> <span class="at">-vcf</span> var.vcf.gz <span class="at">-ref-alt</span> <span class="at">-o</span> anno.bam Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The above command will annotate genes, exons, junctions, and genetic variants all at once. However, it is also possible to annotate different features sequentially. For example, the command below produces the same results as the previous one, but allows for stepwise annotation.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> anno <span class="at">-gtf</span> gencode.v44.annotation.gtf.gz <span class="at">-exon</span> <span class="at">-psi</span> <span class="at">-o</span> anno1.bam Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam </span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> anno <span class="at">-vcf</span> var.vcf.gz <span class="at">-ref-alt</span> <span class="at">-o</span> anno2.bam anno1.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Notes:</strong></p>
<ul>
<li><code>-exon</code>: Generates exon and junction names.</li>
<li><code>-psi</code>: Generate labels for exon skipped reads, named after the skipped exons.</li>
<li><code>-ref-alt</code>: For genetic variants, both reference allele and alternative allele will be annotated. If not set, reference allele will be ignored. This option is important for alelle-specific gene expression.</li>
</ul>
<p>Comparing raw and annotated records.</p>
<pre class="{bash}"><code>#| echo: true
#| code-overflow: wrap
# A original record
samtools view Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam  |head -n 3000|grep 'A00836:286:HMTMVDMXX:2:2354:8422:1391'</code></pre>
<pre class="{bash}"><code># the same record after annotation
samtools view anno.bam | head -n 3000|grep 'A00836:286:HMTMVDMXX:2:2354:8422:1391'</code></pre>
<p><strong>Notes:</strong></p>
<ul>
<li>PISA anno generates <code>GN</code>, <code>TX</code>, <code>RE</code>, and <code>GX</code> tags when the <code>-gtf</code> option is set. Original tag values are replaced, and PISA uses a slightly different annotation strategy than CellRanger, see <a href="https://shiquan.github.io/PISA.html#difference-in-alignment-annotation-between-pisa-and-cellranger">details</a> here. In the example above, the <code>RE</code> tag has been changed from <code>E</code> (exon) to <code>S</code> (spliced).</li>
<li>When the <code>-exon</code> option is used, <code>EX</code> and <code>JC</code> tags are annotated only if the read has a <code>GN</code> tag.</li>
<li>When the <code>-psi</code> option is used, <code>ER</code> tag is annotated, the label for ER is the exon name which indicate the read skip this exon.</li>
</ul>
</section>
<section id="generate-cell-x-feature-expression-matrix" class="level3">
<h3 class="anchored" data-anchor-id="generate-cell-x-feature-expression-matrix">3. Generate Cell X feature expression matrix</h3>
<div class="sourceCode" id="cb46"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exp </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exon</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exclude</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> junction</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> var</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tags</span> CB <span class="at">-anno-tag</span> GN <span class="at">-umi</span> UB <span class="at">-outdir</span> exp anno.bam</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tags</span> CB <span class="at">-anno-tag</span> EX <span class="at">-umi</span> UB <span class="at">-outdir</span> exon anno.bam</span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tags</span> CB <span class="at">-anno-tag</span> ER <span class="at">-umi</span> UB <span class="at">-outdir</span> exclude anno.bam</span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tags</span> CB <span class="at">-anno-tag</span> JC <span class="at">-umi</span> UB <span class="at">-outdir</span> junction anno.bam</span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tags</span> CB <span class="at">-anno-tag</span> VR <span class="at">-umi</span> UB <span class="at">-outdir</span> var anno.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Notes:</strong> These commands will generate raw counts for all droplets. Although the <code>-list</code> parameter can be used to export only selected cells with <code>PISA anno</code>, cell selection may sometimes lack robustness. Therefore, it is safer to store the raw matrix file for long-term storage.</p>
<pre class="{bash}"><code>ls exp/</code></pre>
<pre class="{bash}"><code>zless exp/barcodes.tsv.gz|head</code></pre>
<pre class="{bash}"><code>zless exp/features.tsv.gz|head</code></pre>
<pre class="{bash}"><code>zless exp/matrix.mtx.gz|head</code></pre>
<pre class="{bash}"><code>zless exon/features.tsv.gz|head</code></pre>
<pre class="{bash}"><code>zless junction/features.tsv.gz|head</code></pre>
<p>The format for exon and junction names is “chr:pos reference_allele&gt;alternative_allele/strand” for alternative alleles, or “chr:pos=/strand” for reference allele.</p>
<pre class="{bash}"><code># the format of expressed genetic variant is 
zless var/features.tsv.gz|head</code></pre>
<p><strong>Notes:</strong> By default, the parameters for <code>PISA anno</code> are strand-sensitive. For non-strand-sensitive libraries, such as Smartseq2, the strand should be ignored by using the <code>-is</code> option.</p>
</section>
<section id="questions" class="level3">
<h3 class="anchored" data-anchor-id="questions">Questions?</h3>
<p>If you have any questions regarding this workflows, please feel free to report them through the <a href="https://github.com/shiquan/PISA/issues">github issues</a>.</p>
<p>If you have any ideas or suggestions regarding PISA annotation, please refer to the <a href="https://github.com/shiquan/Yano/discussions">Yano discussion forum</a>. Since PISA is commonly used alongside Yano, it may not be necessary to host separate forums for each.</p>
<p>This vignette demonstrates how to use Yano build functions to interactively select or pick cells/spots from a plot using your mouse and keyboard. While tools like CellxGene are widely used for this purpose, to the best of our knowledge, an efficient solution for performing this task within the R environment is still lacking.</p>
</section>
<section id="prepare-the-data." class="level2">
<h2 class="anchored" data-anchor-id="prepare-the-data.">Prepare the data.</h2>
<p>We use demo data from SeuratData for simplity.</p>
<p>```{r message=FALSE} require(Yano) require(SeuratData)</p>
<p>InstallData(“stxBrain”)</p>
<p>brain &lt;- LoadData(“stxBrain”, type = “anterior1”) brain &lt;- SCTransform(brain, assay = “Spatial”, verbose = FALSE) brain &lt;- RunPCA(brain, assay = “SCT”, verbose = FALSE) brain &lt;- FindNeighbors(brain, reduction = “pca”, dims = 1:30) brain &lt;- FindClusters(brain, verbose = FALSE) brain &lt;- RunUMAP(brain, reduction = “pca”, dims = 1:30)</p>
<p>SpatialPlot(brain)</p>
<pre><code>
## Select spots from spatial plot
```r
sel.1 &lt;- SpatialSelector(brain)</code></pre>
<p><strong><em>Press <code>ESC</code> on your keyboard to exist the selection!</em></strong></p>
<p><img src="spatial_selector.gif" class="img-fluid" style="width:50.0%"></p>
<p>Now the selected cells will be exported to <code>sel.1</code>. If you want return a object, try to set <code>return.object=TRUE</code> in the function.</p>
</section>
<section id="select-cells-from-dimension-reduction-plot" class="level2">
<h2 class="anchored" data-anchor-id="select-cells-from-dimension-reduction-plot">Select cells from dimension reduction plot</h2>
<p>We can also select cells from dimension reduction plot.</p>
<pre class="{r}"><code>DimPlot(brain)</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>sel<span class="fl">.1</span> <span class="ot">&lt;-</span> <span class="fu">DimSelector</span>(brain)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong><em>Press <code>ESC</code> on your keyboard to exist the selection!</em></strong></p>
<p><img src="dim_selector.gif" class="img-fluid" style="width:50.0%"></p>
</section>
<section id="select-cells-based-on-a-feature-expression" class="level2">
<h2 class="anchored" data-anchor-id="select-cells-based-on-a-feature-expression">Select cells based on a feature expression</h2>
<pre class="{r}"><code>FeaturePlot(brain, features = c('Ttr'), order=TRUE)</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureSelector</span>(brain, <span class="at">feature =</span> <span class="fu">c</span>(<span class="st">'Ttr'</span>), <span class="at">order=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Please note in <code>FeatureSelector()</code> only one feature is support at each selection. Therefore, I designed the parameter <code>feature</code> instead of orignal <code>features</code> here.</p>
<p><img src="feature_selector.gif" class="img-fluid" style="width:50.0%"></p>
</section>
<section id="select-cells-from-image-based-spatial-data" class="level2">
<h2 class="anchored" data-anchor-id="select-cells-from-image-based-spatial-data">Select cells from Image-based spatial data</h2>
<p>Here we will use 10x Genomics Xenium data, generated from <a href="https://satijalab.org/seurat/articles/seurat5_spatial_vignette_2">Seurat’s tutorial</a>. <code>{r warnings=FALSE} xenium.obj &lt;- readRDS("xenium.rds") ImageDimPlot(xenium.obj, border.color = "white", border.size = 0.1, cols = "polychrome")</code></p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ImageDimSelector</span>(xenium.obj, <span class="at">border.color =</span> <span class="st">"white"</span>, <span class="at">border.size =</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><img src="image_selector.gif" class="img-fluid" style="width:80.0%"></p>
</section>
<section id="generate-2d-concave-hull-of-selected-region" class="level2">
<h2 class="anchored" data-anchor-id="generate-2d-concave-hull-of-selected-region">Generate 2D concave hull of selected region</h2>
<p>In spatial transcriptomics, it is sometimes necessary to manually define capsule or membrane regions, as these areas are often thin, mixed with neighboring cells, and difficult to identify accurately. Manually specifying these regions can be beneficial for downstream analyses, such as zonation analysis, where precise spatial organization of cells is critical for understanding tissue architecture and functional gradients.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> <span class="fu">SpatialConcaveHull</span>(brain)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><img src="concave_selector.gif" class="img-fluid" style="width:50.0%"></p>
</section>
<section id="questions-1" class="level2">
<h2 class="anchored" data-anchor-id="questions-1">Questions?</h2>
<p>If you have any questions regarding this vignette, the usage of dST or suggestions, please feel free to report them through the <a href="https://github.com/shiquan/dST/discussions">discussion forum</a>.</p>
<p>This vignette demonstrates how to analyze alternative splicing across multiple Visium samples. The procedure for spatial transcriptomics follows the same principles as scRNA-seq but requires careful handling of imaging data and spatial coordinates for accurate visualization. Additionally, this example illustrates how to visualize genome tracks from multiple BAM files.</p>
</section>
<section id="prepare-the-data.-1" class="level2">
<h2 class="anchored" data-anchor-id="prepare-the-data.-1">0. Prepare the data.</h2>
<p>We will use four Visium slices for 10X data source page. These slices come from two sections, each section contain two slices, one is sagittal anterior, and another one is sagittal posterior of mouse brain. We download the processed BAM files, and annotated them with PISA.</p>
<p><a href="https://www.10xgenomics.com/datasets/mouse-brain-serial-section-1-sagittal-anterior-1-standard-1-1-0">Section 1, sagittal anterior</a></p>
<p><a href="https://www.10xgenomics.com/datasets/mouse-brain-serial-section-1-sagittal-posterior-1-standard-1-1-0">Section 1, sagittal posterior</a></p>
<p><a href="https://www.10xgenomics.com/datasets/mouse-brain-serial-section-2-sagittal-anterior-1-standard-1-1-0">Section 2, sagittal anterior</a></p>
<p><a href="https://www.10xgenomics.com/datasets/mouse-brain-serial-section-2-sagittal-posterior-1-standard-1-1-0">Section 2, sagittal posterior</a></p>
<p>The following commands run on a terminal.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create directory for each section and slice</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> visium/section1/Posterior</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> visium/section1/Anterior</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> visium/section2/Posterior</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> visium/section2/Anterior</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Enter to work directory and download GTF for mouse</span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> visium<span class="kw">;</span></span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M10/gencode.vM10.annotation.gtf.gz</span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the bam file and image files to each directory</span></span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> section1/Posterior</span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Posterior/V1_Mouse_Brain_Sagittal_Posterior_possorted_genome_bam.bam</span>
<span id="cb61-14"><a href="#cb61-14" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Posterior/V1_Mouse_Brain_Sagittal_Posterior_possorted_genome_bam.bam.bai</span>
<span id="cb61-15"><a href="#cb61-15" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Posterior/V1_Mouse_Brain_Sagittal_Posterior_spatial.tar.gz</span>
<span id="cb61-16"><a href="#cb61-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-17"><a href="#cb61-17" aria-hidden="true" tabindex="-1"></a><span class="co"># unpack image files</span></span>
<span id="cb61-18"><a href="#cb61-18" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> zxvf V1_Mouse_Brain_Sagittal_Posterior_spatial.tar.gz</span>
<span id="cb61-19"><a href="#cb61-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-20"><a href="#cb61-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Annotate and count features</span></span>
<span id="cb61-21"><a href="#cb61-21" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> anno <span class="at">-gtf</span> ../../gencode.vM10.annotation.gtf.gz <span class="at">-exon</span> <span class="at">-psi</span> <span class="at">-o</span> anno.bam <span class="at">-t</span> 7 V1_Mouse_Brain_Sagittal_Posterior_possorted_genome_bam.bam</span>
<span id="cb61-22"><a href="#cb61-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-23"><a href="#cb61-23" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exp</span>
<span id="cb61-24"><a href="#cb61-24" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exon</span>
<span id="cb61-25"><a href="#cb61-25" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> junction</span>
<span id="cb61-26"><a href="#cb61-26" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exclude</span>
<span id="cb61-27"><a href="#cb61-27" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> GN <span class="at">-outdir</span> exp anno.bam</span>
<span id="cb61-28"><a href="#cb61-28" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> EX <span class="at">-outdir</span> exon anno.bam</span>
<span id="cb61-29"><a href="#cb61-29" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> JC <span class="at">-outdir</span> junction anno.bam</span>
<span id="cb61-30"><a href="#cb61-30" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> ER <span class="at">-outdir</span> exclude anno.bam</span>
<span id="cb61-31"><a href="#cb61-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-32"><a href="#cb61-32" aria-hidden="true" tabindex="-1"></a><span class="co">## Go back to work directory, and reenter to another slice's directory</span></span>
<span id="cb61-33"><a href="#cb61-33" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="at">-</span><span class="kw">;</span> </span>
<span id="cb61-34"><a href="#cb61-34" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> section1/Anterior</span>
<span id="cb61-35"><a href="#cb61-35" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://s3-us-west-2.amazonaws.com/10x.files/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Anterior/V1_Mouse_Brain_Sagittal_Anterior_possorted_genome_bam.bam</span>
<span id="cb61-36"><a href="#cb61-36" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Anterior/V1_Mouse_Brain_Sagittal_Anterior_possorted_genome_bam.bam.bai</span>
<span id="cb61-37"><a href="#cb61-37" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Anterior/V1_Mouse_Brain_Sagittal_Anterior_spatial.tar.gz</span>
<span id="cb61-38"><a href="#cb61-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-39"><a href="#cb61-39" aria-hidden="true" tabindex="-1"></a><span class="co"># unpack image file</span></span>
<span id="cb61-40"><a href="#cb61-40" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> zxvf V1_Mouse_Brain_Sagittal_Anterior_spatial.tar.gz</span>
<span id="cb61-41"><a href="#cb61-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-42"><a href="#cb61-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Annotate and count features</span></span>
<span id="cb61-43"><a href="#cb61-43" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> anno <span class="at">-gtf</span> ../../gencode.vM10.annotation.gtf.gz <span class="at">-exon</span> <span class="at">-psi</span> <span class="at">-o</span> anno.bam <span class="at">-t</span> 7 V1_Mouse_Brain_Sagittal_Anterior_possorted_genome_bam.bam</span>
<span id="cb61-44"><a href="#cb61-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-45"><a href="#cb61-45" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exp</span>
<span id="cb61-46"><a href="#cb61-46" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exon</span>
<span id="cb61-47"><a href="#cb61-47" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> junction</span>
<span id="cb61-48"><a href="#cb61-48" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exclude</span>
<span id="cb61-49"><a href="#cb61-49" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> GN <span class="at">-outdir</span> exp anno.bam</span>
<span id="cb61-50"><a href="#cb61-50" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> EX <span class="at">-outdir</span> exon anno.bam</span>
<span id="cb61-51"><a href="#cb61-51" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> JC <span class="at">-outdir</span> junction anno.bam</span>
<span id="cb61-52"><a href="#cb61-52" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> ER <span class="at">-outdir</span> exclude anno.bam</span>
<span id="cb61-53"><a href="#cb61-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-54"><a href="#cb61-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Swith to section 2 </span></span>
<span id="cb61-55"><a href="#cb61-55" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="at">-</span><span class="kw">;</span></span>
<span id="cb61-56"><a href="#cb61-56" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> section2/Anterior</span>
<span id="cb61-57"><a href="#cb61-57" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://s3-us-west-2.amazonaws.com/10x.files/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Anterior_Section_2/V1_Mouse_Brain_Sagittal_Anterior_Section_2_possorted_genome_bam.bam</span>
<span id="cb61-58"><a href="#cb61-58" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Anterior_Section_2/V1_Mouse_Brain_Sagittal_Anterior_Section_2_possorted_genome_bam.bam.bai</span>
<span id="cb61-59"><a href="#cb61-59" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Anterior_Section_2/V1_Mouse_Brain_Sagittal_Anterior_Section_2_spatial.tar.gz</span>
<span id="cb61-60"><a href="#cb61-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-61"><a href="#cb61-61" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> zxvf V1_Mouse_Brain_Sagittal_Anterior_Section_2_spatial.tar.gz</span>
<span id="cb61-62"><a href="#cb61-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-63"><a href="#cb61-63" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> anno <span class="at">-gtf</span> ../../gencode.vM10.annotation.gtf.gz <span class="at">-exon</span> <span class="at">-psi</span> <span class="at">-o</span> anno.bam <span class="at">-t</span> 7 V1_Mouse_Brain_Sagittal_Anterior_Section_2_possorted_genome_bam.bam</span>
<span id="cb61-64"><a href="#cb61-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-65"><a href="#cb61-65" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exp</span>
<span id="cb61-66"><a href="#cb61-66" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exon</span>
<span id="cb61-67"><a href="#cb61-67" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> junction</span>
<span id="cb61-68"><a href="#cb61-68" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exclude</span>
<span id="cb61-69"><a href="#cb61-69" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> GN <span class="at">-outdir</span> exp anno.bam</span>
<span id="cb61-70"><a href="#cb61-70" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> EX <span class="at">-outdir</span> exon anno.bam</span>
<span id="cb61-71"><a href="#cb61-71" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> JC <span class="at">-outdir</span> junction anno.bam</span>
<span id="cb61-72"><a href="#cb61-72" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> ER <span class="at">-outdir</span> exclude anno.bam</span>
<span id="cb61-73"><a href="#cb61-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-74"><a href="#cb61-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-75"><a href="#cb61-75" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="at">-</span><span class="kw">;</span></span>
<span id="cb61-76"><a href="#cb61-76" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> section2/Sagittal</span>
<span id="cb61-77"><a href="#cb61-77" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Posterior_Section_2/V1_Mouse_Brain_Sagittal_Posterior_Section_2_possorted_genome_bam.bam</span>
<span id="cb61-78"><a href="#cb61-78" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Posterior_Section_2/V1_Mouse_Brain_Sagittal_Posterior_Section_2_possorted_genome_bam.bam.bai</span>
<span id="cb61-79"><a href="#cb61-79" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Brain_Sagittal_Posterior_Section_2/V1_Mouse_Brain_Sagittal_Posterior_Section_2_spatial.tar.gz</span>
<span id="cb61-80"><a href="#cb61-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-81"><a href="#cb61-81" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> zxvf V1_Mouse_Brain_Sagittal_Posterior_Section_2_spatial.tar.gz</span>
<span id="cb61-82"><a href="#cb61-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-83"><a href="#cb61-83" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> anno <span class="at">-gtf</span> ../../gencode.vM10.annotation.gtf.gz <span class="at">-exon</span> <span class="at">-psi</span> <span class="at">-o</span> anno.bam <span class="at">-t</span> 7 V1_Mouse_Brain_Sagittal_Posterior_Section_2_possorted_genome_bam.bam</span>
<span id="cb61-84"><a href="#cb61-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-85"><a href="#cb61-85" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exp</span>
<span id="cb61-86"><a href="#cb61-86" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exon</span>
<span id="cb61-87"><a href="#cb61-87" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> junction</span>
<span id="cb61-88"><a href="#cb61-88" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exclude</span>
<span id="cb61-89"><a href="#cb61-89" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> GN <span class="at">-outdir</span> exp anno.bam</span>
<span id="cb61-90"><a href="#cb61-90" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> EX <span class="at">-outdir</span> exon anno.bam</span>
<span id="cb61-91"><a href="#cb61-91" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> JC <span class="at">-outdir</span> junction anno.bam</span>
<span id="cb61-92"><a href="#cb61-92" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> ER <span class="at">-outdir</span> exclude anno.bam</span>
<span id="cb61-93"><a href="#cb61-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-94"><a href="#cb61-94" aria-hidden="true" tabindex="-1"></a><span class="co"># Now go back to work directory</span></span>
<span id="cb61-95"><a href="#cb61-95" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ../../..</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="preprocess-all-four-slices-and-intergrate-them-into-one-object" class="level2">
<h2 class="anchored" data-anchor-id="preprocess-all-four-slices-and-intergrate-them-into-one-object">1. Preprocess all four slices and intergrate them into one object</h2>
<p>This section will perform Seurat pipeline for all the sections. A Seurat workflow for one section can be found here https://satijalab.org/seurat/articles/spatial_vignette. The difference between this workflow with Seurat’s is we use counts generated from PISA, while Seurat use counts generated with CellRanger’s pipeline. For the details of different between PISA and CellRanger, please see our <a href="https://shiquan.github.io/PISA.html#difference-in-alignment-annotation-between-pisa-and-cellranger">manual</a>.</p>
<p>```{r warning=FALSE, fig.width=12, fig.height=5} require(Yano) exp.1 &lt;- ReadPISA(“./visium/section1/Anterior/exp/”) image.1 &lt;- Read10X_Image(“./visium/section1/Anterior/spatial/”)</p>
</section>
<section id="quickrecipe-is-actually-an-integration-function-of-seurat-workflow" class="level2">
<h2 class="anchored" data-anchor-id="quickrecipe-is-actually-an-integration-function-of-seurat-workflow">QuickRecipe() is actually an integration function of Seurat workflow</h2>
<p>obj.1 &lt;- QuickRecipe(exp.1[, Cells(image.1)], verbose = FALSE) obj.1[[‘slice1’]] &lt;- image.1[colnames(obj.1),] obj.1$orig.ident &lt;- “sec1_ant”</p>
<p>exp.2 &lt;- ReadPISA(“./visium/section1/Posterior/exp/”) image.2 &lt;- Read10X_Image(“./visium/section1/Posterior/spatial/”) obj.2 &lt;- QuickRecipe(exp.2[,Cells(image.2)], verbose = FALSE) obj.2[[‘slice2’]] &lt;- image.2[colnames(obj.2),] obj.2$orig.ident &lt;- “sec1_pos”</p>
<p>exp.3 &lt;- ReadPISA(“./visium/section2/Anterior/exp/”) image.3 &lt;- Read10X_Image(“./visium/section2/Anterior/spatial/”) obj.3 &lt;- QuickRecipe(exp.3[, Cells(image.3)], verbose = FALSE) obj.3[[‘slice3’]] &lt;- image.3[colnames(obj.3),] obj.3$orig.ident &lt;- “sec2_ant”</p>
<p>exp.4 &lt;- ReadPISA(“./visium/section2/Posterior/exp/”) image.4 &lt;- Read10X_Image(“./visium/section2/Posterior/spatial/”) obj.4 &lt;- QuickRecipe(exp.4[, Cells(image.4)], verbose = FALSE) obj.4[[‘slice4’]] &lt;- image.4[colnames(obj.4),] obj.4$orig.ident &lt;- “sec2_pos”</p>
</section>
</section>
<section id="merge-all-processed-seurat-object" class="level1">
<h1>Merge all processed Seurat object</h1>
<p>obj &lt;- merge(obj.1, y = list(obj.2, obj.3, obj.4), add.cell.ids = c(“sec1_ant”, “sec1_pos”, “sec2_ant”, “sec2_pos”))</p>
<p>obj &lt;- QuickRecipe(obj)</p>
<p>p1 &lt;- DimPlot(obj, group.by = “orig.ident”) p2 &lt;- DimPlot(obj) p1 + p2</p>
<p>SpatialPlot(obj, pt.size.factor = 2)</p>
<p>exon.1 &lt;- ReadPISA(“./visium/section1/Anterior/exon/”, prefix = “sec1_ant_”, cells = Cells(obj)) exon.2 &lt;- ReadPISA(“./visium/section1/Posterior/exon/”, prefix = “sec1_pos_”, cells = Cells(obj)) exon.3 &lt;- ReadPISA(“./visium/section2/Anterior/exon/”, prefix = “sec2_ant_”, cells = Cells(obj)) exon.4 &lt;- ReadPISA(“./visium/section2/Posterior/exon/”, prefix = “sec2_pos_”, cells = Cells(obj)) exon &lt;- mergeMatrix(exon.1, exon.2, exon.3, exon.4)</p>
<p>obj[[‘exon’]] &lt;- CreateAssayObject(exon[, Cells(obj)], min.cells = 20) DefaultAssay(obj) &lt;- “exon” obj &lt;- NormalizeData(obj) obj &lt;- ParseExonName(obj) obj &lt;- RunAutoCorr(obj) obj &lt;- SetAutoCorrFeatures(obj)</p>
<p>obj &lt;- RunBlockCorr(obj, bind.name = “gene_name”, bind.assay = “RNA”)</p>
<p>FbtPlot(obj, val = “gene_name.padj”)</p>
<p>obj[[‘exon’]][[]] %&gt;% filter(gene_name.pval&lt;1e-40)</p>
<pre><code>```{r warning=FALSE, fig.width=15, fig.height=8}
SpatialPlot(obj, features = c("chr11:53548291-53549565/+/Sept8","Sept8"), pt.size.factor = 2)</code></pre>
<section id="visualize-genome-tracks-of-multiple-samples" class="level2">
<h2 class="anchored" data-anchor-id="visualize-genome-tracks-of-multiple-samples">2. Visualize genome tracks of multiple samples</h2>
<p>Since the spot/cell names have already been renamed in the Seurat object, we need to define a binding list that maps the new cell names to their original names, which will be used for reading sequences from the BAM files. In this vignette, we use four BAM files, requiring us to define four corresponding cell vectors. Each vector contains group factor values, where the names of the values correspond to the original cell names in the BAM file.</p>
<p>```{r message=FALSE, fig.width=15, fig.height=10} gtf &lt;- gtf2db(“./visium/gencode.vM10.annotation.gtf.gz”)</p>
<p>bamfiles &lt;- list( sec1_ant = “./visium/section1/Anterior/V1_Mouse_Brain_Sagittal_Anterior_possorted_genome_bam.bam”, sec1_pos = “./visium/section1/Posterior/V1_Mouse_Brain_Sagittal_Posterior_possorted_genome_bam.bam”, sec2_ant = “./visium/section2/Anterior/V1_Mouse_Brain_Sagittal_Anterior_Section_2_possorted_genome_bam.bam”, sec2_pos = “./visium/section2/Posterior/V1_Mouse_Brain_Sagittal_Posterior_Section_2_possorted_genome_bam.bam” )</p>
<p>cell.list &lt;- split(obj<span class="math inline">\(seurat_clusters, obj\)</span>orig.ident) str(cell.list)</p>
<p>cell.list1 &lt;- lapply(cell.list, function(x) { nm &lt;- names(x) names(x) &lt;- gsub(“.*_(.*)“,”\1”,nm) x }) str(cell.list1)</p>
<p>TrackPlot(bamfile = bamfiles, cell.group = cell.list1, gtf = gtf, gene = “Sept8”, highlights = c(53548291,53549565), junc=TRUE)</p>
<pre><code>


---
title: From fastq to feature counts with PISA
date: 2024-07-08
---

The following vignette demonstrates how to parse raw FASTQ files to FASTQ+ format and perform alignment, feature annotation, and counting using PISA. For simplification, we use minimap2 to align reads in this example, but other tools such as STAR can also be used.

## Prepare the raw data and reference

-   Download raw data: &lt;http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_fastqs.tar&gt; (5.2G)
-   Download the human reference databases and gene annotation files:
    -   &lt;https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh38.p13.genome.fa.gz&gt;
    -   &lt;https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.annotation.gtf.gz&gt;

## The third-party programs we will use

-   minimap2 (&lt;https://github.com/lh3/minimap2&gt;)
-   samtools (&lt;https://github.com/samtools/samtools&gt;)

## Step by step tutorial

### 1. Parse cell barcode and UMI

``` bash
# Unpack the raw fastq
tar xvf pbmc_1k_v3_fastqs.tar 
gzip -d GRCh38.p13.genome.fa.gz

# Convert raw fastq to fastq+
PISA parse -rule 'CB,R1:1-16;UR,R1:17-28;R1,R2' pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R1_001.fastq.gz,pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R1_001.fastq.gz pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R2_001.fastq.gz,pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R2_001.fastq.gz -nw -report fastq_qc.csv -1 reads.fq </code></pre>
<pre class="{bash}"><code># Check the fastq+ records
head reads.fq</code></pre>
<pre class="{bash}"><code># Check the report
cat fastq_qc.csv</code></pre>
<section id="align-reads-with-minimap2" class="level3">
<h3 class="anchored" data-anchor-id="align-reads-with-minimap2">2. Align reads with minimap2</h3>
<p>To get identical results with 10X’s Cell Ranger, you need use 10X’s gtf to index genome and annotate BAM as they modified it from original ENSEMBL gtf.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="ex">minimap2</span> <span class="at">-t</span> 8 <span class="at">-ax</span> splice GRCh38.p13.genome.fa reads.fq <span class="op">&gt;</span> aln.sam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Notice : Please do NOT use samtools view to convert the SAM to BAM. Because the barcode information still store in the read name. Using PISA sam2bam instead.</p>
<p>FASTQ+ aligned SAM records can be look like this.</p>
<pre class="{bash}"><code>samtools view aln.sam|head -n 1</code></pre>
</section>
<section id="convert-sam-to-bam" class="level3">
<h3 class="anchored" data-anchor-id="convert-sam-to-bam">3. Convert SAM to BAM</h3>
<p>This step will convert the SAM to BAM, while generating the summary of alignments and parse the tags from the read id to SAM optional field. If <code>-adjust-mapq</code> and <code>-gtf</code> parameters set, mapping quality of multi-reads (reads mapped to one gene region and more than one integenic region) will be adjusted.</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> sam2bam <span class="at">-report</span> alignment_report.csv <span class="at">-@</span> 5 <span class="at">-adjust-mapq</span> <span class="at">-gtf</span> gencode.v32.annotation.gtf.gz <span class="at">-o</span> aln.bam aln.sam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Converted records can be look like this.</p>
<pre class="{bash}"><code>samtools view aln.bam|head -n 1</code></pre>
<p>And the summary.</p>
<pre class="{bash}"><code>cat alignment_report.csv</code></pre>
</section>
<section id="gene-annotations" class="level3">
<h3 class="anchored" data-anchor-id="gene-annotations">4. Gene annotations</h3>
<div class="sourceCode" id="cb71"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> anno <span class="at">-gtf</span> gencode.v32.annotation.gtf.gz <span class="at">-o</span> anno.bam aln.bam <span class="at">-t</span> 5 <span class="at">-report</span> anno.csv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="{bash}"><code>cat anno.csv</code></pre>
</section>
<section id="sort-bam-by-genomics-coordinate-optional" class="level3">
<h3 class="anchored" data-anchor-id="sort-bam-by-genomics-coordinate-optional">5. Sort BAM by genomics coordinate (Optional)</h3>
<p>A sorted and indexed BAM file can be useful for many secondary analyses, but it is not mandatory if you only need the gene expression count matrix. The annotation and expression count steps do not require a sorted BAM. However, if two similar UMIs have the same supporting reads, the first one will always be selected, and the second will be corrected during the <code>corr</code> step. If the records are not sorted, this may lead to non-robust results.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co">## sambamba can also be used to sort bam for better performance. For simplication, we use samtools here.</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> sort <span class="at">-@</span> 5 <span class="at">-o</span> sorted.bam anno.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="umis-correction-for-one-cell-in-one-gene" class="level3">
<h3 class="anchored" data-anchor-id="umis-correction-for-one-cell-in-one-gene">6. UMIs correction for one cell in one gene</h3>
<div class="sourceCode" id="cb74"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> corr <span class="at">-tag</span> UR <span class="at">-new-tag</span> UB <span class="at">-tags-block</span> CB,GN <span class="at">-cr</span> <span class="at">-o</span> final.bam <span class="at">-@</span> 5 sorted.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="generate-cell-x-gene-expression-matrix" class="level3">
<h3 class="anchored" data-anchor-id="generate-cell-x-gene-expression-matrix">7. Generate Cell X Gene expression matrix</h3>
<div class="sourceCode" id="cb75"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> raw_gene_expression</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tags</span> CB <span class="at">-anno-tag</span> GN <span class="at">-umi</span> UB <span class="at">-outdir</span> raw_gene_expression <span class="at">-@</span> 5 final.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><code>PISA count</code> will write the feature counts in MEX format.</p>
<pre class="{bash}"><code>ls raw_gene_expression</code></pre>
</section>
<section id="cell-selection-with-dropletutils" class="level3">
<h3 class="anchored" data-anchor-id="cell-selection-with-dropletutils">8. Cell selection with DropletUtils</h3>
<p>In the last step, we generate a raw gene expression matrix. By using R package <a href="https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html">DropletUtils</a>, we select “valid’ cell barcodes from this raw matrix and generate a filtered gene expression matrix in the end.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load required packages</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(DropletUtils)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(Yano)</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Read raw MEX file into a sparse matrix</span></span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">ReadPISA</span>(<span class="st">"raw_gene_expression/"</span>)</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Read sparse matrix as a SingleCellExperiment object</span></span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">SingleCellExperiment</span>(<span class="fu">list</span>(<span class="st">'counts'</span> <span class="ot">=</span> counts))</span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Run barcodeRanks function from DropletUtils, see ref for details</span></span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>br.out <span class="ot">&lt;-</span> <span class="fu">barcodeRanks</span>(sce)</span>
<span id="cb77-13"><a href="#cb77-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-14"><a href="#cb77-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Get all barcodes above the inflection point</span></span>
<span id="cb77-15"><a href="#cb77-15" aria-hidden="true" tabindex="-1"></a>filter_bcs <span class="ot">&lt;-</span> <span class="fu">rownames</span>(br.out)[br.out<span class="sc">$</span>total <span class="sc">&gt;</span><span class="fu">metadata</span>(br.out)<span class="sc">$</span>inflection]</span>
<span id="cb77-16"><a href="#cb77-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-17"><a href="#cb77-17" aria-hidden="true" tabindex="-1"></a><span class="co"># the filer matrix</span></span>
<span id="cb77-18"><a href="#cb77-18" aria-hidden="true" tabindex="-1"></a>filter_counts <span class="ot">&lt;-</span> counts[,filter_bcs]</span>
<span id="cb77-19"><a href="#cb77-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-20"><a href="#cb77-20" aria-hidden="true" tabindex="-1"></a><span class="co"># write file to new folder in 10X format for further use</span></span>
<span id="cb77-21"><a href="#cb77-21" aria-hidden="true" tabindex="-1"></a><span class="fu">write10xCounts</span>(<span class="st">"filtered_gene_expression"</span>, filter_counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="useful-resorces" class="level2">
<h2 class="anchored" data-anchor-id="useful-resorces">Useful resorces</h2>
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html" class="uri">https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html</a></li>
<li><a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_1k_v3" class="uri">https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_1k_v3</a></li>
</ul>
<p>This vignette demonstrates how Yano can annotate genetic variants using public databases and assist users in interpreting the functional impact of these variants.</p>
</section>
<section id="predict-the-consequence-of-genetic-variants" class="level2">
<h2 class="anchored" data-anchor-id="predict-the-consequence-of-genetic-variants">1. Predict the consequence of genetic variants</h2>
<p>Load the object which generated from <a href="Yano_ASE.Rmd">allele specific gene expression from scRNA-seq</a>. In that case, we used annoVAR to annotate the gene and approximate gene region type for the variants. In this section, we will go further, to predict the consequence of genetic variants. Note that, different from genetic annotation at DNA level, strand information is also consdiered here, so antisense terms will also be generated.</p>
<p>The predicted consequences were derived from transcript annotations. However, gencode annotations, generated as part of the ENCODE project, contain numerous truncated transcript records that are not fully consistent with the reference genome, which can introduce bias when predicting amino acid changes. For gene expression analysis, such as scRNA-seq, it’s more common to use gencode, consider the good coverage. But for clinical research, validated annotations like RefSeq are preferred for genetic variant annotation. In this analysis, RefSeq annotations were downloaded from the UCSC server to provide more accurate and reliable predictions. Furthermore, the human genome reference is also need for the prediction.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://hgdownload.soe.ucsc.edu/goldenPath/archive/hg38/ncbiRefSeq/110/hg38.110.ncbiRefSeq.gtf.gz</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh38.p13.genome.fa.gz</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span> GRCh38.p13.genome.fa.gz</span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a><span class="co"># FASTA need to be indexed</span></span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> faidx GRCh38.p13.genome.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="{r}"><code>require(Yano)
obj &lt;- readRDS("glioblastoma5k.rds")

gtf &lt;- gtf2db("hg38.110.ncbiRefSeq.gtf.gz")
DefaultAssay(obj) &lt;- "var"
obj &lt;- annoVAR(obj, gtf = gtf, fasta = "./GRCh38.p13.genome.fa")
obj[['var']][[]] -&gt; df
head(df)
table(df$consequence)
subset(df, consequence %in% "stop_gained")</code></pre>
<p><code>{r fig.width=14, fig.height=6, fig.align='center'} sel.chrs &lt;- c(1:21, "X") FbtPlot(obj, val = "locus.padj", remove.chr = TRUE, sel.chrs = sel.chrs, col.by = "consequence", pt.size = 2)</code></p>
</section>
<section id="annotate-genetic-variants-with-publish-databases" class="level2">
<h2 class="anchored" data-anchor-id="annotate-genetic-variants-with-publish-databases">2. Annotate genetic variants with publish databases</h2>
<p>Another key advancement in genetic variant studies is the ability to perform annotation using genotype-phenotype and allele frequency databases. For demonstration purposes, we use the <a href="https://www.ncbi.nlm.nih.gov/clinvar/">ClinVar</a> database for annotation in this vignette.</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download the database in VCF format</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span>  https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar.vcf.gz</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a><span class="co"># remeber to index this database, otherwise annoVAR will report error.</span></span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index clinvar.vcf.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Let’s take a look at the format of the ClinVar database. As noted, the chromosome names are in the NCBI style, meaning they do not include the “chr” prefix. Therefore, we need to generate a new column with chromosome names formatted for annotation.</p>
<pre class="{bash}"><code>bcftools view clinvar.vcf.gz| grep -v "^#"| head</code></pre>
<p>We need to create new chromosome names that are compatible with the NCBI database format, because chromosome names in Clinvar database are look like ‘1’, ‘2’.</p>
<pre class="{r}"><code>obj[['var']][['chr0']] &lt;- gsub("chr(.*)","\\1", obj[['var']][[]]$chr)
obj[['var']][[]] %&gt;% head

# here we annotate RS id and clinical signification for variants, to check the VCF header for all possible tags
obj &lt;- annoVAR(obj, vcf = "clinvar.vcf.gz", chr = "chr0", tags = c("RS", "CLNSIG"))

# The RS id and CLNSIG tag now created
obj[['var']][[]] -&gt; df
head(df)

# We note that there are several pathogenic variants detected
table(df[['CLNSIG']])

df %&gt;% filter(CLNSIG %in% c('Pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic|risk_factor'))</code></pre>
<p>In the selected SNVs, we found some reference alleles are highlighted, such as chr15:92992967C=/+. These reference alelles are more like to be benign allele. Let’s query the orignal record in the clinvar database to check what exactly stituation it should be.</p>
<pre class="{bash}"><code>bcftools view clinvar.vcf.gz 15:92992967-92992967</code></pre>
<p>You can easily observe that the pathogenic allele is G, where the change from C to G leads to a nonsense mutation. However, if we examine the header, we see that the ‘Number’ field of CLNSIG is labeled as ‘.’, which indicates that this term is not allele-specific—an inappropriate type for this record. In fact, this issue extends beyond just CLNSIG; all terms in the ClinVar database are labeled as ‘.’.</p>
<p>To correct this, we need to manually change the type from ‘.’ to ‘A’ (allele-specific) for allele-specific tags in the VCF file. For more information, please refer to the <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF specification</a>.</p>
<p>Unfortunately, many databases distributed in VCF format may contain non-uniformly formatted tags. Apart from manually updating these databases or filter reference allele afterward, there is no straightforward solution to this issue.</p>
<p>In addition, the RS tag is annotated, and we can refer to the NCBI SNP database for more information. For example, the mutation chr13:50945445G&gt;A/+ at RS 75184679 can be found at the following link: <a href="https://www.ncbi.nlm.nih.gov/snp/rs75184679">https://www.ncbi.nlm.nih.gov/snp/rs75184679</a>. This mutation is associated with conditions such as “Abnormality of the nervous system” and other diseases.</p>
<p><code>{r message=FALSE, fig.width=12, fig.height=5} FeaturePlot(obj, features = c("chr13:50945445G&gt;A/+", "chr13:50945445G=/+", "RNASEH2B"), order = TRUE, ncol=3)</code> Note that not all reads from this gene cover the SNV, which results in the SNV expression being sparser compared to the overall gene expression. Moreover, it also possible to map the clinical significant labels or other tags to the Manhatten plot.</p>
<p><code>{r fig.width=14, fig.height=6, fig.align='center'} sel.chrs &lt;- c(1:21, "X") FbtPlot(obj, val = "locus.padj", remove.chr = TRUE, sel.chrs = sel.chrs, col.by = "CLNSIG", pt.size = 2, cols = c("#131313","blue","#A6CEE3","#1F78B4","#B2DF8A","#33A02C","#FB9A99","#E31A1C","#FDBF6F","#FF7F00","#CAB2D6","#6A3D9A","#FFFF99","#B15928"))</code></p>
<pre class="{r}"><code>Command(obj)

sessionInfo()</code></pre>
<p>This vignette uses gene, exon, and junction expression files generated from the <a href="anno.Rmd">Annotate Various Features for Alignment</a>. While current state-of-the-art scRNA-seq methods tend to be biased towards the 3’ or 5’ ends of transcripts, it is still possible to obtain coverage information for a subset of exons. Despite the sparsity of gene and exon expression in single cells, our spatial dissimilarity test leverages the spatial distribution properties of features. This means that even features with low overall expression but strong spatial expression patterns across cells can still be highlighted. By performing a spatial dissimilarity test between exon/junction and gene expression, we can predict potential alternative splicing events.</p>
</section>
<section id="prerequisite" class="level2">
<h2 class="anchored" data-anchor-id="prerequisite">0. Prerequisite</h2>
<ul>
<li>Make sure you have installed the R environment and <code>Yano</code> package before proceeding with the testing.</li>
</ul>
</section>
<section id="perform-cell-clustering-with-seurat" class="level2">
<h2 class="anchored" data-anchor-id="perform-cell-clustering-with-seurat">1. Perform cell clustering with Seurat</h2>
<p>Load Yano will automatically load Seurat. <code>{r warning=FALSE} require(Yano)</code></p>
<p><code>{r warning=FALSE} # Read raw gene expression matrix exp &lt;- ReadPISA("./exp/")</code> In this section, we will perform the standard Seurat analysis pipeline. Since the spatial dissimilarity test is not rely on cell clustering so changing the resolution or other parameters for <code>FindClusters</code> and <code>RunUMAP</code> will not impact the outcome of the spatial dissimilarity test. ```{r warning=FALSE} # Create Seurat object and filter droplets with fewer than 1000 genes obj &lt;- CreateSeuratObject(exp, min.features = 1000, min.cells = 10)</p>
</section>
</section>
<section id="filter-low-quality-droplets" class="level1">
<h1>Filter low quality droplets</h1>
<p>obj[[“percent.mt”]] &lt;- PercentageFeatureSet(obj, pattern = “^MT-”) obj &lt;- subset(obj, nFeature_RNA &lt; 9000 &amp; percent.mt &lt; 20)</p>
</section>
<section id="downsampling-to-2000-cells-for-fast-testing" class="level1">
<h1>Downsampling to 2000 cells for fast testing</h1>
<p>obj &lt;- obj[, sample(colnames(obj),2000)]</p>
</section>
<section id="we-run-the-cell-clustering-analysis-with-seurat-pipeline" class="level1">
<h1>We run the cell clustering analysis with Seurat pipeline</h1>
<p>obj &lt;- NormalizeData(obj) %&gt;% FindVariableFeatures() %&gt;% ScaleData() %&gt;% RunPCA(verbose=FALSE) %&gt;% FindNeighbors(dims = 1:10, verbose=FALSE) %&gt;% FindClusters(resolution = 0.5, verbose=FALSE) %&gt;% RunUMAP(dims=1:10, verbose=FALSE)</p>
<p>DimPlot(obj, label=TRUE, label.size = 5, label.box = TRUE)</p>
<pre><code>
## 2. Perform alternative splicing with exon assay
In this section, we will compare exon expression patterns with the expression patterns of their corresponding genes in a spatial context. Here, the term "spatial" refers to the organization of cells in space. In this vignette, we will use the PCA space for the analysis, but the approach can also be applied to lineage trajectories, spatial coordinates or integration space such as harmony. The spatial dissimilarity test is divided into several steps. First, we will load exon data as a new assay in the Seurat object. In the second step, we will perform a spatial autocorrelation test for all exons and select the ones that show significant autocorrelation for further analysis. Next, we will define the binding relationship between exons and their corresponding genes and run the spatial dissimilarity test. After testing, P values and adjusted P values for each exon will be provided.

```{r warning=FALSE, fig.width=14, fig.height=6}
# Read exon count matrix file
exon &lt;- ReadPISA("./exon/")

# Load the exon expression to Seurat object as a new assay, make sure the exon matrix has the same cells.
obj[['exon']] &lt;- CreateAssayObject(exon[, colnames(obj)], min.cells=20)

# Switch work assay to exon
DefaultAssay(obj) &lt;- "exon"
# Empty info for exon features
head(obj[['exon']][[]]) %&gt;% knitr::kable()

obj &lt;- ParseExonName(obj)
# Gene name and location are parsed from exon name
head(obj[['exon']][[]]) %&gt;% knitr::kable()

# Normalize the data for spatial autocorrelation test
obj &lt;- NormalizeData(obj)
obj &lt;- RunAutoCorr(obj)

# Select autocorrleated features for downstream test.
obj &lt;- SetAutoCorrFeatures(obj)
</code></pre>
<p>The permutation process can be computationally expensive. In the example below, I set perm=20 to perform only 20 permutations for quicker results. However, the default setting runs 100 permutations for more accurate evaluation. While it’s possible to increase the number of permutations for even more precision, it may not always be necessary. If you’re running Yano for the first time on your dataset, setting perm=20 can help you save time and provide an initial overview of the entire dataset. ```{r warning=FALSE, fig.width=14, fig.height=6} obj &lt;- RunBlockCorr(obj, bind.name = “gene_name”, bind.assay = “RNA”, perm=20)</p>
</section>
<section id="now-p-values-and-adjusted-p-values-have-been-generated" class="level1">
<h1>Now p values and adjusted p values have been generated</h1>
<p>head(obj[[‘exon’]][[]]) %&gt;% knitr::kable()</p>
</section>
<section id="plot-feature-binding-test-plot" class="level1">
<h1>Plot feature binding test plot</h1>
<p>FbtPlot(obj, val = “gene_name.padj”)</p>
<pre><code>
The chromosome names are too long and tend to overlap in the visualization. To resolve this, you can either resize the labels or remove the 'chr' prefix from the chromosome names. Additionally, since the Y chromosome and mitochondrial are not of particular interest to us in this analysis, they can be excluded from the visualization.
```{r warning=FALSE, fig.width=14, fig.height=6}
sel.chrs &lt;- c(1:21, "X")
FbtPlot(obj, val = "gene_name.padj", remove.chr = TRUE, sel.chrs = sel.chrs)

# Let's see how many exons are expressed in different spatial pattern with their genes
obj[['exon']][[]] %&gt;% filter(gene_name.padj &lt; 0.001) %&gt;% knitr::kable() 

# Random select a gene and its exons and visulize with FeaturePlot.
FeaturePlot(obj, features = c("chr15:43801711-43804427/+/SERF2", "SERF2"), ncol=2)

# The default color and parameters perhaps not easily to tell the difference between exon and its binding gene expression. Let's change the scaled colors and enlarge point size and order by expression.
require(RColorBrewer)
FeaturePlot(obj, features = c("chr15:43801711-43804427/+/SERF2", "SERF2"), ncol=2, order = TRUE, pt.size=1) &amp; scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu")))</code></pre>
<p>We can also map the ratio of exon expression to gene expression on the UMAP. The <code>RatioPlot</code> function is designed for this purpose. As observed, the gene SERF2 is relatively low expressed in groups 0, 5, and 9, while the ratio of the exon chr15:43801711-43804427/+/SERF2 is higher in these groups.</p>
<pre class="{r}"><code>RatioPlot(obj, features = c("chr15:43801711-43804427/+/SERF2"), assay = 'exon', bind.assay = 'RNA', bind.name = "gene_name", order = TRUE, pt.size=1) </code></pre>
<p>In the feature plot and ratio plot above, the exon appears to lack a strong expression pattern across cell groups, whereas the gene SERF2 seems to be highly expressed in many groups, except for groups 0, 5 and 9. This inconsistent expression pattern between the exon and its corresponding gene may suggest differential exon usage. To explore the coverage details of both the exon and the gene body, we will generate a track plot next.</p>
<p>In our package, retrieving gene locations requires loading a GTF file instead of relying on current Bioconductor databases, such as <code>org.Hs.eg.db</code>. This is due to the varying versions of gene annotations provided by different institutes, which can introduce inconsistencies. To avoid potential bias during preprocessing and postprocessing, we strongly recommend using the same GTF file consistently throughout your project. The Yano package includes the <code>gtf2db</code> function, which enables you to load a GTF file into memory for further analysis.</p>
<pre class="{r}"><code>gtf &lt;- gtf2db("./gencode.v44.annotation.gtf.gz")</code></pre>
<p>A track plot is used to study the read coverage per cell group. In the track plot shown below, the cell group is specified by the <code>cell.group</code> parameter. Unlike IGV, where read depth is used, we use UMI depth in this plot. The cell barcode tag and UMI tag are predefined as “CB” and “UB” with parameter <code>cell.tag</code> and <code>umi.tag</code>. For each cell group, the UMI depth has been normalized by the number of cells in that group. This means that the depth at each location can be interpreted as the mean UMI depth per cell for the group. As a result, the tracks are directly comparable across different cell groups. If <code>cell.group</code> is not set, the track plot will generate the raw UMI depth per location.</p>
<p><code>{r message=FALSE, fig.width=12, fig.height=8} TrackPlot(bamfile = "Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam", gtf = gtf, gene = "SERF2", cell.group = Idents(obj), highlights = c(43801711,43804427) )</code> In the track plot, we can easily observe that an exon around position 43,794,000 dominates the expression of the SERF2 gene and is highly expressed in many cell groups. However, the exon ‘chr15:43801711-43804427/+/SERF2’ (highlighted) shows low expression and is not visible in the track plot. To visualize low-expressed exons, we can set the <code>max.depth</code> parameter to 2, which caps the UMI depth at 2. And many genes in the region, we set <code>display.genes</code> to SERF2 only. This adjustment allows the low-expressed exons and their related transcripts to be more clearly represented in the plot. In this case, we can found the the highlighed exon shows different expressed pattern with the gene SERF2.</p>
<p><code>{r message=FALSE, fig.width=12, fig.height=8} TrackPlot(bamfile = "Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam", gtf = gtf, gene = "SERF2", cell.group = Idents(obj), highlights = c(43801711,43804427), max.depth = 2, display.genes = "SERF2")</code></p>
<section id="load-junction-assay" class="level2">
<h2 class="anchored" data-anchor-id="load-junction-assay">3. Load junction assay</h2>
<p>In addition to exon expression, junction expression can provide insights into different expression patterns across transcripts, offering a complementary perspective. Junction expression refers to the UMI counts of reads that span more than one exon. It’s important to note that junctions are named similarly to exons, but the start and end positions are different. The start of the junction corresponds to the end of the previous exon, while the end of the junction represents the start of the next exon. ```{r warning=FALSE, fig.width=14, fig.height=6} junction &lt;- ReadPISA(“./junction/”) obj[[‘junction’]] &lt;- CreateAssayObject(junction[, colnames(obj)], min.cells=20)</p>
<p>DefaultAssay(obj) &lt;- “junction” obj &lt;- NormalizeData(obj)</p>
</section>
</section>
<section id="select-spatial-autocorrelated-junctions" class="level1">
<h1>select spatial autocorrelated junctions</h1>
<p>obj &lt;- RunAutoCorr(obj) obj &lt;- SetAutoCorrFeatures(obj)</p>
</section>
<section id="parse-the-gene-name-and-coordinates-from-junction-names" class="level1">
<h1>Parse the gene name and coordinates from junction names</h1>
<p>obj &lt;- ParseExonName(obj)</p>
</section>
<section id="perform-dissimilarity-test-between-junctions-and-their-binding-genes" class="level1">
<h1>perform dissimilarity test between junctions and their binding genes</h1>
<p>obj &lt;- RunBlockCorr(obj, bind.name = “gene_name”, bind.assay = “RNA”, perm=20)</p>
<p>FbtPlot(obj, val=“gene_name.padj”, remove.chr=TRUE, sel.chrs = sel.chrs)</p>
<p>obj[[‘junction’]][[]] %&gt;% filter(gene_name.padj&lt;1e-5)</p>
</section>
<section id="because-both-exon-and-junction-are-compared-with-gene-so-its-reasonable-to-combine-these-two-assays-in-one-plot" class="level1">
<h1>Because both exon and junction are compared with gene, so it’s reasonable to combine these two assays in one plot</h1>
<p>FbtPlot(obj, val=“gene_name.padj”, assay = c(“exon”, “junction”), col.by = “assay”, shape.by = “assay”, pt.size = 2, remove.chr = TRUE, sel.chrs = sel.chrs, cols = c(“red”, “blue”))</p>
</section>
<section id="we-can-find-there-is-an-exon-and-a-junction-at-chromosome-12-with-very-low-p-value-1e-8-lets-see-which-gene-they-are-located" class="level1">
<h1>We can find there is an exon and a junction at chromosome 12 with very low p value (&lt;1e-8), let’s see which gene they are located</h1>
<p>obj[[‘exon’]][[]] %&gt;% filter(chr == “chr12” &amp; gene_name.pval &lt; 1e-8) %&gt;% knitr::kable() obj[[‘junction’]][[]] %&gt;% filter(chr == “chr12” &amp; gene_name.pval &lt; 1e-8) %&gt;% knitr::kable()</p>
<p>FeaturePlot(obj, features = c(“chr12:56161387-56161465/+/MYL6”,“chr12:56160320-56161387/+/MYL6”, “MYL6”), order = TRUE, pt.size = 2, ncol=3) &amp; scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = “RdBu”)))</p>
</section>
<section id="we-could-also-plot-the-expression-ratio-of-these-exons-or-junctions-on-umap" class="level1">
<h1>We could also plot the expression ratio of these exons or junctions on umap</h1>
<p>p1 &lt;- RatioPlot(obj, assay = “exon”, bind.assay = “RNA”, bind.name = “gene_name”, features = “chr12:56161387-56161465/+/MYL6”) p2 &lt;- RatioPlot(obj, assay = “exon”, bind.assay = “RNA”, bind.name = “gene_name”, features = “chr12:56160626-56160670/+/MYL6”) p3 &lt;- RatioPlot(obj, assay = “junction”, bind.assay = “RNA”, bind.name = “gene_name”, features = “chr12:56160320-56161387/+/MYL6”) cowplot::plot_grid(p1,p2,p3, ncol=3)</p>
<pre><code>
We then visualize the track plot for this gene, including junction reads by setting `junc=TRUE`. The height of the splice paths in the plot represents the expression level of each junction within the specified cell group.
```{r message=FALSE, fig.width=12, fig.height=8}
TrackPlot(bamfile = "Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam", gtf = gtf, gene = "MYL6", cell.group = Idents(obj), junc = TRUE, highlights = list(c(56160320,56161387),c(56161387,56161465)))</code></pre>
<p>You might be wondering why the exon chr12:56161387-56161465/+/MYL6 appears highly expressed in cell group 2 in the track plot, where the overlapping peak is clearly higher than in other groups, but its expression level in the feature plot is not as high as expected.</p>
<p>This discrepancy arises because the exon is overlapping with other exons from different transcripts. We only count reads that are fully contained within the exon as part of the exon’s expression. Therefore, reads that partially overlap with this exon are not included in the count.</p>
<p>In contrast, the overlapping exon chr12:56161387-56161575/+/MYL6 shows higher expression in group 2 compared to other groups. It’s important to note that if a read is fully contained within two or more overlapping exons, PISA will count it for all relevant exons. Check PISA’s manual for details.</p>
<p><code>{r fig.width=10, fig.height=5} p1 &lt;- DimPlot(obj, label=TRUE, label.size = 5, label.box = TRUE) p2 &lt;- FeaturePlot(obj, features = c("chr12:56161387-56161575/+/MYL6"), order = TRUE, pt.size = 1) &amp; scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu"))) p1 + p2</code></p>
<section id="heatmap-analysis-for-highlight-group-specific-alternative-splicing" class="level2">
<h2 class="anchored" data-anchor-id="heatmap-analysis-for-highlight-group-specific-alternative-splicing">4. Heatmap analysis for highlight group specific alternative splicing</h2>
<p>The spatial dissimilarity test method prioritizes alternatively spliced exons and junctions across all cells but does not identify which specific cell groups exhibit these splicing events. To address this, let’s manually extract the scaled expression data for the selected alternatively spliced exons and their corresponding genes, then perform a co-clustering analysis. A comprehensive heatmap will be generated using the <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/index.html">ComplexHeatmap</a> package, providing a visual representation of the exon and gene distribution across cell groups.</p>
<p>```{r message=FALSE, warning=FALSE, fig.width=12, fig.height=5} obj[[‘exon’]][[]] %&gt;% filter(gene_name.padj&lt;0.001) %&gt;% rownames -&gt; exons obj[[‘exon’]][[]] %&gt;% filter(gene_name.padj&lt;0.001) %&gt;% pull(gene_name) -&gt; bind.genes DefaultAssay(obj) &lt;- “RNA” obj &lt;- ScaleData(obj, features = unique(bind.genes)) DefaultAssay(obj) &lt;- “exon” obj &lt;- ScaleData(obj, features = exons)</p>
<p>dat1 &lt;- GetAssayData(obj, assay = ‘exon’, layer = ‘scale.data’) dat2 &lt;- GetAssayData(obj, assay = ‘RNA’, layer = ‘scale.data’) idents &lt;- sort(Idents(obj)) order.cells &lt;- names(idents)</p>
<p>dat2 &lt;- dat2[bind.genes,] rownames(dat2) &lt;- exons</p>
<p>dat &lt;- cbind(dat1, dat2)</p>
<p>require(ComplexHeatmap) d &lt;- dist(dat) hc &lt;- hclust(d) idx &lt;- hc<span class="math inline">\(labels[hc\)</span>order]</p>
<p>ha &lt;- HeatmapAnnotation(group=idents, border = TRUE) ht1 &lt;- Heatmap(dat1[idx, order.cells], cluster_rows = FALSE, cluster_columns = FALSE, show_column_names = FALSE, border = TRUE, top_annotation = ha, name = “exon”, column_title = “exon”) ht2 &lt;- Heatmap(dat2[idx, order.cells], cluster_rows = FALSE, cluster_columns = FALSE, show_column_names = FALSE, border = TRUE, top_annotation = ha, name = “gene”, column_title = “gene”, row_names_max_width = max_text_width(rownames(dat2), gp = gpar(fontsize = 12)))</p>
<p>ht &lt;- ht1 + ht2 draw(ht, heatmap_legend_side = “left”, annotation_legend_side = “left”)</p>
<pre><code>

## 5. Test between exon and exon skipped reads

In the previous sections, we conducted a spatial dissimilarity test between exon/junction expression and gene expression. However, binding features are not always limited to genes; they can also correspond to other types of features. In this section, we perform a test between exon expression and reads that skip this exon (exclude assay). This approach is similar to the Percent Spliced In (PSI) method, which is widely used to analyze alternative splicing in both bulk and single-cell RNA-seq data. The PSI is calculated as: `PSI = exon reads / (exon reads + reads skipping this exon)`. 

```{r  message=FALSE, warning=FALSE, fig.width=12, fig.height=4}
# The reads that skip exons are annotated using the `-psi` option in PISA anno, and these counts are stored in the `exclude` directory. We then load these excluded counts into a new assay.
exclude &lt;- ReadPISA("exclude/")
obj[['exclude']] &lt;- CreateAssayObject(exclude[,colnames(obj)], min.cells = 10)
DefaultAssay(obj) &lt;- "exclude"
# Normalize counts for exon-excluded reads
obj &lt;- NormalizeData(obj)
# Then we switch to exon assay
DefaultAssay(obj) &lt;- "exon"

# Because the feature names in the exclude assay are exactly the same as those in the exon assay, they represent the reads that skip each corresponding exon. Therefore, we set up the binding feature using the exon name itself.
obj[['exon']][['exon_name']] &lt;- rownames(obj)
obj[['exon']][['exon_name']] %&gt;% head

# Then we perform spatial dissimilarity test between exon and exclude, mode 1
obj &lt;- RunBlockCorr(obj, bind.name = "exon_name", bind.assay = "exclude")

# Swith to exon exluded assay
DefaultAssay(obj) &lt;- "exclude"
obj &lt;- RunAutoCorr(obj)
obj &lt;- SetAutoCorrFeatures(obj)
obj &lt;- ParseExonName(obj)
obj[['exclude']][['exon_name']] &lt;- rownames(obj)

obj &lt;- RunBlockCorr(obj, bind.name = "exon_name", bind.assay = "exon")

FbtPlot(obj, val = "exon_name.padj", remove.chr = TRUE, assay = c("exclude", "exon"), shape.by = "assay", col.by = "assay", cols = c("yellow", "green"), pt.size = 2)</code></pre>
<p>```{r message=FALSE, warning=FALSE, fig.width=12, fig.height=4} # Let’s how many exons can be prioritized by both exon assay and exclude assay obj[[‘exclude’]][[]] %&gt;% filter(exon_name.padj&lt;1e-5) %&gt;% rownames -&gt; sel1 obj[[‘exon’]][[]] %&gt;% filter(exon_name.padj&lt;1e-5) %&gt;% rownames -&gt; sel2 intersect(sel1,sel2)</p>
<p>DefaultAssay(obj) &lt;- “exclude” p1 &lt;- FeaturePlot(obj, features = c(“chr19:16095264-16095357/+/TPM4”),order = TRUE) DefaultAssay(obj) &lt;- “exon” p2 &lt;- FeaturePlot(obj, features = c(“chr19:16095264-16095357/+/TPM4”),order = TRUE) p3 &lt;- PSIPlot(obj, exon.assay = “exon”, exclude.assay = “exclude”, features = c(“chr19:16095264-16095357/+/TPM4”),order = TRUE) cowplot::plot_grid(p1,p2,p3, ncol=3)</p>
<pre><code>```{r message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
TrackPlot(bamfile = "Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam", gtf = gtf, gene = "TPM4", cell.group = Idents(obj), highlights = c(16095264,16095357), junc = TRUE, max.depth = 1)</code></pre>
<p>In previous sections, we noted that exon or junction expression is part of gene expression, and inverse expression patterns can strongly indicate alternative splicing. However, exon-skipped reads are largely independent of exon expression, making them more sensitive for detecting alternative splicing and allowing for the prioritization of many events.</p>
<p>Because our spatial dissimilarity test does not account for the spatial dependency of the binding feature, numerous events may be prioritized, especially when the binding feature is sparsely expressed. While some of these events may be true, others might arise due to low coverage. To enhance detection power and reduce potential false positives, intersecting the prioritized exons with the prioritized exon-excluded features can help refine the results.</p>
<p>Mode 3 can be used as an alternative to mode 1 to specifically detect events with strong inverse expression patterns. In mode 3, the exon reads and exon-excluded reads are summed as the binding assay, allowing for a more targeted analysis of such patterns. It is important to note that events detected with mode 3 are always detectable with mode 1, making mode 3 a refined approach for prioritizing inverse expression events.</p>
<p><code>{r  message=FALSE, warning=FALSE, fig.width=12, fig.height=4} DefaultAssay(obj) &lt;- "exclude" obj &lt;- RunBlockCorr(obj, bind.name = "exon_name", bind.assay = "exon", mode = 3, prefix = "mode3") DefaultAssay(obj) &lt;- "exon" obj &lt;- RunBlockCorr(obj, bind.name = "exon_name", bind.assay = "exclude", mode = 3, prefix = "mode3") FbtPlot(obj, val = "mode3.padj", remove.chr = TRUE, assay = c("exclude", "exon"), shape.by = "assay", col.by = "assay", pt.size = 2, cols=c("yellow", "green"))</code></p>
<p>In this case study, we performed a spatial dissimilarity test between various feature pairs. This method provides an overview of the entire cell population and does not rely on prior cell clustering and annotation, making it a powerful tool for analyzing cell data without any prior knowledge. It is recommended to test different features including junctions and exons, and set up different with their corresponding genes in 3’ or 5’ biased scRNA-seq. Additionally, testing between exon-included and exon-skipped reads have more power to detect exon excluded events.</p>
<p>To obtain cell-cluster-specific expression patterns, applying heatmaps and clustering in subsequent analyses is recommended.</p>
</section>
<section id="questions-2" class="level2">
<h2 class="anchored" data-anchor-id="questions-2">Questions?</h2>
<p>If you have any questions regarding this vignette and the usage of Yano, please feel free to report them through the <a href="https://github.com/shiquan/Yano/discussions">discussion forum</a>. When submitting your query, please ensure you attach the commands you used for better clarity and support.</p>
<pre class="{r}"><code>Command(obj)

sessionInfo()</code></pre>
<p>This vignette uses gene and single nucleotide variant (SNV) expression files generated from the <a href="anno.Rmd">Annotate Various Features for Alignment</a>.</p>
</section>
<section id="perform-cell-clustering-with-seurat-1" class="level2">
<h2 class="anchored" data-anchor-id="perform-cell-clustering-with-seurat-1">0. Perform cell clustering with Seurat</h2>
<p>We begin by performing cell clustering based on gene expression, which is identical to the first step outlined in the <a href="Yano_AS.Rmd">Alternative splicing analysis for scRNA-seq</a>. If you’re continuing from the previous vignette, you are free to skip this section.</p>
<p>```{r warning=FALSE, fig.align=‘center’} require(Yano)</p>
<p>exp &lt;- ReadPISA(“./exp/”)</p>
<p>obj &lt;- CreateSeuratObject(exp, min.features = 1000, min.cells = 10) obj[[“percent.mt”]] &lt;- PercentageFeatureSet(obj, pattern = “^MT-”) obj &lt;- subset(obj, nFeature_RNA &lt; 9000 &amp; percent.mt &lt; 20)</p>
</section>
</section>
<section id="downsampling-to-2000-cells-for-fast-testing-1" class="level1">
<h1>Downsampling to 2000 cells for fast testing</h1>
<p>obj &lt;- obj[, sample(colnames(obj),2000)]</p>
</section>
<section id="to-improve-visualization-we-reduced-the-resolution-when-identifying-cell-clusters." class="level1">
<h1>To improve visualization, we reduced the resolution when identifying cell clusters.</h1>
<p>obj &lt;- NormalizeData(obj) %&gt;% FindVariableFeatures() %&gt;% ScaleData() %&gt;% RunPCA(verbose=FALSE) %&gt;% FindNeighbors(dims = 1:10, verbose=FALSE) %&gt;% FindClusters(resolution = 0.1, verbose=FALSE) %&gt;% RunUMAP(dims=1:10, verbose=FALSE) DimPlot(obj, label=TRUE, label.size = 5, label.box = TRUE)</p>
<pre><code>## 1. Perform allele-specific gene expression analysis by using SNV anchors

In this vignette, when a genetic variant is detected, all possible alleles at that location, including both alternative alleles and the reference allele, are annotated and counted. Therefore, genetic variants refer to both alternative alleles and the reference allele in this context.
```{r}
var &lt;- ReadPISA("./var")
obj[['var']] &lt;- CreateAssayObject(var[,colnames(obj)], min.cells = 10)
# Switch working assay to 'var'
DefaultAssay(obj) &lt;- "var"

obj &lt;- NormalizeData(obj)

# you can skip to read GTF again if you already load it
gtf &lt;- gtf2db("./gencode.v44.annotation.gtf.gz")

# At this moment the meta infomation for genetic variant assay is still empty
obj[['var']][[]] %&gt;% head

obj &lt;- annoVAR(obj, gtf = gtf)

# After annotation, the locations are parsed from the feature name, and overlapped gene are also annotated.
obj[['var']][[]] %&gt;% head %&gt;% knitr::kable()

# Select autocorrlated genetic variants 
obj &lt;- RunAutoCorr(obj)
obj &lt;- SetAutoCorrFeatures(obj)</code></pre>
<p>In the alternative splicing analysis, we use mode 1 (the default mode). However, for allele-specific gene expression, we switch to mode 2 to compare the expression of an SNV with other SNVs at the same location. Since no binding assay is specified, the records at the same locus are first aggregated, and then the test feature is subtracted from the aggregated data (mode 2).</p>
<p>```{r fig.width=12, fig.height=5, fig.align=‘center’} obj &lt;- RunBlockCorr(object = obj, bind.name = “locus”, mode = 2)</p>
<p>sel.chrs &lt;- c(1:21, “X”) FbtPlot(obj, val = “locus.padj”, remove.chr = TRUE, sel.chrs = sel.chrs)</p>
</section>
<section id="we-could-also-zoom-in-specific-region" class="level1">
<h1>We could also zoom in specific region</h1>
</section>
<section id="interesting-we-can-found-the-peak-at-chromosome-6-is-located-at-3-of-gene-hla-j" class="level1">
<h1>Interesting, we can found the peak at chromosome 6 is located at 3’ of gene HLA-J</h1>
<p>FbtPlot(obj, val = “locus.padj”, chr=“chr6”, start = 30000000, end = 30020000, gtf = gtf)</p>
</section>
<section id="lets-random-pick-a-pair-of-features-at-the-same-locus" class="level1">
<h1>Let’s random pick a pair of features at the same locus</h1>
<p>obj[[‘var’]][[]] %&gt;% filter(locus.padj &lt; 1e-6) %&gt;% knitr::kable() FeaturePlot(obj, features = c(“chr10:17237326C&gt;T/+”, “chr10:17237326C=/+”, “VIM” ), order=TRUE, pt.size = 1, ncol=3)</p>
<pre><code>It appears that the T allele of the gene VIM is expressed in most cell groups, while the C allele seems to be primarily expressed in groups 1 and 4.

Using our spatial dissimilarity analysis method, we have identified many cases of allele-specific gene expression. This phenomenon can be influenced by several known and unknown mechanisms, including gene imprinting, genetic recombination, and somatic variants during cell or disease development. While we won't explore these mechanisms in detail here, you can refer to our manuscript, which includes several case studies dedicated to this topic.



```{r}
# Save the Seurat object for further use
saveRDS(obj, file = "glioblastoma5k.rds")</code></pre>
<pre class="{r}"><code>Command(obj)

sessionInfo()</code></pre>
<p>In the previous vignette, we performed cell clustering using unsupervised methods. Typically, a PCA or integrated space (e.g., Harmony) is generated during this process, and we then calculate the cell weight matrix directly from the cell embedding space. This approach is convenient for running the pipeline, and provide the general result in all cell population. Furthermore, for specific signaling or biological pathways, it is often more informative to focus on a ‘supervised’ cell ordering. This vignette demonstrates how to perform spatial dissimilarity test in low-dimensional spaces, such as cell trajectory.</p>
</section>
<section id="prepare-raw-data" class="level1">
<h1>Prepare raw data</h1>
<p>We used the human testis single-cell atlas dataset for this analysis. The full dataset is publicly available at <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112013">GSE112013</a>. For this vignette, we only used replicate 1.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Create a work directory</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> humantestis</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> humantestis/</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a><span class="co">## We first download the BAM files and GTF file</span></span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://sra-pub-src-1.s3.amazonaws.com/SRR6860519/14515X1.bam.1 <span class="at">-O</span> donor1_rep1.bam</span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://sra-pub-src-1.s3.amazonaws.com/SRR6860520/14606X1.bam.1 <span class="at">-O</span> donor1_rep2.bam</span>
<span id="cb96-8"><a href="#cb96-8" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://ftp.ensembl.org/pub/release-114/gtf/homo_sapiens/Homo_sapiens.GRCh38.114.chr.gtf.gz</span>
<span id="cb96-9"><a href="#cb96-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-10"><a href="#cb96-10" aria-hidden="true" tabindex="-1"></a><span class="co">## Index the bam files for track plots</span></span>
<span id="cb96-11"><a href="#cb96-11" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index donor1_rep1.bam</span>
<span id="cb96-12"><a href="#cb96-12" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index donor1_rep2.bam</span>
<span id="cb96-13"><a href="#cb96-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-14"><a href="#cb96-14" aria-hidden="true" tabindex="-1"></a><span class="co">## Then we use PISA to generate feature counts</span></span>
<span id="cb96-15"><a href="#cb96-15" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> anno <span class="at">-gtf</span> Homo_sapiens.GRCh38.114.chr.gtf.gz <span class="at">-exon</span> <span class="at">-psi</span> <span class="at">-o</span> d1_rep2_anno.bam donor1_rep2.bam</span>
<span id="cb96-16"><a href="#cb96-16" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> anno <span class="at">-gtf</span> Homo_sapiens.GRCh38.114.chr.gtf.gz <span class="at">-exon</span> <span class="at">-psi</span> <span class="at">-o</span> d1_rep1_anno.bam donor1_rep1.bam</span>
<span id="cb96-17"><a href="#cb96-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-18"><a href="#cb96-18" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exp1</span>
<span id="cb96-19"><a href="#cb96-19" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exp2</span>
<span id="cb96-20"><a href="#cb96-20" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exon1</span>
<span id="cb96-21"><a href="#cb96-21" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exon2</span>
<span id="cb96-22"><a href="#cb96-22" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> junction1</span>
<span id="cb96-23"><a href="#cb96-23" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> junction2</span>
<span id="cb96-24"><a href="#cb96-24" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exclude1</span>
<span id="cb96-25"><a href="#cb96-25" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> exclude2</span>
<span id="cb96-26"><a href="#cb96-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-27"><a href="#cb96-27" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> GN <span class="at">-outdir</span> exp1 d1_rep1_anno.bam</span>
<span id="cb96-28"><a href="#cb96-28" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> GN <span class="at">-outdir</span> exp2 d1_rep2_anno.bam</span>
<span id="cb96-29"><a href="#cb96-29" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> EX <span class="at">-outdir</span> exon1 d1_rep1_anno.bam</span>
<span id="cb96-30"><a href="#cb96-30" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> EX <span class="at">-outdir</span> exon2 d1_rep2_anno.bam</span>
<span id="cb96-31"><a href="#cb96-31" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> JC <span class="at">-outdir</span> junction1 d1_rep1_anno.bam</span>
<span id="cb96-32"><a href="#cb96-32" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> JC <span class="at">-outdir</span> junction2 d1_rep2_anno.bam</span>
<span id="cb96-33"><a href="#cb96-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-34"><a href="#cb96-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Note: Although we generate the excluded reads here, we do not use them in this demonstration for the sake of simplicity. However, it is recommended to test them in real cases, so I’ve kept the data available for you to use in your own analyses.</span></span>
<span id="cb96-35"><a href="#cb96-35" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> ER <span class="at">-outdir</span> exclude1 d1_rep1_anno.bam</span>
<span id="cb96-36"><a href="#cb96-36" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> count <span class="at">-tag</span> CB <span class="at">-umi</span> UB <span class="at">-anno-tag</span> ER <span class="at">-outdir</span> exclude2 d1_rep2_anno.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><code>{r, setup, include=FALSE} knitr::opts_knit$set(root.dir = './humantestis/')</code></p>
</section>
<section id="cell-clustering-and-annotation" class="level1">
<h1>Cell clustering and annotation</h1>
<p>```{r message=FALSE} require(Yano)</p>
<p>exp1 &lt;- ReadPISA(“exp1/”, prefix = “d1-rep1-”) exp2 &lt;- ReadPISA(“exp2/”, prefix = “d1-rep2-”)</p>
<p>exp &lt;- mergeMatrix(exp1, exp2)</p>
<p>obj &lt;- QuickRecipe(exp, min.features = 800, min.cells = 100, resolution=0.2) rm(exp1, exp2, exp)</p>
<p>obj$replicate &lt;- gsub(“d1-(.<em>)-.</em>-1”, “\1”, colnames(obj)) DimPlot(obj, label = TRUE, label.size = 5)</p>
<p>FeaturePlot(obj, features = c(“CD14”, “VIM”, “VWF”, “ACTA2”, “DLK1”, “DAZL”, “MAGEA4”, “UTF1”, “ID4”, “FGFR3”, “KIT”, “DMRT1”, “DMRTB1”, “STRA8”, “SYCP3”, “SPO11”, “MLH3”, “ZPBP”, “TNP1”, “PRM2”), ncol=5, order=TRUE) &amp; NoAxes() &amp; NoLegend()</p>
<p>new.cluster.ids &lt;- c(“Sperm”, “Early primary S’cytes”, “Leydig cells”, “Round S’tids”, “Differentiating S’gonia”, “Elongated S’tids”, “Endothelial cells”, “SSCs”, “Myoid cells”, “Sertoli cells”, “Macrophages”, “Later primary S’cytes”, “Sertoli-like”) names(new.cluster.ids) &lt;- levels(obj) obj &lt;- RenameIdents(obj, new.cluster.ids) DimPlot(obj, reduction = “umap”, label = TRUE, pt.size = 0.5) + NoLegend()</p>
<pre><code>
# Cell development trajetory construction

We now select germline cells from the dataset and construct the cell development trajectory using a principal curve. Although there are many well-documented methods for constructing linear trajectories, here we use a straightforward and perhaps the simplest approach purely for demonstration purposes.
```r
sel &lt;- DimSelector(obj)
obj.sel &lt;- obj[,sel]</code></pre>
<p><img src="trajectory.gif" class="img-fluid" style="width:50.0%"></p>
<p><code>{r include=FALSE} obj.sel &lt;- readRDS("./select.rds")</code></p>
<pre class="{r}"><code>require(princurve)
emb &lt;- Embeddings(obj.sel, 'umap')
emb0 &lt;- as.data.frame(emb)
emb0$idents &lt;- Idents(obj)[rownames(emb0)]
emb0 %&gt;% group_by(idents) %&gt;% summarize(x=median(umap_1),y=median(umap_2)) -&gt; emb1
emb2 &lt;- as.matrix(emb1[,c("x","y")])
rownames(emb2) &lt;- emb1$idents
emb2 &lt;- emb2[c("SSCs", "Differentiating S'gonia", "Early primary S'cytes", "Later primary S'cytes", "Round S'tids", "Elongated S'tids", "Sperm"),]
pc &lt;- princurve::principal_curve(x = emb, start = emb2)
s &lt;- as.data.frame(pc$s)
colnames(s) &lt;- c("x", "y")
emb0 &lt;- cbind(emb0, s)
emb0 &lt;- emb0[pc$ord, ]

ggplot(emb0)+ geom_point(aes(umap_1, umap_2, color = idents)) + geom_path(aes(x,y), color = "blue", linewidth=1)+ geom_point(data = emb2, aes(x, y), size=2) </code></pre>
</section>
<section id="perform-alternative-splicing-analysis-along-the-cell-trajetory" class="level1">
<h1>Perform alternative splicing analysis along the cell trajetory</h1>
<p>```{r message=FALSE} cells &lt;- colnames(obj.sel)</p>
<p>exon1 &lt;- ReadPISA(“exon1/”, prefix = “d1-rep1-”, cells = cells) exon2 &lt;- ReadPISA(“exon2/”, prefix = “d1-rep2-”, cells = cells) exon &lt;- mergeMatrix(exon1, exon2)</p>
<p>junction1 &lt;- ReadPISA(“junction1/”, prefix = “d1-rep1-”, cells = cells) junction2 &lt;- ReadPISA(“junction2/”, prefix = “d1-rep2-”, cells = cells) junction &lt;- mergeMatrix(junction1, junction2)</p>
<p>obj.sel[[‘exon’]] &lt;- CreateAssayObject(exon, min.cells = 100) obj.sel[[‘junction’]] &lt;- CreateAssayObject(junction, min.cells = 100) rm(exon, exon1,exon2, junction, junction1, junction2)</p>
<p>order.cells &lt;- rownames(emb0)</p>
<p>DefaultAssay(obj.sel) &lt;- “exon” obj.sel &lt;- NormalizeData(obj.sel) obj.sel &lt;- ParseExonName(obj.sel)</p>
<pre><code>When setup order cells, the weight matrix name will be named to "trajectory_wm".

```{r fig.width=15, fig.height=10}
obj.sel &lt;- RunAutoCorr(obj.sel, order.cells = order.cells, wm.name = "trajectory_wm")
obj.sel &lt;- RunBlockCorr(obj.sel, wm.name = "trajectory_wm", bind.name = "gene_name", bind.assay = "RNA")

DefaultAssay(obj.sel) &lt;- "junction"
obj.sel &lt;- NormalizeData(obj.sel)
obj.sel &lt;- ParseExonName(obj.sel)
obj.sel &lt;- RunAutoCorr(obj.sel, order.cells = order.cells, wm.name = "trajectory_wm")
obj.sel &lt;- RunBlockCorr(obj.sel, wm.name = "trajectory_wm", bind.name = "gene_name", bind.assay = "RNA")

FbtPlot(obj.sel, val = "gene_name.padj", assay = c("exon", "junction"), shape.by = "assay", color.by = "assay")

Meta(obj.sel, assay = "exon") %&gt;% filter(gene_name.padj &lt;1e-10) %&gt;% knitr::kable()
Meta(obj.sel, assay = "junction") %&gt;% filter(gene_name.padj &lt;1e-10) %&gt;% knitr::kable()

FeaturePlot(obj.sel, features = c("3:52413996-52414048/+/PHF7", "3:52414048-52414496/+/PHF7", "PHF7"), ncol=3, order=TRUE)</code></pre>
</section>
<section id="perform-alternative-splicing-analysis-for-all-cell-space" class="level1">
<h1>Perform alternative splicing analysis for all cell space</h1>
<p>We do not specify a name for the weight matrix here, so it will default to using the PCA reduction map to generate the weight matrix. The resulting matrix will be named “<span class="math inline">\(reduction\)</span>_wm”.</p>
<pre class="{r}"><code>DefaultAssay(obj.sel) &lt;- "exon"
obj.sel &lt;- RunAutoCorr(obj.sel)</code></pre>
<p>Now we have two weight matrices. If no name is specified, the first matrix will always be used by default, so we need to explicitly specify “pca_wm” to avoid misusing the weight matrix.</p>
<pre class="{r}"><code>grep("_wm", names(obj.sel), value=TRUE)
obj.sel &lt;- RunBlockCorr(obj.sel, bind.name = "gene_name", bind.assay = "RNA", wm.name = "pca_wm", prefix="all")
DefaultAssay(obj.sel) &lt;- "junction"
obj.sel &lt;- RunAutoCorr(obj.sel)
obj.sel &lt;- RunBlockCorr(obj.sel, bind.name = "gene_name", bind.assay = "RNA", wm.name = "pca_wm", prefix="all")</code></pre>
<p>By comparing the use of cell trajectory ordering with the PCA space, we found that using the PCA approach can detect more alternatively expressed genes. This is because forcing all cells into a 1D trajectory ordering preserves only the rank of each cell, but loses much of the neighborhood information about cells in the broader context. Since cell trajectories are typically constructed based on cell–cell distances in high-dimensional space, regardless of the algorithm used to build the trajectory, I personally recommend using the cell space directly rather than relying solely on 1D cell ordering, unless you have specific reasons or detailed knowledge.</p>
<p>```{r fig.width=15, fig.height=10} FbtPlot(obj.sel, val = “all.padj”, assay = c(“exon”, “junction”), shape.by = “assay”, color.by = “assay”)</p>
<p>Meta(obj.sel, assay = “exon”) %&gt;% filter(all.padj &lt;1e-10) %&gt;% knitr::kable() Meta(obj.sel, assay = “junction”) %&gt;% filter(all.padj &lt;1e-10) %&gt;% knitr::kable()</p>
<pre><code># Perform alternative splicing on a supervised two dimension space

```{r fig.width=15, fig.height=10}
DefaultAssay(obj.sel) &lt;- "RNA"

s.genes &lt;- cc.genes$s.genes
g2m.genes &lt;- cc.genes$g2m.genes

obj.sel[['Group']] &lt;- Idents(obj.sel)
obj.sel &lt;- CellCycleScoring(obj.sel, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

df &lt;- obj.sel[[]]

p1 &lt;- ggplot(df) + geom_point(aes(x=S.Score, y = G2M.Score, color = Phase))
p2 &lt;- ggplot(df) + geom_point(aes(x=S.Score, y = G2M.Score, color = Group))
p1 + p2

# Create cell cycle coordinate space with the module scores
ccc &lt;- as.matrix(df[, c("S.Score", "G2M.Score")])
head(ccc)
colnames(ccc) &lt;- c("CCC_1", "CCC_2")

obj.sel[['ccc']] &lt;- CreateDimReducObject(embeddings = ccc, assay = "RNA")

DefaultAssay(obj.sel) &lt;- "exon"
obj.sel &lt;- RunAutoCorr(obj.sel, reduction = "ccc", wm.name = "ccc_wm", dims=1:2)
obj.sel &lt;- RunBlockCorr(obj.sel, wm.name = "ccc_wm", bind.name = "gene_name", bind.assay = "RNA", prefix = "ccc")

DefaultAssay(obj.sel) &lt;- "junction"
obj.sel &lt;- RunAutoCorr(obj.sel, reduction = "ccc", wm.name = "ccc_wm", dims=1:2)
obj.sel &lt;- RunBlockCorr(obj.sel, wm.name = "ccc_wm", bind.name = "gene_name", bind.assay = "RNA", prefix = "ccc")
FbtPlot(obj.sel, val = "ccc.padj", assay = c("exon", "junction"), shape.by = "assay", color.by = "assay")

Meta(obj.sel, assay = "exon") %&gt;% filter(ccc.padj&lt;1e-35) %&gt;% knitr::kable()

FeaturePlot(obj.sel, reduction =  "ccc", features = c("7:74196642-74197050/+/EIF4H", "EIF4H"), order = TRUE)</code></pre>
<p>```{r message=FALSE, fig.width=15, fig.height=10} gtf &lt;- gtf2db(“./Homo_sapiens.GRCh38.114.chr.gtf.gz”) bamfiles &lt;- list(rep1 = “./donor1_rep1.bam”, rep2 = “./donor1_rep2.bam”) cell.list &lt;- split(obj.sel<span class="math inline">\(Phase, obj.sel\)</span>replicate)</p>
<p>cell.list1 &lt;- lapply(cell.list, function(x) { nm &lt;- names(x) names(x) &lt;- gsub(“.<em>rep[12]-(.</em>)”,“\1”,nm) x }) str(cell.list1)</p>
<p>TrackPlot(bamfile = bamfiles, cell.group = cell.list1, gtf = gtf, gene = “EIF4H”, highlights = c(74196642,74197050), junc=TRUE)</p>
<pre><code>
```{r}
Commands(obj.sel)

sessionInfo()</code></pre>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>