---
#page-layout: custom
section-divs: false
toc: false
css: index.css
editor: source
---
This website outlines the key tools and software packages developed from my research:

<!-- -->

-   **FASTQ+**: This format enhances the standard FASTQ file by integrating barcode information directly into the read name. This modification facilitates the harmonization of single-cell sequencing data across various platforms. FASTQ+ can be generated with PISA.
-   **PISA**: A suite of programs specifically designed for preprocessing data from single-cell sequencing and spatial transcriptomics. It efficiently processes FASTQ+, BAM, and BED files, streamlining the handling and generates various feature counts, facilitating the downstream analyses necessary for in-depth biological insights.
-   **Yano**: An R package that employs the spatial dissimilarity test method as its core algorithm. Yano is tailored for exploring various biological events in single-cell data, including alternative splicing, allele-specific gene expression, and more, providing comprehensive insights into cellular dynamics.

<strong>Workflows and Short Cases for scRNA-seq and Spatial Transcriptome</strong>

- [From raw reads to gene counts](workflow1.Rmd) (PISA)
- [Annotate various features for alignment](anno.Rmd) (PISA)
- [Alternative splicing analysis for scRNA-seq](Yano_AS.Rmd) (Yano)
- [Allele-specific gene expression analysis for scRNA-seq](Yano_ASE.Rmd) (Yano)
<!--
- [Annotate and visualise genetic varaints for scRNA-seq](Yano_anno.Rmd) (Yano)
- [Alternative splicing analysis for Adult Mouse Brain Coronal Section](Yano_Visium.Rmd) (Yano, Visium)
- [Run PISA and Yano with Smartseq2 data](Smartseq2.Rmd) (Smartseq2)
-->

  