---
title: From fastq to feature counts with PISA
date: 2024-07-08
---

The following vignette demonstrates how to parse raw FASTQ files to FASTQ+ format and perform alignment, feature annotation, and counting using PISA. For simplification, we use minimap2 to align reads in this example, but other tools such as STAR can also be used.

## Prepare the raw data and reference

-   Download raw data: <http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_fastqs.tar> (5.2G)
-   Download the human reference databases and gene annotation files:
    -   <https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh38.p13.genome.fa.gz>
    -   <https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.annotation.gtf.gz>

## The third-party programs we will use

-   minimap2 (<https://github.com/lh3/minimap2>)
-   samtools (<https://github.com/samtools/samtools>)

## Step by step tutorial

### 1. Parse cell barcode and UMI

``` bash
# Unpack the raw fastq
tar xvf pbmc_1k_v3_fastqs.tar 
gzip -d GRCh38.p13.genome.fa.gz

# Convert raw fastq to fastq+
PISA parse -rule 'CB,R1:1-16;UR,R1:17-28;R1,R2' pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R1_001.fastq.gz,pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R1_001.fastq.gz pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R2_001.fastq.gz,pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R2_001.fastq.gz -nw -report fastq_qc.csv -1 reads.fq 
```

```{bash}
# Check the fastq+ records
head reads.fq
```
```{bash}
# Check the report
cat fastq_qc.csv
```

### 2. Align reads with minimap2

To get identical results with 10X's Cell Ranger, you need use 10X's gtf to index genome and annotate BAM as they modified it from original ENSEMBL gtf.

``` bash
minimap2 -t 8 -ax splice GRCh38.p13.genome.fa reads.fq > aln.sam
```

Notice : Please do NOT use samtools view to convert the SAM to BAM. Because the barcode information still store in the read name. Using PISA sam2bam instead.

FASTQ+ aligned SAM records can be look like this.

```{bash}
samtools view aln.sam|head -n 1
```

### 3. Convert SAM to BAM

This step will convert the SAM to BAM, while generating the summary of alignments and parse the tags from the read id to SAM optional field. If `-adjust-mapq` and `-gtf` parameters set, mapping quality of multi-reads (reads mapped to one gene region and more than one integenic region) will be adjusted.

``` bash
PISA sam2bam -report alignment_report.csv -@ 5 -adjust-mapq -gtf gencode.v32.annotation.gtf.gz -o aln.bam aln.sam
```

Converted records can be look like this.

```{bash}
samtools view aln.bam|head -n 1
```

And the summary.

```{bash}
cat alignment_report.csv
```

### 4. Gene annotations

``` bash
PISA anno -gtf gencode.v32.annotation.gtf.gz -o anno.bam aln.bam -t 5 -report anno.csv
```

```{bash}
cat anno.csv
```

### 5. Sort BAM by genomics coordinate (Optional)

A sorted and indexed BAM file can be useful for many secondary analyses, but it is not mandatory if you only need the gene expression count matrix. The annotation and expression count steps do not require a sorted BAM. However, if two similar UMIs have the same supporting reads, the first one will always be selected, and the second will be corrected during the `corr` step. If the records are not sorted, this may lead to non-robust results.

``` bash
## sambamba can also be used to sort bam for better performance. For simplication, we use samtools here.
samtools sort -@ 5 -o sorted.bam anno.bam
```

### 6. UMIs correction for one cell in one gene

``` bash
PISA corr -tag UR -new-tag UB -tags-block CB,GN -cr -o final.bam -@ 5 sorted.bam
```

### 7. Generate Cell X Gene expression matrix

``` bash
mkdir raw_gene_expression
PISA count -tags CB -anno-tag GN -umi UB -outdir raw_gene_expression -@ 5 final.bam
```

`PISA count` will write the feature counts in MEX format.

```{bash}
ls raw_gene_expression
```

### 8. Cell selection with DropletUtils

In the last step, we generate a raw gene expression matrix. By using R package [DropletUtils](https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html), we select "valid' cell barcodes from this raw matrix and generate a filtered gene expression matrix in the end.

``` r
# load required packages
require(DropletUtils)
require(Yano)

# Read raw MEX file into a sparse matrix
counts <- ReadPISA("raw_gene_expression/")

# Read sparse matrix as a SingleCellExperiment object
sce <- SingleCellExperiment(list('counts' = counts))

# Run barcodeRanks function from DropletUtils, see ref for details
br.out <- barcodeRanks(sce)

# Get all barcodes above the inflection point
filter_bcs <- rownames(br.out)[br.out$total >metadata(br.out)$inflection]

# the filer matrix
filter_counts <- counts[,filter_bcs]

# write file to new folder in 10X format for further use
write10xCounts("filtered_gene_expression", filter_counts)
```

## Useful resorces

-   <https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html>
-   <https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_1k_v3>
