---
title: Annotating and prioritizing genetic variants for single cell RNA sequencing
# date: 2024-07-08
format:
  html:
    code-overflow: wrap
---
This vignette demonstrates how Yano can annotate genetic variants using public databases and assist users in interpreting the functional impact of these variants.

## 1. Predict the consequence of genetic variants

Load the object which generated from [allele specific gene expression from scRNA-seq](Yano_ASE.Rmd). In that case, we used annoVAR to annotate the gene and approximate gene region type for the variants. In this section, we will go further, to predict the consequence of genetic variants. Note that, different from genetic annotation at DNA level, strand information is also consdiered here, so antisense terms will also be generated. 

The predicted consequences were derived from transcript annotations. However, gencode annotations, generated as part of the ENCODE project, contain numerous truncated transcript records that are not fully consistent with the reference genome, which can introduce bias when predicting amino acid changes. For gene expression analysis, such as scRNA-seq, it's more common to use gencode, consider the good coverage. But for clinical research, validated annotations like RefSeq are preferred for genetic variant annotation. In this analysis, RefSeq annotations were downloaded from the UCSC server to provide more accurate and reliable predictions. Furthermore, the human genome reference is also need for the prediction.
```bash
wget -c https://hgdownload.soe.ucsc.edu/goldenPath/archive/hg38/ncbiRefSeq/110/hg38.110.ncbiRefSeq.gtf.gz
wget -c https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh38.p13.genome.fa.gz
unzip GRCh38.p13.genome.fa.gz

# FASTA need to be indexed
samtools faidx GRCh38.p13.genome.fa
```

```{r}
require(Yano)
obj <- readRDS("glioblastoma5k.rds")

gtf <- gtf2db("hg38.110.ncbiRefSeq.gtf.gz")
DefaultAssay(obj) <- "var"
obj <- annoVAR(obj, gtf = gtf, fasta = "./GRCh38.p13.genome.fa")
obj[['var']][[]] -> df
head(df)
table(df$consequence)
subset(df, consequence %in% "stop_gained")
```

```{r fig.width=14, fig.height=6, fig.align='center'}
sel.chrs <- c(1:21, "X")
FbtPlot(obj, val = "locus.padj", remove.chr = TRUE, sel.chrs = sel.chrs, col.by = "consequence", pt.size = 2)
```

## 2. Annotate genetic variants with publish databases
Another key advancement in genetic variant studies is the ability to perform annotation using genotype-phenotype and allele frequency databases. For demonstration purposes, we use the [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/) database for annotation in this vignette.
```bash
# download the database in VCF format
wget -c  https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar.vcf.gz

# remeber to index this database, otherwise annoVAR will report error.
bcftools index clinvar.vcf.gz
```

Let's take a look at the format of the ClinVar database. As noted, the chromosome names are in the NCBI style, meaning they do not include the "chr" prefix. Therefore, we need to generate a new column with chromosome names formatted for annotation.
```{bash}
bcftools view clinvar.vcf.gz| grep -v "^#"| head
```

We need to create new chromosome names that are compatible with the NCBI database format, because chromosome names in Clinvar database are look like '1', '2'.
```{r}
obj[['var']][['chr0']] <- gsub("chr(.*)","\\1", obj[['var']][[]]$chr)
obj[['var']][[]] %>% head

# here we annotate RS id and clinical signification for variants, to check the VCF header for all possible tags
obj <- annoVAR(obj, vcf = "clinvar.vcf.gz", chr = "chr0", tags = c("RS", "CLNSIG"))

# The RS id and CLNSIG tag now created
obj[['var']][[]] -> df
head(df)

# We note that there are several pathogenic variants detected
table(df[['CLNSIG']])

df %>% filter(CLNSIG %in% c('Pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic|risk_factor'))
```

In the selected SNVs, we found some reference alleles are highlighted, such as chr15:92992967C=/+. These reference alelles are more like to be benign allele. Let's query the orignal record in the clinvar database to check what exactly stituation it should be.

```{bash}
bcftools view clinvar.vcf.gz 15:92992967-92992967
```

You can easily observe that the pathogenic allele is G, where the change from C to G leads to a nonsense mutation. However, if we examine the header, we see that the 'Number' field of CLNSIG is labeled as '.', which indicates that this term is not allele-specificâ€”an inappropriate type for this record. In fact, this issue extends beyond just CLNSIG; all terms in the ClinVar database are labeled as '.'.

To correct this, we need to manually change the type from '.' to 'A' (allele-specific) for allele-specific tags in the VCF file. For more information, please refer to the [VCF specification](https://samtools.github.io/hts-specs/VCFv4.2.pdf).

Unfortunately, many databases distributed in VCF format may contain non-uniformly formatted tags. Apart from manually updating these databases or filter reference allele afterward, there is no straightforward solution to this issue.

In addition, the RS tag is annotated, and we can refer to the NCBI SNP database for more information. For example, the mutation chr13:50945445G>A/+ at RS 75184679 can be found at the following link: [https://www.ncbi.nlm.nih.gov/snp/rs75184679](https://www.ncbi.nlm.nih.gov/snp/rs75184679). This mutation is associated with conditions such as "Abnormality of the nervous system" and other diseases.

```{r message=FALSE, fig.width=12, fig.height=5}
FeaturePlot(obj, features = c("chr13:50945445G>A/+", "chr13:50945445G=/+", "RNASEH2B"), order = TRUE, ncol=3)
```
Note that not all reads from this gene cover the SNV, which results in the SNV expression being sparser compared to the overall gene expression. Moreover, it also possible to map the clinical significant labels or other tags to the Manhatten plot.

```{r fig.width=14, fig.height=6, fig.align='center'}
sel.chrs <- c(1:21, "X")
FbtPlot(obj, val = "locus.padj", remove.chr = TRUE, sel.chrs = sel.chrs, col.by = "CLNSIG", pt.size = 2, cols = c("#131313","blue","#A6CEE3","#1F78B4","#B2DF8A","#33A02C","#FB9A99","#E31A1C","#FDBF6F","#FF7F00","#CAB2D6","#6A3D9A","#FFFF99","#B15928"))
```


```{r}
Command(obj)

sessionInfo()
```

