---
title: Yano
description: Spatial dissimilarity analysis in single cells
sidebar: false
---

## Introduction

Yano represents an R/C toolkit designed for conducting spatial dissimilarity analysis on single-cell RNA sequencing data. This method revolves around the core concept of examining the distinct expression patterns of a given feature (e.g. exon, snp allele) in relation to its associated binding feature (typically a gene or another allele at the same genomic locus) within the context of cell lineage (1D), spatial position (2D), or the multi-dimensional PCA space. The discernible differences in feature expression patterns and their binding features provide insights into a range of biological phenomena, including alternative splicing, cis-antisense RNA regulation, allele-specific gene expression, and more.

Yano is seamlessly integrated with Seurat, building upon the Seurat object's framework. Users can perform conventional cell clustering analyses using the state-of-the-art Seurat pipeline and then incorporate exon, SNP counts as new "assays" within the Seurat objects. Subsequently, Yano facilitates the assessment of spatial dissimilarity between these two assays. For more details about the method, please refer to our manuscript.

## INSTALL

```r
# Dependencies
install.packages(c("Seurat", "R.utils", "viridis", "devtools"), repos = "https://cloud.r-project.org")
```

```r
# Install Yano 
devtools::install_github("shiquan/Yano")
```
***Notice:*** Multithread mode is disabled on macOS by default due to the lack of OpenMP support. However, `data.table` provides a useful tutorial on how to enable OpenMP on macOS. For more information, please refer to this guide: [https://github.com/Rdatatable/data.table/wiki/Installation#enable-openmp-for-macos](https://github.com/Rdatatable/data.table/wiki/Installation#enable-openmp-for-macos).

## Get started

A typical workflow with Yano starts by using a built-in dataset.

```{r message=FALSE, warning=FALSE, fig.align='center', fig.width=12, fig.height=6}
require(Yano)
data("glbt_small")
DefaultAssay(glbt_small) <- "RNA"
glbt_small <- NormalizeData(glbt_small) %>% RunUMAP(dim = 1:20)

DimPlot(glbt_small, label = TRUE, label.size = 5)

DefaultAssay(glbt_small) <- "exon"
glbt_small <- NormalizeData(glbt_small)
glbt_small <- ParseExonName(glbt_small)
grep("_wm$",names(glbt_small), value=TRUE)
glbt_small <- RunAutoCorr(glbt_small)
grep("_wm$",names(glbt_small), value=TRUE)
glbt_small <- SetAutoCorrFeatures(glbt_small)
glbt_small <- RunBlockCorr(glbt_small, bind.name = "gene_name", bind.assay = "RNA")

# Manhattan plot for spatial dissimilarity test result
FbtPlot(glbt_small, val = "gene_name.padj")

FeaturePlot(glbt_small, features = c("chr19:16095264-16095454/+/TPM4", "TPM4"), order=TRUE)

# Track plot for gene coverage at different cell types
db <- gtf2db("./gencode.v44.annotation.gtf.gz")
TrackPlot(bamfile="./Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam", gtf =db, gene = "TPM4", junc = TRUE, cell.group = Idents(glbt_small), highlights = c(16095264,16095454))
```

See short cases for more details.

## Short cases

- [Alternative splicing analysis for scRNA-seq](https://shiquan.github.io/Yano_AS.html)
- [Allele-specific gene expression analysis for scRNA-seq](https://shiquan.github.io/Yano_ASE.html)
- [Annotating and prioritizing genetic variants for scRNA-seq](https://shiquan.github.io/Yano_anno.html)
- [Perform alternative splicing analysis for multiple Visium samples](https://shiquan.github.io/Visium.html)

## Changelog
<details>
<summary>1.0 2025/02/19</summary>

First stable release.

</details>

<details>
<summary>0.0.0.9999 2023/03/22</summary>

Init version.

</details>

## Issues or questions

- [https://github.com/shiquan/Yano/issues](https://github.com/shiquan/Yano/issues)
- [https://github.com/shiquan/Yano/discussions](https://github.com/shiquan/Yano/discussions)

## Functions

Function(s) | Description
|---|---|
|[annoBED](Yano/annoBED.md)|Annotate BED with preload GTF.|
|[annoVAR](Yano/annoVAR.md)|Annotate genetic variants with preload GTF and/or VCF databases.|
|[AutoCorrFeatures](Yano/AutoCorrFeatures.md)|Select spatial autocorrelated features.|
|[FbtPlot](Yano/FbtPlot.md)|Generate a mahattan plot for features.|
|[FindAltExp](Yano/FindAltExp.md)|Find group specific alternative expressed features.|
|[FindAllAltExp](Yano/FindAllAltExp.md)|Find alternative expressed features for all groups.|
|[GetWeights](Yano/GetWeights.md)|Get weight matrix.|
|[gtf2db](Yano/gtf2db.md)|Load GTF database into memory.|
|[mergeMatrix](Yano/mergeMatrix-SMatrix-method.md)|Merge two or more count matrix into one.|
|[ParseBED](Yano/ParseBED.md)|Parse chromsome name, locations and strand from BED names.|
|[ParseExonName](Yano/ParseExonName.md)|Parse chromsome name, locations, strand and gene name from Exon/Junction names.|
|[ParseVarName](Yano/ParseVarName.md)|Parse chromsome name, locations and strand from genetic variant names.|
|[PSIPlot](Yano/PSIPlot.md)|Plot Percent Spliced-IN value on a reduction map.|
|[RatioPlot](Yano/RatioPlot.md)|Plot ratio of feature and binding feature expression on a reduction map.|
|[ReadPISA](Yano/ReadPISA.md)|Load raw matrix which generated by PISA.|
|[RunAutoCorr](Yano/RunAutoCorr.md)|Perform spatial autocorrelation test for features.|
|[RunBlockCorr](Yano/RunBlockCorr.md)|Perform spatial dissimilarity test for features.|
|[RunDEXSeq](Yano/RunDEXSeq.md)|Run alternative expression analysis with DEXSeq.|
|[QuickRecipe](Yano/QuickRecipe-Seurat-method.md)|Run Seurat pipeline.|
|[SetAutoCorrFeatures](Yano/SetAutoCorrFeatures.md)|Setup spatial autocorrelated features.|
|[TrackPlot](Yano/TrackPlot.md)|Generate tracks plot from BAM(s).|
|[varanno](Yano/varanno.md)|Low level version of annoVAR, need specify chr, position manually.|

