---
title: Alelle specific gene expression analysis for single cell RNA sequencing
format:
  html:
    code-overflow: wrap
---
This vignette uses gene and single nucleotide variant (SNV) expression files generated from the [Annotate Various Features for Alignment](anno.Rmd).


## 0. Perform cell clustering with Seurat

We begin by performing cell clustering based on gene expression, which is identical to the first step outlined in the [Alternative splicing analysis for scRNA-seq](Yano_AS.Rmd). If you're continuing from the previous vignette, you are free to skip this section.

```{r warning=FALSE, fig.align='center'}
require(Yano)

exp <- ReadPISA("./exp/")

obj <- CreateSeuratObject(exp, min.features = 1000, min.cells = 10)
obj[["percent.mt"]] <- PercentageFeatureSet(obj, pattern = "^MT-")
obj <- subset(obj, nFeature_RNA < 9000 & percent.mt < 20)

# Downsampling to 2000 cells for fast testing
obj <- obj[, sample(colnames(obj),2000)]

# Clustering analysis
obj <- NormalizeData(obj) %>% FindVariableFeatures() %>% ScaleData() %>%  RunPCA(verbose=FALSE) %>% FindNeighbors(dims = 1:10, verbose=FALSE) %>% FindClusters(resolution = 0.1, verbose=FALSE) %>% RunUMAP(dims=1:10, verbose=FALSE)
DimPlot(obj, label=TRUE, label.size = 5, label.box = TRUE)
```
## 1. Perform allele-specific gene expression analysis by using SNV anchors

In this vignette, when a genetic variant is detected, all possible alleles at that location, including both alternative alleles and the reference allele, are annotated and counted. Therefore, genetic variants refer to both alternative alleles and the reference allele in this context.
```{r}
var <- ReadPISA("./var")
obj[['var']] <- CreateAssayObject(var[,colnames(obj)], min.cells = 10)
DefaultAssay(obj) <- "var"

obj <- NormalizeData(obj)

# you can skip to read GTF again if you already load it
gtf <- gtf2db("./gencode.v44.annotation.gtf.gz")

# At this moment the meta infomation for genetic variant assay is still empty
obj[['var']][[]] %>% head

obj <- annoVAR(obj, gtf = gtf)

# After annotation, the locations are parsed from the feature name, and overlapped gene are also annotated.
obj[['var']][[]] %>% head %>% knitr::kable()

# Select autocorrlated genetic variants 
obj <- RunAutoCorr(obj)
obj <- SetAutoCorrFeatures(obj)
```

In the alternative splicing analysis, we use mode 1 (the default mode). However, for allele-specific gene expression, we switch to mode 2 to compare the expression of an SNV with other SNVs at the same location. Since no binding assay is specified, the records at the same locus are first aggregated, and then the test feature is subtracted from the aggregated data (mode 2).

```{r fig.width=12, fig.height=5, fig.align='center'}
obj <- RunBlockCorr(object = obj, bind.name = "locus", mode = 2)

sel.chrs <- c(1:21, "X")
FbtPlot(obj, val = "locus.padj", remove.chr = TRUE, sel.chrs = sel.chrs)

# We could also zoom in specific region
# Interesting, we can found the peak at chromosome 6 is located at 3' of gene HLA-J
FbtPlot(obj, val = "locus.padj", chr="chr6", start = 30000000, end = 30020000, gtf = gtf)

# Let's random pick a pair of features at the same locus
obj[['var']][[]] %>% filter(locus.padj < 1e-6) %>% knitr::kable()
FeaturePlot(obj, features = c("chr10:17237326C>T/+", "chr10:17237326C=/+", "VIM" ), order=TRUE, pt.size = 1, ncol=3)
```
It appears that the T allele of the gene VIM is expressed in most cell groups, while the C allele seems to be primarily expressed in groups 1 and 4.

Using our spatial dissimilarity analysis method, we have identified many cases of allele-specific gene expression. This phenomenon can be influenced by several known and unknown mechanisms, including gene imprinting, genetic recombination, and somatic variants during cell or disease development. While we won't explore these mechanisms in detail here, you can refer to our manuscript, which includes several case studies dedicated to this topic.



```{r}
# Save the Seurat object for further use
saveRDS(obj, file = "glioblastoma5k.rds")
```

```{r}
Command(obj)
```
