---
title: Alelle specific gene expression analysis for scRNA-seq
date: 2024-07-08
format:
  html:
    code-overflow: wrap
---
This vignette uses gene and single nucleotide variant (SNV) expression files generated from the [Annotate Various Features for Alignment](anno.Rmd).


## 0. Perform cell clustering with Seurat

We begin by performing cell clustering based on gene expression, which is identical to the first step outlined in the [Alternative splicing analysis for scRNA-seq](Yano_AS.Rmd). If you're continuing from the previous vignette, you are free to skip this section.

```{r warning=FALSE}
require(Yano)

exp <- ReadPISA("./exp/")

obj <- CreateSeuratObject(exp, min.features = 1000, min.cells = 10)
obj[["percent.mt"]] <- PercentageFeatureSet(obj, pattern = "^MT-")
obj <- subset(obj, nFeature_RNA < 9000 & percent.mt < 20)

# Downsampling to 2000 cells for fast testing
obj <- obj[, sample(colnames(obj),2000)]

# Clustering analysis
obj <- NormalizeData(obj) %>% FindVariableFeatures() %>% ScaleData() %>%  RunPCA(verbose=FALSE) %>% FindNeighbors(dims = 1:10, verbose=FALSE) %>% FindClusters(resolution = 0.1, verbose=FALSE) %>% RunUMAP(dims=1:10, verbose=FALSE)
DimPlot(obj, label=TRUE, label.size = 5, label.box = TRUE)
```
## 1. Perform allele-specific gene expression analysis by using SNV anchors

In this vignette, when a genetic variant is detected, all possible alleles at that location, including both alternative alleles and the reference allele, are annotated and counted. Therefore, genetic variants refer to both alternative alleles and the reference allele in this context.
```{r}
var <- ReadPISA("./var")
obj[['var']] <- CreateAssayObject(var[,colnames(obj)], min.cells = 10)
DefaultAssay(obj) <- "var"

obj <- NormalizeData(obj)

# you can skip to read GTF again if you already load it
gtf <- gtf2db("./gencode.v44.annotation.gtf.gz")

# At this moment the meta infomation for genetic variant assay is still empty
obj[['var']][[]] %>% head

obj <- annoVAR(obj, gtf = gtf)

# After annotation, the locations are parsed from the feature name, and overlapped gene are also annotated.
obj[['var']][[]] %>% head %>% knitr::kable()

# Select autocorrlated genetic variants 
obj <- RunAutoCorr(obj)
obj <- SetAutoCorrFeatures(obj)
```

In the alternative splicing analysis, we use mode 1 (the default mode). However, for allele-specific gene expression, we switch to mode 2 to compare the expression of an SNV with other SNVs at the same location. Since no binding assay is specified, the records at the same locus are first aggregated, and then the test feature is subtracted from the aggregated data (mode 2).

```{r fig.width=12, fig.height=5}
obj <- RunBlockCorr(object = obj, bind.name = "locus", mode = 2)

sel.chrs <- c(1:21, "X")
FbtPlot(obj, val = "locus.padj", remove.chr = TRUE, sel.chrs = sel.chrs)

# We could also zoom in specific region
# Interesting, we can found the peak at chromosome 6 is located at 3' of gene HLA-J
FbtPlot(obj, val = "locus.padj", chr="chr6", start = 30000000, end = 30020000, gtf = gtf)

# Let's random pick a pair of features at the same locus
obj[['var']][[]] %>% filter(locus.padj < 1e-6) %>% knitr::kable()
FeaturePlot(obj, features = c("chr10:17237326C>T/+", "chr10:17237326C=/+", "VIM" ), order=TRUE, pt.size = 1, ncol=3)
```
It appears that the T allele of the gene VIM is expressed in most cell groups, while the C allele seems to be primarily expressed in groups 1 and 4.

Using our spatial dissimilarity analysis method, we have identified many cases of allele-specific gene expression. This phenomenon can be influenced by several known and unknown mechanisms, including gene imprinting, genetic recombination, and somatic variants during cell or disease development. While we won't explore these mechanisms in detail here, you can refer to our manuscript, which includes several case studies dedicated to this topic.

## 2. Annotate genetic variants
One key advancement in genetic variant studies is the ability to perform functional interpretation and annotation using genotype-phenotype and allele frequency databases. For demonstration purposes, we use the [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/) database for annotation in this vignette.
```bash
# download the database in VCF format
wget -c  https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar.vcf.gz

# remeber to index this database, otherwise annoVAR will report error.
bcftools index clinvar.vcf.gz
```

Let's take a look at the format of the ClinVar database. As noted, the chromosome names are in the NCBI style, meaning they do not include the "chr" prefix. Therefore, we need to generate a new column with chromosome names formatted for annotation.
```{bash}
bcftools view clinvar.vcf.gz| grep -v "^#"| head
```


```{r}
obj[['var']][['chr0']] <- gsub("chr(.*)","\\1", obj[['var']][[]]$chr)
obj[['var']][[]] %>% head

# here we annotate RS id and clinical signification for variants, to check the VCF header for all possible tags
obj <- annoVAR(obj, vcf = "clinvar.vcf.gz", chr = "chr0", tags = c("RS", "CLNSIG"))

# The RS id and CLNSIG tag now created
obj[['var']][[]] -> df
head(df)

# We note that there are several pathogenic variants detected
table(df[['CLNSIG']])

df %>% filter(CLNSIG %in% c('Pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic|risk_factor'))
```

In the selected SNVs, we found some reference alleles are highlighted, such as chr15:92992967C=/+. These reference alelles are more like to be benign allele. Let's query the orignal record in the clinvar database to check what exactly stituation it should be.

```{bash}
bcftools view clinvar.vcf.gz 15:92992967-92992967
```

You can easily observe that the pathogenic allele is G, where the change from C to G leads to a nonsense mutation. However, if we examine the header, we see that the 'Number' field of CLNSIG is labeled as '.', which indicates that this term is not allele-specificâ€”an inappropriate type for this record. In fact, this issue extends beyond just CLNSIG; all terms in the ClinVar database are labeled as '.'.

To correct this, we need to manually change the type from '.' to 'A' (allele-specific) for allele-specific tags in the VCF file. For more information, please refer to the [VCF specification](https://samtools.github.io/hts-specs/VCFv4.2.pdf).

Unfortunately, many databases distributed in VCF format may contain non-uniformly formatted tags. Apart from manually updating these databases or filter reference allele afterward, there is no straightforward solution to this issue.

In addition, the RS tag is annotated, and we can refer to the NCBI SNP database for more information. For example, the mutation chr13:50945445G>A/+ at RS 75184679 can be found at the following link: [https://www.ncbi.nlm.nih.gov/snp/rs75184679](https://www.ncbi.nlm.nih.gov/snp/rs75184679). This mutation is associated with conditions such as "Abnormality of the nervous system" and other diseases.

```{r message=FALSE, fig.width=12, fig.height=5}
FeaturePlot(obj, features = c("chr13:50945445G>A/+", "chr13:50945445G=/+", "RNASEH2B"), order = TRUE, ncol=3)
```
Note that not all reads from this gene cover the SNV, which results in the SNV expression being sparser compared to the overall gene expression. Here, we highlight the genetic variant location. Unfortunately, the track plot does not yet support the visualization of sequence variants.
```{r message=FALSE, fig.width=12, fig.height=8}
TrackPlot(bamfile = "Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam", chr = "chr13", start=50935000, end=50949000, cell.group = Idents(obj), junc = TRUE, highlights = c(50945445,50945445), gtf=gtf )
```


If you have any questions regarding this vignette, please feel free to report them through the [discussion forum](https://github.com/shiquan/Yano/discussions). When submitting your query, please ensure you attach the commands you used for better clarity and support.
```{r}
Command(obj)
```

