

# ParseExonName

[**Source code**](https://github.com/shiquan/Yano/tree/master/R/#L)

## Description

Parse chromosome, start, end, strand, and gene name from the exon name.
The exon name generated by ‘PISA anno’ is formated like
chr:start-end/\[+-\]/gene_name.

## Usage

<pre><code class='language-R'>ParseExonName(object = NULL, assay = NULL)
</code></pre>

## Arguments

<table>
<tr>
<td style="white-space: nowrap; font-family: monospace; vertical-align: top">
<code id="object">object</code>
</td>
<td>
Seurat object.
</td>
</tr>
<tr>
<td style="white-space: nowrap; font-family: monospace; vertical-align: top">
<code id="assay">assay</code>
</td>
<td>
Exon assay name. Default use current actived assay.
</td>
</tr>
</table>

## Examples

``` r
library("Yano")

data("glbt_small")
DefaultAssay(glbt_small) <- "exon"
# Check the meta table before parsing
head(glbt_small[['exon']][[]])
```

    data frame with 0 columns and 6 rows

``` r
glbt_small <- ParseExonName(glbt_small)

# Now see the meta table after parsing
head(glbt_small[['exon']][[]])
```

                                        chr     start       end gene_name strand
    chr1:154169305-154169383/-/TPM3    chr1 154169305 154169383      TPM3      -
    chr11:35197162-35197793/+/CD44    chr11  35197162  35197793      CD44      +
    chr11:75421727-75422280/+/RPS3    chr11  75421727  75422280      RPS3      +
    chr11:123060825-123061329/-/HSPA8 chr11 123060825 123061329     HSPA8      -
    chr11:123061588-123061833/-/HSPA8 chr11 123061588 123061833     HSPA8      -
    chr11:123061869-123062022/-/HSPA8 chr11 123061869 123062022     HSPA8      -
