<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.53">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>PISA User Guide</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
<link rel="stylesheet" href="index.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
          <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Overview</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./fastq.html"> 
<span class="menu-text">FASTQ+</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./PISA.html" aria-current="page"> 
<span class="menu-text">PISA</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./Yano.html"> 
<span class="menu-text">Yano</span></a>
  </li>  
</ul>
          <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="GitHub" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label="GitHub"><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/shiquan">
            My GitHub Page
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/shiquan/Yano/discussions">
            Discussions
            </a>
          </li>
      </ul>
    </div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
            <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#synopsis" id="toc-synopsis" class="nav-link active" data-scroll-target="#synopsis">Synopsis</a></li>
  <li><a href="#install" id="toc-install" class="nav-link" data-scroll-target="#install">Install</a></li>
  <li><a href="#get-started" id="toc-get-started" class="nav-link" data-scroll-target="#get-started">Get Started</a></li>
  <li><a href="#list-of-commands" id="toc-list-of-commands" class="nav-link" data-scroll-target="#list-of-commands">List of commands</a></li>
  <li><a href="#commands-and-options" id="toc-commands-and-options" class="nav-link" data-scroll-target="#commands-and-options">Commands and options</a>
  <ul class="collapse">
  <li><a href="#common-options-for-all-tools" id="toc-common-options-for-all-tools" class="nav-link" data-scroll-target="#common-options-for-all-tools">Common options for all tools</a></li>
  <li><a href="#parse" id="toc-parse" class="nav-link" data-scroll-target="#parse">parse</a></li>
  <li><a href="#fsort" id="toc-fsort" class="nav-link" data-scroll-target="#fsort">fsort</a></li>
  <li><a href="#stream" id="toc-stream" class="nav-link" data-scroll-target="#stream">stream</a></li>
  <li><a href="#addtags" id="toc-addtags" class="nav-link" data-scroll-target="#addtags">addtags</a></li>
  <li><a href="#sam2bam" id="toc-sam2bam" class="nav-link" data-scroll-target="#sam2bam">sam2bam</a></li>
  <li><a href="#rmdup" id="toc-rmdup" class="nav-link" data-scroll-target="#rmdup">rmdup</a></li>
  <li><a href="#pick" id="toc-pick" class="nav-link" data-scroll-target="#pick">pick</a></li>
  <li><a href="#anno" id="toc-anno" class="nav-link" data-scroll-target="#anno">anno</a></li>
  <li><a href="#corr" id="toc-corr" class="nav-link" data-scroll-target="#corr">corr</a></li>
  <li><a href="#attrcnt" id="toc-attrcnt" class="nav-link" data-scroll-target="#attrcnt">attrcnt</a></li>
  <li><a href="#extract" id="toc-extract" class="nav-link" data-scroll-target="#extract">extract</a></li>
  <li><a href="#count" id="toc-count" class="nav-link" data-scroll-target="#count">count</a></li>
  <li><a href="#bam2fq" id="toc-bam2fq" class="nav-link" data-scroll-target="#bam2fq">bam2fq</a></li>
  <li><a href="#bam2frag" id="toc-bam2frag" class="nav-link" data-scroll-target="#bam2frag">bam2frag</a></li>
  <li><a href="#depth" id="toc-depth" class="nav-link" data-scroll-target="#depth">depth</a></li>
  <li><a href="#callept" id="toc-callept" class="nav-link" data-scroll-target="#callept">callept</a></li>
  <li><a href="#count2" id="toc-count2" class="nav-link" data-scroll-target="#count2">count2</a></li>
  <li><a href="#mergebed" id="toc-mergebed" class="nav-link" data-scroll-target="#mergebed">mergebed</a></li>
  <li><a href="#annobed" id="toc-annobed" class="nav-link" data-scroll-target="#annobed">annobed</a></li>
  <li><a href="#flatten" id="toc-flatten" class="nav-link" data-scroll-target="#flatten">flatten</a></li>
  <li><a href="#gtffmt" id="toc-gtffmt" class="nav-link" data-scroll-target="#gtffmt">gtffmt</a></li>
  <li><a href="#gtf2bed" id="toc-gtf2bed" class="nav-link" data-scroll-target="#gtf2bed">gtf2bed</a></li>
  </ul></li>
  <li><a href="#deprecated-commands" id="toc-deprecated-commands" class="nav-link" data-scroll-target="#deprecated-commands">Deprecated commands</a>
  <ul class="collapse">
  <li><a href="#parse0" id="toc-parse0" class="nav-link" data-scroll-target="#parse0">parse0</a></li>
  </ul></li>
  <li><a href="#difference-between-pisa-and-cellranger" id="toc-difference-between-pisa-and-cellranger" class="nav-link" data-scroll-target="#difference-between-pisa-and-cellranger">Difference between PISA and CellRanger</a></li>
  <li><a href="#changelog" id="toc-changelog" class="nav-link" data-scroll-target="#changelog">Changelog</a></li>
  <li><a href="#todo-list" id="toc-todo-list" class="nav-link" data-scroll-target="#todo-list">TODO list</a></li>
  <li><a href="#reporting-issues" id="toc-reporting-issues" class="nav-link" data-scroll-target="#reporting-issues">Reporting Issues</a></li>
  <li><a href="#citation" id="toc-citation" class="nav-link" data-scroll-target="#citation">Citation</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content column-page-left" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">PISA User Guide</h1>
</div>



<div class="quarto-title-meta column-page-left">

    
  
    
  </div>
  


</header>


<section id="synopsis" class="level2">
<h2 class="anchored" data-anchor-id="synopsis">Synopsis</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">PISA</span> <span class="pp">[</span><span class="ss">tool</span><span class="pp">]</span> <span class="pp">[</span><span class="ss">options</span><span class="pp">]</span> <span class="pp">[</span><span class="ss">input</span><span class="pp">-</span><span class="ss">file</span><span class="pp">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="install" class="level2">
<h2 class="anchored" data-anchor-id="install">Install</h2>
<p>PISA source code can be downloaded at <a href="https://github.com/shiquan/PISA/releases" class="uri">https://github.com/shiquan/PISA/releases</a>, or the development version from <a href="https://github.com/shiquan/PISA/" class="uri">https://github.com/shiquan/PISA/</a>. To compile PISA from sources run <strong>make</strong> in the source directory.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> git clone https://github.com/shiquan/PISA</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd PISA</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> make</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="get-started" class="level2">
<h2 class="anchored" data-anchor-id="get-started">Get Started</h2>
<p>The code snippet below demonstrates how to use the PISA tools to process test data and generate various feature counts. You can find the test data in the PISA/demo directory. This example provides a practical approach to familiarize yourself with the functionality of PISA and to validate its operations with provided sample data.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>$ cd demo</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>$ ls</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>aln.sam.gz  barcodes.txt  demo_1.fq.gz  demo_2.fq.gz  demo.gtf.gz  peaks.bed  README.md  var.vcf.gz</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>$ zcat demo_1.fq.gz|head -n 4</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>@A00984:220:HNJ7KDRXX:1:1118:2510:4586</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>AAGCATCCACACAGAGCACCCCGTTCTT</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>+</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>FFFFFFFFFFFFFFFFFFFFFFFFFFFF</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>$ zcat demo_2.fq.gz|head -n 4</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>@A00984:220:HNJ7KDRXX:1:1118:2510:4586</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>+</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="fu"># Convert raw FASTQ to FASTQ+ format</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>$ PISA parse -rule 'CR,R1:1-16,barcodes.txt,CB,1;UR,R1:17-28;R1,R2' demo_1.fq.gz demo_2.fq.gz -1 demo.fq</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>Number of Fragments,825</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>Fragments pass QC,825</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>Fragments with Exactly Matched Barcodes,805</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>Fragments with Failed Barcodes,0</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:21:35</span><span class="co">]</span> Real time: 0.003 sec; CPU: 0.009 sec; Peak RSS: 0.010 GB.</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>$ head -n 4 demo.fq</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>@A00984:220:HNJ7KDRXX:1:1118:2510:4586|||CR:Z:AAGCATCCACACAGAG|||CB:Z:AAGCATCCACACAGAG|||UR:Z:CACCCCGTTCTT</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>+</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a><span class="fu"># Alignment results of FASTQ+ </span></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>$ samtools view aln.sam.gz|head -n 1</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>A00984:220:HNJ7KDRXX:1:1118:2510:4586|||CR:Z:AAGCATCCACACAGAG|||CB:Z:AAGCATCCACACAGAG|||UR:Z:CACCCCGTTCTT   0   chr11   35139165    255 26S65M  *   0   0   GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF NH:i:1  HI:i:1  AS:i:61 nM:i:1</span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a><span class="fu"># Convert format alignment records from SAM to BAM</span></span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>$ PISA sam2bam aln.sam.gz -o aln.bam</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>Raw reads,825</span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>Mapped reads,820 (99.39%)</span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>Plus strand,820</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a>Minus strand,0</span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>Mitochondria ratio,0.00%</span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:26:31</span><span class="co">]</span> Real time: 0.005 sec; CPU: 0.009 sec; Peak RSS: 0.010 GB.</span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>$ samtools view aln.bam|head -n 1</span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a>A00984:220:HNJ7KDRXX:1:1118:2510:4586   0   chr11   35139165    255 26S65M  *   0   0   GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF NH:i:1  HI:i:1  AS:i:61 nM:i:1  CR:Z:AAGCATCCACACAGAG   CB:Z:AAGCATCCACACAGAG   UR:Z:CACCCCGTTCTT</span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a><span class="fu"># Annotate gene names for BAM</span></span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a>$ PISA anno -gtf ./demo.gtf.gz aln.bam -o anno_gtf.bam</span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:28:38</span><span class="co">]</span> GTF loading..</span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:28:38</span><span class="co">]</span> Load 2 genes.</span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:28:38</span><span class="co">]</span> Load time : 0.003 sec</span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a>Reads Mapped to Genome (Map Quality &gt;= 0),99.4%</span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a>Reads Mapped to Exonic Regions,99.3%</span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a>Reads Mapped to Intronic Regions,0.0%</span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a>Reads Mapped to both Exonic and Intronic Regions,0.7%</span>
<span id="cb3-57"><a href="#cb3-57" aria-hidden="true" tabindex="-1"></a>Reads Mapped Antisense to Gene,0.0%</span>
<span id="cb3-58"><a href="#cb3-58" aria-hidden="true" tabindex="-1"></a>Reads Mapped to Intergenic Regions,0.0%</span>
<span id="cb3-59"><a href="#cb3-59" aria-hidden="true" tabindex="-1"></a>Reads Mapped to Gene but Failed to Interpret Type,0.0%</span>
<span id="cb3-60"><a href="#cb3-60" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:28:38</span><span class="co">]</span> Real time: 0.026 sec; CPU: 0.086 sec; Speed : 9528 records/sec; Peak RSS: 0.034 GB.</span>
<span id="cb3-61"><a href="#cb3-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-62"><a href="#cb3-62" aria-hidden="true" tabindex="-1"></a><span class="fu"># Correct UMIs amongst other UMIs from the same cell and mapped to the same gene, and create new tag UB for corrected UMIs</span></span>
<span id="cb3-63"><a href="#cb3-63" aria-hidden="true" tabindex="-1"></a>$ PISA corr -tag UR -new-tag UB -tags-block CB,GN anno_gtf.bam -o corr.bam</span>
<span id="cb3-64"><a href="#cb3-64" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:36:21</span><span class="co">]</span> Building index ..</span>
<span id="cb3-65"><a href="#cb3-65" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:36:21</span><span class="co">]</span> Build time : 0.002 sec</span>
<span id="cb3-66"><a href="#cb3-66" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:36:21</span><span class="co">]</span> Real time: 0.077 sec; CPU: 0.085 sec</span>
<span id="cb3-67"><a href="#cb3-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-68"><a href="#cb3-68" aria-hidden="true" tabindex="-1"></a>$ samtools view corr.bam|head -n 1</span>
<span id="cb3-69"><a href="#cb3-69" aria-hidden="true" tabindex="-1"></a>A00984:220:HNJ7KDRXX:1:1118:2510:4586   0   chr11   35139165    255 26S65M  *   0   0   GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF NH:i:1  HI:i:1  AS:i:61 nM:i:1  CR:Z:AAGCATCCACACAGAG   CB:Z:AAGCATCCACACAGAG   UR:Z:CACCCCGTTCTT   RE:A:E  GX:Z:ENSG00000026508.18 GN:Z:CD44   TX:Z:ENST00000263398.10,ENST00000428726.7,ENST00000526025.2 UB:Z:CACCCCGTTCTT</span>
<span id="cb3-70"><a href="#cb3-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-71"><a href="#cb3-71" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count gene X cell features</span></span>
<span id="cb3-72"><a href="#cb3-72" aria-hidden="true" tabindex="-1"></a>$ mkdir exp</span>
<span id="cb3-73"><a href="#cb3-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-74"><a href="#cb3-74" aria-hidden="true" tabindex="-1"></a>$ PISA count  -cb CB -anno-tag GN -outdir exp -umi UB corr.bam </span>
<span id="cb3-75"><a href="#cb3-75" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:38:44</span><span class="co">]</span> Real time: 0.033 sec; CPU: 0.013 sec; Peak RSS: 0.010 GB.</span>
<span id="cb3-76"><a href="#cb3-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-77"><a href="#cb3-77" aria-hidden="true" tabindex="-1"></a><span class="fu"># Gene expression matrix generated in the Market Exchange format </span></span>
<span id="cb3-78"><a href="#cb3-78" aria-hidden="true" tabindex="-1"></a>$ ls exp/</span>
<span id="cb3-79"><a href="#cb3-79" aria-hidden="true" tabindex="-1"></a>barcodes.tsv.gz  features.tsv.gz  matrix.mtx.gz</span>
<span id="cb3-80"><a href="#cb3-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-81"><a href="#cb3-81" aria-hidden="true" tabindex="-1"></a><span class="fu"># Not just gene features, we can also annotate variants and functional regions to reads</span></span>
<span id="cb3-82"><a href="#cb3-82" aria-hidden="true" tabindex="-1"></a>$ PISA anno -bed peaks.bed -tag PK -vcf var.vcf.gz -vtag VR  corr.bam -o anno_vcf_bed.bam    </span>
<span id="cb3-83"><a href="#cb3-83" aria-hidden="true" tabindex="-1"></a>Reads Mapped to Genome (Map Quality &gt;= 0),99.4%</span>
<span id="cb3-84"><a href="#cb3-84" aria-hidden="true" tabindex="-1"></a>Reads Mapped to BED regions / Peaks,0.0%</span>
<span id="cb3-85"><a href="#cb3-85" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:43:01</span><span class="co">]</span> Real time: 0.027 sec; CPU: 0.090 sec; Speed : 9085 records/sec; Peak RSS: 0.034 GB.</span>
<span id="cb3-86"><a href="#cb3-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-87"><a href="#cb3-87" aria-hidden="true" tabindex="-1"></a>$ samtools view anno_vcf_bed.bam|grep "VR:Z"|grep "PK:Z"|head -n 1</span>
<span id="cb3-88"><a href="#cb3-88" aria-hidden="true" tabindex="-1"></a>A00984:220:HNJ7KDRXX:1:2266:27597:30843 0   chr11   35229688    255 91M *   0   0   CCCAGGGTTAATAGGGCCTGGTCCCTGGGAGGAAATTTGAATGGGTCCATTTTGCCCTTCCATAGCCTAATCCCGGGGCATTGTTTTCCAC</span>
<span id="cb3-89"><a href="#cb3-89" aria-hidden="true" tabindex="-1"></a>FFFF,FFFFFFFFFF,FFFFFFF:,:FFFFFF,FF:FFFFFFFFFFF:,,:F:F:F,F:F:F,FFFF,F:FFFF,FF,FF:FFFFFFF:F: NH:i:1  HI:i:1  AS:i:85 nM:i:2  CR:Z:ATTGTTCCAAGTCCCG   CB:Z:ATTGTTCCAAGTCCCG   UR:Z:TCTTTAAGTCAG   RE:A:E  GX:Z:ENSG00000026508.18 GN:Z:CD44   TX:Z:ENST00000263398.10,ENST00000428726.7,ENST00000425428.6,ENST00000433892.6,ENST00000525469.1 UB:Z:TCTTTAAGTCAG   PK:Z:demo_peak_14a;demo_peak_14b    VR:Z:chr11:35229771C&gt;T</span>
<span id="cb3-90"><a href="#cb3-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-91"><a href="#cb3-91" aria-hidden="true" tabindex="-1"></a><span class="fu"># Summarize the reads, UMIs, genes, peaks, and variants per cell</span></span>
<span id="cb3-92"><a href="#cb3-92" aria-hidden="true" tabindex="-1"></a>$ PISA attrcnt -cb CB -tags UB,GN,PK,VR anno_vcf_bed.bam -dedup |head -n 5</span>
<span id="cb3-93"><a href="#cb3-93" aria-hidden="true" tabindex="-1"></a>BARCODE Raw UB  GN  PK  VR</span>
<span id="cb3-94"><a href="#cb3-94" aria-hidden="true" tabindex="-1"></a>AAGCATCCACACAGAG    503 132 1   15  1</span>
<span id="cb3-95"><a href="#cb3-95" aria-hidden="true" tabindex="-1"></a>ATTGTTCCAAGTCCCG    533 124 1   13  1</span>
<span id="cb3-96"><a href="#cb3-96" aria-hidden="true" tabindex="-1"></a>AAGCATCCACACNGAG    3   1   1   1   0</span>
<span id="cb3-97"><a href="#cb3-97" aria-hidden="true" tabindex="-1"></a>AAGCNTCCACACAGAG    3   1   1   1   0</span>
<span id="cb3-98"><a href="#cb3-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-99"><a href="#cb3-99" aria-hidden="true" tabindex="-1"></a><span class="fu"># Deduplicate BAM file for each cell</span></span>
<span id="cb3-100"><a href="#cb3-100" aria-hidden="true" tabindex="-1"></a>$ PISA rmdup -tags CB corr.bam -o rmdup.bam -nw </span>
<span id="cb3-101"><a href="#cb3-101" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:59:39</span><span class="co">]</span> Deduplicating chr11</span>
<span id="cb3-102"><a href="#cb3-102" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:59:39</span><span class="co">]</span> All reads,820</span>
<span id="cb3-103"><a href="#cb3-103" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:59:39</span><span class="co">]</span> Duplicate reads,125</span>
<span id="cb3-104"><a href="#cb3-104" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:59:39</span><span class="co">]</span> Duplicate ratio,0.1524</span>
<span id="cb3-105"><a href="#cb3-105" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 12:59:39</span><span class="co">]</span> Real time: 0.008 sec; CPU: 0.015 sec; Peak RSS: 0.010 GB.</span>
<span id="cb3-106"><a href="#cb3-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-107"><a href="#cb3-107" aria-hidden="true" tabindex="-1"></a><span class="fu"># Deduplicate BAM file for each molecular</span></span>
<span id="cb3-108"><a href="#cb3-108" aria-hidden="true" tabindex="-1"></a>$ PISA rmdup -tags CB,UR corr.bam -o rmdup1.bam -nw</span>
<span id="cb3-109"><a href="#cb3-109" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:00:35</span><span class="co">]</span> Deduplicating chr11</span>
<span id="cb3-110"><a href="#cb3-110" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:00:35</span><span class="co">]</span> All reads,820</span>
<span id="cb3-111"><a href="#cb3-111" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:00:35</span><span class="co">]</span> Duplicate reads,0</span>
<span id="cb3-112"><a href="#cb3-112" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:00:35</span><span class="co">]</span> Duplicate ratio,0.0000</span>
<span id="cb3-113"><a href="#cb3-113" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:00:35</span><span class="co">]</span> Real time: 0.009 sec; CPU: 0.015 sec; Peak RSS: 0.011 GB.</span>
<span id="cb3-114"><a href="#cb3-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-115"><a href="#cb3-115" aria-hidden="true" tabindex="-1"></a><span class="fu"># Select all reads annotated to gene CD44</span></span>
<span id="cb3-116"><a href="#cb3-116" aria-hidden="true" tabindex="-1"></a><span class="fu"># Generate a gene candidate list</span></span>
<span id="cb3-117"><a href="#cb3-117" aria-hidden="true" tabindex="-1"></a>$  echo "CD44" &gt; gene.txt</span>
<span id="cb3-118"><a href="#cb3-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-119"><a href="#cb3-119" aria-hidden="true" tabindex="-1"></a>$ PISA pick -tags GN -list gene.txt anno_vcf_bed.bam  -o picked.bam</span>
<span id="cb3-120"><a href="#cb3-120" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:03:01</span><span class="co">]</span> Real time: 0.009 sec; CPU: 0.016 sec</span>
<span id="cb3-121"><a href="#cb3-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-122"><a href="#cb3-122" aria-hidden="true" tabindex="-1"></a><span class="fu"># Select reads with more features</span></span>
<span id="cb3-123"><a href="#cb3-123" aria-hidden="true" tabindex="-1"></a>$ awk '{printf("%s\tCD44\n", $1)}' barcodes.txt &gt; candidates.txt</span>
<span id="cb3-124"><a href="#cb3-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-125"><a href="#cb3-125" aria-hidden="true" tabindex="-1"></a>$ cat candidates.txt </span>
<span id="cb3-126"><a href="#cb3-126" aria-hidden="true" tabindex="-1"></a>AAGCATCCACACAGAG    CD44</span>
<span id="cb3-127"><a href="#cb3-127" aria-hidden="true" tabindex="-1"></a>ATTGTTCCAAGTCCCG    CD44</span>
<span id="cb3-128"><a href="#cb3-128" aria-hidden="true" tabindex="-1"></a>GCACATAGTCAGTTTG    CD44</span>
<span id="cb3-129"><a href="#cb3-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-130"><a href="#cb3-130" aria-hidden="true" tabindex="-1"></a>$ PISA pick -tags CB,GN -list candidates.txt anno_vcf_bed.bam -o picked.bam </span>
<span id="cb3-131"><a href="#cb3-131" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:09:28</span><span class="co">]</span> Real time: 0.007 sec; CPU: 0.013 sec</span>
<span id="cb3-132"><a href="#cb3-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-133"><a href="#cb3-133" aria-hidden="true" tabindex="-1"></a><span class="fu"># Convert BAM to FASTQ+</span></span>
<span id="cb3-134"><a href="#cb3-134" aria-hidden="true" tabindex="-1"></a>$ PISA bam2fq -tags CB,GN picked.bam -o pick.fq</span>
<span id="cb3-135"><a href="#cb3-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-136"><a href="#cb3-136" aria-hidden="true" tabindex="-1"></a>$ head -n 4 pick.fq </span>
<span id="cb3-137"><a href="#cb3-137" aria-hidden="true" tabindex="-1"></a>@A00984:220:HNJ7KDRXX:1:1118:2510:4586|||CB:Z:AAGCATCCACACAGAG|||GN:Z:CD44</span>
<span id="cb3-138"><a href="#cb3-138" aria-hidden="true" tabindex="-1"></a>GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC</span>
<span id="cb3-139"><a href="#cb3-139" aria-hidden="true" tabindex="-1"></a>+</span>
<span id="cb3-140"><a href="#cb3-140" aria-hidden="true" tabindex="-1"></a>FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF</span>
<span id="cb3-141"><a href="#cb3-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-142"><a href="#cb3-142" aria-hidden="true" tabindex="-1"></a><span class="fu"># Sort FASTQ+ reads by CB and GN</span></span>
<span id="cb3-143"><a href="#cb3-143" aria-hidden="true" tabindex="-1"></a>$ PISA fsort -tags CB,GN pick.fq -o fsort.fq.gz</span>
<span id="cb3-144"><a href="#cb3-144" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:22:50</span><span class="co">]</span> Write 795 records to fsort.fq.gz.0000.bgz.</span>
<span id="cb3-145"><a href="#cb3-145" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:22:50</span><span class="co">]</span> Unlink fsort.fq.gz.0000.bgz</span>
<span id="cb3-146"><a href="#cb3-146" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:22:50</span><span class="co">]</span> Create fsort.fq.gz from 1 files.</span>
<span id="cb3-147"><a href="#cb3-147" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:22:50</span><span class="co">]</span> Real time: 0.021 sec; CPU: 0.020 sec</span>
<span id="cb3-148"><a href="#cb3-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-149"><a href="#cb3-149" aria-hidden="true" tabindex="-1"></a>$ zcat fsort.fq.gz|head -n 8</span>
<span id="cb3-150"><a href="#cb3-150" aria-hidden="true" tabindex="-1"></a>@A00984:220:HNJ7KDRXX:1:1118:2510:4586|||CB:Z:AAGCATCCACACAGAG|||GN:Z:CD44</span>
<span id="cb3-151"><a href="#cb3-151" aria-hidden="true" tabindex="-1"></a>GCAGTGGTATCAACGCAGAGTACATGGGGAGCCTCATTGCCCAGCGGACCCCAGCCTCTGCCAGGTTCGGTCCGCCATCCTCGTCCCGTCC</span>
<span id="cb3-152"><a href="#cb3-152" aria-hidden="true" tabindex="-1"></a>+</span>
<span id="cb3-153"><a href="#cb3-153" aria-hidden="true" tabindex="-1"></a>FFFFF:FFFFFF:FFFFFFFFFF,FFFFFFFFF:FFFFFFFFF,FFFF:FFF,FFFFF:FFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFF</span>
<span id="cb3-154"><a href="#cb3-154" aria-hidden="true" tabindex="-1"></a>@A00984:220:HNJ7KDRXX:1:2143:21640:21496|||CB:Z:AAGCATCCACACAGAG|||GN:Z:CD44</span>
<span id="cb3-155"><a href="#cb3-155" aria-hidden="true" tabindex="-1"></a>CCTGCCCCGCGCCCAGAGATCCTCCAGCTCCTTTCGCCCGCGCCCTACGTTCGCTCCGGACACCATGGACAAGTATTGGTGGAACACAGCC</span>
<span id="cb3-156"><a href="#cb3-156" aria-hidden="true" tabindex="-1"></a>+</span>
<span id="cb3-157"><a href="#cb3-157" aria-hidden="true" tabindex="-1"></a>,,FFFFFFFF,F,,:F,F,F:FFF,FFFFFF,F::F,FF,F:FF:,,FF:FFF,:FFFF:FFF,::FFFF,F:F,FFFF,,F,FFF,F,::</span>
<span id="cb3-158"><a href="#cb3-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-159"><a href="#cb3-159" aria-hidden="true" tabindex="-1"></a><span class="fu"># Assembly reads mapped to CD44 of the same cell into contigs</span></span>
<span id="cb3-160"><a href="#cb3-160" aria-hidden="true" tabindex="-1"></a><span class="fu"># This step requires Trinity software and seqtk already installed in your environment</span></span>
<span id="cb3-161"><a href="#cb3-161" aria-hidden="true" tabindex="-1"></a>$ PISA stream -tags CB,GN -script 'Trinity --seqType fq --SS_lib_type F --single ${FQ} --max_memory 1G 2&gt;/dev/null 1&gt;/dev/null; seqtk rename trinity_out_dir.Trinity.fasta ${UBI}_ 2&gt;/dev/null' -t 10 -fa  -nw ./fsort.fq.gz -o assem.fa</span>
<span id="cb3-162"><a href="#cb3-162" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:24:21</span><span class="co">]</span> Real time: 5.607 sec; CPU: 0.010 sec; Peak RSS: 0.010 GB.</span>
<span id="cb3-163"><a href="#cb3-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-164"><a href="#cb3-164" aria-hidden="true" tabindex="-1"></a>$ seqtk seq assem.fa -l 100 |head </span>
<span id="cb3-165"><a href="#cb3-165" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;Z_AAGCATCCACACAGAG_Z_CD44_1|||CB:Z:AAGCATCCACACAGAG|||GN:Z:CD44 len=439 path=</span><span class="co">[</span><span class="ot">0:0-438</span><span class="co">]</span></span>
<span id="cb3-166"><a href="#cb3-166" aria-hidden="true" tabindex="-1"></a><span class="at">GAAATTAGGGCCCAATTAATAATCAGCAAGAATTTGATCGTTCCAGTTCCCACTTGGAGGCCTTTCATCCCTCGGGTGTGCTATGGATGGCTTCTAACAA</span></span>
<span id="cb3-167"><a href="#cb3-167" aria-hidden="true" tabindex="-1"></a><span class="at">AAACTACACATATGTATTCCTGATCGCCAACCTTTCCCCCACCAGCTAAGGACATTTCCCAGGGTTAATAGGGCCTGGTCCCTGGGAGGAAATTTGAATG</span></span>
<span id="cb3-168"><a href="#cb3-168" aria-hidden="true" tabindex="-1"></a><span class="at">GGTCCATTTTGCCCTTCCATAGCCTAATCCCTGGGCATTGTTTTCCACTGAGGTTGGGGGTTGGGGTGTACTAGTTACACATCTTCAACAGACCCCCTCT</span></span>
<span id="cb3-169"><a href="#cb3-169" aria-hidden="true" tabindex="-1"></a><span class="at">AGAAATTTTTCAGATGCTTCTGGGAGACACCCAAAGGGTGAAGCTATTTATCTGTAGTAAACTATTTATCTGTGTTTTTGAAATATTAAACCCTGGATCA</span></span>
<span id="cb3-170"><a href="#cb3-170" aria-hidden="true" tabindex="-1"></a><span class="at">GTCCTTTGATCAGTATAATTTTTTAAAGTTACTTTGTCA</span></span>
<span id="cb3-171"><a href="#cb3-171" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;Z_AAGCATCCACACAGAG_Z_CD44_2|||CB:Z:AAGCATCCACACAGAG|||GN:Z:CD44 len=384 path=</span><span class="co">[</span><span class="ot">0:0-383</span><span class="co">]</span></span>
<span id="cb3-172"><a href="#cb3-172" aria-hidden="true" tabindex="-1"></a><span class="at">CCTGGTAGAATTGGCTTTTCTAGCAGAACCTTTCCAAAAGTTTTATATTGAGATTCATAACAACACCAAGAATTGATTTTGTAGCCAACATTCATTCAAT</span></span>
<span id="cb3-173"><a href="#cb3-173" aria-hidden="true" tabindex="-1"></a><span class="at">ACTGTTATATCAGAGGAGTAGGAGAGAGGAAACATTTGACTTATCTGGAAAAGCAAAATGTACTTAAGAATAAGAATAACATGGTCCATTCACCTTTATG</span></span>
<span id="cb3-174"><a href="#cb3-174" aria-hidden="true" tabindex="-1"></a><span class="at">TTATAGATATGTCTTTGTGTAAATCATTTGTTTTGAGTTTTCAAAGAATAGCCCATTGTTCATTCTTGTGCTGTACAATGACCACTGTTATTGTTACTTT</span></span>
<span id="cb3-175"><a href="#cb3-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-176"><a href="#cb3-176" aria-hidden="true" tabindex="-1"></a><span class="fu"># Align assembled reads to reference and convert to BAM file</span></span>
<span id="cb3-177"><a href="#cb3-177" aria-hidden="true" tabindex="-1"></a><span class="fu"># Here I use minimap2 for simplicity</span></span>
<span id="cb3-178"><a href="#cb3-178" aria-hidden="true" tabindex="-1"></a>$ minimap2 -x splice -a ~/Documents/datasets/GRCh38/fasta/genome.fasta assem.fa 1&gt; asm_aln.sam</span>
<span id="cb3-179"><a href="#cb3-179" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::mm_idx_gen::50.495*1.81</span><span class="co">]</span> collected minimizers</span>
<span id="cb3-180"><a href="#cb3-180" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::mm_idx_gen::71.980*2.16</span><span class="co">]</span> sorted minimizers</span>
<span id="cb3-181"><a href="#cb3-181" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::main::71.980*2.16</span><span class="co">]</span> loaded/built the index for 194 target sequence(s)</span>
<span id="cb3-182"><a href="#cb3-182" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::mm_mapopt_update::75.057*2.11</span><span class="co">]</span> mid_occ = 767</span>
<span id="cb3-183"><a href="#cb3-183" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::mm_idx_stat</span><span class="co">]</span> kmer size: 15; skip: 5; is_hpc: 0; #seq: 194</span>
<span id="cb3-184"><a href="#cb3-184" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::mm_idx_stat::77.004*2.09</span><span class="co">]</span> distinct minimizers: 167225302 (35.42% are singletons); average occurrences: 6.036; average spacing: 3.071; total length: 3099750718</span>
<span id="cb3-185"><a href="#cb3-185" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::worker_pipeline::77.010*2.09</span><span class="co">]</span> mapped 13 sequences</span>
<span id="cb3-186"><a href="#cb3-186" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::main</span><span class="co">]</span> Version: 2.21-r1071</span>
<span id="cb3-187"><a href="#cb3-187" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::main</span><span class="co">]</span> CMD: minimap2 -x splice -a /home/shiquan/Documents/datasets/GRCh38/fasta/genome.fasta assem.fa</span>
<span id="cb3-188"><a href="#cb3-188" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">M::main</span><span class="co">]</span> Real time: 77.358 sec; CPU: 161.000 sec; Peak RSS: 18.519 GB</span>
<span id="cb3-189"><a href="#cb3-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-190"><a href="#cb3-190" aria-hidden="true" tabindex="-1"></a>$ PISA sam2bam asm_aln.sam -o asm_aln.bam</span>
<span id="cb3-191"><a href="#cb3-191" aria-hidden="true" tabindex="-1"></a>Raw reads,13</span>
<span id="cb3-192"><a href="#cb3-192" aria-hidden="true" tabindex="-1"></a>Mapped reads,13 (100.00%)</span>
<span id="cb3-193"><a href="#cb3-193" aria-hidden="true" tabindex="-1"></a>Plus strand,13</span>
<span id="cb3-194"><a href="#cb3-194" aria-hidden="true" tabindex="-1"></a>Minus strand,0</span>
<span id="cb3-195"><a href="#cb3-195" aria-hidden="true" tabindex="-1"></a>Mitochondria ratio,0.00%</span>
<span id="cb3-196"><a href="#cb3-196" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">2022-04-22 13:30:20</span><span class="co">]</span> Real time: 0.001 sec; CPU: 0.006 sec; Peak RSS: 0.010 GB.</span>
<span id="cb3-197"><a href="#cb3-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-198"><a href="#cb3-198" aria-hidden="true" tabindex="-1"></a>$ samtools view asm_aln.bam|head -n 1</span>
<span id="cb3-199"><a href="#cb3-199" aria-hidden="true" tabindex="-1"></a>Z_AAGCATCCACACAGAG_Z_CD44_1 0   chr11   35229531    60  439M    *   0   0   GAAATTAGGGCCCAATTAATAATCAGCAAGAATTTGATCGTTCCAGTTCCCACTTGGAGGCCTTTCATCCCTCGGGTGTGCTATGGATGGCTTCTAACAA</span>
<span id="cb3-200"><a href="#cb3-200" aria-hidden="true" tabindex="-1"></a>AAACTACACATATGTATTCCTGATCGCCAACCTTTCCCCCACCAGCTAAGGACATTTCCCAGGGTTAATAGGGCCTGGTCCCTGGGAGGAAATTTGAATG</span>
<span id="cb3-201"><a href="#cb3-201" aria-hidden="true" tabindex="-1"></a>GGTCCATTTTGCCCTTCCATAGCCTAATCCCTGGGCATTGTTTTCCACTGAGGTTGGGGGTTGGGGTGTACTAGTTACACATCTTCAACAGACCCCCTCT</span>
<span id="cb3-202"><a href="#cb3-202" aria-hidden="true" tabindex="-1"></a>AGAAATTTTTCAGATGCTTCTGGGAGACACCCAAAGGGTGAAGCTATTTATCTGTAGTAAACTATTTATCTGTGTTTTTGAAATATTAAACCCTGGATCA</span>
<span id="cb3-203"><a href="#cb3-203" aria-hidden="true" tabindex="-1"></a>GTCCTTTGATCAGTATAATTTTTTAAAGTTACTTTGTCA *   NM:i:1  ms:i:436    AS:i:436    nn:i:0  tp:A:P  cm:i:137    s1:i:430    s2:i:0  de:f:0.0023 rl:i:0  CB:Z:AAGCATCCACACAGAG   GN:Z:CD44</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="list-of-commands" class="level2">
<h2 class="anchored" data-anchor-id="list-of-commands">List of commands</h2>
<table class="table-striped table-hover caption-top table">
<colgroup>
<col style="width: 23%">
<col style="width: 76%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Tool</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"></td>
<td style="text-align: left;"><strong><em>The following tools are used to process FASTQ/FASTQ+ files.</em></strong></td>
</tr>
<tr class="even">
<td style="text-align: left;">parse</td>
<td style="text-align: left;">Parse barcodes from FASTQ reads to FASTQ+.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">fsort</td>
<td style="text-align: left;">Sort FASTQ+ records by barcodes.</td>
</tr>
<tr class="even">
<td style="text-align: left;">stream</td>
<td style="text-align: left;">Perform user-defined process for each read block.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">addtags</td>
<td style="text-align: left;">Add tag string to FASTQ reads.</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;"><strong><em>The following tools are used to process BAM files.</em></strong></td>
</tr>
<tr class="odd">
<td style="text-align: left;">sam2bam</td>
<td style="text-align: left;">Parse FASTQ+ read name and convert SAM to BAM.</td>
</tr>
<tr class="even">
<td style="text-align: left;">rmdup</td>
<td style="text-align: left;">Remove PCR duplicates per molecular.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">pick</td>
<td style="text-align: left;">Pick alignments with tags.</td>
</tr>
<tr class="even">
<td style="text-align: left;">anno</td>
<td style="text-align: left;">Annotate functional regions or gene names.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">corr</td>
<td style="text-align: left;">Correct error prone UMIs. 1 mismatch considered.</td>
</tr>
<tr class="even">
<td style="text-align: left;">attrcnt</td>
<td style="text-align: left;">Count raw reads and tag values per cell.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">extract</td>
<td style="text-align: left;">Extract tag value from BAM.</td>
</tr>
<tr class="even">
<td style="text-align: left;">count</td>
<td style="text-align: left;">Count feature X cell matrix from BAMs.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">bam2fq</td>
<td style="text-align: left;">Convert BAM to FASTQ+ file with selected tags.</td>
</tr>
<tr class="even">
<td style="text-align: left;">bam2frag</td>
<td style="text-align: left;">Generate fragment file.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">depth</td>
<td style="text-align: left;">Coverage depth/UMI for target regions.</td>
</tr>
<tr class="even">
<td style="text-align: left;">addtags</td>
<td style="text-align: left;">Add tag string to BAM alignments.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">callept</td>
<td style="text-align: left;">Call expressed peak tags (EPTs) for RNA library.</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;"><strong><em>The following tool used to process fragment file.</em></strong></td>
</tr>
<tr class="odd">
<td style="text-align: left;">count2</td>
<td style="text-align: left;">Count peak X cell matrix from fragment file.</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;"><strong><em>The following tools used to process BED files.</em></strong></td>
</tr>
<tr class="odd">
<td style="text-align: left;">mergebed</td>
<td style="text-align: left;">Merge BED files.</td>
</tr>
<tr class="even">
<td style="text-align: left;">annobed</td>
<td style="text-align: left;">Annotate BED files with genes and functional elements.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">flatten</td>
<td style="text-align: left;">Flatten overlapped regions to nonoverlaps.</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;"><strong><em>The following tools used to process GTF files.</em></strong></td>
</tr>
<tr class="odd">
<td style="text-align: left;">gtffmt</td>
<td style="text-align: left;">Format and reorder GTF file.</td>
</tr>
<tr class="even">
<td style="text-align: left;">gtf2bed</td>
<td style="text-align: left;">Convert GTF to BED.</td>
</tr>
</tbody>
</table>
</section>
<section id="commands-and-options" class="level2">
<h2 class="anchored" data-anchor-id="commands-and-options">Commands and options</h2>
<section id="common-options-for-all-tools" class="level3">
<h3 class="anchored" data-anchor-id="common-options-for-all-tools">Common options for all tools</h3>
<dl>
<dt>
<strong>-h</strong>
</dt>
<dd>
Help information
</dd>
<dt>
<strong>-o</strong> <em>FILE</em>
</dt>
<dd>
Output file.
</dd>
<dt>
<strong>-tags</strong> <em>tag(s)</em>
</dt>
<dd>
Barcode tags to group reads, usually be cell barcode tag.
</dd>
<dt>
<strong>-umi</strong> <em>tag</em>
</dt>
<dd>
UMI tag.
</dd>
<dt>
<strong>-t/-@</strong> <em>number</em>
</dt>
<dd>
Multithreads to process data.
</dd>
<dt>
<strong>-report</strong> <em>FILE.csv</em>
</dt>
<dd>
Summary report in csv format.
</dd>
</dl>
</section>
<section id="parse" class="level3">
<h3 class="anchored" data-anchor-id="parse">parse</h3>
<p>The parse tool is specifically designed to convert FASTQ files into the extended FASTQ+ format. It utilizes the -rule option to define the library structure, accommodating various sequencing setups. For ease of use, this tool has included predefined common library structures in the release; these can be applied using the -x option. This tool is optimized for speed and supports the correction of barcodes that have up to one mismatch.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Parse cell barcode and UMI string from raw FASTQ.</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>$ PISA parse -rule CB,R1:1-10,whitelist.txt,CB,1;R1,R1:11-60;R2,R2 -report fastq.csv lane1_1.fq.gz,lane02_1.fq.gz lane1_2.fq.gz,lane2_2.fq.gz</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a> -1       <span class="co">[</span><span class="ot">fastq</span><span class="co">]</span>   Read 1 output.</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a> -2       <span class="co">[</span><span class="ot">fastq</span><span class="co">]</span>   Read 2 output.</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a> -rule    <span class="co">[</span><span class="ot">STRING</span><span class="co">]</span>  Read structure in line. See Notice.</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a> -p                 Read 1 and read 2 interleaved in the input file.</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a> -q       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Drop reads if average sequencing quality below this value.</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a> -dropN             Drop reads if N base in sequence or barcode.</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a> -report  <span class="co">[</span><span class="ot">csv</span><span class="co">]</span>     Summary report.</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a> -t       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Threads. <span class="co">[</span><span class="ot">4</span><span class="co">]</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a> -order             Keep input order.</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a> -x                 Predefined code for specific library.</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="ss">          * </span>C4      Library structure for DNBelab C4 RNA kit v1.</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>-rule accept tag rule STRING to parse input fastq following format "TAG,location,whitelist,corrected TAG,allow mismatch".</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>   For each tag rule, location part should be format like R<span class="co">[</span><span class="ot">12</span><span class="co">]</span>:start-end. Both start and end location start from 1.</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>   TAG and locaion parts are mandatory, and whitelist, corrected TAG and mismatch are optional.</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>   Futhermore, multiply tags seperated by ';'. In location part, R1 stands for raw read 1, R2 stands for raw read 2.</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>   In tag part, R1 stands for output read 1 while R2 stands for output read 2. Here are some examples.</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>$ PISA parse -rule 'CR,R1:1-18,barcodes.txt,CB,1;UR,R1:19-30;R1,R2:1-100' -1 read_1.fq raw_read_1.fq raw_read_2.fq</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a><span class="fu"># CR,R1:1-18,barcodes.txt,CB,1  - CR tag start from 1 to 18 in read 1, and barcodes.txt are barcode whitelist,</span></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="fu">#   each barcode per line. Cell barcode will be corrected while hamming distance &lt;= 1.</span></span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a><span class="fu">#   Corrected cell barcode tag is CB. </span></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a><span class="fu"># UR,R1:19-30 - UR tag start from 19-30 in read 1.</span></span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a><span class="fu"># R1,R2:1-100 - Sequence from 1 to 100 in read 2 output to read 1 file. </span></span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>$ PISA parse -rule 'CR,R1:1-10,bc1.txt,CB,1;CR,R1:11-20,bc2.txt,CB,1;R1,R2:1-100' -1 read_1.fq raw_read_1.fq raw_read_2.fq</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a><span class="fu"># CR,R1:1-10,bc1.txt,CB,1;CR,R1:11-20,bc2.txt,CB,1 - This cell barcode consist of two segments, first segment start</span></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a><span class="fu">#   from 1 to 10 in read 1, and whitelist is bc1.txt, and second segment start from 11 to 20, and whitelist is bc2.txt.</span></span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a><span class="fu">#   These two segments will be combined after correction, because the corrected tag are the same.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Option -report can specify a quality control report in CSV format. Here is the explanation of each term in this file.</p>
<div class="center">
<table class="caption-top table">
<colgroup>
<col style="width: 36%">
<col style="width: 63%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"><strong>Terms</strong></th>
<th style="text-align: left;"><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Number of Fragments</td>
<td style="text-align: left;">The number of records in the FASTQ(s). For paired reads, each pair only count once.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Fragments pass QC</td>
<td style="text-align: left;">Reads or paired reads pass QC.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Fragments with Exactly Matched Barcodes</td>
<td style="text-align: left;">Barcodes exactly matched with any barcode in the candidate list.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Fragments with Failed Barcodes</td>
<td style="text-align: left;">No barcode found in the candidate list with similar search.</td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="fsort" class="level3">
<h3 class="anchored" data-anchor-id="fsort">fsort</h3>
<p>The fsort tool is engineered to order FASTQ+ records based on specified tags, which can be defined using the -tags option. This tool efficiently handles file sorting: for files smaller than 1 gigabyte, it performs the sort directly. However, for larger FASTQ files, where caching the entire file in memory is impractical, fsort employs a different strategy. It splits the large file into smaller segments, sorts each segment individually, and then merges the sorted segments. This method ensures efficient handling of large datasets while maintaining the integrity and order of the FASTQ+ records.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Sort reads by tags.</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>$ PISA fsort -tags CB,UR -list cell_barcodes_top10K.txt -@ 5 -o sorted.fq.gz in.fq</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a> -tags    <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>     Tags, such as CB,UR. Order of these tags is sensitive.</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a> -@       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Threads to compress file.</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a> -o       <span class="co">[</span><span class="ot">fq.gz</span><span class="co">]</span>    bgzipped output fastq file.</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a> -m       <span class="co">[</span><span class="ot">mem</span><span class="co">]</span>      Memory per thread. <span class="co">[</span><span class="ot">1G</span><span class="co">]</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a> -p                  Input fastq is smart pairing.</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a> -T       <span class="co">[</span><span class="ot">prefix</span><span class="co">]</span>   Write temporary files to PREFIX.nnnn.tmp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="stream" class="level3">
<h3 class="anchored" data-anchor-id="stream">stream</h3>
<p>The stream tool is designed as a framework to process FASTQ+ files, where each FASTQ+ blockdefined by having identical tags and grouped together in the fileis handled individually. The -script option allows users to specify a custom bash script that processes each block. This user-defined script reads a small FASTQ+ file, generating FASTQ or FASTA output that is then sent to stdout. The stream tool captures this output via a pipe and updates the tags to ensure that each block of reads retains its original tags. Finally, it consolidates all outputs into a single file. In essence, this tool efficiently divides and processes each FASTQ+ block through a user-defined method, then seamlessly merges the results.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Perform user-defined script for each FASTQ+ block.</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>$ PISA stream -script run.sh reads.fq.gz</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a> -tags    <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>     Tags to define read blocks.</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a> -script  <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     User defined bash script, process $FQ and generate results to stdout.</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a> -min     <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Mininal reads per block to process.  <span class="co">[</span><span class="ot">2</span><span class="co">]</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a> -keep               Output unprocessed FASTQ+ records.</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a> -fa                 Stream FASTQ output instead of FASTQ.</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a> -tmpdir</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a> -t                  Threads.</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a> -o       <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Path to output file.</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a> -nw                 Disable warning messages.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="write-a-script-for-pisa-stream" class="level4">
<h4 class="anchored" data-anchor-id="write-a-script-for-pisa-stream">Write a script for PISA stream</h4>
<p>The PISA stream tool generates a temporary file named _block.fq for each block of reads, storing this file in a designated temporary directory. The path to this file is set in the environment variable ${FQ} for accessibility by subprocesses. Additionally, to ensure the uniqueness of each block, an alias named the unique block index is exported to the environment as ${UBI}.</p>
<p>The following example script demonstrates how to convert FASTQ+ to FASTA and rename the sequence ID. It is crucial for users to ensure that the scripts final output (either FASTQ+ or FASTA) is directed to stdout. All other script steps should avoid producing output to stdout or stderr, except for the last step. This precaution is necessary because PISA captures the scripts output through a pipe, and any unintended characters could disrupt the data format. Scripts can be written in a bash file or specified inline within the command.</p>
<div class="framed">
<div class="sourceCode" id="cb7"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>$ cat run.sh</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>seqtk rename ${FQ} &gt; test.fa; seqtk rename test.fa ${UBI}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="addtags" class="level3">
<h3 class="anchored" data-anchor-id="addtags">addtags</h3>
<p>The addtags tool is designed to put the new tags or update tags for FASTQ+ or BAM files.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Just add tags to reads.</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>$ PISA addtags -str CB:Z:CELL1,LB:Z:PoII2 -o out.bam in.bam</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>$ PISA addtags -str CB:Z:CELL1,LB:Z:PoII2 -o out.fastq in.fastq</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>-o    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Output file, bam or fastq, depends on input format</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>-str  <span class="co">[</span><span class="ot">string</span><span class="co">]</span>  TAGs.</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>-@    <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Threads to pack file.</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>-mapq <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Mapping quality score to filter mapped reads.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="sam2bam" class="level3">
<h3 class="anchored" data-anchor-id="sam2bam">sam2bam</h3>
<p>After alignment, the sequence ID from the FASTQ+ records is retained in the RNAME field of the SAM file. Given that the RNAME field is limited to 254 characters, we also restrict the sequence ID and any optional tag fields in FASTQ+ to this length to ensure compatibility. The sam2bam tool processes these details by parsing the tags from the RNAME and appending them to the end of the SAM optional fields.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Parse FASTQ+ read name and convert SAM to BAM.</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>$ PISA sam2bam -report alignment.csv -@ 5 -adjust-mapq -gtf genes.gtf -o aln.bam in.sam<span class="co">[</span><span class="ot">.gz</span><span class="co">]</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a> -o       <span class="co">[</span><span class="ot">BAM</span><span class="co">]</span>       Output file <span class="co">[</span><span class="ot">stdout</span><span class="co">]</span>.</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a> -t       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>       Work threads.</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a> -mito    <span class="co">[</span><span class="ot">string</span><span class="co">]</span>    Mitochondria name. Used to stat ratio of mitochondria reads.</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a> -maln    <span class="co">[</span><span class="ot">BAM</span><span class="co">]</span>       Export mitochondria reads into this file instead of standard output file.</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a> -@       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>       Threads to compress bam file.</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a> -report  <span class="co">[</span><span class="ot">csv</span><span class="co">]</span>       Alignment report.</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>Note :</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Reads map to multiple loci usually be marked as low quality and filtered at downstream analysis.</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  But for RNAseq library, if reads map to an exonic locus but also align to 1 or more non-exonic loci,</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>  the exonic locus can be prioritized as primary alignments, and mapping quality adjusts to 255. Tag</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>  MM:i:1 will also be added for this record. Following options used to adjust mapping quality.</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Input SAM need be sorted by read name, and aligner should output all hits of a read in this SAM.</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a> -adjust-mapq         Enable adjusts mapping quality score.</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a> -gtf     <span class="co">[</span><span class="ot">GTF</span><span class="co">]</span>       GTF annotation file. This file is required to check the exonic regions.</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a> -qual    <span class="co">[</span><span class="ot">255</span><span class="co">]</span>       Updated quality score.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Option -t is to set the threads to parse the SAM records. The -@ option is to set the threads to compress alignments in BGZF format. The default compress level of BGZF is 6 in the htslib, but here PISA has reset this value to 2 to decrease the CPU times. Its not a good design to have both -t and -@, but will require a lot work to redesign the multithreads strategy, I have put this work in my todo list.</p>
<p>Option -report can specify a quality control report in CSV format. Here is the explanation of each term in this file.</p>
<div class="center">
<table class="caption-top table">
<colgroup>
<col style="width: 26%">
<col style="width: 73%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"><strong>Terms</strong></th>
<th style="text-align: left;"><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Raw reads</td>
<td style="text-align: left;">Raw reads in the BAM files, secondary alignment will be skipped.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Mapped reads</td>
<td style="text-align: left;">Reads mapped to reference, and the ratio of raw reads.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Plus strand</td>
<td style="text-align: left;">Reads mapped to forward strand of reference.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Minus strand</td>
<td style="text-align: left;">Reads mapped to backward strand of reference.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Mitochondria ratio</td>
<td style="text-align: left;">Ratio of reads mapped to chromosome mitochondria. The default mito name is chrM, user should change it by -mito option if reference is different. Otherwise this value will always be 0.</td>
</tr>
</tbody>
</table>
</div>
<section id="mapq-adjust-method" class="level4">
<h4 class="anchored" data-anchor-id="mapq-adjust-method">MapQ adjust method</h4>
<p>For RNA libraries, if a read from cDNA maps to an exonic locus but also maps to one or more non-exonic regions, the exonic locus can be prioritized as primary alignments, and mapping quality adjusts to 255. In the below records below, read DP8400008965TLL1C001R0102043364 mapped to three loci, and the aligner random pick one as the primary alignment and others as secondary. Each of these alignments has low mapping quality (MAPQ == 2, is usually filtered at downstream analysis). Our adjustment method will check if only one of these alignments overlaps with exonic regions. In our example, the last alignment overlapped with gene EEF1A1, and the other two hit intergenic regions. After adjustment, the last record has been flagged as a primary hit, and the mapping quality adjusted to 255, and other alignments of the same read are updated as secondary, MAPQ adjust to 0. MM:i:1 tag also is added to the primary record after adjustment. Option -adjust-mapq is reimplemented to mirror the 10X CellRangers MAPQ adjustment method (<a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview#alignment" class="uri">https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview#alignment</a>).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a> #  Output by aligner:</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>DP8400008965TLL1C001R0102043364 0   9   133020979   2   100M    *   0   0   GTTAATGATAACAATGCATCGTAAAACCTTCAGAAGGAAAGGAGAATGTTTTGTGGACCACTTTGGTTTTCTTGTTTGCGTGTGGCAGTTTTAAGTTTTT    ...</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>DP8400008965TLL1C001R0102043364 272 7   22510408    2   100M    *   0   0   AAAAACTTAAAACTGCCACACGCAAACAAGAAAACCAAAGTGGTCCACAAAACATTCTCCTTTCCTTCTGAAGGTTTTACGATGCATTGTTATCATTAAC    ...</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>DP8400008965TLL1C001R0102043364 272 6   73517606    2   100M    *   0   0   AAAAACTTAAAACTGCCACACGCAAACAAGAAAACCAAAGTGGTCCACAAAACATTCTCCTTTCCTTCTGAAGGTTTTACGATGCATTGTTATCATTAAC    ...</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="fu"># After adjustment (Seq and Qual in secondary alignments masked as *):</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>DP8400008965TLL1C001R0102043364 256 9   133020979   0   100M    *   0   0   *   ...</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>DP8400008965TLL1C001R0102043364 272 7   22510408    0   100M    *   0   0   *   ...</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>DP8400008965TLL1C001R0102043364 16  6   73517606    255 100M    *   0   0   AAAAACTTAAAACTGCCACACGCAAACAAGAAAACCAAAGTGGTCCACAAAACATTCTCCTTTCCTTCTGAAGGTTTTACGATGCATTGTTATCATTAAC    ... MM:i:1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>An example list here to show how to enable mapq adjuestment.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>PISA sam2bam -report alignment.csv -o out.bam -adjust-mapq -gtf hg38.gtf in.sam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="rmdup" class="level3">
<h3 class="anchored" data-anchor-id="rmdup">rmdup</h3>
<p>To effectively manage PCR duplication in single-cell experiments, it is essential to consider both cell and molecular barcodes. During the feature counting stage facilitated by PISA count, deduplication is efficiently handled by relying solely on unique UMIs. This reliance makes traditional PCR deduplication unnecessary for libraries that use UMIs. Nonetheless, producing a deduplicated BAM file remains beneficial for other analytical processes, such as variant calling, or simply to reduce file size.</p>
<p>The rmdup tool is specifically designed to remove duplicate reads that share identical barcodes, such as UMIs and cell barcodes. This selective deduplication approach ensures that only truly redundant data is removed, thus preserving the integrity and completeness of the dataset for comprehensive downstream analyses.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Deduplicate PCR reads with same barcodes.</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>$ PISA rmdup -tags CB,UR -o rmdup.bam in.bam</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>   -tags  <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>       Barcode tags to group reads.</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>   -@     <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>        Threads to unpack BAM.</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>   -o     <span class="co">[</span><span class="ot">BAM</span><span class="co">]</span>        Output bam.</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>   -q     <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>        Map Quality Score cutoff.</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>   -k                  Keep duplicates, make flag instead of remove them.</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>   -nw                 Disable warnings.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>In this version, PISA rmdup only supports single-end reads. For paired-end reads, such as scATAC data, PCR deduplication can be performed by the PISA bam2frag tool.</p>
</section>
<section id="pick" class="level3">
<h3 class="anchored" data-anchor-id="pick">pick</h3>
<p>The PISA pick tool is designed to select alignments with predefined tags and candidate values.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>$ PISA pick -tags CB,GN -list cell_barcodes.txt in.bam</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a> -tags    <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>       Barcode tags.</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a> -list    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>       Barcode white list, tag values in related column will be apply.</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a> -o       <span class="co">[</span><span class="ot">BAM</span><span class="co">]</span>        Output file.</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a> -q       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>        Map Quality Score cutoff.</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a> -@       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>        Threads to unpack BAM.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Depending on the number of tags specified by the user, the candidate list for tags can consist of either a single column or multiple columns. If multiple tags are specified but only one column is present in the list, the program will primarily compare the value of the first tag in the alignments with the list.</p>
</section>
<section id="anno" class="level3">
<h3 class="anchored" data-anchor-id="anno">anno</h3>
<p>Connecting alignment results with genomic features is essential in single-cell data analysis. PISA categorizes features into three main types: gene annotation, functional regions, and genetic or sequence variations. For gene annotation, the PISA anno tool efficiently organizes all exons, transcripts, and genes from a GTF database into a sorted hierarchical tree structure.</p>
<p>Based on their alignment status, reads are then classified into one of nine distinct types, see illustrate below <a href="#fig:fig01" data-reference-type="ref" data-reference="fig:fig01">Figure 1</a>. This detailed categorization helps in accurately assessing the transcriptional landscape and understanding the complex genomic architecture within single-cell datasets.</p>
<div id="fig:fig01" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="anno_type.png" class="img-fluid figure-img"></p>
<figcaption>Figure 1: Read types.</figcaption>
</figure>
</div>
<div class="sourceCode" id="cb14"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu"># annotate strand-specific reads</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>$ PISA anno -gtf genes.gtf -o anno.bam sorted.bam</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="fu"># annotate non-strand-specific reads, for Smartseq or bulk RNAseq</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>$ PISA anno -is -gtf genes.gtf -o anno.bam sorted.bam</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="fu"># also label gene name for intronic reads, i.e. RNA velocity analysis</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>$ PISA anno -intron -gtf genes.gtf -o anno.bam sorted.bam</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="fu"># annotate expressed peaks and genetic variants (both reference allele and alternative allele) </span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>$ PISA anno -bed peak.bed -vcf in.vcf.gz -vcf-ss -ref-alt -o anno.bam in.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Besides these annotation methods, PISA anno also supports a -chr-species method. This method requires a binding list for chromosome and related label relationships. The software will check the chromosome and add the related tag for each chromosome. This method, combined with PISA attrcnt can help to summarize the mixed two cell lines from different species.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># A binding list is tab-separated two columns txt file.</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>$ cat binding_list.txt</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>GRCh38_chr1    Human</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>GRCh38_chr21   Human</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>mm10_chr21     Mouse</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>$ PISA anno -chr-species binding.txt -btag SP -o anno_species.bam sorted.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The full options and descriptions list below.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Annotate SAM/BAM records with overlapped function regions. Such as gene, transcript etc.</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>$ PISA anno -bed peak.bed -tag PK -vcf in.vcf.gz -vtag VF -vcf-ss -ref-alt -o anno.bam in.bam</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>$ PISA anno -gtf genes.gtf -o anno.bam in.bam</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>$ PISA anno -gtf genes.gtf -o anno.bam -sam in.sam</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">BAM</span><span class="co">]</span>       Output bam file.</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a> -report   <span class="co">[</span><span class="ot">csv</span><span class="co">]</span>       Summary report.</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a> -@        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>       Threads to compress bam file.</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a> -q        <span class="co">[</span><span class="ot">0</span><span class="co">]</span>         Map Quality Score cutoff. MapQ smaller than this value will not be annotated.</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a> -t        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>       Threads to annotate.</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a> -chunk    <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>       Chunk size per thread.</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a> -anno-only            Export annotated reads only.</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a> -sam                  Input is SAM file, parse tags from read name.</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a> -rev                  Annotation in reverse strand; Some probe ligation library for FFPE samples create reverse fragments.</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a> -is                   Disable strand sensitive annotation of gene, genomic region and genetic variants.</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>Options for BED file :</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a> -bed      <span class="co">[</span><span class="ot">BED</span><span class="co">]</span>       Function regions. Three or four columns bed file. Col 4 could be empty or names of this region.</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a> -tag      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>       Attribute tag name. Set with -bed. Default is PK.</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>Options for mixed samples.</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a> -chr-species  <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>  Chromosome name and related species binding list.</span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a> -btag     <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>       Species tag name. Set with -chr-species. Default is SP.</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>Options for GTF file :</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a> -gtf      <span class="co">[</span><span class="ot">GTF</span><span class="co">]</span>       GTF annotation file. gene_id,transcript_id is required for each record.</span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a> -tags     <span class="co">[</span><span class="ot">TAGs</span><span class="co">]</span>      Attribute names, more details see <span class="in">`Notice`</span> below. <span class="co">[</span><span class="ot">TX,GN,GX,RE,EX,JC</span><span class="co">]</span></span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a> -splice               Reads covered exon-intron edge (ExonIntron type) will also be annotated with all tags.</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a> -intron/-velo         Reads covered intron regions will also be annotated with all tags.</span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a> -exon                 Generate exon level and junction annotation. Put exon name (chr:start-end/[<span class="co">[</span><span class="ot">+-</span><span class="co">]</span>/Gene) in EX tag.\ </span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>                       Also generate junction name (chr:exon1_end-exon2_start/<span class="co">[</span><span class="ot">+-</span><span class="co">]</span>/Gene) in JC tag.</span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a> -flatten              Split overlapped exons into nonoverlapped bins.</span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a> -psi                  Annotate exclude reads tag (ER) for each exon.</span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a> -tss                  Annotate reads start from TSS, designed for capped library. **experiment**</span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a> -ctag     <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>       Tag name for TSS annotation. Need set with -tss.</span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a>Options for VCF file :</span>
<span id="cb16-41"><a href="#cb16-41" aria-hidden="true" tabindex="-1"></a> -vcf      <span class="co">[</span><span class="ot">VCF/BCF</span><span class="co">]</span>   Varaints file in vcf or bcf format. In default, only annotate alternative alleles.</span>
<span id="cb16-42"><a href="#cb16-42" aria-hidden="true" tabindex="-1"></a> -vtag     <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>       Tag name for variants. Set with -vcf. Default is VR.</span>
<span id="cb16-43"><a href="#cb16-43" aria-hidden="true" tabindex="-1"></a> -ref-alt              Annotate ref allele.</span>
<span id="cb16-44"><a href="#cb16-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-45"><a href="#cb16-45" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb16-46"><a href="#cb16-46" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>If input is SAM format, will try to parse the tags in the read name.</span>
<span id="cb16-47"><a href="#cb16-47" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>For GTF mode, this program will set tags in default, you could also reset them by -tags.</span>
<span id="cb16-48"><a href="#cb16-48" aria-hidden="true" tabindex="-1"></a>   TX : Transcript id.</span>
<span id="cb16-49"><a href="#cb16-49" aria-hidden="true" tabindex="-1"></a>   GN : Gene name.</span>
<span id="cb16-50"><a href="#cb16-50" aria-hidden="true" tabindex="-1"></a>   GX : Gene ID.</span>
<span id="cb16-51"><a href="#cb16-51" aria-hidden="true" tabindex="-1"></a>   RE : Region type, E (exon), N (intron), C (exon and intron), S (junction reads cover isoforms properly), V (ambiguous reads),</span>
<span id="cb16-52"><a href="#cb16-52" aria-hidden="true" tabindex="-1"></a>        I (intergenic), A (antisense or antisense exon), B (antisense intron), X (one or more exons are excluded in transcrpit)</span>
<span id="cb16-53"><a href="#cb16-53" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>The following tags set with -exon.</span>
<span id="cb16-54"><a href="#cb16-54" aria-hidden="true" tabindex="-1"></a>   EX : Exon name tag.</span>
<span id="cb16-55"><a href="#cb16-55" aria-hidden="true" tabindex="-1"></a>   JC : Isoform junction name.</span>
<span id="cb16-56"><a href="#cb16-56" aria-hidden="true" tabindex="-1"></a>   FL : Flatten exon name. Only generate it with -flatten.</span>
<span id="cb16-57"><a href="#cb16-57" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>The following tags set with -psi.</span>
<span id="cb16-58"><a href="#cb16-58" aria-hidden="true" tabindex="-1"></a>   ER : Excluded exons.</span>
<span id="cb16-59"><a href="#cb16-59" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>PSI = EX/(EX+ER); EX is the exon tag, which indicate include reads in exon.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="corr" class="level3">
<h3 class="anchored" data-anchor-id="corr">corr</h3>
<p>The diversity of UMIs of each gene in one cell is used to evaluate the gene expression level, but the error of UMI comes from sequencing or PCR may introduce bias. PISA corr is designed to correct the UMI or barcode sequence based on Hamming distance. In default, two groups of UMI from the same gene of one cell with Hamming distance equal to 1, will be considered to originate from the same transcript. The group with high frequency will be selected as a real one, and another one will be corrected to the high one.</p>
<p>Because PISA corr does not require a sorted BAM, this tool will first build a correction list by caching all the raw UMIs and barcodes and then correct them in memory. After these steps, this tool will reread the file and update these records by order. This design can avoid potential bias for the same gene from different chromosomes (i.e., the HLA genes in alternative locus). However, this design also required a lot of memory for a big BAM.</p>
<p>CellRanger also introduces an algorithm to correct reads with the same UMI of one cell but mapping to more than one gene (<a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview" class="uri">https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview</a>). PISA implements this method but does not enable it in default. The -cr option can be used by users to enable this function. The reason for not enable it by default is PISA corr not only is used to correct UMIs, but also can be used to correct any other types of barcodes. The following example shows how to use PISA corr to correct cell barcodes for reads in the same gene.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>// Reads with same gene tag (GN) and UMI (UB) will be grouped and calculate the Hamming distance between each other. </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>// Only if Hamming distance == 1 will be corrected.</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>PISA corr -tag CR -new-tag CB -tags-block GN,UB -o cell_barcode_corrected.bam in.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Full options of PISA corr list below.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Correct similar barcodes (hamming distance == 1).</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>$ PISA corr -tag UR -new-tag UB -tags-block CB,GN -@ 5 -o corr.bam in.bam</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">BAM</span><span class="co">]</span>       Output bam.</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a> -tag      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>       Tag to correct.</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a> -new-tag  <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>       Create a new tag for corrected barcodes.</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a> -tags-block  <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>   Tags to define read group. For example, if set to GN (gene), reads in the same gene will be grouped together.</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a> -cr                   Enable CellRanger like UMI correction method. See <span class="in">`Examples`</span> for details.</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a> -e                    Maximal hamming distance to define similar barcode, default is 1.</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a> -@        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>       Thread to compress BAM file.</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>Examples :</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a> // Two groups of reads have same cell barcode (CB) and gene (GN) but their raw UMIs (UR) differ by only one base. The UMI of less </span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a> // supported is corrected to the UMI with higher support. UB save the checked or corrected UMI.</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>$ PISA corr -tag UR -new-tag UB -tags-block CB,GN in.bam -o corr.bam </span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a> // Same with above. Besides, if two or more groups of reads have same CB and UB but different GN, the GN with the most supporting reads</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a> // is kept for UMI counting, and the other read groups are discarded. In case of a tie for maximal read support, all read groups are</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a> // discarded, as the gene cannot be confidently assigned (Cell Ranger method).</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>$ PISA corr -cr -tag UR -new-tag UB -tags-block CB,GN in.bam -o corr.bam </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="attrcnt" class="level3">
<h3 class="anchored" data-anchor-id="attrcnt">attrcnt</h3>
<p>PISA attrcnt is used to summarize the meta information of the library. We start introduce this tool with few examples.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count reads per cell, -cb option is required to specify cell barcode tag</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>$ PISA attrcnt -cb CB in.bam</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>// the summary information output in tsv format</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>BARCODE Raw                           // the title</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>AGCTATGCTTCTAGTGTAAC-1  38            // cell barcode and raw reads per cell, separated by a tab</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>GCCGCTGATCGGCCTGCACA-1  12</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>GTCGCGATTCTGCTCAGAAG-1  13</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>GTCAGTACCATGCTCAGAAG-1  27</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>CAACTCGTCGGTTGTCTGAC-1  23</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count raw reads and reads in the peaks per cell</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>$ PISA attrcnt -cb CB -tags PK             // PK tag is annotated for reads in peak</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>             -o summary.tsv              // Summary file</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>             example/anno/demo_1.bam </span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>$ head summary.tsv </span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>BARCODE Raw PK</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>GGCATTATCGGCTCGGTATG    2   2</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>TGAGTTGTGTATACTCCTAC    2   2</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>CCGCGGCACTACACACCAGA    1   1</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>ATTAGTGGTCTCCTGGTCGG    1   1</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>TTATTGGACCACGTTGAATA    1   1</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>GGAATGCCACAAGCGCCGTA    1   1</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>AGTCTATCGTGGCCTGCACA    5   5</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>AGGCACACCTCGCTGAATTC    3   3</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>GTCAGGATCGATAACATACG    2   2</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count UMIs per cell</span></span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a>$ PISA attrcnt -cb CB -tags UB      // UB is tag of corrected UMIs</span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a>             -dedup               // -dedup option used to remove duplication of tag values, if not set, this </span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a>                                  // command will export reads with UB tag but not the unique UMIs per cell</span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a>             -o summary.tsv anno.bam</span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count UMIs and Genes per cell</span></span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a>$ PISA attrcnt -cb CB -tags UB,GN  // GN is tag for annotated gene; -tags can accept multiple tag names, and seperated by ","</span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true" tabindex="-1"></a>             -dedup -o summary.tsv anno.bam</span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true" tabindex="-1"></a>$ head summary.tsv</span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true" tabindex="-1"></a>BARCODE Raw UB  GN</span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true" tabindex="-1"></a>AGCTATGCTTCTAGTGTAAC-1  38  35  0</span>
<span id="cb19-42"><a href="#cb19-42" aria-hidden="true" tabindex="-1"></a>GCCGCTGATCGGCCTGCACA-1  12  12  0</span>
<span id="cb19-43"><a href="#cb19-43" aria-hidden="true" tabindex="-1"></a>GTCGCGATTCTGCTCAGAAG-1  13  11  0</span>
<span id="cb19-44"><a href="#cb19-44" aria-hidden="true" tabindex="-1"></a>GTCAGTACCATGCTCAGAAG-1  27  26  0</span>
<span id="cb19-45"><a href="#cb19-45" aria-hidden="true" tabindex="-1"></a>CAACTCGTCGGTTGTCTGAC-1  23  21  0</span>
<span id="cb19-46"><a href="#cb19-46" aria-hidden="true" tabindex="-1"></a>GGTACACCACAGTAGTTACG-1  12  10  0</span>
<span id="cb19-47"><a href="#cb19-47" aria-hidden="true" tabindex="-1"></a>GCGCGCCGAGGGACACTCTT-1  1   1   0</span>
<span id="cb19-48"><a href="#cb19-48" aria-hidden="true" tabindex="-1"></a>CTCTAAGCATCGAGGTTAAC-1  162 138 50</span>
<span id="cb19-49"><a href="#cb19-49" aria-hidden="true" tabindex="-1"></a>TTCGTAGCACCGATACTAGC-1  51  48  46</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Full list of options list below.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count the frequency of tag values.</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>$ PISA attrcnt -cb CB -tags UR,GN -dedup -all-tags in.bam</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a> -cb       <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Cell Barcode, or other tag used for grouping reads.</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a> -list     <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Cell barcode white list.</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a> -tags     <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>     Tags to count.</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a> -dedup               Deduplicate the atrributes in each tag.</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a> -all-tags            Only records with all tags be count.</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a> -group    <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Group tag, count all tags for each group seperately.</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Output count table.</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a> -q        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Map Quality to filter bam.</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a> -no-header           Ignore header in the output.</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a> -@        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Thread to unpack bam.</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a> -ttag     <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Region type tag. <span class="co">[</span><span class="ot">RE</span><span class="co">]</span></span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a> -ttype               Region type used to count. Set <span class="in">`E,S`</span> to count exon enclosed reads. Set <span class="in">`N,C`</span> to count intron overlapped reads.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="extract" class="level3">
<h3 class="anchored" data-anchor-id="extract">extract</h3>
<p>PISA extract is designed to extract the values of tags from BAM records and generate a tab-separated file.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Extract tag values from alignments.</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>$ PISA extract -tags CB,UR,GN -o tags.tsv in.bam</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a> -tags     <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>     Tags to be extracted.</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Output file. tsv format</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a> -n                   Print read name.</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a> -q                   Map Quality Score threshold.</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a> -all                 Only export if all tags have value.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="count" class="level3">
<h3 class="anchored" data-anchor-id="count">count</h3>
<p>The PISA count tool is designed to generate a counts matrix for various features or tags, traditionally outputting a gene-by-cell digit matrix. Starting with version 0.4, this tool now supports the MEX format output, which is highly recommended for use in downstream analyses due to its superior performance. Its important to note that MEX format consists of three files, so you should use the -outdir option to specify the directory where the output files will be saved.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count reads or fragments matrix for single-cell datasets.</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>$ PISA count -cb CB -anno-tag GN -umi UB -outdir exp aln.bam</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>$ PISA count <span class="co">[</span><span class="ot">options</span><span class="co">]</span> aln1.bam,aln2.bam</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>$ PISA count -cb RG -sample-list bam_files.txt -outdir exp</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>$ PISA count -tags Cx,Cy -anno-tag GN -umi UB -outdir exp -velo aln.bam</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a> -tags/-cb <span class="co">[</span><span class="ot">TAGs</span><span class="co">]</span>     A cell barcode tag or two tags of spatial coordinates for spatial data.</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a> -anno-tag <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Annotation tag, gene or peak.</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a> -genome-bin <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>    If genome bin size set, genome bin count matrix will be generated, conflict with -anno-tag and -chr.</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a> -is                  Ignore strand for bin counting.</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a> -chr                 Count chromosome expression level, conflict with -anno-tag and -genome-bin.</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a> -list     <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Barcode white list, used as column names at matrix. If not set, all barcodes will be count.</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a> -outdir   <span class="co">[</span><span class="ot">DIR</span><span class="co">]</span>      Output matrix in MEX format into this folder.</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a> -umi      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      UMI tag. Count once if more than one record has same UMI in one gene or peak.</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a> -one-hit             Skip if a read hits more than 1 gene or peak.</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a> -q        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Minimal map quality to filter. Default is 20.</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a> -t        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Threads.</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a> -ttag     <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Region type tag. <span class="co">[</span><span class="ot">RE</span><span class="co">]</span></span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a> -velo                Generate spliced and unspliced matrix files for RNA velocity analysis.</span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a> -ttype    <span class="co">[</span><span class="ot">TYPE</span><span class="co">]</span>     Region type used to count. Set <span class="in">`E,S`</span> to count exon enclosed reads. Set <span class="in">`N,C`</span> to count intron overlapped reads.</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a> -sample-list <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>  A list of bam files. Each path per line.</span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a>Options for Stereoseq:</span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a> -stereoseq           Stereoseq pipeline pack UMI to hex string. Need set this option to decode UMIs.</span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a> -spatial-bin <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>   Bin size for spatial coordiate. Can be set if -tags specify spatial coordinates.<span class="co">[</span><span class="ot">1</span><span class="co">]</span></span>
<span id="cb22-28"><a href="#cb22-28" aria-hidden="true" tabindex="-1"></a> -dup                 Do NOT skip duplicate reads.</span>
<span id="cb22-29"><a href="#cb22-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-30"><a href="#cb22-30" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb22-31"><a href="#cb22-31" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Region type (RE), which label functional region reads mapped, is annotated by <span class="in">`PISA anno`</span>. Optional -ttype can be set</span>
<span id="cb22-32"><a href="#cb22-32" aria-hidden="true" tabindex="-1"></a>   to one of region types(E/S/C/N) or combination to count reads mapped to these functional regions only.</span>
<span id="cb22-33"><a href="#cb22-33" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>If you want count from more than one bam file, there are two ways to set the parameter. By seperating bam files with ',' or by</span>
<span id="cb22-34"><a href="#cb22-34" aria-hidden="true" tabindex="-1"></a>   setting -sample-list option.</span>
<span id="cb22-35"><a href="#cb22-35" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>If -velo set, spliced and unspliced folders will be created at outdir.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="for-smartseq-user" class="level4">
<h4 class="anchored" data-anchor-id="for-smartseq-user">For Smartseq user</h4>
<p>PISA count also support counting gene expression from multiple bam files.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>$ PISA count -file-barcode             // use alias name for each bam file as cell barcode. If this flag is not set -cb must be specified.</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>-tags CB                   // Cell barcode</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>-sample-list bam_list.txt // bam file path and alias name</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>-outdir exp/ -anno-tag GN </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="fu"># The `-sample-list' specify multiply files, each BAM path per line.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="description-of-mex-file" class="level4">
<h4 class="anchored" data-anchor-id="description-of-mex-file">Description of MEX file</h4>
<p>The Market Exchange (MEX) format (<a href="https://math.nist.gov/MatrixMarket/formats.html" class="uri">https://math.nist.gov/MatrixMarket/formats.html</a>) is designed for representing the sparse matrix. The -outdir option specifies the output directory for one MEX fold. The MEX format consists of three files, one is cell barcodes, one is feature names (genes or peak names), and the third one defines the expression or signal values.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>$ ls</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>barcodes.tsv.gz  features.tsv.gz  matrix.mtx.gz</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>$ zcat barcodes.tsv.gz|head</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>AACCTGGTGAAGTTGTCGAA</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>AAGGAACTAAGCGCAGCACC</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>CGATAGAATACTTCTTCGTA</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>TACTATCCTCTAGCTGCTAC</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>TGACCATCCTACAGTCCACC</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>CAGATTCAACTACGAAGTGC</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>TTCGTAGCACTCTTCATCTC</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>GGCACCTTGCTTAACGTAGG</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>ACTTCGGATACGTATCGCCT</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>GACTCGCTAGTAGTCGGAAT</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>$ zcat features.tsv.gz|head</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>RP11-34P13.7</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>RP11-34P13.8</span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>RP11-34P13.9</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>FO538757.3</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>FO538757.2</span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>AP006222.2</span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>RP4-669L17.10</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>RP5-857K21.4</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>RP11-206L10.4</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a>RP11-206L10.9</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a>$ zcat matrix.mtx.gz|head</span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a>%%MatrixMarket matrix coordinate integer general</span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a>% Generated by PISA v0.4-alpha-72-g09c4ded</span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a>23900   782761  11533380</span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a>1   1   2</span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a>1   2   2</span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a>1   3   2</span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a>1   4   1</span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a>1   5   2</span>
<span id="cb24-37"><a href="#cb24-37" aria-hidden="true" tabindex="-1"></a>1   6   1</span>
<span id="cb24-38"><a href="#cb24-38" aria-hidden="true" tabindex="-1"></a>1   7   2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The MEX file can be read by R package Yano::ReadPISA.</p>
</section>
</section>
<section id="bam2fq" class="level3">
<h3 class="anchored" data-anchor-id="bam2fq">bam2fq</h3>
<p>PISA bam2fq is designed to convert alignment records to FATSQ+ records. Option -tags specify which tags will be kept in the FASTQ+. Full options list below.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Convert BAM into fastq.</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>$ PISA bam2fq -tags CB,UB,GN -o out.fq aln.bam</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a> -tags     <span class="co">[</span><span class="ot">TAGS</span><span class="co">]</span>     Export tags in read name.</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a> -f                   Filter records if specified tags not all exist.</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a> -fa                  Output fasta instead of fastq.</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">fastq</span><span class="co">]</span>    Output file.</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a> -@        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Threads to unpack BAM.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="bam2frag" class="level3">
<h3 class="anchored" data-anchor-id="bam2frag">bam2frag</h3>
<p>The fragment file is a five columns tab-separated flat file which is designed for scATAC-seq. The first column is the chromosome name, the second column is the start location of this fragment (0 based), and the third column is the end position in 1 based of this fragment. The fourth column is the cell barcode of this fragment and the last column is how many duplicates of this fragment.</p>
<p>PISA bam2frag requires the input BAM file to be sorted by coordinate. This tool will check the cell barcode and the fragment position for each paired reads, duplicates in one cell will only keep one record, and the numeber of copies for this fragment will be updated in column four.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Convert sam record to fragment file.</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>$ PISA bam2frag -cb CB -list cell_barcodes.txt -o out.tsv.gz in.bam</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a> -o       <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Output file. This file will be bgzipped and indexed.</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a> -cb      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>     Cell barcode tag.</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a> -list    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Cell barcode white list.</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a> -q       <span class="co">[</span><span class="ot">20</span><span class="co">]</span>      Mapping quality score to filter reads.</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a> -isize   <span class="co">[</span><span class="ot">2000</span><span class="co">]</span>    Skip if insert size greater than this. <span class="co">[</span><span class="ot">2KB</span><span class="co">]</span></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a> -bed     <span class="co">[</span><span class="ot">BED</span><span class="co">]</span>     Only convert fragments overlapped with target regions.</span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a> -black-region <span class="co">[</span><span class="ot">BED</span><span class="co">]</span> Skip convert fragments overlapped with black regions.</span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a> -stat    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Transposition events per cell.</span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a> -@       <span class="co">[</span><span class="ot">4</span><span class="co">]</span>       Thread to unpack and pack files.<span class="co">[</span><span class="ot">4</span><span class="co">]</span></span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a> -disable-offset    Disable Tn5 offset for each fragment.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="depth" class="level3">
<h3 class="anchored" data-anchor-id="depth">depth</h3>
<p>PISA depth generates coverage information for each position of the predefined region(s). The significant difference between PISA depth and samtools depth (<a href="http://www.htslib.org/doc/samtools-depth.html" class="uri">http://www.htslib.org/doc/samtools-depth.html</a>) is that PISA depth use UMI rather than reads. Besides, PISA depth is strand sensitive, and can produce results for target cells.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count coverage depth or unique UMIs for genome locations.</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>Usage : PISA depth <span class="co">[</span><span class="ot">options</span><span class="co">]</span> sorted.bam <span class="co">[</span><span class="ot">region</span><span class="co">]</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>$ PISA depth -cb CB -umi UB -tags GN -region in.bed -o depth.tsv sorted.bam</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>$ PISA depth -cb CB -umi UB sorted.bam chr1:1-2:+</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>Options : </span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a> -tag      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Tag used for grouping reads.</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a> -list     <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Candidate list for -tag.</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a> -umi      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      UMI tag. If set, only count unique UMIs for each location.</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a> -bed      <span class="co">[</span><span class="ot">BED</span><span class="co">]</span>      Target BED region file. If the strand in column six set, only count reads with the same strand.</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Output depth file. <span class="co">[</span><span class="ot">stdout</span><span class="co">]</span>.</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a> -q        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Minimal map quality to filter. <span class="co">[</span><span class="ot">20</span><span class="co">]</span></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a> -@        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Threads to unpack bam. <span class="co">[</span><span class="ot">4</span><span class="co">]</span></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Require sorted and indexed BAM as input.</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Compare with `samtools depth', PISA depth considers UMIs and strand of reads.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="callept" class="level3">
<h3 class="anchored" data-anchor-id="callept">callept</h3>
<p>The term EPT stands for Expressed Peak Tag. EPTs can be identified by PISA within aligned reads from RNA libraries. Once identified, the relationship between each EPT and its corresponding gene can be annotated and analyzed using the Yano package. This analysis helps in understanding the functional implications of EPTs in gene expression regulation.</p>
<p>The callept tool is designed to call expressed peak tags (EPTs) for indexed BAM files. The peaks are defined with UMI depth &gt;= cutoff.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>$ PISA cellept -o epts.bed sorted.bam</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>$ PISA cellept -tag CB -list cells.txt -umi UB -o epts.bed sorted.bam</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a> -tag      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      Tag used for grouping reads.</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a> -list     <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Candidate list for -tag.</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a> -umi      <span class="co">[</span><span class="ot">TAG</span><span class="co">]</span>      UMI tag. If set, only count unique UMIs for each location.</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a> -is                  Ignore strand.</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a> -gap      <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Maximum gap to merge nearby peaks. <span class="co">[</span><span class="ot">50</span><span class="co">]</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a> -min-length  <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>   Minimum peak length. <span class="co">[</span><span class="ot">50</span><span class="co">]</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a> -cutoff   <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Cutoff of depth. <span class="co">[</span><span class="ot">10</span><span class="co">]</span></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a> -o        <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Output EPTs in bed format. <span class="co">[</span><span class="ot">stdout</span><span class="co">]</span>.</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a> -q        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Minimal map quality to filter. <span class="co">[</span><span class="ot">20</span><span class="co">]</span></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a> -t        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Threads. <span class="co">[</span><span class="ot">4</span><span class="co">]</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Requires sorted and indexed BAM as input.</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>Compares with <span class="in">`MACS2`</span> and other peak callers, PISA callept considers UMIs and strand of reads.</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>For paired reads, strand of read 2 will be reversed to revert fragment strand.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="count2" class="level3">
<h3 class="anchored" data-anchor-id="count2">count2</h3>
<p>PISA count generates the MEX format matrix for fragments per peak.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Count fragments per peak per cell matrix.</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>$ PISA count2 -bed peaks.bed -t 10 -list barcodes.txt -outdir exp fragments.tsv.gz</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a> -list     <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>     Barcode white list, used as column names at matrix. If not set, all barcodes will be count.</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a> -outdir   <span class="co">[</span><span class="ot">DIR</span><span class="co">]</span>      Output matrix in MEX format into this fold.</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a> -prefix   <span class="co">[</span><span class="ot">STR</span><span class="co">]</span>      Prefix of output files.</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a> -t        <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>      Threads.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="mergebed" class="level3">
<h3 class="anchored" data-anchor-id="mergebed">mergebed</h3>
<p>Merge overlapped regions by strand and name.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>$ PISA mergebed -o merged.bed sample1.bed sample2.bed</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>$ PISA mergebed -up 500 -down 500 -o flank.bed peaks.bed</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>-o    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Output bed file.</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>-s              Ignore strand.</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>-up   <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Enlarge regions upstream.</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>-down <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Enlarge regions downstream.</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>-name           Merge regions by bed name.</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>This tool accepts 3 columns or 6 columns bed file(s), strand (+/-) is set in column 6.</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>By default, merging is done with respect to strandness unless -s is set.</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>-up/-down is set respect to strandness, so upstream of plus strand is downstream of minus strand.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="annobed" class="level3">
<h3 class="anchored" data-anchor-id="annobed">annobed</h3>
<p>Annotate the type of region and output in a BED like file. The region type can be defined into.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>$ PISA annobed -gtf genes.gtf -o anno.bed in.bed</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>-gtf    <span class="co">[</span><span class="ot">GTF</span><span class="co">]</span>     GTF database.</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>-o      <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Output bed file.</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>-report <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Summary report. Export in STDERR by default.</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>-is               Ignore strand.</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>-gene-name        Set annatated gene as bed name (column 4).</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>-skip-chrs        Skip chromosomes if not exist in GTF. Defa</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>-up  <span class="co">[</span><span class="ot">1000</span><span class="co">]</span>         Annotate intergenic regions at upstream of gene.</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>-down  <span class="co">[</span><span class="ot">1000</span><span class="co">]</span>       Annotate intergenic regions at downstream of gene.</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>Output format :</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>chromosome,start(0based),end(1based),name,score,strand,number of covered genes, cover gene name(s),type,nearest gene name,distance to nearby gene</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a>Notice :</span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>This tool accepts 3 columns or 6 columns bed file(s), strand (+/-) is set in column 6.</span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a><span class="ss"> * </span>By default, annotation is done with respect to strandness unless -s is set.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="flatten" class="level3">
<h3 class="anchored" data-anchor-id="flatten">flatten</h3>
<p>Convert overlapped into flatten records. If strand exist, the flatten records will be strand sensitive.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>$ PISA flatten -o flatten.bed overlapped.bed</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>-o    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Output bed file.</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>For example:</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>reg1 ===========</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>reg2       ===========</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>flattening of regions:</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>reg1 ======</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>reg2       =====</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>reg3            ======</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="gtffmt" class="level3">
<h3 class="anchored" data-anchor-id="gtffmt">gtffmt</h3>
<p>Format and reorder GTF records.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>$ PISA gtffmt in.gtf</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a> -o      <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Output GTF file</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a> -f                Only export gene, transcript, exon and CDS records.</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a> -key    <span class="co">[</span><span class="ot">all</span><span class="co">]</span>     Export selected keys in optional fields.</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a> -report <span class="co">[</span><span class="ot">stderr</span><span class="co">]</span>  Summary report file.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="gtf2bed" class="level3">
<h3 class="anchored" data-anchor-id="gtf2bed">gtf2bed</h3>
<p>Convert GTF file to BED format.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>$ PISA gtf2bed -o merged.bed in.gtf.gz</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>Options:</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  -o    <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>                          Output bed file.</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  -type <span class="co">[</span><span class="ot">gene|transcript|exon</span><span class="co">]</span>          Covert to bed.</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  -name <span class="co">[</span><span class="ot">none|gene|transcript|exon</span><span class="co">]</span>     Set name for bed.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="deprecated-commands" class="level2">
<h2 class="anchored" data-anchor-id="deprecated-commands">Deprecated commands</h2>
<section id="parse0" class="level3">
<h3 class="anchored" data-anchor-id="parse0">parse0</h3>
<p>Since v1.0, the old parse tool has been replaced by parse2, and renamed to parse0 for backup purposes. Due to its lower performance compared to parse, parse0 will be deleted from v2 onwards. The parse0 tool requires a configuration file to describe the library structure, including cell barcode locations, UMI locations, and tag names. Additionally, it generates various quality control reports and outputs cell barcode distributions to stdout. Below is the complete options list and a general workflow for filtering and counting reads.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>$ PISA parse -config read_struct.json -report fastq.csv -cbdis cell_dist.tsv \</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>          -1 out.fq lane1_1.fq.gz,lane02_1.fq.gz  lane1_2.fq.gz,lane2_2.fq.gz</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>Options :</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a> -1       <span class="co">[</span><span class="ot">fastq</span><span class="co">]</span>   Read 1 output. Default is stdout.</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a> -2       <span class="co">[</span><span class="ot">fastq</span><span class="co">]</span>   Read 2 output.</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a> -config  <span class="co">[</span><span class="ot">json</span><span class="co">]</span>    Read structure configure file in JSON format. Required.</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a> -run     <span class="co">[</span><span class="ot">string</span><span class="co">]</span>  Run code, used for different library.</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a> -cbdis   <span class="co">[</span><span class="ot">FILE</span><span class="co">]</span>    Read count per cell barcode.</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a> -p                 Read 1 and read 2 interleaved in the input file.</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a> -q       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Drop reads if average sequencing quality below this value.</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a> -dropN             Drop reads if N base in sequence or barcode.</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a> -report  <span class="co">[</span><span class="ot">csv</span><span class="co">]</span>     Summary report.</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a> -t       <span class="co">[</span><span class="ot">INT</span><span class="co">]</span>     Threads. <span class="co">[</span><span class="ot">4</span><span class="co">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>-config requires a JSON file to tell software the locations of cell barcodes and/or UMIs. An example configured file can be found at demo/demo.json.</p>
<div class="framed">
<div class="sourceCode" id="cb36"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>    "cell barcode tag":"CB",  </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    "cell barcode":[</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>        {</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>            "location":"R1:1-16" # the location is 1 based</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>    "UMI tag":"UR",</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>    "UMI":{</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>        "location":"R1:17-28",</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>    "read 1":{</span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>        "location":"R2:1-91",</span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Option -report can specify a quality control report in CSV format. Here is the explanation of each term in this file.</p>
<div class="center">
<table class="caption-top table">
<colgroup>
<col style="width: 36%">
<col style="width: 63%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"><strong>Terms</strong></th>
<th style="text-align: left;"><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Number of Fragments</td>
<td style="text-align: left;">The number of records in the FASTQ(s). For paired reads, each pair only count once.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Fragments pass QC</td>
<td style="text-align: left;">Reads or paired reads pass QC.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Fragments with Exactly Matched Barcodes</td>
<td style="text-align: left;">Cell barcode exactly matched with any barcode in the candidate list.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Fragments with Failed Barcodes</td>
<td style="text-align: left;">No cell barcode found in the candidate list with similar search.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Fragments Filtered on Low Quality</td>
<td style="text-align: left;">Mean sequence quality of the records smaller than threshold.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Fragments Filtered on Unknown Sample Barcodes</td>
<td style="text-align: left;">Sample barcodes not matched with any barcode in the candidate list.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Q30 bases in Cell Barcode</td>
<td style="text-align: left;">Ratio of bases in cell barcode sequence with quality <span class="math inline">\(&gt;=\)</span> 30.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Q30 bases in Sample Barcode</td>
<td style="text-align: left;">Ratio of bases in sample barcode sequence with quality <span class="math inline">\(&gt;=\)</span> 30.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Q30 bases in UMI</td>
<td style="text-align: left;">Ratio of bases in UMI sequence with quality <span class="math inline">\(&gt;=\)</span> 30.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Q30 bases in Reads</td>
<td style="text-align: left;">Ratio of bases in read sequence with quality <span class="math inline">\(&gt;=\)</span> 30.</td>
</tr>
</tbody>
</table>
</div>
<p>Here is an example to parse barcodes and reads with a predefined configure JSON file, the test files can be found at demo directory.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>$ PISA parse0 -config demo/demo.json -report demo/parse.csv demo/demo_1.fq.gz demo/demo_2.fq.gz &gt; demo/demo.fq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="difference-between-pisa-and-cellranger" class="level2">
<h2 class="anchored" data-anchor-id="difference-between-pisa-and-cellranger">Difference between PISA and CellRanger</h2>
</section>
<section id="changelog" class="level2">
<h2 class="anchored" data-anchor-id="changelog">Changelog</h2>
<details>
<summary>
v1.3 2024/10/10
</summary>
<ul>
<li>Remove <code>-vcf-ss</code> option for <code>PISA anno</code>. Enable strand sensitive mode in default for genetic variant annotation.</li>
</ul>
</details>
<details>
<summary>
v1.2 2024/9/7
</summary>
<ul>
<li>Fix a bug of <code>PISA corr</code>. This bug was caused by the recent update to <code>dict.c</code>. The<code>PISA cor</code>r function was not updated accordingly. This issue will not influence the result.</li>
</ul>
</details>
<details>
<summary>
v1.1 2024/6/11
</summary>
<ul>
<li>-psi and -flatten added for <code>PISA anno</code>;</li>
<li>Add new tools <code>gtf2bed</code> and <code>flatten</code>.</li>
</ul>
</details>
<details>
<summary>
v1.0 2023/12/2
</summary>
<ul>
<li>EPT calling method added.</li>
<li><code>mergebed</code> and <code>annobed</code> for BED file added.</li>
<li>New parameters added to <code>anno</code>.</li>
</ul>
</details>
<details>
<summary>
v1.0a 2023/10/20
</summary>
<ul>
<li>Major update.</li>
<li>New <code>callept</code> method introduced. Check EPT paper for details.</li>
<li>New <code>mergebed</code> and <code>annobed</code> methods added.</li>
<li>New <code>-vcf-ss</code> and <code>-alt-ref</code> now added to <code>PISA anno</code></li>
<li><code>PISA count</code> can count bin and chromosome expression.</li>
</ul>
</details>
<details>
<summary>
v0.12 2022/06/18
</summary>
<ul>
<li>Fix the number of records in MEX file.</li>
</ul>
</details>
<details>
<summary>
v0.12b 2022/04/26
</summary>
<ul>
<li>Bugs fixed.</li>
<li>Change compress level of BGZF from 6 to 2, speed up few tools.</li>
<li>Now accept unorder GTFs</li>
<li>New tool <code>gtfmt</code> introduced to format a GTF file.</li>
<li>Add manual and FASTQ+ specification.</li>
</ul>
</details>
<details>
<summary>
v0.12a 2022/03/31
</summary>
<ul>
<li>Add <code>parse2</code>.</li>
</ul>
</details>
<details>
<summary>
v0.11a 2022/03/13
</summary>
<ul>
<li>Add <code>counts2</code>, count peaks X cells matrix from the fragment file.</li>
</ul>
</details>
<details>
<summary>
v0.10 2022/01/06
</summary>
<ul>
<li>Update <code>bam2frag</code>, export a fragment file compatible with 10X cellranger-ATAC.</li>
</ul>
</details>
<details>
<summary>
v0.10b 2021/12/09
</summary>
<ul>
<li><code>PISA count</code> now has <code>-velo</code> option to export unspliced and spliced matrix together. For velocity analysis, remember to use <code>-intron</code> to annotate reads.</li>
<li><code>PISA parse</code> support multi-threads.</li>
</ul>
</details>
<details>
<summary>
v0.10a 2021/11/06
</summary>
<ul>
<li><code>PISA count</code> support count spliced and unspliced reads.</li>
<li><code>PISA count</code> support count from multiple bam files.</li>
</ul>
</details>
<details>
<summary>
v0.9 2021/10/14
</summary>
<ul>
<li>Rewrite <code>rmdup</code>. Not support paired reads for now.</li>
</ul>
</details>
<details>
<summary>
v0.8 2021/07/20
</summary>
<ul>
<li>Reduce memory usage of <code>count</code></li>
<li>Fix region query bug of <code>anno -bed</code></li>
<li>Add <code>anno -vcf</code> method</li>
</ul>
</details>
<details>
<summary>
v0.7 2020/11/20
</summary>
<ul>
<li>Introduce the PCR deduplicate method <code>rmdup</code>.</li>
<li>Mask read and qual field as * instead of sequence for secondary alignments in the BAM file.</li>
</ul>
</details>
<details>
<summary>
v0.6 2020/10/29
</summary>
<ul>
<li><code>PISA attrcnt</code>, Skip secondary alignments before counting reads</li>
<li><code>PISA anno</code> fix segments fault bugs when loading malformed GTF</li>
</ul>
</details>
<details>
<summary>
v0.5 2020/08/27
</summary>
<ul>
<li>Add <code>PISA bam2frag</code> function (experimental).</li>
<li><code>PISA anno</code> Skip secondary alignments when counting total reads.</li>
</ul>
</details>
<details>
<summary>
v0.4 2020/07/14
</summary>
<ul>
<li><code>PISA sam2bam</code> add mapping quality adjustment method;</li>
<li>Rewrite UMI correction index structure to reduce memory use;</li>
<li>Fix bugs.</li>
</ul>
</details>
<details>
<summary>
v0.4alpha 2020/05/2
</summary>
<ul>
<li><code>PISA anno</code> use UCSC bin scheme instead of linear search for reads query gene regions. Fix the bug of misannotated antisense reads.</li>
<li><code>PISA count</code> use MEX output instead of plain cell vs gene table.</li>
</ul>
</details>
<details>
<summary>
v0.3 2020/03/26
</summary>
<p>Fix bugs and improve preformance.</p>
</details>
<details>
<summary>
0.0.0.9999 2019/05/19
</summary>
<p>Init version.</p>
</details>
</section>
<section id="todo-list" class="level2">
<h2 class="anchored" data-anchor-id="todo-list">TODO list</h2>
<ul>
<li>Improve multi-threads performance.</li>
<li><s>Support Stereo-seq and Visium </s></li>
<li><s>Speed up <code>PISA count</code> and <code>PISA corr</code>.</s></li>
<li><em>Implement parse strategy for cell hash and CITEseq (frozen)</em>.</li>
<li><s>Assemble reads original from one molecule;</s></li>
<li><s>Implement new designed and more user-friendly <code>parse</code>;</s></li>
<li><em>Support loom output (frozen)</em>;</li>
<li><s>Export unspliced matrix for velocity;</s></li>
<li><s>Upgrade <code>PISA parse</code> for faster process fastqs.</s></li>
</ul>
</section>
<section id="reporting-issues" class="level2">
<h2 class="anchored" data-anchor-id="reporting-issues">Reporting Issues</h2>
<p>If you have any suggestion or report issues, please using Github issues page <a href="https://github.com/shiquan/PISA/issues" class="uri">https://github.com/shiquan/PISA/issues</a>.</p>
</section>
<section id="citation" class="level2">
<h2 class="anchored" data-anchor-id="citation">Citation</h2>
<p>Shi Q, Liu S, Kristiansen K, Liu L. The FASTQ+ format and PISA. Bioinformatics. 2022 Sep 30;38(19):4639-4642. doi: 10.1093/bioinformatics/btac562. PMID: 35993907.</p>


</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>