<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.53">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Annotating and prioritizing genetic variants for single cell RNA sequencing</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./Yano_ASE.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
          <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Overview</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./fastq.html"> 
<span class="menu-text">FASTQ+</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./PISA.html"> 
<span class="menu-text">PISA</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./Yano.html"> 
<span class="menu-text">Yano</span></a>
  </li>  
</ul>
          <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="GitHub" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label="GitHub"><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/shiquan">
            My GitHub Page
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/shiquan/Yano/discussions">
            Discussions
            </a>
          </li>
      </ul>
    </div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
            <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./workflow1.html">Workflows &amp; Short cases</a></li><li class="breadcrumb-item"><a href="./Yano_anno.html">Annotating genetic variants</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Workflows &amp; Short cases</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./workflow1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">From FASTQ to counts</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./anno.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Annotate various features</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Yano_AS.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Alternative splicing analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Yano_ASE.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Allele-specific gene expression analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Yano_anno.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Annotating genetic variants</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#predict-the-consequence-of-genetic-variants" id="toc-predict-the-consequence-of-genetic-variants" class="nav-link active" data-scroll-target="#predict-the-consequence-of-genetic-variants">Predict the consequence of genetic variants</a></li>
  <li><a href="#annotate-genetic-variants" id="toc-annotate-genetic-variants" class="nav-link" data-scroll-target="#annotate-genetic-variants">Annotate genetic variants</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content column-body" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./workflow1.html">Workflows &amp; Short cases</a></li><li class="breadcrumb-item"><a href="./Yano_anno.html">Annotating genetic variants</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Annotating and prioritizing genetic variants for single cell RNA sequencing</h1>
</div>



<div class="quarto-title-meta column-body">

    
  
    
  </div>
  


</header>


<p>This vignette demonstrates how Yano can annotate genetic variants using public databases and assist users in interpreting the functional impact of these variants.</p>
<section id="predict-the-consequence-of-genetic-variants" class="level2">
<h2 class="anchored" data-anchor-id="predict-the-consequence-of-genetic-variants">Predict the consequence of genetic variants</h2>
<p>Load the object which generated from <a href="./Yano_ASE.html">allele specific gene expression from scRNA-seq</a>. In that case, we used annoVAR to annotate the gene and approximate gene region type for the variants. The predicted consequences were derived from transcript annotations. However, gencode annotations, generated as part of the ENCODE project, contain numerous truncated transcript records that are not fully consistent with the reference genome, which can introduce bias when predicting amino acid changes. For clinical research, validated annotations like RefSeq are preferred for genetic variant annotation. In this analysis, RefSeq annotations were downloaded from the UCSC server to provide more accurate and reliable predictions. Furthermore, the human genome reference is also need for the prediction.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://hgdownload.soe.ucsc.edu/goldenPath/archive/hg38/ncbiRefSeq/110/hg38.110.ncbiRefSeq.gtf.gz</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh38.p13.genome.fa.gz</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span> GRCh38.p13.genome.fa.gz</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># FASTA need to be indexed</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> faidx GRCh38.p13.genome.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(Yano)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Yano</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching packages ────────────────────────────────────────────── Yano 1.0 ──
✔ dplyr   1.1.4     ✔ Seurat  5.1.0
✔ ggplot2 3.5.1     </code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"glioblastoma5k.rds"</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>gtf <span class="ot">&lt;-</span> <span class="fu">gtf2db</span>(<span class="st">"hg38.110.ncbiRefSeq.gtf.gz"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[2024-11-28 12:22:41] [32mGTF loading..[0m
[2024-11-28 12:23:02] [32mLoad 66991 genes.[0m
[2024-11-28 12:23:02] [32mLoad time : 21.191 sec[0m</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"var"</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">annoVAR</span>(obj, <span class="at">gtf =</span> gtf, <span class="at">fasta =</span> <span class="st">"./GRCh38.p13.genome.fa"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Parse names ..</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>CDS region of transcript NM_001261418.2 cannot translate to codons properly.
CDS region of transcript XM_047433500.1 cannot translate to codons properly.
CDS region of transcript NM_001350237.2 cannot translate to codons properly.
CDS region of transcript XM_017002563.2 cannot translate to codons properly.
CDS region of transcript XM_047432264.1 cannot translate to codons properly.
CDS region of transcript XM_017002566.2 cannot translate to codons properly.
CDS region of transcript NM_001350238.2 cannot translate to codons properly.
CDS region of transcript XM_017002570.2 cannot translate to codons properly.
CDS region of transcript XM_017002567.2 cannot translate to codons properly.
CDS region of transcript XM_047432272.1 cannot translate to codons properly.
CDS region of transcript XM_017002562.2 cannot translate to codons properly.
CDS region of transcript NM_001350241.2 cannot translate to codons properly.
CDS region of transcript XM_047432255.1 cannot translate to codons properly.
CDS region of transcript XM_017002560.3 cannot translate to codons properly.
CDS region of transcript NM_001350239.2 cannot translate to codons properly.
CDS region of transcript NM_001032291.3 cannot translate to codons properly.
CDS region of transcript NM_001394005.1 cannot translate to codons properly.
CDS region of transcript NM_001350240.2 cannot translate to codons properly.
CDS region of transcript NM_032636.8 cannot translate to codons properly.
CDS region of transcript XM_047432251.1 cannot translate to codons properly.
CDS region of transcript XM_047432274.1 cannot translate to codons properly.
CDS region of transcript XM_017002569.2 cannot translate to codons properly.
CDS region of transcript XM_017002564.2 cannot translate to codons properly.
CDS region of transcript NM_001350242.2 cannot translate to codons properly.
CDS region of transcript NM_001394002.1 cannot translate to codons properly.
CDS region of transcript NM_001363309.2 cannot translate to codons properly.
CDS region of transcript XM_047432275.1 cannot translate to codons properly.
CDS region of transcript XM_005271283.4 cannot translate to codons properly.
CDS region of transcript XM_017015641.3 cannot translate to codons properly.
CDS region of transcript NM_001385672.1 cannot translate to codons properly.
CDS region of transcript XM_011536897.3 cannot translate to codons properly.
CDS region of transcript NM_001018005.2 cannot translate to codons properly.
CDS region of transcript XM_006720667.5 cannot translate to codons properly.
CDS region of transcript NM_001301244.2 cannot translate to codons properly.
CDS region of transcript XM_047433003.1 cannot translate to codons properly.
CDS region of transcript NM_001330346.2 cannot translate to codons properly.
CDS region of transcript XM_005254650.4 cannot translate to codons properly.
CDS region of transcript NM_004850.5 cannot translate to codons properly.
CDS region of transcript NM_001321643.2 cannot translate to codons properly.
CDS region of transcript NM_080794.4 cannot translate to codons properly.
CDS region of transcript NM_001243385.2 cannot translate to codons properly.
CDS region of transcript XM_047441349.1 cannot translate to codons properly.
CDS region of transcript XM_024452237.2 cannot translate to codons properly.
CDS region of transcript XM_047441350.1 cannot translate to codons properly.
CDS region of transcript XM_047447996.1 cannot translate to codons properly.
CDS region of transcript XM_047416720.1 cannot translate to codons properly.
CDS region of transcript XM_047416719.1 cannot translate to codons properly.
CDS region of transcript XM_005248250.4 cannot translate to codons properly.
CDS region of transcript XM_011513966.3 cannot translate to codons properly.
CDS region of transcript NM_001077654.3 cannot translate to codons properly.
CDS region of transcript NM_001385620.1 cannot translate to codons properly.
CDS region of transcript XM_047418494.1 cannot translate to codons properly.
CDS region of transcript XM_017010597.2 cannot translate to codons properly.
CDS region of transcript XM_006716181.5 cannot translate to codons properly.
CDS region of transcript XM_047421050.1 cannot translate to codons properly.
CDS region of transcript XM_047422695.1 cannot translate to codons properly.
CDS region of transcript rna-CYTB cannot translate to codons properly.
CDS region of transcript rna-COX3 cannot translate to codons properly.
CDS region of transcript rna-ND1 cannot translate to codons properly.
CDS region of transcript rna-ND2 cannot translate to codons properly.
CDS region of transcript rna-ND4 cannot translate to codons properly.
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270442.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270729.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000205.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000205.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000195.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000218.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270737.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270713.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270733.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270733.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270733.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270733.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome KI270733.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000219.1 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m
[33m[warnings] Chromosome GL000216.2 not found in GTF, use wrong database? [0m</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'var'</span>]][[]] <span class="ot">-&gt;</span> df</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   chr   start ref alt strand          locus gene_name     type
chr1:633192G=/+   chr1  633192   G   G      +  chr1:633192/+  MTCO2P12     utr5
chr1:1000079A&gt;G/- chr1 1000079   A   G      - chr1:1000079/-      HES4     utr5
chr1:1000731C=/+  chr1 1000731   C   C      + chr1:1000731/+      HES4 upstream
chr1:1000814A&gt;G/+ chr1 1000814   A   G      + chr1:1000814/+      HES4 upstream
chr1:1000830C&gt;A/+ chr1 1000830   C   A      + chr1:1000830/+      HES4 upstream
chr1:1014228G=/+  chr1 1014228   G   G      + chr1:1014228/+     ISG15     exon
                  moransi.pval     moransi autocorr.variable    locus.D
chr1:633192G=/+   3.235392e-01 0.001330283             FALSE         NA
chr1:1000079A&gt;G/- 1.270509e-01 0.004159267             FALSE         NA
chr1:1000731C=/+  5.063894e-03 0.010004135              TRUE         NA
chr1:1000814A&gt;G/+ 6.632273e-06 0.017655831              TRUE         NA
chr1:1000830C&gt;A/+ 3.142598e-06 0.018340286              TRUE         NA
chr1:1014228G=/+  3.650720e-80 0.079578829              TRUE 0.04032407
                    locus.r locus.pval locus.mean  locus.var locus.padj
chr1:633192G=/+          NA         NA         NA         NA         NA
chr1:1000079A&gt;G/-        NA         NA         NA         NA         NA
chr1:1000731C=/+         NA         NA         NA         NA         NA
chr1:1000814A&gt;G/+        NA         NA         NA         NA         NA
chr1:1000830C&gt;A/+        NA         NA         NA         NA         NA
chr1:1014228G=/+  0.8618829  0.9999916  0.1330176 0.02047667          1
                            consequence
chr1:633192G=/+                     ref
chr1:1000079A&gt;G/-                  utr5
chr1:1000731C=/+  antisense_upstream_1K
chr1:1000814A&gt;G/+ antisense_upstream_1K
chr1:1000830C&gt;A/+ antisense_upstream_1K
chr1:1014228G=/+                    ref</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(df<span class="sc">$</span>consequence)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
antisense_downstream_1K          antisense_exon        antisense_intron 
                     84                     623                    4446 
  antisense_upstream_1K           downstream_1K               exon_loss 
                    192                     350                       3 
  frameshift_elongation   frameshift_truncation              intergenic 
                      1                       3                    2851 
                 intron                missense          noncoding_exon 
                  11827                     661                     935 
       noncoding_intron noncoding_splice_region                     ref 
                    914                       4                   12064 
        splice_acceptor            splice_donor           splice_region 
                      5                       9                      26 
           splice_sites              start_loss          start_retained 
                     12                       2                       1 
            stop_gained              synonymous             upstream_1K 
                      5                     831                     111 
                   utr3                    utr5 
                   5018                     625 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">subset</span>(df, consequence <span class="sc">%in%</span> <span class="st">"stop_gained"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                      chr     start ref alt strand            locus gene_name
chr11:62602409G&gt;A/- chr11  62602409   G   A      - chr11:62602409/-      EML3
chr16:15086183G&gt;A/- chr16  15086183   G   A      - chr16:15086183/-      RRN3
chr21:21480040G&gt;A/+ chr21  21480040   G   A      + chr21:21480040/+     NCAM2
chr5:140557175G&gt;T/-  chr5 140557175   G   T      - chr5:140557175/-      SRA1
chr7:55473014G&gt;A/-   chr7  55473014   G   A      -  chr7:55473014/-     VOPP1
                      type moransi.pval      moransi autocorr.variable locus.D
chr11:62602409G&gt;A/-   exon  0.336093969  0.001276426             FALSE      NA
chr16:15086183G&gt;A/-   exon  0.655940637 -0.002113975             FALSE      NA
chr21:21480040G&gt;A/+ intron  0.001069736  0.012035608              TRUE      NA
chr5:140557175G&gt;T/-   exon  0.268485438  0.001999330             FALSE      NA
chr7:55473014G&gt;A/-    exon  0.006479300  0.009925890              TRUE      NA
                    locus.r locus.pval locus.mean locus.var locus.padj
chr11:62602409G&gt;A/-      NA         NA         NA        NA         NA
chr16:15086183G&gt;A/-      NA         NA         NA        NA         NA
chr21:21480040G&gt;A/+      NA         NA         NA        NA         NA
chr5:140557175G&gt;T/-      NA         NA         NA        NA         NA
chr7:55473014G&gt;A/-       NA         NA         NA        NA         NA
                    consequence
chr11:62602409G&gt;A/- stop_gained
chr16:15086183G&gt;A/- stop_gained
chr21:21480040G&gt;A/+ stop_gained
chr5:140557175G&gt;T/- stop_gained
chr7:55473014G&gt;A/-  stop_gained</code></pre>
</div>
</div>
</section>
<section id="annotate-genetic-variants" class="level2">
<h2 class="anchored" data-anchor-id="annotate-genetic-variants">Annotate genetic variants</h2>
<p>One key advancement in genetic variant studies is the ability to perform functional interpretation and annotation using genotype-phenotype and allele frequency databases. For demonstration purposes, we use the <a href="https://www.ncbi.nlm.nih.gov/clinvar/">ClinVar</a> database for annotation in this vignette.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download the database in VCF format</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span>  https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar.vcf.gz</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co"># remeber to index this database, otherwise annoVAR will report error.</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index clinvar.vcf.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Let’s take a look at the format of the ClinVar database. As noted, the chromosome names are in the NCBI style, meaning they do not include the “chr” prefix. Therefore, we need to generate a new column with chromosome names formatted for annotation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view clinvar.vcf.gz<span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> <span class="st">"^#"</span><span class="kw">|</span> <span class="fu">head</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1   69134   2205837 A   G   .   .   ALLELEID=2193183;CLNDISDB=MedGen:CN169374;CLNDN=not_specified;CLNHGVS=NC_000001.11:g.69134A&gt;G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Likely_benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1
1   69314   3205580 T   G   .   .   ALLELEID=3374047;CLNDISDB=MedGen:CN169374;CLNDN=not_specified;CLNHGVS=NC_000001.11:g.69314T&gt;G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1
1   69423   3205581 G   A   .   .   ALLELEID=3374048;CLNDISDB=MedGen:CN169374;CLNDN=not_specified;CLNHGVS=NC_000001.11:g.69423G&gt;A;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1
1   69581   2252161 C   G   .   .   ALLELEID=2238986;CLNDISDB=MedGen:CN169374;CLNDN=not_specified;CLNHGVS=NC_000001.11:g.69581C&gt;G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1
1   69682   2396347 G   A   .   .   ALLELEID=2386655;CLNDISDB=MedGen:CN169374;CLNDN=not_specified;CLNHGVS=NC_000001.11:g.69682G&gt;A;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1
1   69731   3205582 T   C   .   .   ALLELEID=3374049;CLNDISDB=MedGen:CN169374;CLNDN=not_specified;CLNHGVS=NC_000001.11:g.69731T&gt;C;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1
1   69769   2288999 T   C   .   .   ALLELEID=2278803;CLNDISDB=MedGen:CN169374;CLNDN=not_specified;CLNHGVS=NC_000001.11:g.69769T&gt;C;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1
1   69995   2351346 G   C   .   .   ALLELEID=2333177;CLNDISDB=MedGen:CN169374;CLNDN=not_specified;CLNHGVS=NC_000001.11:g.69995G&gt;C;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1
1   925946  1924157 C   G   .   .   ALLELEID=1983057;CLNDISDB=MedGen:C3661900;CLNDN=not_provided;CLNHGVS=NC_000001.11:g.925946C&gt;G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=SAMD11:148398;MC=SO:0001583|missense_variant;ORIGIN=1
1   925952  1019397 G   A   .   .   ALLELEID=1003021;CLNDISDB=MedGen:C3661900;CLNDN=not_provided;CLNHGVS=NC_000001.11:g.925952G&gt;A;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=SAMD11:148398;MC=SO:0001583|missense_variant;ORIGIN=1;RS=1640863258</code></pre>
</div>
</div>
<p>We need to create new chromosome names that are compatible with the NCBI database format, because chromosome names in Clinvar database are look like ‘1’, ‘2’.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'var'</span>]][[<span class="st">'chr0'</span>]] <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"chr(.*)"</span>,<span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, obj[[<span class="st">'var'</span>]][[]]<span class="sc">$</span>chr)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'var'</span>]][[]] <span class="sc">%&gt;%</span> head</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   chr   start ref alt strand          locus gene_name     type
chr1:633192G=/+   chr1  633192   G   G      +  chr1:633192/+  MTCO2P12     utr5
chr1:1000079A&gt;G/- chr1 1000079   A   G      - chr1:1000079/-      HES4     utr5
chr1:1000731C=/+  chr1 1000731   C   C      + chr1:1000731/+      HES4 upstream
chr1:1000814A&gt;G/+ chr1 1000814   A   G      + chr1:1000814/+      HES4 upstream
chr1:1000830C&gt;A/+ chr1 1000830   C   A      + chr1:1000830/+      HES4 upstream
chr1:1014228G=/+  chr1 1014228   G   G      + chr1:1014228/+     ISG15     exon
                  moransi.pval     moransi autocorr.variable    locus.D
chr1:633192G=/+   3.235392e-01 0.001330283             FALSE         NA
chr1:1000079A&gt;G/- 1.270509e-01 0.004159267             FALSE         NA
chr1:1000731C=/+  5.063894e-03 0.010004135              TRUE         NA
chr1:1000814A&gt;G/+ 6.632273e-06 0.017655831              TRUE         NA
chr1:1000830C&gt;A/+ 3.142598e-06 0.018340286              TRUE         NA
chr1:1014228G=/+  3.650720e-80 0.079578829              TRUE 0.04032407
                    locus.r locus.pval locus.mean  locus.var locus.padj
chr1:633192G=/+          NA         NA         NA         NA         NA
chr1:1000079A&gt;G/-        NA         NA         NA         NA         NA
chr1:1000731C=/+         NA         NA         NA         NA         NA
chr1:1000814A&gt;G/+        NA         NA         NA         NA         NA
chr1:1000830C&gt;A/+        NA         NA         NA         NA         NA
chr1:1014228G=/+  0.8618829  0.9999916  0.1330176 0.02047667          1
                            consequence chr0
chr1:633192G=/+                     ref    1
chr1:1000079A&gt;G/-                  utr5    1
chr1:1000731C=/+  antisense_upstream_1K    1
chr1:1000814A&gt;G/+ antisense_upstream_1K    1
chr1:1000830C&gt;A/+ antisense_upstream_1K    1
chr1:1014228G=/+                    ref    1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># here we annotate RS id and clinical signification for variants, to check the VCF header for all possible tags</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">annoVAR</span>(obj, <span class="at">vcf =</span> <span class="st">"clinvar.vcf.gz"</span>, <span class="at">chr =</span> <span class="st">"chr0"</span>, <span class="at">tags =</span> <span class="fu">c</span>(<span class="st">"RS"</span>, <span class="st">"CLNSIG"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Parse names ..</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The RS id and CLNSIG tag now created</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'var'</span>]][[]] <span class="ot">-&gt;</span> df</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   chr   start ref alt strand          locus gene_name     type
chr1:633192G=/+   chr1  633192   G   G      +  chr1:633192/+  MTCO2P12     utr5
chr1:1000079A&gt;G/- chr1 1000079   A   G      - chr1:1000079/-      HES4     utr5
chr1:1000731C=/+  chr1 1000731   C   C      + chr1:1000731/+      HES4 upstream
chr1:1000814A&gt;G/+ chr1 1000814   A   G      + chr1:1000814/+      HES4 upstream
chr1:1000830C&gt;A/+ chr1 1000830   C   A      + chr1:1000830/+      HES4 upstream
chr1:1014228G=/+  chr1 1014228   G   G      + chr1:1014228/+     ISG15     exon
                  moransi.pval     moransi autocorr.variable    locus.D
chr1:633192G=/+   3.235392e-01 0.001330283             FALSE         NA
chr1:1000079A&gt;G/- 1.270509e-01 0.004159267             FALSE         NA
chr1:1000731C=/+  5.063894e-03 0.010004135              TRUE         NA
chr1:1000814A&gt;G/+ 6.632273e-06 0.017655831              TRUE         NA
chr1:1000830C&gt;A/+ 3.142598e-06 0.018340286              TRUE         NA
chr1:1014228G=/+  3.650720e-80 0.079578829              TRUE 0.04032407
                    locus.r locus.pval locus.mean  locus.var locus.padj
chr1:633192G=/+          NA         NA         NA         NA         NA
chr1:1000079A&gt;G/-        NA         NA         NA         NA         NA
chr1:1000731C=/+         NA         NA         NA         NA         NA
chr1:1000814A&gt;G/+        NA         NA         NA         NA         NA
chr1:1000830C&gt;A/+        NA         NA         NA         NA         NA
chr1:1014228G=/+  0.8618829  0.9999916  0.1330176 0.02047667          1
                            consequence chr0   RS CLNSIG
chr1:633192G=/+                     ref    1 &lt;NA&gt;   &lt;NA&gt;
chr1:1000079A&gt;G/-                  utr5    1 &lt;NA&gt;   &lt;NA&gt;
chr1:1000731C=/+  antisense_upstream_1K    1 &lt;NA&gt;   &lt;NA&gt;
chr1:1000814A&gt;G/+ antisense_upstream_1K    1 &lt;NA&gt;   &lt;NA&gt;
chr1:1000830C&gt;A/+ antisense_upstream_1K    1 &lt;NA&gt;   &lt;NA&gt;
chr1:1014228G=/+                    ref    1 1921 Benign</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We note that there are several pathogenic variants detected</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(df[[<span class="st">'CLNSIG'</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
                                             association 
                                                       8 
                                                  Benign 
                                                    1702 
                                    Benign/Likely_benign 
                                                      47 
            Conflicting_classifications_of_pathogenicity 
                                                       8 
Conflicting_classifications_of_pathogenicity|risk_factor 
                                                       2 
                                           Likely_benign 
                                                      56 
                no_classification_for_the_single_variant 
                                                       8 
                                            not_provided 
                                                       1 
                                                   other 
                                                      11 
                            Pathogenic/Likely_pathogenic 
                                                       3 
                                  Pathogenic|risk_factor 
                                                       1 
                                             risk_factor 
                                                       5 
                                  Uncertain_significance 
                                                      26 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">%&gt;%</span> <span class="fu">filter</span>(CLNSIG <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">'Pathogenic'</span>, <span class="st">'Pathogenic/Likely_pathogenic'</span>, <span class="st">'Pathogenic|risk_factor'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                      chr     start ref alt strand            locus gene_name
chr1:97450058C=/-    chr1  97450058   C   C      -  chr1:97450058/-      DPYD
chr1:196690107C=/+   chr1 196690107   C   C      + chr1:196690107/+       CFH
chr13:50945445G=/+  chr13  50945445   G   G      + chr13:50945445/+  RNASEH2B
chr13:50945445G&gt;A/+ chr13  50945445   G   A      + chr13:50945445/+  RNASEH2B
                          type moransi.pval      moransi autocorr.variable
chr1:97450058C=/-       intron 4.368919e-01 0.0001360072             FALSE
chr1:196690107C=/+  multigenes 3.903101e-61 0.0647296819              TRUE
chr13:50945445G=/+        exon 1.374427e-06 0.0191248673              TRUE
chr13:50945445G&gt;A/+       exon 7.057338e-07 0.0197286674              TRUE
                       locus.D   locus.r locus.pval locus.mean  locus.var
chr1:97450058C=/-           NA        NA         NA         NA         NA
chr1:196690107C=/+  0.07481304 0.7473314  0.9902322  0.1350459 0.02537402
chr13:50945445G=/+  0.11670969 0.3809105  0.8034837  0.1316099 0.01736865
chr13:50945445G&gt;A/+ 0.12248441 0.3809105  0.7251275  0.1352893 0.02133448
                    locus.padj consequence chr0       RS
chr1:97450058C=/-           NA         ref    1  3918290
chr1:196690107C=/+           1         ref    1  1061170
chr13:50945445G=/+           1         ref   13 75184679
chr13:50945445G&gt;A/+          1    missense   13 75184679
                                          CLNSIG
chr1:97450058C=/-   Pathogenic/Likely_pathogenic
chr1:196690107C=/+        Pathogenic|risk_factor
chr13:50945445G=/+  Pathogenic/Likely_pathogenic
chr13:50945445G&gt;A/+ Pathogenic/Likely_pathogenic</code></pre>
</div>
</div>
<p>In the selected SNVs, we found some reference alleles are highlighted, such as chr15:92992967C=/+. These reference alelles are more like to be benign allele. Let’s query the orignal record in the clinvar database to check what exactly stituation it should be.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view clinvar.vcf.gz 15:92992967-92992967</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>##fileformat=VCFv4.1
##FILTER=&lt;ID=PASS,Description="All filters passed"&gt;
##fileDate=2024-10-01
##source=ClinVar
##reference=GRCh38
##ID=&lt;Description="ClinVar Variation ID"&gt;
##INFO=&lt;ID=AF_ESP,Number=1,Type=Float,Description="allele frequencies from GO-ESP"&gt;
##INFO=&lt;ID=AF_EXAC,Number=1,Type=Float,Description="allele frequencies from ExAC"&gt;
##INFO=&lt;ID=AF_TGP,Number=1,Type=Float,Description="allele frequencies from TGP"&gt;
##INFO=&lt;ID=ALLELEID,Number=1,Type=Integer,Description="the ClinVar Allele ID"&gt;
##INFO=&lt;ID=CLNDN,Number=.,Type=String,Description="ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB"&gt;
##INFO=&lt;ID=CLNDNINCL,Number=.,Type=String,Description="For included Variant : ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB"&gt;
##INFO=&lt;ID=CLNDISDB,Number=.,Type=String,Description="Tag-value pairs of disease database name and identifier submitted for germline classifications, e.g. OMIM:NNNNNN"&gt;
##INFO=&lt;ID=CLNDISDBINCL,Number=.,Type=String,Description="For included Variant: Tag-value pairs of disease database name and identifier for germline classifications, e.g. OMIM:NNNNNN"&gt;
##INFO=&lt;ID=CLNHGVS,Number=.,Type=String,Description="Top-level (primary assembly, alt, or patch) HGVS expression."&gt;
##INFO=&lt;ID=CLNREVSTAT,Number=.,Type=String,Description="ClinVar review status of germline classification for the Variation ID"&gt;
##INFO=&lt;ID=CLNSIG,Number=.,Type=String,Description="Aggregate germline classification for this single variant; multiple values are separated by a vertical bar"&gt;
##INFO=&lt;ID=CLNSIGCONF,Number=.,Type=String,Description="Conflicting germline classification for this single variant; multiple values are separated by a vertical bar"&gt;
##INFO=&lt;ID=CLNSIGINCL,Number=.,Type=String,Description="Germline classification for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:classification; multiple values are separated by a vertical bar"&gt;
##INFO=&lt;ID=CLNVC,Number=1,Type=String,Description="Variant type"&gt;
##INFO=&lt;ID=CLNVCSO,Number=1,Type=String,Description="Sequence Ontology id for variant type"&gt;
##INFO=&lt;ID=CLNVI,Number=.,Type=String,Description="the variant's clinical sources reported as tag-value pairs of database and variant identifier"&gt;
##INFO=&lt;ID=DBVARID,Number=.,Type=String,Description="nsv accessions from dbVar for the variant"&gt;
##INFO=&lt;ID=GENEINFO,Number=1,Type=String,Description="Gene(s) for the variant reported as gene symbol:gene id. The gene symbol and id are delimited by a colon (:) and each pair is delimited by a vertical bar (|)"&gt;
##INFO=&lt;ID=MC,Number=.,Type=String,Description="comma separated list of molecular consequence in the form of Sequence Ontology ID|molecular_consequence"&gt;
##INFO=&lt;ID=ONCDN,Number=.,Type=String,Description="ClinVar's preferred disease name for the concept specified by disease identifiers in ONCDISDB"&gt;
##INFO=&lt;ID=ONCDNINCL,Number=.,Type=String,Description="For included variant: ClinVar's preferred disease name for the concept specified by disease identifiers in ONCDISDBINCL"&gt;
##INFO=&lt;ID=ONCDISDB,Number=.,Type=String,Description="Tag-value pairs of disease database name and identifier submitted for oncogenicity classifications, e.g. MedGen:NNNNNN"&gt;
##INFO=&lt;ID=ONCDISDBINCL,Number=.,Type=String,Description="For included variant: Tag-value pairs of disease database name and identifier for oncogenicity classifications, e.g. OMIM:NNNNNN"&gt;
##INFO=&lt;ID=ONC,Number=.,Type=String,Description="Aggregate oncogenicity classification for this single variant; multiple values are separated by a vertical bar"&gt;
##INFO=&lt;ID=ONCINCL,Number=.,Type=String,Description="Oncogenicity classification for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:classification; multiple values are separated by a vertical bar"&gt;
##INFO=&lt;ID=ONCREVSTAT,Number=.,Type=String,Description="ClinVar review status of oncogenicity classification for the Variation ID"&gt;
##INFO=&lt;ID=ONCCONF,Number=.,Type=String,Description="Conflicting oncogenicity classification for this single variant; multiple values are separated by a vertical bar"&gt;
##INFO=&lt;ID=ORIGIN,Number=.,Type=String,Description="Allele origin. One or more of the following values may be added: 0 - unknown; 1 - germline; 2 - somatic; 4 - inherited; 8 - paternal; 16 - maternal; 32 - de-novo; 64 - biparental; 128 - uniparental; 256 - not-tested; 512 - tested-inconclusive; 1073741824 - other"&gt;
##INFO=&lt;ID=RS,Number=.,Type=String,Description="dbSNP ID (i.e. rs number)"&gt;
##INFO=&lt;ID=SCIDN,Number=.,Type=String,Description="ClinVar's preferred disease name for the concept specified by disease identifiers in SCIDISDB"&gt;
##INFO=&lt;ID=SCIDNINCL,Number=.,Type=String,Description="For included variant: ClinVar's preferred disease name for the concept specified by disease identifiers in SCIDISDBINCL"&gt;
##INFO=&lt;ID=SCIDISDB,Number=.,Type=String,Description="Tag-value pairs of disease database name and identifier submitted for somatic clinial impact classifications, e.g. MedGen:NNNNNN"&gt;
##INFO=&lt;ID=SCIDISDBINCL,Number=.,Type=String,Description="For included variant: Tag-value pairs of disease database name and identifier for somatic clinical impact classifications, e.g. OMIM:NNNNNN"&gt;
##INFO=&lt;ID=SCIREVSTAT,Number=.,Type=String,Description="ClinVar review status of somatic clinical impact for the Variation ID"&gt;
##INFO=&lt;ID=SCI,Number=.,Type=String,Description="Aggregate somatic clinical impact for this single variant; multiple values are separated by a vertical bar"&gt;
##INFO=&lt;ID=SCIINCL,Number=.,Type=String,Description="Somatic clinical impact classification for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:classification; multiple values are separated by a vertical bar"&gt;
##contig=&lt;ID=1&gt;
##contig=&lt;ID=2&gt;
##contig=&lt;ID=3&gt;
##contig=&lt;ID=4&gt;
##contig=&lt;ID=5&gt;
##contig=&lt;ID=6&gt;
##contig=&lt;ID=7&gt;
##contig=&lt;ID=8&gt;
##contig=&lt;ID=9&gt;
##contig=&lt;ID=10&gt;
##contig=&lt;ID=11&gt;
##contig=&lt;ID=12&gt;
##contig=&lt;ID=13&gt;
##contig=&lt;ID=14&gt;
##contig=&lt;ID=15&gt;
##contig=&lt;ID=16&gt;
##contig=&lt;ID=17&gt;
##contig=&lt;ID=18&gt;
##contig=&lt;ID=19&gt;
##contig=&lt;ID=20&gt;
##contig=&lt;ID=21&gt;
##contig=&lt;ID=22&gt;
##contig=&lt;ID=X&gt;
##contig=&lt;ID=Y&gt;
##contig=&lt;ID=MT&gt;
##contig=&lt;ID=NT_113889.1&gt;
##contig=&lt;ID=NT_187633.1&gt;
##contig=&lt;ID=NT_187661.1&gt;
##contig=&lt;ID=NT_187693.1&gt;
##contig=&lt;ID=NW_009646201.1&gt;
##bcftools_viewVersion=1.13+htslib-1.13
##bcftools_viewCommand=view clinvar.vcf.gz 15:92992967-92992967; Date=Thu Nov 28 12:23:09 2024
#CHROM  POS ID  REF ALT QUAL    FILTER  INFO
15  92992967    633515  C   G   .   .   ALLELEID=621980;CLNDISDB=MedGen:CN517202;CLNDN=not_provided;CLNHGVS=NC_000015.10:g.92992967C&gt;G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Pathogenic;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=CHD2:1106;MC=SO:0001587|nonsense;ORIGIN=1;RS=2272457
15  92992967    257709  C   T   .   .   AF_ESP=0.2485;AF_EXAC=0.22658;AF_TGP=0.19349;ALLELEID=255446;CLNDISDB=MeSH:D030342,MedGen:C0950123|MONDO:MONDO:0014150,MedGen:C3809278,OMIM:615369,Orphanet:1942,Orphanet:2382|MedGen:CN169374|MedGen:C3661900;CLNDN=Inborn_genetic_diseases|Developmental_and_epileptic_encephalopathy_94|not_specified|not_provided;CLNHGVS=NC_000015.10:g.92992967C&gt;T;CLNREVSTAT=criteria_provided,_multiple_submitters,_no_conflicts;CLNSIG=Benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;CLNVI=ClinGen:CA7748248;GENEINFO=CHD2:1106;MC=SO:0001819|synonymous_variant;ORIGIN=1;RS=2272457</code></pre>
</div>
</div>
<p>You can easily observe that the pathogenic allele is G, where the change from C to G leads to a nonsense mutation. However, if we examine the header, we see that the ‘Number’ field of CLNSIG is labeled as ‘.’, which indicates that this term is not allele-specific—an inappropriate type for this record. In fact, this issue extends beyond just CLNSIG; all terms in the ClinVar database are labeled as ‘.’.</p>
<p>To correct this, we need to manually change the type from ‘.’ to ‘A’ (allele-specific) for allele-specific tags in the VCF file. For more information, please refer to the <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF specification</a>.</p>
<p>Unfortunately, many databases distributed in VCF format may contain non-uniformly formatted tags. Apart from manually updating these databases or filter reference allele afterward, there is no straightforward solution to this issue.</p>
<p>In addition, the RS tag is annotated, and we can refer to the NCBI SNP database for more information. For example, the mutation chr13:50945445G&gt;A/+ at RS 75184679 can be found at the following link: <a href="https://www.ncbi.nlm.nih.gov/snp/rs75184679">https://www.ncbi.nlm.nih.gov/snp/rs75184679</a>. This mutation is associated with conditions such as “Abnormality of the nervous system” and other diseases.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(obj, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"chr13:50945445G&gt;A/+"</span>, <span class="st">"chr13:50945445G=/+"</span>, <span class="st">"RNASEH2B"</span>), <span class="at">order =</span> <span class="cn">TRUE</span>, <span class="at">ncol=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Could not find RNASEH2B in the default search locations, found in
'RNA' assay instead</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_anno_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
<p>Note that not all reads from this gene cover the SNV, which results in the SNV expression being sparser compared to the overall gene expression. Moreover, it also possible to map the clinical significant labels or other tags to the Manhatten plot.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>sel.chrs <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">21</span>, <span class="st">"X"</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FbtPlot</span>(obj, <span class="at">val =</span> <span class="st">"locus.padj"</span>, <span class="at">remove.chr =</span> <span class="cn">TRUE</span>, <span class="at">sel.chrs =</span> sel.chrs, <span class="at">col.by =</span> <span class="st">"CLNSIG"</span>, <span class="at">pt.size =</span> <span class="dv">2</span>, <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"#131313"</span>,<span class="st">"blue"</span>,<span class="st">"#A6CEE3"</span>,<span class="st">"#1F78B4"</span>,<span class="st">"#B2DF8A"</span>,<span class="st">"#33A02C"</span>,<span class="st">"#FB9A99"</span>,<span class="st">"#E31A1C"</span>,<span class="st">"#FDBF6F"</span>,<span class="st">"#FF7F00"</span>,<span class="st">"#CAB2D6"</span>,<span class="st">"#6A3D9A"</span>,<span class="st">"#FFFF99"</span>,<span class="st">"#B15928"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="Yano_anno_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="1344"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Command</span>(obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "NormalizeData.RNA"        "FindVariableFeatures.RNA"
 [3] "ScaleData.RNA"            "RunPCA.RNA"              
 [5] "FindNeighbors.RNA.pca"    "FindClusters"            
 [7] "RunUMAP.RNA.pca"          "NormalizeData.var"       
 [9] "RunAutoCorr.var.pca"      "SetAutoCorrFeatures.var" 
[11] "RunBlockCorr.var.pca"     "ParseVarName.var"        
[13] "annoVAR.var"             </code></pre>
</div>
</div>


</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation column-body">
  <div class="nav-page nav-page-previous">
      <a href="./Yano_ASE.html" class="pagination-link" aria-label="Allele-specific gene expression analysis">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Allele-specific gene expression analysis</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->




</body></html>