<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Alternative splicing analysis for single cell RNA sequencing</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./Yano_ASE.html" rel="next">
<link href="./selector.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-de84f8d6bb715db06a919283c2d1e787.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark-9a023c9ade86a60361e96e3e3f11bc54.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-2cae7a44443c7384b30cd79a64306367.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark-2d6ecf2d45d93a56a874280a57c27dd9.min.css" rel="prefetch" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
          <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Overview</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./fastq.html"> 
<span class="menu-text">FASTQ+</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./PISA.html"> 
<span class="menu-text">PISA</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./Yano.html"> 
<span class="menu-text">Yano</span></a>
  </li>  
</ul>
          <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="GitHub" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label="GitHub"><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/shiquan">
            My GitHub Page
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/shiquan/Yano/discussions">
            Discussions
            </a>
          </li>
      </ul>
    </div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
            <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./workflow1.html">Workflows &amp; Short cases</a></li><li class="breadcrumb-item"><a href="./Yano_AS.html">Alternative splicing analysis</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Workflows &amp; Short cases</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./workflow1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">From FASTQ to counts</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./anno.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Annotate various features</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./selector.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Select cells from reduction maps</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Yano_AS.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Alternative splicing analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Yano_ASE.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Allele-specific gene expression analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Yano_anno.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Annotating genetic variants</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Visium.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Analysis multiple Visium samples</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Yano_trajectory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Cell trajectory analysis</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#prerequisite" id="toc-prerequisite" class="nav-link active" data-scroll-target="#prerequisite">0. Prerequisite</a></li>
  <li><a href="#perform-cell-clustering-with-seurat" id="toc-perform-cell-clustering-with-seurat" class="nav-link" data-scroll-target="#perform-cell-clustering-with-seurat">1. Perform cell clustering with Seurat</a></li>
  <li><a href="#perform-alternative-splicing-with-exon-assay" id="toc-perform-alternative-splicing-with-exon-assay" class="nav-link" data-scroll-target="#perform-alternative-splicing-with-exon-assay">2. Perform alternative splicing with exon assay</a></li>
  <li><a href="#load-junction-assay" id="toc-load-junction-assay" class="nav-link" data-scroll-target="#load-junction-assay">3. Load junction assay</a></li>
  <li><a href="#heatmap-analysis-for-highlight-group-specific-alternative-splicing" id="toc-heatmap-analysis-for-highlight-group-specific-alternative-splicing" class="nav-link" data-scroll-target="#heatmap-analysis-for-highlight-group-specific-alternative-splicing">4. Heatmap analysis for highlight group specific alternative splicing</a></li>
  <li><a href="#test-between-exon-and-exon-skipped-reads" id="toc-test-between-exon-and-exon-skipped-reads" class="nav-link" data-scroll-target="#test-between-exon-and-exon-skipped-reads">5. Test between exon and exon skipped reads</a></li>
  <li><a href="#questions" id="toc-questions" class="nav-link" data-scroll-target="#questions">Questions?</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content column-body" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./workflow1.html">Workflows &amp; Short cases</a></li><li class="breadcrumb-item"><a href="./Yano_AS.html">Alternative splicing analysis</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Alternative splicing analysis for single cell RNA sequencing</h1>
</div>



<div class="quarto-title-meta column-body">

    
  
    
  </div>
  


</header>


<p>This vignette uses gene, exon, and junction expression files generated from the <a href="./anno.html">Annotate Various Features for Alignment</a>. While current state-of-the-art scRNA-seq methods tend to be biased towards the 3’ or 5’ ends of transcripts, it is still possible to obtain coverage information for a subset of exons. Despite the sparsity of gene and exon expression in single cells, our spatial dissimilarity test leverages the spatial distribution properties of features. This means that even features with low overall expression but strong spatial expression patterns across cells can still be highlighted. By performing a spatial dissimilarity test between exon/junction and gene expression, we can predict potential alternative splicing events.</p>
<section id="prerequisite" class="level2">
<h2 class="anchored" data-anchor-id="prerequisite">0. Prerequisite</h2>
<ul>
<li>Make sure you have installed the R environment and <code>Yano</code> package before proceeding with the testing.</li>
</ul>
</section>
<section id="perform-cell-clustering-with-seurat" class="level2">
<h2 class="anchored" data-anchor-id="perform-cell-clustering-with-seurat">1. Perform cell clustering with Seurat</h2>
<p>Load Yano will automatically load Seurat.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(Yano)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Yano</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching packages ────────────────────────────────────────────── Yano 1.2 ──
✔ dplyr   1.1.4     ✔ Seurat  5.3.0
✔ ggplot2 3.5.2     </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read raw gene expression matrix</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>exp <span class="ot">&lt;-</span> <span class="fu">ReadPISA</span>(<span class="st">"./exp/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this section, we will perform the standard Seurat analysis pipeline. Since the spatial dissimilarity test is not rely on cell clustering so changing the resolution or other parameters for <code>FindClusters</code> and <code>RunUMAP</code> will not impact the outcome of the spatial dissimilarity test.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create Seurat object and filter droplets with fewer than 1000 genes</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(exp, <span class="at">min.features =</span> <span class="dv">1000</span>, <span class="at">min.cells =</span> <span class="dv">10</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter low quality droplets</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">"percent.mt"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(obj, <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">subset</span>(obj, nFeature_RNA <span class="sc">&lt;</span> <span class="dv">9000</span> <span class="sc">&amp;</span> percent.mt <span class="sc">&lt;</span> <span class="dv">20</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Downsampling to 2000 cells for fast testing</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> obj[, <span class="fu">sample</span>(<span class="fu">colnames</span>(obj),<span class="dv">2000</span>)]</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="co"># We run the cell clustering analysis with Seurat pipeline</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(obj) <span class="sc">%&gt;%</span> <span class="fu">FindVariableFeatures</span>() <span class="sc">%&gt;%</span> <span class="fu">ScaleData</span>() <span class="sc">%&gt;%</span>  <span class="fu">RunPCA</span>(<span class="at">verbose=</span><span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> <span class="fu">FindNeighbors</span>(<span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> <span class="fu">FindClusters</span>(<span class="at">resolution =</span> <span class="fl">0.5</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> <span class="fu">RunUMAP</span>(<span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(obj, <span class="at">label=</span><span class="cn">TRUE</span>, <span class="at">label.size =</span> <span class="dv">5</span>, <span class="at">label.box =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="perform-alternative-splicing-with-exon-assay" class="level2">
<h2 class="anchored" data-anchor-id="perform-alternative-splicing-with-exon-assay">2. Perform alternative splicing with exon assay</h2>
<p>In this section, we will compare exon expression patterns with the expression patterns of their corresponding genes in a spatial context. Here, the term “spatial” refers to the organization of cells in space. In this vignette, we will use the PCA space for the analysis, but the approach can also be applied to lineage trajectories, spatial coordinates or integration space such as harmony. The spatial dissimilarity test is divided into several steps. First, we will load exon data as a new assay in the Seurat object. In the second step, we will perform a spatial autocorrelation test for all exons and select the ones that show significant autocorrelation for further analysis. Next, we will define the binding relationship between exons and their corresponding genes and run the spatial dissimilarity test. After testing, P values and adjusted P values for each exon will be provided.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read exon count matrix file</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>exon <span class="ot">&lt;-</span> <span class="fu">ReadPISA</span>(<span class="st">"./exon/"</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the exon expression to Seurat object as a new assay, make sure the exon matrix has the same cells.</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'exon'</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(exon[, <span class="fu">colnames</span>(obj)], <span class="at">min.cells=</span><span class="dv">20</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Switch work assay to exon</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"exon"</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Empty info for exon features</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(obj[[<span class="st">'exon'</span>]][[]]) <span class="sc">%&gt;%</span> knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<tbody>
<tr class="odd">
<td style="text-align: left;">chr1:135141-135895/-/ENSG00000268903</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr1:629640-630683/+/MTND2P28</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr1:631074-632616/+/MTCO1P12</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr1:632757-633438/+/MTCO2P12</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr1:633696-634376/+/MTATP6P1</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr1:634376-634922/+/MTCO3P12</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">ParseExonName</span>(obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Working on assay exon</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene name and location are parsed from exon name</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(obj[[<span class="st">'exon'</span>]][[]]) <span class="sc">%&gt;%</span> knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 46%">
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 8%">
<col style="width: 20%">
<col style="width: 8%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">chr</th>
<th style="text-align: right;">start</th>
<th style="text-align: right;">end</th>
<th style="text-align: left;">gene_name</th>
<th style="text-align: left;">strand</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">chr1:135141-135895/-/ENSG00000268903</td>
<td style="text-align: left;">chr1</td>
<td style="text-align: right;">135141</td>
<td style="text-align: right;">135895</td>
<td style="text-align: left;">ENSG00000268903</td>
<td style="text-align: left;">-</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr1:629640-630683/+/MTND2P28</td>
<td style="text-align: left;">chr1</td>
<td style="text-align: right;">629640</td>
<td style="text-align: right;">630683</td>
<td style="text-align: left;">MTND2P28</td>
<td style="text-align: left;">+</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr1:631074-632616/+/MTCO1P12</td>
<td style="text-align: left;">chr1</td>
<td style="text-align: right;">631074</td>
<td style="text-align: right;">632616</td>
<td style="text-align: left;">MTCO1P12</td>
<td style="text-align: left;">+</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr1:632757-633438/+/MTCO2P12</td>
<td style="text-align: left;">chr1</td>
<td style="text-align: right;">632757</td>
<td style="text-align: right;">633438</td>
<td style="text-align: left;">MTCO2P12</td>
<td style="text-align: left;">+</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr1:633696-634376/+/MTATP6P1</td>
<td style="text-align: left;">chr1</td>
<td style="text-align: right;">633696</td>
<td style="text-align: right;">634376</td>
<td style="text-align: left;">MTATP6P1</td>
<td style="text-align: left;">+</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr1:634376-634922/+/MTCO3P12</td>
<td style="text-align: left;">chr1</td>
<td style="text-align: right;">634376</td>
<td style="text-align: right;">634922</td>
<td style="text-align: left;">MTCO3P12</td>
<td style="text-align: left;">+</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize the data for spatial autocorrelation test</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(obj)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">RunAutoCorr</span>(obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Working on assay : exon</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Run autocorrelation test for 117043 features.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Runtime : 19.20793 secs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>70909 autocorrelated features.</code></pre>
</div>
</div>
<p>The permutation process can be computationally expensive. In the example below, I set perm=20 to perform only 20 permutations for quicker results. However, the default setting runs 100 permutations for more accurate evaluation. While it’s possible to increase the number of permutations for even more precision, it may not always be necessary. If you’re running Yano for the first time on your dataset, setting perm=20 can help you save time and provide an initial overview of the entire dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">RunSDT</span>(obj, <span class="at">bind.name =</span> <span class="st">"gene_name"</span>, <span class="at">bind.assay =</span> <span class="st">"RNA"</span>, <span class="at">perm=</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Working on assay exon.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Working on binding assay RNA.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Use predefined weight matrix "pca_wm".</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Processing 70909 features.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Processing 14881 binding features.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Retrieve binding data from assay RNA.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Use "data" layer for test features and binding features.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Using 7 threads.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Runtime : 2.358137 mins.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now p values and adjusted p values have been generated</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(obj[[<span class="st">'exon'</span>]][[]]) <span class="sc">%&gt;%</span> knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 21%">
<col style="width: 2%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 9%">
<col style="width: 4%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 10%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 8%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">chr</th>
<th style="text-align: right;">start</th>
<th style="text-align: right;">end</th>
<th style="text-align: left;">gene_name</th>
<th style="text-align: left;">strand</th>
<th style="text-align: right;">moransi.pval</th>
<th style="text-align: right;">moransi</th>
<th style="text-align: left;">autocorr.variable</th>
<th style="text-align: right;">gene_name.D</th>
<th style="text-align: right;">gene_name.t</th>
<th style="text-align: right;">gene_name.pval</th>
<th style="text-align: right;">gene_name.padj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">chr1:135141-135895/-/ENSG00000268903</td>
<td style="text-align: left;">chr1</td>
<td style="text-align: right;">135141</td>
<td style="text-align: right;">135895</td>
<td style="text-align: left;">ENSG00000268903</td>
<td style="text-align: left;">-</td>
<td style="text-align: right;">0.0000001</td>
<td style="text-align: right;">0.0222427</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.0002562</td>
<td style="text-align: right;">-7.275754</td>
<td style="text-align: right;">0.9999997</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr1:629640-630683/+/MTND2P28</td>
<td style="text-align: left;">chr1</td>
<td style="text-align: right;">629640</td>
<td style="text-align: right;">630683</td>
<td style="text-align: left;">MTND2P28</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000013</td>
<td style="text-align: right;">0.0203117</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.0017026</td>
<td style="text-align: right;">-8.030082</td>
<td style="text-align: right;">0.9999999</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr1:631074-632616/+/MTCO1P12</td>
<td style="text-align: left;">chr1</td>
<td style="text-align: right;">631074</td>
<td style="text-align: right;">632616</td>
<td style="text-align: left;">MTCO1P12</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.7821273</td>
<td style="text-align: right;">-0.0038815</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr1:632757-633438/+/MTCO2P12</td>
<td style="text-align: left;">chr1</td>
<td style="text-align: right;">632757</td>
<td style="text-align: right;">633438</td>
<td style="text-align: left;">MTCO2P12</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.5700335</td>
<td style="text-align: right;">-0.0012501</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr1:633696-634376/+/MTATP6P1</td>
<td style="text-align: left;">chr1</td>
<td style="text-align: right;">633696</td>
<td style="text-align: right;">634376</td>
<td style="text-align: left;">MTATP6P1</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.3171474</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.0032809</td>
<td style="text-align: right;">-7.659511</td>
<td style="text-align: right;">0.9999998</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr1:634376-634922/+/MTCO3P12</td>
<td style="text-align: left;">chr1</td>
<td style="text-align: right;">634376</td>
<td style="text-align: right;">634922</td>
<td style="text-align: left;">MTCO3P12</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.8377380</td>
<td style="text-align: right;">-0.0046027</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot feature binding test plot</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FbtPlot</span>(obj, <span class="at">val =</span> <span class="st">"gene_name.padj"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="1344"></p>
</figure>
</div>
</div>
</div>
<p>The chromosome names are too long and tend to overlap in the visualization. To resolve this, you can either resize the labels or remove the ‘chr’ prefix from the chromosome names. Additionally, since the Y chromosome and mitochondrial are not of particular interest to us in this analysis, they can be excluded from the visualization.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>sel.chrs <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">21</span>, <span class="st">"X"</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FbtPlot</span>(obj, <span class="at">val =</span> <span class="st">"gene_name.padj"</span>, <span class="at">remove.chr =</span> <span class="cn">TRUE</span>, <span class="at">sel.chrs =</span> sel.chrs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="1344"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's see how many exons are expressed in different spatial pattern with their genes</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'exon'</span>]][[]] <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene_name.padj <span class="sc">&lt;</span> <span class="fl">0.001</span>) <span class="sc">%&gt;%</span> knitr<span class="sc">::</span><span class="fu">kable</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 22%">
<col style="width: 3%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 8%">
<col style="width: 3%">
<col style="width: 6%">
<col style="width: 5%">
<col style="width: 9%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 8%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">chr</th>
<th style="text-align: right;">start</th>
<th style="text-align: right;">end</th>
<th style="text-align: left;">gene_name</th>
<th style="text-align: left;">strand</th>
<th style="text-align: right;">moransi.pval</th>
<th style="text-align: right;">moransi</th>
<th style="text-align: left;">autocorr.variable</th>
<th style="text-align: right;">gene_name.D</th>
<th style="text-align: right;">gene_name.t</th>
<th style="text-align: right;">gene_name.pval</th>
<th style="text-align: right;">gene_name.padj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">chr11:35197162-35197793/+/CD44</td>
<td style="text-align: left;">chr11</td>
<td style="text-align: right;">35197162</td>
<td style="text-align: right;">35197793</td>
<td style="text-align: left;">CD44</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.3402391</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3346163</td>
<td style="text-align: right;">9.289870</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000317</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr11:75421727-75422280/+/RPS3</td>
<td style="text-align: left;">chr11</td>
<td style="text-align: right;">75421727</td>
<td style="text-align: right;">75422280</td>
<td style="text-align: left;">RPS3</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0493364</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.2921820</td>
<td style="text-align: right;">7.352049</td>
<td style="text-align: right;">3e-07</td>
<td style="text-align: right;">0.0006783</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr11:123060825-123061329/-/HSPA8</td>
<td style="text-align: left;">chr11</td>
<td style="text-align: right;">123060825</td>
<td style="text-align: right;">123061329</td>
<td style="text-align: left;">HSPA8</td>
<td style="text-align: left;">-</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.1313983</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.2745198</td>
<td style="text-align: right;">9.259444</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000317</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr12:6943817-6944173/+/C12orf57</td>
<td style="text-align: left;">chr12</td>
<td style="text-align: right;">6943817</td>
<td style="text-align: right;">6944173</td>
<td style="text-align: left;">C12orf57</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0652261</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3459513</td>
<td style="text-align: right;">8.895493</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000540</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr12:53303034-53306861/+/ENSG00000288663</td>
<td style="text-align: left;">chr12</td>
<td style="text-align: right;">53303034</td>
<td style="text-align: right;">53306861</td>
<td style="text-align: left;">ENSG00000288663</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0259674</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3661311</td>
<td style="text-align: right;">7.157851</td>
<td style="text-align: right;">4e-07</td>
<td style="text-align: right;">0.0009298</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr12:53306680-53306861/+/ENSG00000288663</td>
<td style="text-align: left;">chr12</td>
<td style="text-align: right;">53306680</td>
<td style="text-align: right;">53306861</td>
<td style="text-align: left;">ENSG00000288663</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0244351</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3681630</td>
<td style="text-align: right;">7.186346</td>
<td style="text-align: right;">4e-07</td>
<td style="text-align: right;">0.0009075</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr12:56161387-56161465/+/MYL6</td>
<td style="text-align: left;">chr12</td>
<td style="text-align: right;">56161387</td>
<td style="text-align: right;">56161465</td>
<td style="text-align: left;">MYL6</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.2516071</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.7751959</td>
<td style="text-align: right;">23.317591</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr12:79872808-79872938/-/PPP1R12A</td>
<td style="text-align: left;">chr12</td>
<td style="text-align: right;">79872808</td>
<td style="text-align: right;">79872938</td>
<td style="text-align: left;">PPP1R12A</td>
<td style="text-align: left;">-</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0556598</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3002225</td>
<td style="text-align: right;">8.083645</td>
<td style="text-align: right;">1e-07</td>
<td style="text-align: right;">0.0001958</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr15:43801518-43801569/+/SERF2</td>
<td style="text-align: left;">chr15</td>
<td style="text-align: right;">43801518</td>
<td style="text-align: right;">43801569</td>
<td style="text-align: left;">SERF2</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0974296</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3683830</td>
<td style="text-align: right;">7.801022</td>
<td style="text-align: right;">1e-07</td>
<td style="text-align: right;">0.0002985</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr15:43801711-43804427/+/SERF2</td>
<td style="text-align: left;">chr15</td>
<td style="text-align: right;">43801711</td>
<td style="text-align: right;">43804427</td>
<td style="text-align: left;">SERF2</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.1182703</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.4634172</td>
<td style="text-align: right;">9.627454</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000214</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr18:49481681-49482410/-/RPL17-C18orf32</td>
<td style="text-align: left;">chr18</td>
<td style="text-align: right;">49481681</td>
<td style="text-align: right;">49482410</td>
<td style="text-align: left;">RPL17-C18orf32</td>
<td style="text-align: left;">-</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.2771721</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.9619640</td>
<td style="text-align: right;">40.306543</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr19:16093684-16093753/+/TPM4</td>
<td style="text-align: left;">chr19</td>
<td style="text-align: right;">16093684</td>
<td style="text-align: right;">16093753</td>
<td style="text-align: left;">TPM4</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.1013825</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3247428</td>
<td style="text-align: right;">10.275068</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000080</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr19:16095264-16095357/+/TPM4</td>
<td style="text-align: left;">chr19</td>
<td style="text-align: right;">16095264</td>
<td style="text-align: right;">16095357</td>
<td style="text-align: left;">TPM4</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.1949699</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.5515461</td>
<td style="text-align: right;">19.018562</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr19:16095264-16096744/+/TPM4</td>
<td style="text-align: left;">chr19</td>
<td style="text-align: right;">16095264</td>
<td style="text-align: right;">16096744</td>
<td style="text-align: left;">TPM4</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.2477580</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.5977548</td>
<td style="text-align: right;">23.910418</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr19:16095264-16095454/+/TPM4</td>
<td style="text-align: left;">chr19</td>
<td style="text-align: right;">16095264</td>
<td style="text-align: right;">16095454</td>
<td style="text-align: left;">TPM4</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.2432229</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.6049650</td>
<td style="text-align: right;">28.141821</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr19:16095264-16095893/+/TPM4</td>
<td style="text-align: left;">chr19</td>
<td style="text-align: right;">16095264</td>
<td style="text-align: right;">16095893</td>
<td style="text-align: left;">TPM4</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.2506281</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.5989865</td>
<td style="text-align: right;">24.155913</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr19:16095264-16095591/+/TPM4</td>
<td style="text-align: left;">chr19</td>
<td style="text-align: right;">16095264</td>
<td style="text-align: right;">16095591</td>
<td style="text-align: left;">TPM4</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.2518667</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.6011723</td>
<td style="text-align: right;">24.235526</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr19:16095264-16095496/+/TPM4</td>
<td style="text-align: left;">chr19</td>
<td style="text-align: right;">16095264</td>
<td style="text-align: right;">16095496</td>
<td style="text-align: left;">TPM4</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.2582948</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.6168771</td>
<td style="text-align: right;">28.105457</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr19:18169087-18170494/+/ENSG00000268173</td>
<td style="text-align: left;">chr19</td>
<td style="text-align: right;">18169087</td>
<td style="text-align: right;">18170494</td>
<td style="text-align: left;">ENSG00000268173</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0506799</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.4966975</td>
<td style="text-align: right;">20.652615</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr2:218344808-218346793/+/PNKD</td>
<td style="text-align: left;">chr2</td>
<td style="text-align: right;">218344808</td>
<td style="text-align: right;">218346793</td>
<td style="text-align: left;">PNKD</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0899427</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.4868878</td>
<td style="text-align: right;">11.778748</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000009</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr2:218344808-218346784/+/PNKD</td>
<td style="text-align: left;">chr2</td>
<td style="text-align: right;">218344808</td>
<td style="text-align: right;">218346784</td>
<td style="text-align: left;">PNKD</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0889706</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.4851371</td>
<td style="text-align: right;">12.634455</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000003</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr2:218344808-218346756/+/PNKD</td>
<td style="text-align: left;">chr2</td>
<td style="text-align: right;">218344808</td>
<td style="text-align: right;">218346756</td>
<td style="text-align: left;">PNKD</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0887807</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.4852883</td>
<td style="text-align: right;">12.825710</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000003</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr2:218344808-218346791/+/PNKD</td>
<td style="text-align: left;">chr2</td>
<td style="text-align: right;">218344808</td>
<td style="text-align: right;">218346791</td>
<td style="text-align: left;">PNKD</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0899427</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.4868878</td>
<td style="text-align: right;">11.778748</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000009</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr2:218344808-218346771/+/PNKD</td>
<td style="text-align: left;">chr2</td>
<td style="text-align: right;">218344808</td>
<td style="text-align: right;">218346771</td>
<td style="text-align: left;">PNKD</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0879248</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.4843662</td>
<td style="text-align: right;">12.855317</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000003</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chr5:179811988-179813641/+/SQSTM1</td>
<td style="text-align: left;">chr5</td>
<td style="text-align: right;">179811988</td>
<td style="text-align: right;">179813641</td>
<td style="text-align: left;">SQSTM1</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0003873</td>
<td style="text-align: right;">0.0141935</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3137492</td>
<td style="text-align: right;">9.096496</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000399</td>
</tr>
<tr class="even">
<td style="text-align: left;">chr5:179812407-179813598/+/SQSTM1</td>
<td style="text-align: left;">chr5</td>
<td style="text-align: right;">179812407</td>
<td style="text-align: right;">179813598</td>
<td style="text-align: left;">SQSTM1</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0003694</td>
<td style="text-align: right;">0.0142486</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3140973</td>
<td style="text-align: right;">8.869969</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000540</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chrX:119582588-119582676/+/UBE2A</td>
<td style="text-align: left;">chrX</td>
<td style="text-align: right;">119582588</td>
<td style="text-align: right;">119582676</td>
<td style="text-align: left;">UBE2A</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0315370</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3095027</td>
<td style="text-align: right;">7.926565</td>
<td style="text-align: right;">1e-07</td>
<td style="text-align: right;">0.0002523</td>
</tr>
<tr class="even">
<td style="text-align: left;">chrX:119583127-119583347/+/UBE2A</td>
<td style="text-align: left;">chrX</td>
<td style="text-align: right;">119583127</td>
<td style="text-align: right;">119583347</td>
<td style="text-align: left;">UBE2A</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0391280</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3247510</td>
<td style="text-align: right;">9.360876</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000297</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chrX:119583127-119584101/+/UBE2A</td>
<td style="text-align: left;">chrX</td>
<td style="text-align: right;">119583127</td>
<td style="text-align: right;">119584101</td>
<td style="text-align: left;">UBE2A</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0363548</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3127182</td>
<td style="text-align: right;">8.108500</td>
<td style="text-align: right;">1e-07</td>
<td style="text-align: right;">0.0001945</td>
</tr>
<tr class="even">
<td style="text-align: left;">chrX:119583127-119583417/+/UBE2A</td>
<td style="text-align: left;">chrX</td>
<td style="text-align: right;">119583127</td>
<td style="text-align: right;">119583417</td>
<td style="text-align: left;">UBE2A</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0428587</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3280701</td>
<td style="text-align: right;">9.444139</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000274</td>
</tr>
<tr class="odd">
<td style="text-align: left;">chrX:119583127-119583621/+/UBE2A</td>
<td style="text-align: left;">chrX</td>
<td style="text-align: right;">119583127</td>
<td style="text-align: right;">119583621</td>
<td style="text-align: left;">UBE2A</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0413061</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3235995</td>
<td style="text-align: right;">8.773698</td>
<td style="text-align: right;">0e+00</td>
<td style="text-align: right;">0.0000612</td>
</tr>
<tr class="even">
<td style="text-align: left;">chrX:151404916-151405157/+/VMA21</td>
<td style="text-align: left;">chrX</td>
<td style="text-align: right;">151404916</td>
<td style="text-align: right;">151405157</td>
<td style="text-align: left;">VMA21</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0550083</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.3621630</td>
<td style="text-align: right;">7.798756</td>
<td style="text-align: right;">1e-07</td>
<td style="text-align: right;">0.0002985</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Random select a gene and its exons and visulize with FeaturePlot.</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(obj, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"chr15:43801711-43804427/+/SERF2"</span>, <span class="st">"SERF2"</span>), <span class="at">ncol=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid figure-img" width="1344"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The default color and parameters perhaps not easily to tell the difference between exon and its binding gene expression. Let's change the scaled colors and enlarge point size and order by expression.</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(RColorBrewer)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: RColorBrewer</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(obj, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"chr15:43801711-43804427/+/SERF2"</span>, <span class="st">"SERF2"</span>), <span class="at">ncol=</span><span class="dv">2</span>, <span class="at">order =</span> <span class="cn">TRUE</span>, <span class="at">pt.size=</span><span class="dv">1</span>) <span class="sc">&amp;</span> <span class="fu">scale_colour_gradientn</span>(<span class="at">colours =</span> <span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="at">n =</span> <span class="dv">11</span>, <span class="at">name =</span> <span class="st">"RdBu"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-6-3.png" class="img-fluid figure-img" width="1344"></p>
</figure>
</div>
</div>
</div>
<p>We can also map the ratio of exon expression to gene expression on the UMAP. The <code>RatioPlot</code> function is designed for this purpose. As observed, the gene SERF2 is relatively low expressed in groups 0, 5, and 9, while the ratio of the exon chr15:43801711-43804427/+/SERF2 is higher in these groups.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">RatioPlot</span>(obj, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"chr15:43801711-43804427/+/SERF2"</span>), <span class="at">assay =</span> <span class="st">'exon'</span>, <span class="at">bind.assay =</span> <span class="st">'RNA'</span>, <span class="at">bind.name =</span> <span class="st">"gene_name"</span>, <span class="at">order =</span> <span class="cn">TRUE</span>, <span class="at">pt.size=</span><span class="dv">1</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In the feature plot and ratio plot above, the exon appears to lack a strong expression pattern across cell groups, whereas the gene SERF2 seems to be highly expressed in many groups, except for groups 0, 5 and 9. This inconsistent expression pattern between the exon and its corresponding gene may suggest differential exon usage. To explore the coverage details of both the exon and the gene body, we will generate a track plot next.</p>
<p>In our package, retrieving gene locations requires loading a GTF file instead of relying on current Bioconductor databases, such as <code>org.Hs.eg.db</code>. This is due to the varying versions of gene annotations provided by different institutes, which can introduce inconsistencies. To avoid potential bias during preprocessing and postprocessing, we strongly recommend using the same GTF file consistently throughout your project. The Yano package includes the <code>gtf2db</code> function, which enables you to load a GTF file into memory for further analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>gtf <span class="ot">&lt;-</span> <span class="fu">gtf2db</span>(<span class="st">"./gencode.v44.annotation.gtf.gz"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[2025-06-19 18:04:38] GTF loading..
[2025-06-19 18:04:59] Load 62700 genes.</code></pre>
</div>
</div>
<p>A track plot is used to study the read coverage per cell group. In the track plot shown below, the cell group is specified by the <code>cell.group</code> parameter. Unlike IGV, where read depth is used, we use UMI depth in this plot. The cell barcode tag and UMI tag are predefined as “CB” and “UB” with parameter <code>cell.tag</code> and <code>umi.tag</code>. For each cell group, the UMI depth has been normalized by the number of cells in that group. This means that the depth at each location can be interpreted as the mean UMI depth per cell for the group. As a result, the tracks are directly comparable across different cell groups. If <code>cell.group</code> is not set, the track plot will generate the raw UMI depth per location.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">TrackPlot</span>(<span class="at">bamfile =</span> <span class="st">"Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam"</span>, <span class="at">gtf =</span> gtf, <span class="at">gene =</span> <span class="st">"SERF2"</span>, <span class="at">cell.group =</span> <span class="fu">Idents</span>(obj), <span class="at">highlights =</span> <span class="fu">c</span>(<span class="dv">43801711</span>,<span class="dv">43804427</span>) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
<p>In the track plot, we can easily observe that an exon around position 43,794,000 dominates the expression of the SERF2 gene and is highly expressed in many cell groups. However, the exon ‘chr15:43801711-43804427/+/SERF2’ (highlighted) shows low expression and is not visible in the track plot. To visualize low-expressed exons, we can set the <code>max.depth</code> parameter to 2, which caps the UMI depth at 2. And many genes in the region, we set <code>display.genes</code> to SERF2 only. This adjustment allows the low-expressed exons and their related transcripts to be more clearly represented in the plot. In this case, we can found the the highlighed exon shows different expressed pattern with the gene SERF2.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">TrackPlot</span>(<span class="at">bamfile =</span> <span class="st">"Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam"</span>, <span class="at">gtf =</span> gtf, <span class="at">gene =</span> <span class="st">"SERF2"</span>, <span class="at">cell.group =</span> <span class="fu">Idents</span>(obj), <span class="at">highlights =</span> <span class="fu">c</span>(<span class="dv">43801711</span>,<span class="dv">43804427</span>), <span class="at">max.depth =</span> <span class="dv">2</span>, <span class="at">display.genes =</span> <span class="st">"SERF2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="load-junction-assay" class="level2">
<h2 class="anchored" data-anchor-id="load-junction-assay">3. Load junction assay</h2>
<p>In addition to exon expression, junction expression can provide insights into different expression patterns across transcripts, offering a complementary perspective. Junction expression refers to the UMI counts of reads that span more than one exon. It’s important to note that junctions are named similarly to exons, but the start and end positions are different. The start of the junction corresponds to the end of the previous exon, while the end of the junction represents the start of the next exon.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>junction <span class="ot">&lt;-</span> <span class="fu">ReadPISA</span>(<span class="st">"./junction/"</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'junction'</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(junction[, <span class="fu">colnames</span>(obj)], <span class="at">min.cells=</span><span class="dv">20</span>)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"junction"</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(obj)</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a><span class="co"># select spatial autocorrelated junctions</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">RunAutoCorr</span>(obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Working on assay : junction</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Run autocorrelation test for 14786 features.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Runtime : 2.648213 secs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>7531 autocorrelated features.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Parse the gene name and coordinates from junction names</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">ParseExonName</span>(obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Working on assay junction</code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># perform dissimilarity test between junctions and their binding genes</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">RunSDT</span>(obj, <span class="at">bind.name =</span> <span class="st">"gene_name"</span>, <span class="at">bind.assay =</span> <span class="st">"RNA"</span>, <span class="at">perm=</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Working on assay junction.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Working on binding assay RNA.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Use predefined weight matrix "pca_wm".</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Processing 7531 features.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Processing 3775 binding features.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Retrieve binding data from assay RNA.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Use "data" layer for test features and binding features.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Using 7 threads.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Runtime : 15.44265 secs.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FbtPlot</span>(obj, <span class="at">val=</span><span class="st">"gene_name.padj"</span>, <span class="at">remove.chr=</span><span class="cn">TRUE</span>, <span class="at">sel.chrs =</span> sel.chrs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="1344"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'junction'</span>]][[]] <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene_name.padj<span class="sc">&lt;</span><span class="fl">1e-5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                               moransi.pval   moransi autocorr.variable   chr
chr12:56160320-56161387/+/MYL6            0 0.1921649              TRUE chr12
chr19:16093753-16095264/+/TPM4            0 0.1909011              TRUE chr19
                                  start      end gene_name strand gene_name.D
chr12:56160320-56161387/+/MYL6 56160320 56161387      MYL6      +   0.7466745
chr19:16093753-16095264/+/TPM4 16093753 16095264      TPM4      +   0.5364398
                               gene_name.t gene_name.pval gene_name.padj
chr12:56160320-56161387/+/MYL6    13.04541   3.121098e-11   1.175250e-07
chr19:16093753-16095264/+/TPM4    18.46030   6.795704e-14   5.117845e-10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Because both exon and junction are compared with gene, so it's reasonable to combine these two assays in one plot</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FbtPlot</span>(obj, <span class="at">val=</span><span class="st">"gene_name.padj"</span>, <span class="at">assay =</span> <span class="fu">c</span>(<span class="st">"exon"</span>, <span class="st">"junction"</span>), <span class="at">col.by =</span> <span class="st">"assay"</span>, <span class="at">shape.by =</span> <span class="st">"assay"</span>, <span class="at">pt.size =</span> <span class="dv">2</span>, <span class="at">remove.chr =</span> <span class="cn">TRUE</span>, <span class="at">sel.chrs =</span> sel.chrs, <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"blue"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-11-2.png" class="img-fluid figure-img" width="1344"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can find there is an exon and a junction at chromosome 12 with very low p value (&lt;1e-8), let's see which gene they are located</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'exon'</span>]][[]] <span class="sc">%&gt;%</span> <span class="fu">filter</span>(chr <span class="sc">==</span> <span class="st">"chr12"</span> <span class="sc">&amp;</span> gene_name.pval <span class="sc">&lt;</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span> knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 18%">
<col style="width: 3%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 4%">
<col style="width: 7%">
<col style="width: 5%">
<col style="width: 10%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 8%">
<col style="width: 8%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">chr</th>
<th style="text-align: right;">start</th>
<th style="text-align: right;">end</th>
<th style="text-align: left;">gene_name</th>
<th style="text-align: left;">strand</th>
<th style="text-align: right;">moransi.pval</th>
<th style="text-align: right;">moransi</th>
<th style="text-align: left;">autocorr.variable</th>
<th style="text-align: right;">gene_name.D</th>
<th style="text-align: right;">gene_name.t</th>
<th style="text-align: right;">gene_name.pval</th>
<th style="text-align: right;">gene_name.padj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">chr12:56161387-56161465/+/MYL6</td>
<td style="text-align: left;">chr12</td>
<td style="text-align: right;">56161387</td>
<td style="text-align: right;">56161465</td>
<td style="text-align: left;">MYL6</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.2516071</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0.7751959</td>
<td style="text-align: right;">23.31759</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'junction'</span>]][[]] <span class="sc">%&gt;%</span> <span class="fu">filter</span>(chr <span class="sc">==</span> <span class="st">"chr12"</span> <span class="sc">&amp;</span> gene_name.pval <span class="sc">&lt;</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span> knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 18%">
<col style="width: 7%">
<col style="width: 5%">
<col style="width: 10%">
<col style="width: 3%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 4%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 8%">
<col style="width: 8%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">moransi.pval</th>
<th style="text-align: right;">moransi</th>
<th style="text-align: left;">autocorr.variable</th>
<th style="text-align: left;">chr</th>
<th style="text-align: right;">start</th>
<th style="text-align: right;">end</th>
<th style="text-align: left;">gene_name</th>
<th style="text-align: left;">strand</th>
<th style="text-align: right;">gene_name.D</th>
<th style="text-align: right;">gene_name.t</th>
<th style="text-align: right;">gene_name.pval</th>
<th style="text-align: right;">gene_name.padj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">chr12:56160320-56161387/+/MYL6</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.1921649</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: left;">chr12</td>
<td style="text-align: right;">56160320</td>
<td style="text-align: right;">56161387</td>
<td style="text-align: left;">MYL6</td>
<td style="text-align: left;">+</td>
<td style="text-align: right;">0.7466745</td>
<td style="text-align: right;">13.04541</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1e-07</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(obj, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"chr12:56161387-56161465/+/MYL6"</span>,<span class="st">"chr12:56160320-56161387/+/MYL6"</span>, <span class="st">"MYL6"</span>), <span class="at">order =</span> <span class="cn">TRUE</span>, <span class="at">pt.size =</span> <span class="dv">2</span>, <span class="at">ncol=</span><span class="dv">3</span>) <span class="sc">&amp;</span> <span class="fu">scale_colour_gradientn</span>(<span class="at">colours =</span> <span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="at">n =</span> <span class="dv">11</span>, <span class="at">name =</span> <span class="st">"RdBu"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-11-3.png" class="img-fluid figure-img" width="1344"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We could also plot the expression ratio of these exons or junctions on umap</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">RatioPlot</span>(obj, <span class="at">assay =</span> <span class="st">"exon"</span>, <span class="at">bind.assay =</span> <span class="st">"RNA"</span>, <span class="at">bind.name =</span> <span class="st">"gene_name"</span>, <span class="at">features =</span> <span class="st">"chr12:56161387-56161465/+/MYL6"</span>)</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">RatioPlot</span>(obj, <span class="at">assay =</span> <span class="st">"exon"</span>, <span class="at">bind.assay =</span> <span class="st">"RNA"</span>, <span class="at">bind.name =</span> <span class="st">"gene_name"</span>, <span class="at">features =</span> <span class="st">"chr12:56160626-56160670/+/MYL6"</span>)</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">RatioPlot</span>(obj, <span class="at">assay =</span> <span class="st">"junction"</span>, <span class="at">bind.assay =</span> <span class="st">"RNA"</span>, <span class="at">bind.name =</span> <span class="st">"gene_name"</span>, <span class="at">features =</span> <span class="st">"chr12:56160320-56161387/+/MYL6"</span>)</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1,p2,p3, <span class="at">ncol=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-11-4.png" class="img-fluid figure-img" width="1344"></p>
</figure>
</div>
</div>
</div>
<p>We then visualize the track plot for this gene, including junction reads by setting <code>junc=TRUE</code>. The height of the splice paths in the plot represents the expression level of each junction within the specified cell group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">TrackPlot</span>(<span class="at">bamfile =</span> <span class="st">"Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam"</span>, <span class="at">gtf =</span> gtf, <span class="at">gene =</span> <span class="st">"MYL6"</span>, <span class="at">cell.group =</span> <span class="fu">Idents</span>(obj), <span class="at">junc =</span> <span class="cn">TRUE</span>, <span class="at">highlights =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="dv">56160320</span>,<span class="dv">56161387</span>),<span class="fu">c</span>(<span class="dv">56161387</span>,<span class="dv">56161465</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
<p>You might be wondering why the exon chr12:56161387-56161465/+/MYL6 appears highly expressed in cell group 2 in the track plot, where the overlapping peak is clearly higher than in other groups, but its expression level in the feature plot is not as high as expected.</p>
<p>This discrepancy arises because the exon is overlapping with other exons from different transcripts. We only count reads that are fully contained within the exon as part of the exon’s expression. Therefore, reads that partially overlap with this exon are not included in the count.</p>
<p>In contrast, the overlapping exon chr12:56161387-56161575/+/MYL6 shows higher expression in group 2 compared to other groups. It’s important to note that if a read is fully contained within two or more overlapping exons, PISA will count it for all relevant exons. Check PISA’s manual for details.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(obj, <span class="at">label=</span><span class="cn">TRUE</span>, <span class="at">label.size =</span> <span class="dv">5</span>, <span class="at">label.box =</span> <span class="cn">TRUE</span>)</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">FeaturePlot</span>(obj, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"chr12:56161387-56161575/+/MYL6"</span>), <span class="at">order =</span> <span class="cn">TRUE</span>, <span class="at">pt.size =</span> <span class="dv">1</span>) <span class="sc">&amp;</span> <span class="fu">scale_colour_gradientn</span>(<span class="at">colours =</span> <span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="at">n =</span> <span class="dv">11</span>, <span class="at">name =</span> <span class="st">"RdBu"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Could not find chr12:56161387-56161575/+/MYL6 in the default search
locations, found in 'exon' assay instead</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="heatmap-analysis-for-highlight-group-specific-alternative-splicing" class="level2">
<h2 class="anchored" data-anchor-id="heatmap-analysis-for-highlight-group-specific-alternative-splicing">4. Heatmap analysis for highlight group specific alternative splicing</h2>
<p>The spatial dissimilarity test method prioritizes alternatively spliced exons and junctions across all cells but does not identify which specific cell groups exhibit these splicing events. To address this, let’s manually extract the scaled expression data for the selected alternatively spliced exons and their corresponding genes, then perform a co-clustering analysis. A comprehensive heatmap will be generated using the <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/index.html">ComplexHeatmap</a> package, providing a visual representation of the exon and gene distribution across cell groups.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'exon'</span>]][[]] <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene_name.padj<span class="sc">&lt;</span><span class="fl">0.001</span>) <span class="sc">%&gt;%</span> rownames <span class="ot">-&gt;</span> exons</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'exon'</span>]][[]] <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene_name.padj<span class="sc">&lt;</span><span class="fl">0.001</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(gene_name) <span class="ot">-&gt;</span> bind.genes</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(obj, <span class="at">features =</span> <span class="fu">unique</span>(bind.genes))</span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"exon"</span></span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(obj, <span class="at">features =</span> exons)</span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a>dat1 <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(obj, <span class="at">assay =</span> <span class="st">'exon'</span>, <span class="at">layer =</span> <span class="st">'scale.data'</span>)</span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a>dat2 <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(obj, <span class="at">assay =</span> <span class="st">'RNA'</span>, <span class="at">layer =</span> <span class="st">'scale.data'</span>)</span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a>idents <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">Idents</span>(obj))</span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a>order.cells <span class="ot">&lt;-</span> <span class="fu">names</span>(idents)</span>
<span id="cb76-12"><a href="#cb76-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-13"><a href="#cb76-13" aria-hidden="true" tabindex="-1"></a>dat2 <span class="ot">&lt;-</span> dat2[bind.genes,]</span>
<span id="cb76-14"><a href="#cb76-14" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(dat2) <span class="ot">&lt;-</span> exons</span>
<span id="cb76-15"><a href="#cb76-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-16"><a href="#cb76-16" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">cbind</span>(dat1, dat2)</span>
<span id="cb76-17"><a href="#cb76-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-18"><a href="#cb76-18" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ComplexHeatmap)</span>
<span id="cb76-19"><a href="#cb76-19" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">dist</span>(dat)</span>
<span id="cb76-20"><a href="#cb76-20" aria-hidden="true" tabindex="-1"></a>hc <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d)</span>
<span id="cb76-21"><a href="#cb76-21" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> hc<span class="sc">$</span>labels[hc<span class="sc">$</span>order]</span>
<span id="cb76-22"><a href="#cb76-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-23"><a href="#cb76-23" aria-hidden="true" tabindex="-1"></a>ha <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(<span class="at">group=</span>idents, <span class="at">border =</span> <span class="cn">TRUE</span>)</span>
<span id="cb76-24"><a href="#cb76-24" aria-hidden="true" tabindex="-1"></a>ht1 <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(dat1[idx, order.cells], <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>, <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>, <span class="at">show_column_names =</span> <span class="cn">FALSE</span>, <span class="at">border =</span> <span class="cn">TRUE</span>,  <span class="at">top_annotation =</span> ha, <span class="at">name =</span> <span class="st">"exon"</span>, <span class="at">column_title =</span> <span class="st">"exon"</span>)</span>
<span id="cb76-25"><a href="#cb76-25" aria-hidden="true" tabindex="-1"></a>ht2 <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(dat2[idx, order.cells], <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>, <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>, <span class="at">show_column_names =</span> <span class="cn">FALSE</span>, <span class="at">border =</span> <span class="cn">TRUE</span>,  <span class="at">top_annotation =</span> ha, <span class="at">name =</span> <span class="st">"gene"</span>, <span class="at">column_title =</span> <span class="st">"gene"</span>, <span class="at">row_names_max_width =</span> <span class="fu">max_text_width</span>(<span class="fu">rownames</span>(dat2), <span class="at">gp =</span> <span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">12</span>)))</span>
<span id="cb76-26"><a href="#cb76-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-27"><a href="#cb76-27" aria-hidden="true" tabindex="-1"></a>ht <span class="ot">&lt;-</span> ht1 <span class="sc">+</span> ht2</span>
<span id="cb76-28"><a href="#cb76-28" aria-hidden="true" tabindex="-1"></a><span class="fu">draw</span>(ht, <span class="at">heatmap_legend_side =</span> <span class="st">"left"</span>,  <span class="at">annotation_legend_side =</span> <span class="st">"left"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="test-between-exon-and-exon-skipped-reads" class="level2">
<h2 class="anchored" data-anchor-id="test-between-exon-and-exon-skipped-reads">5. Test between exon and exon skipped reads</h2>
<p>In the previous sections, we conducted a spatial dissimilarity test between exon/junction expression and gene expression. However, binding features are not always limited to genes; they can also correspond to other types of features. In this section, we perform a test between exon expression and reads that skip this exon (exclude assay). This approach is similar to the Percent Spliced In (PSI) method, which is widely used to analyze alternative splicing in both bulk and single-cell RNA-seq data. The PSI is calculated as: <code>PSI = exon reads / (exon reads + reads skipping this exon)</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The reads that skip exons are annotated using the `-psi` option in PISA anno, and these counts are stored in the `exclude` directory. We then load these excluded counts into a new assay.</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>exclude <span class="ot">&lt;-</span> <span class="fu">ReadPISA</span>(<span class="st">"exclude/"</span>)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'exclude'</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(exclude[,<span class="fu">colnames</span>(obj)], <span class="at">min.cells =</span> <span class="dv">10</span>)</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"exclude"</span></span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize counts for exon-excluded reads</span></span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(obj)</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Then we switch to exon assay</span></span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"exon"</span></span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Because the feature names in the exclude assay are exactly the same as those in the exon assay, they represent the reads that skip each corresponding exon. Therefore, we set up the binding feature using the exon name itself.</span></span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'exon'</span>]][[<span class="st">'exon_name'</span>]] <span class="ot">&lt;-</span> <span class="fu">rownames</span>(obj)</span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'exon'</span>]][[<span class="st">'exon_name'</span>]] <span class="sc">%&gt;%</span> head</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                                                exon_name
chr1:135141-135895/-/ENSG00000268903 chr1:135141-135895/-/ENSG00000268903
chr1:629640-630683/+/MTND2P28               chr1:629640-630683/+/MTND2P28
chr1:631074-632616/+/MTCO1P12               chr1:631074-632616/+/MTCO1P12
chr1:632757-633438/+/MTCO2P12               chr1:632757-633438/+/MTCO2P12
chr1:633696-634376/+/MTATP6P1               chr1:633696-634376/+/MTATP6P1
chr1:634376-634922/+/MTCO3P12               chr1:634376-634922/+/MTCO3P12</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Then we perform spatial dissimilarity test between exon and exclude, mode 1</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">RunSDT</span>(obj, <span class="at">bind.name =</span> <span class="st">"exon_name"</span>, <span class="at">bind.assay =</span> <span class="st">"exclude"</span>)</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Swith to exon exluded assay</span></span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"exclude"</span></span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">RunAutoCorr</span>(obj)</span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">ParseExonName</span>(obj)</span>
<span id="cb79-8"><a href="#cb79-8" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'exclude'</span>]][[<span class="st">'exon_name'</span>]] <span class="ot">&lt;-</span> <span class="fu">rownames</span>(obj)</span>
<span id="cb79-9"><a href="#cb79-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-10"><a href="#cb79-10" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">RunSDT</span>(obj, <span class="at">bind.name =</span> <span class="st">"exon_name"</span>, <span class="at">bind.assay =</span> <span class="st">"exon"</span>)</span>
<span id="cb79-11"><a href="#cb79-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-12"><a href="#cb79-12" aria-hidden="true" tabindex="-1"></a><span class="fu">FbtPlot</span>(obj, <span class="at">val =</span> <span class="st">"exon_name.padj"</span>, <span class="at">remove.chr =</span> <span class="cn">TRUE</span>, <span class="at">assay =</span> <span class="fu">c</span>(<span class="st">"exclude"</span>, <span class="st">"exon"</span>), <span class="at">shape.by =</span> <span class="st">"assay"</span>, <span class="at">col.by =</span> <span class="st">"assay"</span>, <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"yellow"</span>, <span class="st">"green"</span>), <span class="at">pt.size =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's how many exons can be prioritized by both exon assay and exclude assay</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'exclude'</span>]][[]] <span class="sc">%&gt;%</span> <span class="fu">filter</span>(exon_name.padj<span class="sc">&lt;</span><span class="fl">1e-5</span>) <span class="sc">%&gt;%</span> rownames <span class="ot">-&gt;</span> sel1</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>obj[[<span class="st">'exon'</span>]][[]] <span class="sc">%&gt;%</span> <span class="fu">filter</span>(exon_name.padj<span class="sc">&lt;</span><span class="fl">1e-5</span>) <span class="sc">%&gt;%</span> rownames <span class="ot">-&gt;</span> sel2</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a><span class="fu">intersect</span>(sel1,sel2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "chr1:19342752-19342864/-/CAPZB"    "chr1:153990914-153991034/+/RPS27" 
 [3] "chr10:7806974-7807010/+/ATP5F1C"   "chr10:128047570-128047683/+/PTPRE"
 [5] "chr12:56160626-56160670/+/MYL6"    "chr12:56160626-56160945/+/MYL6"   
 [7] "chr19:16095264-16095357/+/TPM4"    "chr19:16095264-16095454/+/TPM4"   
 [9] "chr19:16095264-16095496/+/TPM4"    "chr19:16095264-16095893/+/TPM4"   
[11] "chr19:16095264-16096744/+/TPM4"    "chr19:16095264-16095591/+/TPM4"   
[13] "chr2:197490527-197490664/-/HSPD1"  "chr20:37238402-37238449/+/RPN2"   
[15] "chr20:49280903-49280964/+/ZFAS1"   "chr3:197953488-197953660/+/RPL35A"
[17] "chr5:83519349-83522309/+/VCAN"     "chr6:85677795-85677875/-/SNHG5"   
[19] "chr6:85677791-85677875/-/SNHG5"    "chr9:35684732-35684807/-/TPM2"    
[21] "chr9:35684732-35684802/-/TPM2"     "chrX:154400464-154400626/+/RPL10" </code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"exclude"</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">FeaturePlot</span>(obj, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"chr19:16095264-16095357/+/TPM4"</span>),<span class="at">order =</span> <span class="cn">TRUE</span>)</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"exon"</span></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">FeaturePlot</span>(obj, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"chr19:16095264-16095357/+/TPM4"</span>),<span class="at">order =</span> <span class="cn">TRUE</span>)</span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">PSIPlot</span>(obj, <span class="at">exon.assay =</span> <span class="st">"exon"</span>, <span class="at">exclude.assay =</span> <span class="st">"exclude"</span>, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"chr19:16095264-16095357/+/TPM4"</span>),<span class="at">order =</span> <span class="cn">TRUE</span>)</span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1,p2,p3, <span class="at">ncol=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">TrackPlot</span>(<span class="at">bamfile =</span> <span class="st">"Parent_SC3v3_Human_Glioblastoma_possorted_genome_bam.bam"</span>, <span class="at">gtf =</span> gtf, <span class="at">gene =</span> <span class="st">"TPM4"</span>, <span class="at">cell.group =</span> <span class="fu">Idents</span>(obj), <span class="at">highlights =</span> <span class="fu">c</span>(<span class="dv">16095264</span>,<span class="dv">16095357</span>), <span class="at">junc =</span> <span class="cn">TRUE</span>, <span class="at">max.depth =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
<p>In previous sections, we noted that exon or junction expression is part of gene expression, and inverse expression patterns can strongly indicate alternative splicing. However, exon-skipped reads are largely independent of exon expression, making them more sensitive for detecting alternative splicing and allowing for the prioritization of many events.</p>
<p>Because our spatial dissimilarity test does not account for the spatial dependency of the binding feature, numerous events may be prioritized, especially when the binding feature is sparsely expressed. While some of these events may be true, others might arise due to low coverage. To enhance detection power and reduce potential false positives, intersecting the prioritized exons with the prioritized exon-excluded features can help refine the results.</p>
<p>Mode 3 can be used as an alternative to mode 1 to specifically detect events with strong inverse expression patterns. In mode 3, the exon reads and exon-excluded reads are summed as the binding assay, allowing for a more targeted analysis of such patterns. It is important to note that events detected with mode 3 are always detectable with mode 1, making mode 3 a refined approach for prioritizing inverse expression events.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"exclude"</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">RunSDT</span>(obj, <span class="at">bind.name =</span> <span class="st">"exon_name"</span>, <span class="at">bind.assay =</span> <span class="st">"exon"</span>, <span class="at">mode =</span> <span class="dv">3</span>, <span class="at">prefix =</span> <span class="st">"mode3"</span>)</span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"exon"</span></span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">RunSDT</span>(obj, <span class="at">bind.name =</span> <span class="st">"exon_name"</span>, <span class="at">bind.assay =</span> <span class="st">"exclude"</span>, <span class="at">mode =</span> <span class="dv">3</span>, <span class="at">prefix =</span> <span class="st">"mode3"</span>)</span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a><span class="fu">FbtPlot</span>(obj, <span class="at">val =</span> <span class="st">"mode3.padj"</span>, <span class="at">remove.chr =</span> <span class="cn">TRUE</span>, <span class="at">assay =</span> <span class="fu">c</span>(<span class="st">"exclude"</span>, <span class="st">"exon"</span>), <span class="at">shape.by =</span> <span class="st">"assay"</span>, <span class="at">col.by =</span> <span class="st">"assay"</span>, <span class="at">pt.size =</span> <span class="dv">2</span>, <span class="at">cols=</span><span class="fu">c</span>(<span class="st">"yellow"</span>, <span class="st">"green"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Yano_AS_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
<p>In this case study, we performed a spatial dissimilarity test between various feature pairs. This method provides an overview of the entire cell population and does not rely on prior cell clustering and annotation, making it a powerful tool for analyzing cell data without any prior knowledge. It is recommended to test different features including junctions and exons, and set up different with their corresponding genes in 3’ or 5’ biased scRNA-seq. Additionally, testing between exon-included and exon-skipped reads have more power to detect exon excluded events.</p>
<p>To obtain cell-cluster-specific expression patterns, applying heatmaps and clustering in subsequent analyses is recommended.</p>
</section>
<section id="questions" class="level2">
<h2 class="anchored" data-anchor-id="questions">Questions?</h2>
<p>If you have any questions regarding this vignette and the usage of Yano, please feel free to report them through the <a href="https://github.com/shiquan/Yano/discussions">discussion forum</a>. When submitting your query, please ensure you attach the commands you used for better clarity and support.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Command</span>(obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "NormalizeData.RNA"            "FindVariableFeatures.RNA"    
 [3] "RunPCA.RNA"                   "FindNeighbors.RNA.pca"       
 [5] "FindClusters"                 "RunUMAP.RNA.pca"             
 [7] "ParseExonName.exon"           "NormalizeData.exon"          
 [9] "RunAutoCorr.exon.pca"         "SetAutoCorrFeatures.exon"    
[11] "NormalizeData.junction"       "RunAutoCorr.junction.pca"    
[13] "SetAutoCorrFeatures.junction" "ParseExonName.junction"      
[15] "RunSDT.junction"              "ScaleData.RNA"               
[17] "ScaleData.exon"               "NormalizeData.exclude"       
[19] "RunAutoCorr.exclude.pca"      "SetAutoCorrFeatures.exclude" 
[21] "ParseExonName.exclude"        "RunSDT.exclude"              
[23] "RunSDT.exon"                 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.0 (2025-04-11)
Platform: aarch64-apple-darwin24.4.0
Running under: macOS Sequoia 15.5

Matrix products: default
BLAS:   /opt/homebrew/Cellar/openblas/0.3.29/lib/libopenblasp-r0.3.29.dylib 
LAPACK: /opt/homebrew/Cellar/r/4.5.0/lib/R/lib/libRlapack.dylib;  LAPACK version 3.12.1

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
[1] ComplexHeatmap_2.24.0 RColorBrewer_1.1-3    future_1.49.0        
[4] dplyr_1.1.4           Seurat_5.3.0          SeuratObject_5.1.0   
[7] sp_2.2-0              ggplot2_3.5.2         Yano_1.2             

loaded via a namespace (and not attached):
  [1] shape_1.4.6.1          jsonlite_2.0.0         magrittr_2.0.3        
  [4] spatstat.utils_3.1-4   farver_2.1.2           rmarkdown_2.29        
  [7] GlobalOptions_0.1.2    vctrs_0.6.5            ROCR_1.0-11           
 [10] spatstat.explore_3.4-2 htmltools_0.5.8.1      sctransform_0.4.2     
 [13] parallelly_1.44.0      KernSmooth_2.23-26     htmlwidgets_1.6.4     
 [16] ica_1.0-3              plyr_1.8.9             plotly_4.10.4         
 [19] zoo_1.8-14             igraph_2.1.4           mime_0.13             
 [22] lifecycle_1.0.4        iterators_1.0.14       pkgconfig_2.0.3       
 [25] Matrix_1.7-3           R6_2.6.1               fastmap_1.2.0         
 [28] clue_0.3-66            fitdistrplus_1.2-2     shiny_1.10.0          
 [31] digest_0.6.37          colorspace_2.1-1       S4Vectors_0.46.0      
 [34] patchwork_1.3.0        tensor_1.5             RSpectra_0.16-2       
 [37] irlba_2.3.5.1          labeling_0.4.3         progressr_0.15.1      
 [40] spatstat.sparse_3.1-0  httr_1.4.7             polyclip_1.10-7       
 [43] abind_1.4-8            compiler_4.5.0         withr_3.0.2           
 [46] doParallel_1.0.17      viridis_0.6.5          fastDummies_1.7.5     
 [49] R.utils_2.13.0         MASS_7.3-65            rjson_0.2.23          
 [52] gtools_3.9.5           tools_4.5.0            lmtest_0.9-40         
 [55] httpuv_1.6.16          future.apply_1.11.3    goftest_1.2-3         
 [58] R.oo_1.27.1            glue_1.8.0             nlme_3.1-168          
 [61] promises_1.3.2         Rtsne_0.17             cluster_2.1.8.1       
 [64] reshape2_1.4.4         generics_0.1.4         gtable_0.3.6          
 [67] spatstat.data_3.1-6    R.methodsS3_1.8.2      tidyr_1.3.1           
 [70] data.table_1.17.2      BiocGenerics_0.54.0    spatstat.geom_3.3-6   
 [73] RcppAnnoy_0.0.22       ggrepel_0.9.6          RANN_2.6.2            
 [76] foreach_1.5.2          pillar_1.10.2          stringr_1.5.1         
 [79] spam_2.11-1            RcppHNSW_0.6.0         later_1.4.2           
 [82] circlize_0.4.16        splines_4.5.0          lattice_0.22-6        
 [85] survival_3.8-3         deldir_2.0-4           tidyselect_1.2.1      
 [88] miniUI_0.1.2           pbapply_1.7-2          knitr_1.50            
 [91] gridExtra_2.3          IRanges_2.42.0         scattermore_1.2       
 [94] stats4_4.5.0           xfun_0.52              matrixStats_1.5.0     
 [97] stringi_1.8.7          lazyeval_0.2.2         yaml_2.3.10           
[100] evaluate_1.0.3         codetools_0.2-20       tibble_3.2.1          
[103] cli_3.6.5              uwot_0.2.3             xtable_1.8-4          
[106] reticulate_1.42.0      Rcpp_1.0.14            globals_0.18.0        
[109] spatstat.random_3.3-3  png_0.1-8              spatstat.univar_3.1-3 
[112] parallel_4.5.0         dotCall64_1.2          listenv_0.9.1         
[115] viridisLite_0.4.2      scales_1.4.0           ggridges_0.5.6        
[118] purrr_1.0.4            crayon_1.5.3           GetoptLong_1.0.5      
[121] rlang_1.1.6            cowplot_1.1.3         </code></pre>
</div>
</div>


</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation column-body">
  <div class="nav-page nav-page-previous">
      <a href="./selector.html" class="pagination-link" aria-label="Select cells from reduction maps">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Select cells from reduction maps</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./Yano_ASE.html" class="pagination-link" aria-label="Allele-specific gene expression analysis">
        <span class="nav-page-text">Allele-specific gene expression analysis</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>